{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This generates gblocks of designs from fasta files. The clusters can be specified in to 8 or 12 to fit a 96 well plate. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take difference between minimum and maximum of design based on name. Next, use kmeans to cluster designs into N-number of\n",
    "cluster points. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Restriction import *\n",
    "from Bio.Seq import Seq\n",
    "from datetime import date\n",
    "from Bio.Alphabet import generic_dna, generic_protein\n",
    "today = date.today()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inputfile to be defined by the user\n",
    "ref_dna_file = \"AP-HT_input_vector_BsaI_scrubbed.fa\"\n",
    "designs_aa_fasta_file = \"init_designs.fasta\"\n",
    "codon_table_file = \"codon-usage_Plentus.txt\"\n",
    "# number of clusters to divide the sequences into\n",
    "cluster_size = 12\n",
    "ofile = str(today)+\"_data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_sequence(designs_aa_fasta_file):\n",
    "    sequences_ = {}\n",
    "    with open(designs_aa_fasta_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if (line[0] == \">\"):\n",
    "                tmp = line.split()\n",
    "                ids = tmp[0]\n",
    "                sequences_[ids] = \"\"\n",
    "            else:\n",
    "                tmp = line.strip()\n",
    "                sequences_[ids] = sequences_[ids] + tmp\n",
    "    return sequences_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure that the names are correct with respect to the reference sequence\n",
    "import sys\n",
    "sys.path.append(\"/Users/pgreisen/pythonscripts/develop/insert_mutation_into_DNA_string\")\n",
    "import InsertSubstitutions\n",
    "ris = InsertSubstitutions.InsertSubstitutions()\n",
    "\n",
    "# get reference DNA\n",
    "ref_dna = get_fasta_sequence(ref_dna_file)\n",
    "# get the AA sequence of the designs\n",
    "sequences = get_fasta_sequence(designs_aa_fasta_file)\n",
    "ref_key = list(ref_dna.keys())[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename sequence to reference DNA\n",
    "ref_dna_bp = Seq(ref_dna[ref_key])\n",
    "ref_dna_peptide_seq = ref_dna_bp.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"/Users/pgreisen/pythonscripts/develop/RenameToNative\")\n",
    "import RenameRelativeToNativeSeq as rrn\n",
    "r_ = rrn.DiffFasta()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_renamed = {}\n",
    "for i in sequences.keys():\n",
    "    var_ = r_.diff_sequence_a_b(ref_dna_peptide_seq, sequences[i])\n",
    "    sequences_renamed[var_[:-1]] = sequences[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insert substitutions into DNA string\n",
    "dna_seqs = {}\n",
    "for key in sequences_renamed.keys():\n",
    "    stop_of_list = -1\n",
    "    mut_header = \"\"\n",
    "    tmpheader = key.split('_')\n",
    "    integer_ = isinstance(tmpheader[-1], int)\n",
    "    if(integer_ == False):\n",
    "        stop_of_list = len(tmpheader)\n",
    "    for i in tmpheader[0:stop_of_list]:\n",
    "        mut_header += i+\"_\"\n",
    "    fastaheader, sequence = ris.run( ref_dna_file, codon_table_file, mut_header[:-1])\n",
    "    assert len(sequence) == len(ref_dna[ref_key])\n",
    "    dna_seqs[fastaheader] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# parameters needed for the run\n",
    "# bsaI sites for cloning\n",
    "preseq = \"GTCACGGTCTCA\"\n",
    "postseq = \"CGAGACCAGTCA\"\n",
    "# range of basepair offset\n",
    "bp_offset_nterm = 3\n",
    "bp_offset_cterm = 3\n",
    "bp_offset = 3\n",
    "# variable parameter for the gene length\n",
    "# g-blocks to synthetize\n",
    "cutoff = [250,500,750,1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prices taken from here (Tubes): \n",
    "# https://www.idtdna.com/pages/products/genes-and-gene-fragments/gblocks-gene-fragments\n",
    "idt_gblock = {\n",
    "    250 : 71.10,\n",
    "    500 : 71.10,\n",
    "    750 : 80.10,\n",
    "    1000 : 89.10,\n",
    "    1250 : 143.10\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_price(length):\n",
    "    ln = int(length)\n",
    "    if ln <= 250:\n",
    "        return 71.10\n",
    "    elif ln <= 500: \n",
    "        return 71.10\n",
    "    elif ln <= 750:\n",
    "        return 80.10\n",
    "    elif ln <= 1000:\n",
    "        return 89.10\n",
    "    elif ln <= 1250:\n",
    "        return 143.10\n",
    "    elif ln <= 1500:\n",
    "        return 179.1\n",
    "    else:\n",
    "        return \"Error\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fasta_file_w_variants(df):\n",
    "    lengt = []\n",
    "    with open(ofile+\".fasta\",'w') as f:\n",
    "        for i, varname, dnaseq, st, end in zip(df.index, df[\"VarName\"],df[\"seq(DNA)\"],df[\"start_cut_DNA\"],df[\"end_cut_DNA\"]):\n",
    "            f.write(varname+\"\\n\")\n",
    "            lengt.append(len(preseq+dnaseq[st:end]+postseq))\n",
    "            idt_seq_ = preseq+dnaseq[st:end]+postseq\n",
    "            a = Analysis(restrictionbatch=RestrictionBatch(['BsaI']),sequence=Seq(idt_seq_) )\n",
    "            prot_sites_ = a.full()\n",
    "            prot_key = list(prot_sites_.keys())[0]\n",
    "            cutsites = ', '.join(map(str, prot_sites_[prot_key]))\n",
    "            df.iloc[i, df.columns.get_loc(\"bsaI_sites\")] = cutsites\n",
    "            df.iloc[i, df.columns.get_loc(\"IDT_gene\")] = idt_seq_\n",
    "            f.write(idt_seq_+\"\\n\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fastafile(ifile):\n",
    "    sequences = {}\n",
    "    with open(ifile,'r') as f:\n",
    "        for line in f:\n",
    "            if(line[0] == \">\"):\n",
    "                tmp = line.split()\n",
    "                ids = tmp[0]\n",
    "                sequences[ids] = \"\"\n",
    "            else:\n",
    "                tmp = line.strip()\n",
    "                sequences[ids] = sequences[ids] + tmp\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reassign_missing_variants(df):\n",
    "    '''\n",
    "    '''\n",
    "    reassign_cluster = {}\n",
    "    # sub-divide the rest into exciting cluster\n",
    "    for i,j,k,c,l in zip(df.index, df[\"start_DNA_w_padding\"], df[\"end_DNA_w_padding\"], df[\"Clustered\"],df[\"Size_of_gene\"]):\n",
    "        if(c == 0):\n",
    "            for key in dict_w_start_stop:\n",
    "                if( (j >= dict_w_start_stop[key][0]) & (k <= dict_w_start_stop[key][1])):\n",
    "                    # add and sort such that we can add it to the\n",
    "                    # shortest gene cluster\n",
    "                    tmpkey = \"index_\"+str(i)\n",
    "                    reassign_cluster[tmpkey] = (i,j,k,c,key,dict_w_start_stop[key][2])\n",
    "    # sort dictionary\n",
    "    listofTuples = sorted(reassign_cluster.items() ,  key=lambda x: x[-1] )\n",
    "    r_c = dict(listofTuples)\n",
    "    for key in reassign_cluster.keys():\n",
    "        df.iloc[r_c[key][0], df.columns.get_loc(\"Clustered\")] = 1\n",
    "        df.iloc[r_c[key][0], df.columns.get_loc(\"Cluster_name\")] = r_c[key][4]\n",
    "        df.iloc[r_c[key][0], df.columns.get_loc(\"Size_of_gene\")] = r_c[key][5]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dataframe with the AA and DNA sequences\n",
    "df_sequences = pd.DataFrame.from_dict(dna_seqs,orient='index',columns=['seq(DNA)'] )\n",
    "df_sequences.reset_index(inplace=True)\n",
    "df_sequences = df_sequences.rename(columns={'index':'VarName'})\n",
    "df_sequences[\"VarName\"] = df_sequences[\"VarName\"].str.replace(\">\",\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_residue_numbers(seq_df):\n",
    "    # initialize new columns\n",
    "    seq_df[\"start_DNA\"] = 0 \n",
    "    seq_df[\"end_DNA\"] = 0 \n",
    "    # loop over keys in hash\n",
    "    # Contain all the positions mutated\n",
    "    residue_nr = []\n",
    "    # contains residue numbers for each variant\n",
    "    tmpresiduenr = {}\n",
    "    for index, varname in zip(seq_df.index, seq_df[\"VarName\"]):\n",
    "        new_array = []\n",
    "        nametmp_ = varname.split(\"_\")\n",
    "        for j in nametmp_:\n",
    "            new_array.append( int(j[1:-1] ) )\n",
    "        mn = min(new_array)\n",
    "        mx = max(new_array)\n",
    "        # The -3 is due to counting from 0 and protein vs DNA\n",
    "        seq_df.iloc[index,seq_df.columns.get_loc(\"start_DNA\")] = 3*mn-3\n",
    "        # \n",
    "        seq_df.iloc[index,seq_df.columns.get_loc(\"end_DNA\")] = 3*mx\n",
    "    return seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sequences = set_residue_numbers(df_sequences)\n",
    "df_sequences[\"start_DNA_w_padding\"] = df_sequences[\"start_DNA\"] - bp_offset\n",
    "df_sequences[\"end_DNA_w_padding\"] = df_sequences[\"end_DNA\"] + bp_offset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sequences.sort_values(by=[\"start_DNA_w_padding\",\"end_DNA_w_padding\"],inplace=True)\n",
    "df_sequences[\"length_DNA_w_padding\"] = df_sequences[\"end_DNA_w_padding\"] - df_sequences[\"start_DNA_w_padding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sequences[\"Clustered\"] = 0\n",
    "df_sequences[\"Cluster_name\"] = \"\"\n",
    "df_sequences[\"Size_of_gene\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_clusters(df,min_size_cluster,cutoff,gb=\"start_DNA_w_padding\", parameter=\"length_DNA_w_padding\"):\n",
    "    # remove lengths greater than cutoff\n",
    "    temp = df[(df[parameter] < cutoff) & (df[\"Clustered\"] != 1) ]\n",
    "    # count how many \n",
    "    tmp_= temp.groupby(gb).count().reset_index()\n",
    "    # only clusters of certain size are investigated\n",
    "    tmp_ = tmp_[tmp_[\"VarName\"] >= min_size_cluster]\n",
    "    return tmp_\n",
    "\n",
    "def set_clusters(df,init_values, cutoff):\n",
    "    for i,j,l,c in zip(df.index, df[\"start_DNA_w_padding\"], df[\"length_DNA_w_padding\"],df[\"Clustered\"]):\n",
    "        \n",
    "        # bool value so you dont overwrite existing clusters\n",
    "        if(c == 1):\n",
    "            continue\n",
    "        elif((j in init_values) & (l < cutoff)):\n",
    "            df.iloc[i, df.columns.get_loc(\"Clustered\")] = 1\n",
    "            df.iloc[i, df.columns.get_loc(\"Cluster_name\")] = \"cluster_\"+str(j)+\"_\"+str(j+cutoff)+\"_\"+str(cutoff)\n",
    "            df.iloc[i, df.columns.get_loc(\"Size_of_gene\")] = cutoff\n",
    "        else:\n",
    "            continue\n",
    "    return df\n",
    "df = df_sequences.copy()\n",
    "df = df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x1a1c23c3c8>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAD8CAYAAACRkhiPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHGZJREFUeJzt3XuYXHWd5/H3p+/d6QYC3RmSNCGBBFhQnwAFw6zAIgxXXRBlTZxRGMd9ooy4uo4y8LAriOOsuiIujw4IawZYMQRvXCSRcQTEUW4dIJAAkSYEaDokIZGQTvqSVH/3jzodyqb63umqzvm8nuc8XfU9vzr9rZPT59PnUh1FBGZmlk5lxW7AzMyKxyFgZpZiDgEzsxRzCJiZpZhDwMwsxRwCZmYp5hAwM0sxh4CZWYo5BMzMUqyi2A0MpbGxMWbPnl3sNszMJo0VK1a8ERFNwxlb8iEwe/ZsWlpait2GmdmkIenl4Y716SAzsxRzCJiZpZhDwMwsxRwCZmYp5hAwM0sxh4CZWYo5BMzMUqzkPycwntrbtvC7379ARXkZJ558OE1N+xS7JbNh68lupie7kTLVUFsxC6n8T+Y/8Mpa7lr7LHXlVXz22D9net2+Ay7rG//9Np5teYmpTQ38jxsvpLFxKk9teY5bXv4pvdHL+TNO55QD/2JPvyVLfPWfl/HoMy+zT301//SlDzB72rQJ+94q9f9jOJPJxFg/LJbNZvnG1+5h5b+v4sT2p5i+fTOv1zdR9fGPsujS88epU7M9oze62bDt52zb8gj7/HwNlWu3kJ07g4aPf5Xa/d/F1h07OPPOW9i25Q0+8PhKZm/YzMt/1kj2wxfwv89d8CfLeuCeJ/jm535EbW8PJ29/nhk736S9cj8ev3EW1MN7f93K9LatrG/el8dOO5zrT/wOtbW1RXrne78HH/sDl197D3U7uzjtlZU0b3uDtoZGXjnxVG749qdHvVxJKyIiM6zBETHoBCwGNgKr8mpLgaeSaR3wVN68y4FWYA1wZl79rKTWClw21Pftm4499tgYq+u/+6v40nEXx46K6ugsr4qA6Cyviu3l1fHA1TeMeflme9L6t+6Il+9eGNn66sjW5bbfbF1lZOurYueDy+J9S2+KD116cWyrqY6O6tz8juqq2FZTHT9dfP2fLOusOX8ffz99QWxXZexQZQTEjuqK6Kwuj87qiuisqcj9fNRUxPa6yvjq9QuK9K7T4YSF34pFp30mOiqqY3uyb9peXhUdFdXx4yv+z6iXC7TEMPexwwmBk4Fj8kOg3/xrgC8nj48EVgLVwBzgRaA8mV4EDgGqkjFHDqfBsYbArl274mPv/3psr6jOvd1+046K6oht28b0Pcz2lJ27tsYf1n4usvWFt99d9bWR+dYVsa2m8PyOmre3769//odx/sGfje3Jzn840/a6yti88eUir4W90zX/8m9x6of/MToG2Dd1jGHfNJIQGPLCcEQ8BGwZ4JBDwEeAJUnpPOD2iOiOiJeS3/qPT6bWiFgbET3A7cnYPW7Llu0cv66FsoFOe0XA0qUT0YrZiPX0vkHDnWugt/D2m81m+YefLkfD2L5XP7aWk7c/j0bw/ct6gz/e9oMRdm3D8bsn1nLaKysH/LfTBO2bxnp30EnAhoh4IXk+E3g1b35bUhuoXpCkRZJaJLVs2rRpTA3W19cwo3MzNdmegvNrsz3Q2jqm72G2p5Srlsq1WyjbUXj7rers4ZDXNzGlu/D8Kd1vb9/7HFDPjJ1vUhs7h/39a7p2Me21zSNv3IbUtH89zdveoG6AfVPdBO2bxhoCH+XtowCg4C8ZMUi9oIi4MSIyEZFpahrWX0MdUG1tFWXz5tJZXlVwfk9VNcydO6bvYbanVJUfCHMPobeusuD8mFLHqzMOZHt14e27q+bt7fuqxX9De+V+dKrwsgq/voKGI44eeeM2pGu+dC5tDY3sGGDf1FU5MfumUYeApArgQ+QuEvdpAw7Ke94MtA9SnxAnXX8lqPBBcGVVJSxYUHCeWbFJouHCr0HZACdxysqZeu13iAG27+rKqt3bd2PjVLaces7Av30VUFlR5Z+PPWTKlClsPecDA//bVU/MvmksRwJ/CTwfEW15tbuBhZKqJc0B5gGPAY8D8yTNkVQFLEzGTojGOTOJe+9lZ00dXRXVAOyqriXqG9Dy5VBfP1GtmI1Y1dRDiHvvJerr6K3Lbb8xpY5oaEDLlnHycf+RN3/2Ezpra9iRHBH01NTsnp+/fV+99PM89sVr2FFWtfuIoLuimt7aWnpqKumqyX10qKu2kt76KZQvv88/H3vQN666kEf/13fZXlG9+4igq7IaGiZu3zTk5wQkLQFOARqBDcCVEfEDSTcDj0TEDf3GXwH8LbAL+HxELE/q5wDfIXen0OKI+NpwGhyPzwns1tGRu9DS2po7zFqwwBu4TR5Dbb8j2b4LjQX/fBTLOO+bRvI5gVR8WMzMLE1GEgL+20FmZinmEDAzSzGHgJlZijkEzMxSzCFgZpZiDgEzsxRzCJiZpZhDwMwsxRwCZmYp5hAwM0sxh4CZWYo5BMzMUswhYGaWYg4BM7MUcwiYmaWYQ8DMLMUcAmZmKeYQMDNLMYeAmVmKOQTMzFJsyBCQtFjSRkmr+tU/K2mNpNWSvpnUZkvqlPRUMt2QN/5YSc9IapV0nSSN/9sxM7ORqBjGmJuB7wK39hUkvQ84D3hPRHRLmpY3/sWImF9gOdcDi4BHgGXAWcDyUfZtZmbjYMgjgYh4CNjSr3wx8PWI6E7GbBxsGZKmA/tExMMREeQC5YOja9nMzMbLaK8JHAacJOlRSb+RdFzevDmSnkzqJyW1mUBb3pi2pGZmZkU0nNNBA71uKnACcBxwh6RDgPXArIjYLOlY4E5JRwGFzv/HQAuXtIjcqSNmzZo1yhbNzGwooz0SaAN+FjmPAb1AY0R0R8RmgIhYAbxI7qihDWjOe30z0D7QwiPixojIRESmqalplC2amdlQRhsCdwKnAkg6DKgC3pDUJKk8qR8CzAPWRsR6YJukE5K7gi4E7hpz92ZmNiZDng6StAQ4BWiU1AZcCSwGFie3jfYAF0VESDoZuFrSLiALfDoi+i4qX0zuTqNacncF+c4gM7MiU+5mndKVyWSipaWl2G2YmU0aklZERGY4Y/2JYTOzFHMImJmlmEPAzCzFHAJmZinmEDAzSzGHgJlZijkEzMxSzCFgZpZiDgEzsxRzCJiZpZhDwMwsxRwCZmYp5hAwM0sxh4CZWYo5BMzMUswhYGaWYg4BM7MUcwiYmaWYQ8DMLMUcAmZmKTZkCEhaLGmjpFX96p+VtEbSaknfzKtfLqk1mXdmXv2spNYq6bLxfRtmZjYaFcMYczPwXeDWvoKk9wHnAe+JiG5J05L6kcBC4ChgBvBvkg5LXvY94HSgDXhc0t0R8ex4vREzMxu5IUMgIh6SNLtf+WLg6xHRnYzZmNTPA25P6i9JagWOT+a1RsRaAEm3J2MdAmZmRTTaawKHASdJelTSbyQdl9RnAq/mjWtLagPVzcysiIZzOmig100FTgCOA+6QdAigAmODwmETAy1c0iJgEcCsWbNG2aKZmQ1ltEcCbcDPIucxoBdoTOoH5Y1rBtoHqRcUETdGRCYiMk1NTaNs0czMhjLaELgTOBUgufBbBbwB3A0slFQtaQ4wD3gMeByYJ2mOpCpyF4/vHmvzZmY2NkOeDpK0BDgFaJTUBlwJLAYWJ7eN9gAXRUQAqyXdQe6C7y7gMxGRTZZzCXAfUA4sjojVe+D9mJnZCCi37y5dmUwmWlpait2GmdmkIWlFRGSGM9afGDYzSzGHgJlZijkEzMxSzCFgZpZiDgEzsxRzCJiZpZhDwMwsxRwCZmYp5hAwM0sxh4CZWYo5BMzMUswhYGaWYg4BM7MUcwiYmaWYQ8DMLMUcAmZmKeYQMDNLMYeAmVmKOQTMzFLMIWBmlmIOATOzFBsyBCQtlrRR0qq82lWSXpP0VDKdk9RnS+rMq9+Q95pjJT0jqVXSdZK0Z96SmZkN13COBG4GzipQvzYi5ifTsrz6i3n1T+fVrwcWAfOSqdAyzcxsAg0ZAhHxELBlLN9E0nRgn4h4OCICuBX44FiWaWZmYzeWawKXSHo6OV00Na8+R9KTkn4j6aSkNhNoyxvTltTMzKyIRhsC1wOHAvOB9cA1SX09MCsijga+APxI0j5AofP/MdDCJS2S1CKpZdOmTaNs0czMhjKqEIiIDRGRjYhe4Cbg+KTeHRGbk8crgBeBw8j95t+ct4hmoH2Q5d8YEZmIyDQ1NY2mRTMzG4ZRhUByjr/P+cCqpN4kqTx5fAi5C8BrI2I9sE3SCcldQRcCd42pczMzG7OKoQZIWgKcAjRKagOuBE6RNJ/cKZ11wKeS4ScDV0vaBWSBT0dE30Xli8ndaVQLLE8mMxsn2WyWtzZ3UFlVQf1+U0b8+oigM7sNCGrLG5D8MaI0UO5mndKVyWSipaWl2G2YlayI4NmH1/DwbQ9w0Mp/Z7+3NqJ58zj0ny7lgMNnD2sZW3raef61ZUz5yb8yZd0b7Dr0YPb/2BeYNu3de7Z52yMkrYiIzLDGOgTMJrfVv3+eJ6++gY888H0UQeXObnoqqkFi58/vYsr7zxj09Vt3buTZe7/GcR//PuqF8h3d7KqrIiQ67vohU0+7YILeiY2XkYSAj/fMJrFsNsvvb7ufjzzwfap6uqjc2Q1A1a5uqnZ2UX3B+dDRMegy1rY/yHEf/z4VHd2U78i9vmJHD5Xbu2n44MeJbdv29NuwInIImE1ib72xjYNX/g4NcEQf2SwsXTroMip+/HPoHWBmby/Z228bY5dWyhwCZpNYeWU5+25t330E0F/lzm5obR10GfXrtlCxo/DrK3b0UPbi2jH3aaXLIWA2iTVMrUczOgf85GUANDcPMDen5vCj2VVXVXBetq6GsnmHj6lHK20OAbNJLLIbOWL+m2NaxgEXfgHKCu8KysorYMGCMS3fSptDwGwy691Mw46ugn+XBZK/19LWNsDcnKp9pxH3/oLe+jp21VXnFltXQzTUo2XLob5+XFu20jLkh8XMrIRVNBOH1ENdOdqRfcfsqKtEc+cOuZjKk0+D9RsoW7oUWlspmzs3dwTgANjrOQTMJrGysn3o/cgC+PJXCg8orxz+6Zz6evjkJ8evOZsUfDrIbLKb/iXijr8i6suJutyPdNSVE/W1aNl9/m3eBuUjAbNJrqysBs6+ld5XvwRLvg9rNxGHnUTZR//GAWBDcgiY7SXK9ns3XPxdoPB/4GFWiE8HmZmlmEPAzCzFHAJmZinmEDAzSzGHgJlZijkEzMxSzCFgZpZiDgEzsxQbMgQkLZa0UdKqvNpVkl6T9FQynZM373JJrZLWSDozr35WUmuVdNn4vxUzMxup4RwJ3AycVaB+bUTMT6ZlAJKOBBYCRyWv+WdJ5ZLKge8BZwNHAh9NxpqZWREN+WcjIuIhSbOHubzzgNsjoht4SVIrcHwyrzUi1gJIuj0Z++yIOzYzs3EzlmsCl0h6OjldNDWpzQRezRvTltQGqpuZWRGNNgSuBw4F5gPrgWuSeqG/WxWD1AuStEhSi6SWTZs2jbJFMzMbyqhCICI2REQ2InqBm3j7lE8bcFDe0GagfZD6QMu/MSIyEZFpamoaTYtmZjYMowoBSdPznp4P9N05dDewUFK1pDnAPOAx4HFgnqQ5kqrIXTy+e/Rtm5nZeBjywrCkJcApQKOkNuBK4BRJ88md0lkHfAogIlZLuoPcBd9dwGciIpss5xLgPqAcWBwRq8f93ZiZ2YgoYsBT8yUhk8lES0tLsdswM5s0JK2IiMxwxvoTw2ZmKeYQMDNLMYeAmVmKOQTMzFLMIWBmlmIOATOzFHMImJmlmEPAzCzFHAJmZinmEDAzSzGHgJlZijkEzMxSzCFgZpZiDgEzsxRzCJiZpZhDwMwsxRwCZmYp5hAwM0sxh4CZWYo5BMzMUswhYGaWYkOGgKTFkjZKWlVg3hclhaTG5PkpkrZKeiqZvpw39ixJayS1SrpsfN+GmZmNxnCOBG4GzupflHQQcDrwSr9Zv42I+cl0dTK2HPgecDZwJPBRSUeOpXEzMxu7IUMgIh4CthSYdS1wKRDD+D7HA60RsTYieoDbgfNG0qiZmY2/UV0TkHQu8FpErCww+y8krZS0XNJRSW0m8GremLakZmZmRVQx0hdIqgOuAM4oMPsJ4OCI6JB0DnAnMA9QgbEDHkFIWgQsApg1a9ZIWzQzs2EazZHAocAcYKWkdUAz8ISkAyPirYjoAIiIZUBlctG4DTgobxnNQPtA3yAiboyITERkmpqaRtGimZkNx4iPBCLiGWBa3/MkCDIR8YakA4ENERGSjicXMpuBN4F5kuYArwELgb8ah/7NzGwMhnOL6BLgYeBwSW2SPjnI8AuAVZJWAtcBCyNnF3AJcB/wHHBHRKwee/tmZjYWihjOzT3Fk8lkoqWlpdhtmJlNGpJWRERmOGP9iWEzsxRzCJiZpZhDwMwsxRwCZmYp5hAwM0sxh4CZWYo5BMzMUswhYGaWYg4BM7MUcwiYmaWYQ8DMLMUcAmZmKeYQMDNLMYeAmVmKOQTMzFLMIWBmlmIOATOzFHMImJmlmEPAzCzFHAJmZik2rBCQtFjSRkmrCsz7oqSQ1Jg8l6TrJLVKelrSMXljL5L0QjJdNH5vw8zMRmO4RwI3A2f1L0o6CDgdeCWvfDYwL5kWAdcnY/cHrgT+HDgeuFLS1NE2bmZmYzesEIiIh4AtBWZdC1wKRF7tPODWyHkE2E/SdOBM4FcRsSUi/gj8igLBYmZmE2fU1wQknQu8FhEr+82aCbya97wtqQ1UNzOzIqkYzYsk1QFXAGcUml2gFoPUCy1/EblTScyaNWs0LZqZ2TCM9kjgUGAOsFLSOqAZeELSgeR+wz8ob2wz0D5I/R0i4saIyEREpqmpaZQtmpnZUEYVAhHxTERMi4jZETGb3A7+mIh4HbgbuDC5S+gEYGtErAfuA86QNDW5IHxGUjMzsyIZ7i2iS4CHgcMltUn65CDDlwFrgVbgJuDvACJiC/BV4PFkujqpmZlZkSii4Gn5kpHJZKKlpaXYbZiZTRqSVkREZjhj/YlhM7MUcwiYmaWYQ8DMLMUcAmZmKTaqD4tNJh3d3fxh02a6du7iwH3qmb3/VMpU6HNrZpNXRNC2eSuvbX6Lyooy5h7YyL5TagqO7cz28OOXVvCHtzbQVN3Ahw8+mub6/Se4YysVe3UIrHj1Ne555HHe9dCDHNDezhMzZnD/ueey8OQTqa+uKnZ7ZuOie+cu7vjdSl568WXe/fBDHLChnQemz+TAT32SE45915+MfeaPbXzu0aVk33qL03/7NPuu38ytMxqZ9YmL+djRpxfpHVgx7bW3iLZvfYu7briJv/3Kl1H0UtXVRU9NDb0Sv7nue5z5Xz+xB7o1m3j3PP4sr9+zjIu+dSX0bevVuW19yx0/YcZ/PgeAbDbL2b++jjlPPs+1X7kFRVDb1cOOmipC8PpPl3Do2R8q8rux8eBbRIEnn1vDJ77yP6nu3EFVVxcAVV1d1HR28p/+22d4c9MbRe7QbOw6e3by9KoXuOhbV1KVv613d1HT1UnTwv8CHR0A3P/6Grrf/CPf/sot1HV2U9vVA0BdVw9TOnuYecFf7x5r6bHXhkD9nXdSNsBRjnqD7JIlE9yR2fjr6OzmXQ8/CNFbcH5ke2HpUgDWbd/M6b99ZsCfC3pj91hLj732msD0ja/v/q2ov+ruLuLVVwrOM5tM6mur2f/19gG39aruLmhtBWD2lAOoXL959xFAfzVd3bvHWnrstUcC046eT3d14bsjdtbWUnPEERPckdn4q62qpOHdRw24rffW1cHcuQCceuDhbDloOjsGuCkimzfW0mOvvTDMtm1kp0+nfPv2d8yKhgbU3g719ePQoVlxdW3+I2WzmqnaseOdMxsaIG9bX/3K88w+Yj5TOruHHGuTly8MAzQ0UP7LXxINDbnfcMj9VhQNDWjZMm/otteoOWAqlb/8Jb319WRrc9t6TJmS26n329aPmnUEFcuXs3NKHd211QD0TqkrONbSYe89EujT0ZG72NXamjvUXbDAG7rtnUayrfvnYq82kiOBvT8EzMxSxqeDzMxsWBwCZmYp5hAwM0sxh4CZWYo5BMzMUswhYGaWYg4BM7MUK/nPCUjaBGwHJtvffm7EPU8E9zwxJmPPMDn7Ho+eD46IpuEMLPkQAJDUMtwPPpQK9zwx3PPEmIw9w+Tse6J79ukgM7MUcwiYmaXYZAmBG4vdwCi454nhnifGZOwZJmffE9rzpLgmYGZme8ZkORIwM7M9oOghIOkgSQ9Iek7SakmfS+r7S/qVpBeSr1OTuiRdJ6lV0tOSjimhnq+S9Jqkp5LpnLzXXJ70vEbSmRPdc9JDjaTHJK1M+v5KUp8j6dFkXS+VVJXUq5Pnrcn82SXU882SXspb1/OTetG3j6SPcklPSvpF8rxk1/EgPZf0Ok56WSfpmaS/lqRWsvuOQXou3r4jIoo6AdOBY5LHDcAfgCOBbwKXJfXLgG8kj88BlgMCTgAeLaGerwK+WGD8kcBKoBqYA7wIlBehbwH1yeNK4NFkHd4BLEzqNwAXJ4//DrghebwQWFpCPd8MXFBgfNG3j6SPLwA/An6RPC/ZdTxIzyW9jpNe1gGN/Wolu+8YpOei7TuKfiQQEesj4onk8TbgOWAmcB5wSzLsFuCDyePzgFsj5xFgP0nTS6TngZwH3B4R3RHxEtAKHL/nO/1TyTrrSJ5WJlMApwI/Ser913Xfv8FPgNMkaYLaBQbteSBF3z4kNQPvB/5v8lyU8DqGd/Y8hKKv4yGU7L5jFPb4vqPoIZAvORQ+mtxve38WEesht9MFpiXDZgKv5r2sjcF3wHtUv54BLkkONRf3HYZSQj0nh/xPARuBX5H7zeLNiNhVoLfdfSfztwIHTGzH7+w5IvrW9deSdX2tpOr+PSeKsa6/A1wK9CbPD6DE1zHv7LlPqa7jPgH8q6QVkhYltVLfdxTqGYq07yiZEJBUD/wU+HxEvDXY0AK1otziVKDn64FDgfnAeuCavqEFXl6UniMiGxHzgWZyv1H8h0LDkq8l0Xf/niW9C7gcOAI4Dtgf+IdkeFF7lvQBYGNErMgvFxhaMut4gJ6hRNdxP++NiGOAs4HPSDp5kLGl0nehnou27yiJEJBUSW5neltE/Cwpb+g7VEu+bkzqbcBBeS9vBtonqtc+hXqOiA3JDqsXuIm3D9tKoud8EfEm8CC5c6P7SapIZuX3trvvZP6+wJaJ7fRteT2flZySi4joBv6F0lnX7wXOlbQOuJ3caaDvUNrr+B09S/phCa/j3SKiPfm6Efg5uR5Let9RqOdi7juKHgLJ+c8fAM9FxLfzZt0NXJQ8vgi4K69+YXKl/wRga9+h30QZqOd+5xfPB1Ylj+8GFiZ3gswB5gGPTVS/ef01SdoveVwL/CW56xkPABckw/qv675/gwuA+yO5WjVRBuj5+bwfcpE755u/rou2fUTE5RHRHBGzyV3ovT8i/poSXscD9PyxUl3HfSRNkdTQ9xg4I+mxlPcdBXsu6r5jPK8yj2YCTiR3ePM08FQynUPuvOivgReSr/sn4wV8j9y57GeATAn1/P+Snp5O/vGm573miqTnNcDZRVrX7wGeTPpbBXw5qR+SbFitwI+B6qRekzxvTeYfUkI935+s61XAD3n7DqKibx95vZ/C23falOw6HqTnkl7HyTpdmUyrgSuSeinvOwbquWj7Dn9i2MwsxYp+OsjMzIrHIWBmlmIOATOzFHMImJmlmEPAzCzFHAJmZinmEDAzSzGHgJlZiv1/aLDCXnzImUMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "from pandas import DataFrame\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "df_cluster = DataFrame(df,columns=['start_DNA_w_padding', 'end_DNA_w_padding'])\n",
    "kmeans = KMeans(n_clusters=cluster_size).fit(df_cluster)\n",
    "centroids = kmeans.cluster_centers_\n",
    "plt.scatter(df_cluster['start_DNA_w_padding'], df_cluster['end_DNA_w_padding'], c= kmeans.labels_.astype(float), s=50, alpha=0.5)\n",
    "plt.scatter(centroids[:, 0], centroids[:, 1], c='red', s=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_map = pd.DataFrame()\n",
    "cluster_map['data_index'] = df.index.values\n",
    "cluster_map['cluster'] = kmeans.labels_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge dataframes\n",
    "df_kmeans = pd.merge(right=df, right_index=True, left=cluster_map, left_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_kmeans.drop(labels=['data_index', 'index',\\\n",
    "                       'start_DNA_w_padding', 'end_DNA_w_padding','length_DNA_w_padding',\\\n",
    "                       'Clustered','Size_of_gene','Cluster_name'], axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "nr_clusters = list(set(kmeans.labels_))\n",
    "df_kmeans['start_cut_DNA'] = 0\n",
    "df_kmeans['end_cut_DNA'] = 0\n",
    "df_kmeans['Size_of_gene'] = 0\n",
    "df_kmeans['Cluster_name'] = \"\"\n",
    "\n",
    "for cluster_ in nr_clusters:\n",
    "    # get each cluster\n",
    "    subcluster = df_kmeans[df_kmeans['cluster'] == cluster_]\n",
    "    # start cut DNA\n",
    "    mn_ = min(subcluster['start_DNA'])\n",
    "    # end cut DNA\n",
    "    mx_ = max(subcluster['end_DNA'])\n",
    "    # size of gene\n",
    "    length_ = mx_-mn_\n",
    "    # cluster name\n",
    "    clustername_ = \"Cluster_\"+str(mn_)+\"_\"+str(mx_)+\"_cluster_nr_\"+str(cluster_)\n",
    "    for i in df_kmeans.index:\n",
    "        if(df_kmeans.iloc[i,df_kmeans.columns.get_loc('cluster')] == cluster_):\n",
    "            df_kmeans.iloc[i,df_kmeans.columns.get_loc('start_cut_DNA')] = mn_ - bp_offset\n",
    "            df_kmeans.iloc[i,df_kmeans.columns.get_loc('end_cut_DNA')] = mx_ + bp_offset\n",
    "            df_kmeans.iloc[i,df_kmeans.columns.get_loc('Size_of_gene')] = length_ + bp_offset\n",
    "            df_kmeans.iloc[i,df_kmeans.columns.get_loc('Cluster_name')]= clustername_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The price of the order is:  7660.799999999999\n",
      "Number of unique cluster:  12\n",
      "Size of the different clusters:  Cluster_name\n",
      "Cluster_222_1374_cluster_nr_5      1\n",
      "Cluster_225_1698_cluster_nr_2      3\n",
      "Cluster_420_1698_cluster_nr_10     1\n",
      "Cluster_429_1374_cluster_nr_4      1\n",
      "Cluster_429_1527_cluster_nr_11     2\n",
      "Cluster_429_1533_cluster_nr_0      1\n",
      "Cluster_429_1698_cluster_nr_6      5\n",
      "Cluster_441_1374_cluster_nr_7      3\n",
      "Cluster_441_1698_cluster_nr_1     10\n",
      "Cluster_447_1533_cluster_nr_9      1\n",
      "Cluster_447_1698_cluster_nr_8      9\n",
      "Cluster_552_1698_cluster_nr_3     11\n",
      "Name: VarName, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Calculate price\n",
    "df_kmeans[\"IDT_price\"] = df_kmeans[\"Size_of_gene\"].apply(get_price)\n",
    "print(\"The price of the order is: \",df_kmeans[\"IDT_price\"].sum())\n",
    "print(\"Number of unique cluster: \", len(df_kmeans[\"Cluster_name\"].unique()))\n",
    "tmp = df_kmeans.groupby(\"Cluster_name\").count()\n",
    "print(\"Size of the different clusters: \",tmp[\"VarName\"] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sets the IDT gene and bsaI cut-sites\n",
    "df_kmeans[\"bsaI_sites\"] = \"\"\n",
    "df_kmeans[\"IDT_gene\"] = \"\"\n",
    "df_kmeans = write_fasta_file_w_variants(df_kmeans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna, generic_protein\n",
    "df_kmeans[\"AA_seq\"] = \"\"\n",
    "for i,j in zip(df_kmeans.index, df_kmeans[\"IDT_gene\"]):\n",
    "    tmpseq_ = j[len(preseq):-len(postseq)]\n",
    "    my_seq = Seq(tmpseq_)\n",
    "    df_kmeans.iloc[i,df_kmeans.columns.get_loc(\"AA_seq\")] = my_seq.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_kmeans.to_excel(ofile+\".xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "with open(\"aa_test.fasta\",'w') as f:\n",
    "    for i,j in zip(df_kmeans[\"VarName\"],df_kmeans[\"AA_seq\"]):\n",
    "        f.write(\">\"+i+\"\\n\")\n",
    "        f.write(str(j)+\"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
