{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Unrooted tree from sequence alignments ##"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "For Anaconda Python distribution, install `biopython` via `conda`:\n",
    "    \n",
    "`conda install biopython`\n",
    "\n",
    "and  `igraph` and  `plotly`, via `pip`:\n",
    "    \n",
    "`pip install python-igraph`    \n",
    "\n",
    "`pip install plotly`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data source:  `aln` files from [http://kinase.com/human/kinome/groups/](http://kinase.com/human/kinome/groups/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator\n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "from Bio import AlignIO\n",
    "\n",
    "\n",
    "inner_node_color = '#ccccff'\n",
    "#  A dictionary to color nodes labeled by keys\n",
    "color_dict = {'RSK':'#e60000','SGK':'#ffff00','PKC':'#32cd32','DMPK':'#e600e6','NDR':'#3366ff',\n",
    "                'GRK':'#8080ff','PKA':'magenta','MAST':'green','YANK':'pink'}\n",
    "\n",
    "\n",
    "def set_node_color(label):\n",
    "    \n",
    "    node_color = 'blue'#default color\n",
    "    for Key in color_dict:\n",
    "        if Key in label:\n",
    "            node_color = color_dict[Key]\n",
    "    return node_color\n",
    "\n",
    "\n",
    "def get_tree():\n",
    "\n",
    "    aln = AlignIO.read('agc.aln', 'clustal')\n",
    "    #Computations for the unrooted tree\n",
    "    calculator = DistanceCalculator('identity')\n",
    "    dm = calculator.get_distance(aln)\n",
    "    constructor = DistanceTreeConstructor()\n",
    "    tree = constructor.nj(dm)\n",
    "    return tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tree=get_tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting python-igraph\n",
      "  Downloading python-igraph-0.7.1.post6.tar.gz (377kB)\n",
      "\u001b[K    100% |████████████████████████████████| 378kB 56kB/s ta 0:00:01\n",
      "\u001b[?25hBuilding wheels for collected packages: python-igraph\n",
      "  Running setup.py bdist_wheel for python-igraph ... \u001b[?25l-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \bdone\n",
      "\u001b[?25h  Stored in directory: /Users/pgreisen/Library/Caches/pip/wheels/2b/9b/50/34d9d2b8f74c7a1c70f6337686d66f8fc8bbab6d2f196fe6ef\n",
      "Successfully built python-igraph\n",
      "Installing collected packages: python-igraph\n",
      "Successfully installed python-igraph-0.7.1.post6\n",
      "Help on package igraph:\n",
      "\n",
      "NAME\n",
      "    igraph - IGraph library.\n",
      "\n",
      "FILE\n",
      "    /Users/pgreisen/anaconda2/lib/python2.7/site-packages/igraph/__init__.py\n",
      "\n",
      "DESCRIPTION\n",
      "    @undocumented: deprecated, _graphmethod, _add_proxy_methods, _layout_method_wrapper,\n",
      "                   _3d_version_for\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _igraph\n",
      "    app (package)\n",
      "    clustering\n",
      "    compat\n",
      "    configuration\n",
      "    cut\n",
      "    datatypes\n",
      "    drawing (package)\n",
      "    force_directed_layout\n",
      "    formula\n",
      "    json_formatter\n",
      "    layout\n",
      "    matching\n",
      "    notebook\n",
      "    remote (package)\n",
      "    statistics\n",
      "    summary\n",
      "    test (package)\n",
      "    utils\n",
      "    vendor (package)\n",
      "\n",
      "CLASSES\n",
      "    __builtin__.object\n",
      "        ARPACKOptions\n",
      "        BFSIter\n",
      "        Edge\n",
      "        Graph\n",
      "            Graph\n",
      "        Vertex\n",
      "    igraph.core.EdgeSeq(__builtin__.object)\n",
      "        EdgeSeq\n",
      "    igraph.core.VertexSeq(__builtin__.object)\n",
      "        VertexSeq\n",
      "    \n",
      "    class ARPACKOptions(__builtin__.object)\n",
      "     |  Class representing the parameters of the ARPACK module.\n",
      "     |  \n",
      "     |  ARPACK is a Fortran implementation of the implicitly restarted\n",
      "     |  Arnoldi method, an algorithm for calculating some of the\n",
      "     |  eigenvalues and eigenvectors of a given matrix. igraph uses this\n",
      "     |  package occasionally, and this class can be used to fine-tune the\n",
      "     |  behaviour of ARPACK in such cases.\n",
      "     |  \n",
      "     |  The class has several attributes which are not documented here,\n",
      "     |  since they are usually of marginal use to the ordinary user.\n",
      "     |  See the source code of the original ARPACK Fortran package\n",
      "     |  (especially the file C{dsaupd.f}) for a detailed explanation of the\n",
      "     |  parameters. Only the most basic attributes are explained here. Most\n",
      "     |  of them are read only unless stated otherwise.\n",
      "     |  \n",
      "     |   - C{bmat}: type of the eigenproblem solved. C{'I'} means standard\n",
      "     |     eigenproblem (A*x = lambda*x), C{'G'} means generalized\n",
      "     |     eigenproblem (A*x = lambda*B*x).\n",
      "     |  \n",
      "     |   - C{n}: dimension of the eigenproblem\n",
      "     |  \n",
      "     |   - C{tol}: precision. If less than or equal to zero, the standard\n",
      "     |     machine precision is used as computed by the LAPACK utility\n",
      "     |     called C{dlamch}. This can be modified.\n",
      "     |  \n",
      "     |   - C{mxiter}: maximum number of update iterations to take. This\n",
      "     |     can be modified. You can also use C{maxiter}.\n",
      "     |  \n",
      "     |   - C{iter}: actual number of update iterations taken\n",
      "     |  \n",
      "     |   - C{numop}: total number of OP*x operations\n",
      "     |  \n",
      "     |   - C{numopb}: total number of B*x operations if C{bmat} is C{'G'}\n",
      "     |  \n",
      "     |   - C{numreo}: total number of steps of re-orthogonalization\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __str__(...)\n",
      "     |      x.__str__() <==> str(x)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes defined here:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "    \n",
      "    class BFSIter(__builtin__.object)\n",
      "     |  igraph BFS iterator object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __iter__(...)\n",
      "     |      x.__iter__() <==> iter(x)\n",
      "     |  \n",
      "     |  next(...)\n",
      "     |      x.next() -> the next value, or raise StopIteration\n",
      "    \n",
      "    class Edge(__builtin__.object)\n",
      "     |  Class representing a single edge in a graph.\n",
      "     |  \n",
      "     |  The edge is referenced by its index, so if the underlying graph\n",
      "     |  changes, the semantics of the edge object might change as well\n",
      "     |  (if the edge indices are altered in the original graph).\n",
      "     |  \n",
      "     |  The attributes of the edge can be accessed by using the edge\n",
      "     |  as a hash:\n",
      "     |  \n",
      "     |    >>> e[\"weight\"] = 2                  #doctest: +SKIP\n",
      "     |    >>> print e[\"weight\"]                #doctest: +SKIP\n",
      "     |    2\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __delitem__(...)\n",
      "     |      x.__delitem__(y) <==> del x[y]\n",
      "     |  \n",
      "     |  __eq__(...)\n",
      "     |      x.__eq__(y) <==> x==y\n",
      "     |  \n",
      "     |  __ge__(...)\n",
      "     |      x.__ge__(y) <==> x>=y\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __gt__(...)\n",
      "     |      x.__gt__(y) <==> x>y\n",
      "     |  \n",
      "     |  __hash__(...)\n",
      "     |      x.__hash__() <==> hash(x)\n",
      "     |  \n",
      "     |  __le__(...)\n",
      "     |      x.__le__(y) <==> x<=y\n",
      "     |  \n",
      "     |  __len__(...)\n",
      "     |      x.__len__() <==> len(x)\n",
      "     |  \n",
      "     |  __lt__(...)\n",
      "     |      x.__lt__(y) <==> x<y\n",
      "     |  \n",
      "     |  __ne__(...)\n",
      "     |      x.__ne__(y) <==> x!=y\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setitem__(...)\n",
      "     |      x.__setitem__(i, y) <==> x[i]=y\n",
      "     |  \n",
      "     |  attribute_names(...)\n",
      "     |      attribute_names() -> list\n",
      "     |      \n",
      "     |      Returns the list of edge attribute names\n",
      "     |  \n",
      "     |  attributes(...)\n",
      "     |      attributes() -> dict\n",
      "     |      \n",
      "     |      Returns a dict of attribute names and values for the edge\n",
      "     |  \n",
      "     |  count_multiple(...)\n",
      "     |      Proxy method to L{Graph.count_multiple()}\n",
      "     |      \n",
      "     |      This method calls the count_multiple method of the L{Graph} class with this edge as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.count_multiple() for details.\n",
      "     |  \n",
      "     |  delete(...)\n",
      "     |      Proxy method to L{Graph.delete_edges()}\n",
      "     |      \n",
      "     |      This method calls the delete_edges method of the L{Graph} class with this edge as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.delete_edges() for details.\n",
      "     |  \n",
      "     |  is_loop(...)\n",
      "     |      Proxy method to L{Graph.is_loop()}\n",
      "     |      \n",
      "     |      This method calls the is_loop method of the L{Graph} class with this edge as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_loop() for details.\n",
      "     |  \n",
      "     |  is_multiple(...)\n",
      "     |      Proxy method to L{Graph.is_multiple()}\n",
      "     |      \n",
      "     |      This method calls the is_multiple method of the L{Graph} class with this edge as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_multiple() for details.\n",
      "     |  \n",
      "     |  is_mutual(...)\n",
      "     |      Proxy method to L{Graph.is_mutual()}\n",
      "     |      \n",
      "     |      This method calls the is_mutual method of the L{Graph} class with this edge as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_mutual() for details.\n",
      "     |  \n",
      "     |  update_attributes(...)\n",
      "     |      update_attributes(E, **F) -> None\n",
      "     |      \n",
      "     |      Updates the attributes of the edge from dict/iterable E and F.\n",
      "     |      \n",
      "     |      If E has a C{keys()} method, it does: C{for k in E: self[k] = E[k]}.\n",
      "     |      If E lacks a C{keys()} method, it does: C{for (k, v) in E: self[k] = v}.\n",
      "     |      In either case, this is followed by: C{for k in F: self[k] = F[k]}.\n",
      "     |      \n",
      "     |      This method thus behaves similarly to the C{update()} method of Python\n",
      "     |      dictionaries.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  graph\n",
      "     |      The graph the edge belongs to\n",
      "     |  \n",
      "     |  index\n",
      "     |      Index of this edge\n",
      "     |  \n",
      "     |  source\n",
      "     |      Source node index of this edge\n",
      "     |  \n",
      "     |  target\n",
      "     |      Target node index of this edge\n",
      "     |  \n",
      "     |  tuple\n",
      "     |      Source and target node index of this edge as a tuple\n",
      "    \n",
      "    class EdgeSeq(igraph.core.EdgeSeq)\n",
      "     |  Class representing a sequence of edges in the graph.\n",
      "     |  \n",
      "     |  This class is most easily accessed by the C{es} field of the\n",
      "     |  L{Graph} object, which returns an ordered sequence of all edges in\n",
      "     |  the graph. The edge sequence can be refined by invoking the\n",
      "     |  L{EdgeSeq.select()} method. L{EdgeSeq.select()} can also be\n",
      "     |  accessed by simply calling the L{EdgeSeq} object.\n",
      "     |  \n",
      "     |  An alternative way to create an edge sequence referring to a given\n",
      "     |  graph is to use the constructor directly:\n",
      "     |  \n",
      "     |    >>> g = Graph.Full(3)\n",
      "     |    >>> es = EdgeSeq(g)\n",
      "     |    >>> restricted_es = EdgeSeq(g, [0, 1])\n",
      "     |  \n",
      "     |  The individual edges can be accessed by indexing the edge sequence\n",
      "     |  object. It can be used as an iterable as well, or even in a list\n",
      "     |  comprehension:\n",
      "     |  \n",
      "     |    >>> g=Graph.Full(3)\n",
      "     |    >>> for e in g.es:\n",
      "     |    ...   print e.tuple\n",
      "     |    ...\n",
      "     |    (0, 1)\n",
      "     |    (0, 2)\n",
      "     |    (1, 2)\n",
      "     |    >>> [max(e.tuple) for e in g.es]\n",
      "     |    [1, 2, 2]\n",
      "     |  \n",
      "     |  The edge sequence can also be used as a dictionary where the keys are the\n",
      "     |  attribute names. The values corresponding to the keys are the values\n",
      "     |  of the given attribute of every edge in the graph:\n",
      "     |  \n",
      "     |    >>> g=Graph.Full(3)\n",
      "     |    >>> for idx, e in enumerate(g.es):\n",
      "     |    ...   e[\"weight\"] = idx*(idx+1)\n",
      "     |    ...\n",
      "     |    >>> g.es[\"weight\"]\n",
      "     |    [0, 2, 6]\n",
      "     |    >>> g.es[\"weight\"] = range(3)\n",
      "     |    >>> g.es[\"weight\"]\n",
      "     |    [0, 1, 2]\n",
      "     |  \n",
      "     |  If you specify a sequence that is shorter than the number of edges in\n",
      "     |  the EdgeSeq, the sequence is reused:\n",
      "     |  \n",
      "     |    >>> g = Graph.Tree(7, 2)\n",
      "     |    >>> g.es[\"color\"] = [\"red\", \"green\"]\n",
      "     |    >>> g.es[\"color\"]\n",
      "     |    ['red', 'green', 'red', 'green', 'red', 'green']\n",
      "     |  \n",
      "     |  You can even pass a single string or integer, it will be considered as a\n",
      "     |  sequence of length 1:\n",
      "     |  \n",
      "     |    >>> g.es[\"color\"] = \"red\"\n",
      "     |    >>> g.es[\"color\"]\n",
      "     |    ['red', 'red', 'red', 'red', 'red', 'red']\n",
      "     |  \n",
      "     |  Some methods of the edge sequences are simply proxy methods to the\n",
      "     |  corresponding methods in the L{Graph} object. One such example is\n",
      "     |  L{EdgeSeq.is_multiple()}:\n",
      "     |  \n",
      "     |    >>> g=Graph(3, [(0,1), (1,0), (1,2)])\n",
      "     |    >>> g.es.is_multiple()\n",
      "     |    [False, True, False]\n",
      "     |    >>> g.es.is_multiple() == g.is_multiple()\n",
      "     |    True\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      EdgeSeq\n",
      "     |      igraph.core.EdgeSeq\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Shorthand notation to select()\n",
      "     |      \n",
      "     |      This method simply passes all its arguments to L{EdgeSeq.select()}.\n",
      "     |  \n",
      "     |  attributes(self)\n",
      "     |      Returns the list of all the edge attributes in the graph\n",
      "     |      associated to this edge sequence.\n",
      "     |  \n",
      "     |  count_multiple(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.count_multiple()}\n",
      "     |      \n",
      "     |      This method calls the C{count_multiple()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.count_multiple() for details.\n",
      "     |  \n",
      "     |  delete = delete_edges(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.delete_edges()}\n",
      "     |      \n",
      "     |      This method calls the C{delete_edges()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.delete_edges() for details.\n",
      "     |  \n",
      "     |  edge_betweenness(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.edge_betweenness()}\n",
      "     |      \n",
      "     |      This method calls the C{edge_betweenness()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.edge_betweenness() for details.\n",
      "     |  \n",
      "     |  find(self, *args, **kwds)\n",
      "     |      Returns the first edge of the edge sequence that matches some\n",
      "     |      criteria.\n",
      "     |      \n",
      "     |      The selection criteria are equal to the ones allowed by L{VertexSeq.select}.\n",
      "     |      See L{VertexSeq.select} for more details.\n",
      "     |      \n",
      "     |      For instance, to find the first edge with weight larger than 5 in graph C{g}:\n",
      "     |      \n",
      "     |          >>> g.es.find(weight_gt=5)           #doctest:+SKIP\n",
      "     |  \n",
      "     |  is_loop(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.is_loop()}\n",
      "     |      \n",
      "     |      This method calls the C{is_loop()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_loop() for details.\n",
      "     |  \n",
      "     |  is_multiple(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.is_multiple()}\n",
      "     |      \n",
      "     |      This method calls the C{is_multiple()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_multiple() for details.\n",
      "     |  \n",
      "     |  is_mutual(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.is_mutual()}\n",
      "     |      \n",
      "     |      This method calls the C{is_mutual()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_mutual() for details.\n",
      "     |  \n",
      "     |  select(self, *args, **kwds)\n",
      "     |      Selects a subset of the edge sequence based on some criteria\n",
      "     |      \n",
      "     |      The selection criteria can be specified by the positional and the\n",
      "     |      keyword arguments. Positional arguments are always processed before\n",
      "     |      keyword arguments.\n",
      "     |      \n",
      "     |        - If the first positional argument is C{None}, an empty sequence is\n",
      "     |          returned.\n",
      "     |      \n",
      "     |        - If the first positional argument is a callable object, the object\n",
      "     |          will be called for every edge in the sequence. If it returns\n",
      "     |          C{True}, the edge will be included, otherwise it will\n",
      "     |          be excluded.\n",
      "     |      \n",
      "     |        - If the first positional argument is an iterable, it must return\n",
      "     |          integers and they will be considered as indices of the current\n",
      "     |          edge set (NOT the whole edge set of the graph -- the\n",
      "     |          difference matters when one filters an edge set that has\n",
      "     |          already been filtered by a previous invocation of\n",
      "     |          L{EdgeSeq.select()}. In this case, the indices do not refer\n",
      "     |          directly to the edges of the graph but to the elements of\n",
      "     |          the filtered edge sequence.\n",
      "     |      \n",
      "     |        - If the first positional argument is an integer, all remaining\n",
      "     |          arguments are expected to be integers. They are considered as\n",
      "     |          indices of the current edge set again.\n",
      "     |      \n",
      "     |      Keyword arguments can be used to filter the edges based on their\n",
      "     |      attributes and properties. The name of the keyword specifies the name\n",
      "     |      of the attribute and the filtering operator, they should be\n",
      "     |      concatenated by an underscore (C{_}) character. Attribute names can\n",
      "     |      also contain underscores, but operator names don't, so the operator is\n",
      "     |      always the largest trailing substring of the keyword name that does not\n",
      "     |      contain an underscore. Possible operators are:\n",
      "     |      \n",
      "     |        - C{eq}: equal to\n",
      "     |      \n",
      "     |        - C{ne}: not equal to\n",
      "     |      \n",
      "     |        - C{lt}: less than\n",
      "     |      \n",
      "     |        - C{gt}: greater than\n",
      "     |      \n",
      "     |        - C{le}: less than or equal to\n",
      "     |      \n",
      "     |        - C{ge}: greater than or equal to\n",
      "     |      \n",
      "     |        - C{in}: checks if the value of an attribute is in a given list\n",
      "     |      \n",
      "     |        - C{notin}: checks if the value of an attribute is not in a given\n",
      "     |          list\n",
      "     |      \n",
      "     |      For instance, if you want to filter edges with a numeric C{weight}\n",
      "     |      property larger than 50, you have to write:\n",
      "     |      \n",
      "     |        >>> g.es.select(weight_gt=50)            #doctest: +SKIP\n",
      "     |      \n",
      "     |      Similarly, to filter edges whose C{type} is in a list of predefined\n",
      "     |      types:\n",
      "     |      \n",
      "     |        >>> list_of_types = [\"inhibitory\", \"excitatory\"]\n",
      "     |        >>> g.es.select(type_in=list_of_types)   #doctest: +SKIP\n",
      "     |      \n",
      "     |      If the operator is omitted, it defaults to C{eq}. For instance, the\n",
      "     |      following selector selects edges whose C{type} property is\n",
      "     |      C{intracluster}:\n",
      "     |      \n",
      "     |        >>> g.es.select(type=\"intracluster\")     #doctest: +SKIP\n",
      "     |      \n",
      "     |      In the case of an unknown operator, it is assumed that the\n",
      "     |      recognized operator is part of the attribute name and the actual\n",
      "     |      operator is C{eq}.\n",
      "     |      \n",
      "     |      Keyword arguments are treated specially if they start with an\n",
      "     |      underscore (C{_}). These are not real attributes but refer to specific\n",
      "     |      properties of the edges, e.g., their centrality.  The rules are as\n",
      "     |      follows:\n",
      "     |      \n",
      "     |        1. C{_source} or {_from} means the source vertex of an edge.\n",
      "     |      \n",
      "     |        2. C{_target} or {_to} means the target vertex of an edge.\n",
      "     |      \n",
      "     |        3. C{_within} ignores the operator and checks whether both endpoints\n",
      "     |           of the edge lie within a specified set.\n",
      "     |      \n",
      "     |        4. C{_between} ignores the operator and checks whether I{one}\n",
      "     |           endpoint of the edge lies within a specified set and the I{other}\n",
      "     |           endpoint lies within another specified set. The two sets must be\n",
      "     |           given as a tuple.\n",
      "     |      \n",
      "     |        5. Otherwise, the rest of the name is interpreted as a method of the\n",
      "     |           L{Graph} object. This method is called with the edge sequence as\n",
      "     |           its first argument (all others left at default values) and edges\n",
      "     |           are filtered according to the value returned by the method.\n",
      "     |      \n",
      "     |      For instance, if you want to exclude edges with a betweenness\n",
      "     |      centrality less than 2:\n",
      "     |      \n",
      "     |        >>> g = Graph.Famous(\"zachary\")\n",
      "     |        >>> excl = g.es.select(_edge_betweenness_ge = 2)\n",
      "     |      \n",
      "     |      To select edges originating from vertices 2 and 4:\n",
      "     |      \n",
      "     |        >>> edges = g.es.select(_source_in = [2, 4])\n",
      "     |      \n",
      "     |      To select edges lying entirely within the subgraph spanned by vertices\n",
      "     |      2, 3, 4 and 7:\n",
      "     |      \n",
      "     |        >>> edges = g.es.select(_within = [2, 3, 4, 7])\n",
      "     |      \n",
      "     |      To select edges with one endpoint in the vertex set containing vertices\n",
      "     |      2, 3, 4 and 7 and the other endpoint in the vertex set containing\n",
      "     |      vertices 8 and 9:\n",
      "     |      \n",
      "     |        >>> edges = g.es.select(_between = ([2, 3, 4, 7], [8, 9]))\n",
      "     |      \n",
      "     |      For properties that take a long time to be computed (e.g., betweenness\n",
      "     |      centrality for large graphs), it is advised to calculate the values\n",
      "     |      in advance and store it in a graph attribute. The same applies when\n",
      "     |      you are selecting based on the same property more than once in the\n",
      "     |      same C{select()} call to avoid calculating it twice unnecessarily.\n",
      "     |      For instance, the following would calculate betweenness centralities\n",
      "     |      twice:\n",
      "     |      \n",
      "     |        >>> edges = g.es.select(_edge_betweenness_gt=10,       # doctest:+SKIP\n",
      "     |        ...                     _edge_betweenness_lt=30)\n",
      "     |      \n",
      "     |      It is advised to use this instead:\n",
      "     |      \n",
      "     |        >>> g.es[\"bs\"] = g.edge_betweenness()\n",
      "     |        >>> edges = g.es.select(bs_gt=10, bs_lt=30)\n",
      "     |      \n",
      "     |      @return: the new, filtered edge sequence\n",
      "     |  \n",
      "     |  subgraph = subgraph_edges(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.subgraph_edges()}\n",
      "     |      \n",
      "     |      This method calls the C{subgraph_edges()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.subgraph_edges() for details.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from igraph.core.EdgeSeq:\n",
      "     |  \n",
      "     |  __delitem__(...)\n",
      "     |      x.__delitem__(y) <==> del x[y]\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __init__(...)\n",
      "     |      x.__init__(...) initializes x; see help(type(x)) for signature\n",
      "     |  \n",
      "     |  __len__(...)\n",
      "     |      x.__len__() <==> len(x)\n",
      "     |  \n",
      "     |  __setitem__(...)\n",
      "     |      x.__setitem__(i, y) <==> x[i]=y\n",
      "     |  \n",
      "     |  attribute_names(...)\n",
      "     |      attribute_names() -> list\n",
      "     |      \n",
      "     |      Returns the attribute name list of the graph's edges\n",
      "     |  \n",
      "     |  get_attribute_values(...)\n",
      "     |      get_attribute_values(attrname) -> list\n",
      "     |      \n",
      "     |      Returns the value of a given edge attribute for all edges.\n",
      "     |      \n",
      "     |      @param attrname: the name of the attribute\n",
      "     |  \n",
      "     |  is_all(...)\n",
      "     |      is_all() -> bool\n",
      "     |      \n",
      "     |      Returns whether the edge sequence contains all the edges exactly once, in\n",
      "     |      the order of their edge IDs.\n",
      "     |      \n",
      "     |      This is used for optimizations in some of the edge selector routines.\n",
      "     |  \n",
      "     |  set_attribute_values(...)\n",
      "     |      set_attribute_values(attrname, values) -> list\n",
      "     |      Sets the value of a given edge attribute for all vertices\n",
      "     |      @param attrname: the name of the attribute\n",
      "     |      @param values: the new attribute values in a list\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from igraph.core.EdgeSeq:\n",
      "     |  \n",
      "     |  graph\n",
      "     |      The graph the edge sequence belongs to\n",
      "     |  \n",
      "     |  indices\n",
      "     |      The edge indices in this edge sequence\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from igraph.core.EdgeSeq:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "    \n",
      "    class Graph(Graph)\n",
      "     |  Generic graph.\n",
      "     |  \n",
      "     |  This class is built on top of L{GraphBase}, so the order of the\n",
      "     |  methods in the Epydoc documentation is a little bit obscure:\n",
      "     |  inherited methods come after the ones implemented directly in the\n",
      "     |  subclass. L{Graph} provides many functions that L{GraphBase} does not,\n",
      "     |  mostly because these functions are not speed critical and they were\n",
      "     |  easier to implement in Python than in pure C. An example is the\n",
      "     |  attribute handling in the constructor: the constructor of L{Graph}\n",
      "     |  accepts three dictionaries corresponding to the graph, vertex and edge\n",
      "     |  attributes while the constructor of L{GraphBase} does not. This extension\n",
      "     |  was needed to make L{Graph} serializable through the C{pickle} module.\n",
      "     |  L{Graph} also overrides some functions from L{GraphBase} to provide a\n",
      "     |  more convenient interface; e.g., layout functions return a L{Layout}\n",
      "     |  instance from L{Graph} instead of a list of coordinate pairs.\n",
      "     |  \n",
      "     |  Graphs can also be indexed by strings or pairs of vertex indices or vertex\n",
      "     |  names.  When a graph is indexed by a string, the operation translates to\n",
      "     |  the retrieval, creation, modification or deletion of a graph attribute:\n",
      "     |  \n",
      "     |    >>> g = Graph.Full(3)\n",
      "     |    >>> g[\"name\"] = \"Triangle graph\"\n",
      "     |    >>> g[\"name\"]\n",
      "     |    'Triangle graph'\n",
      "     |    >>> del g[\"name\"]\n",
      "     |  \n",
      "     |  When a graph is indexed by a pair of vertex indices or names, the graph\n",
      "     |  itself is treated as an adjacency matrix and the corresponding cell of\n",
      "     |  the matrix is returned:\n",
      "     |  \n",
      "     |    >>> g = Graph.Full(3)\n",
      "     |    >>> g.vs[\"name\"] = [\"A\", \"B\", \"C\"]\n",
      "     |    >>> g[1, 2]\n",
      "     |    1\n",
      "     |    >>> g[\"A\", \"B\"]\n",
      "     |    1\n",
      "     |    >>> g[\"A\", \"B\"] = 0\n",
      "     |    >>> g.ecount()\n",
      "     |    2\n",
      "     |  \n",
      "     |  Assigning values different from zero or one to the adjacency matrix will\n",
      "     |  be translated to one, unless the graph is weighted, in which case the\n",
      "     |  numbers will be treated as weights:\n",
      "     |  \n",
      "     |    >>> g.is_weighted()\n",
      "     |    False\n",
      "     |    >>> g[\"A\", \"B\"] = 2\n",
      "     |    >>> g[\"A\", \"B\"]\n",
      "     |    1\n",
      "     |    >>> g.es[\"weight\"] = 1.0\n",
      "     |    >>> g.is_weighted()\n",
      "     |    True\n",
      "     |    >>> g[\"A\", \"B\"] = 2\n",
      "     |    >>> g[\"A\", \"B\"]\n",
      "     |    2\n",
      "     |    >>> g.es[\"weight\"]\n",
      "     |    [1.0, 1.0, 2]\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      Graph\n",
      "     |      Graph\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __add__(self, other)\n",
      "     |      Copies the graph and extends the copy depending on the type of\n",
      "     |      the other object given.\n",
      "     |      \n",
      "     |      @param other: if it is an integer, the copy is extended by the given\n",
      "     |        number of vertices. If it is a string, the copy is extended by a\n",
      "     |        single vertex whose C{name} attribute will be equal to the given\n",
      "     |        string. If it is a tuple with two elements, the copy\n",
      "     |        is extended by a single edge. If it is a list of tuples, the copy\n",
      "     |        is extended by multiple edges. If it is a L{Graph}, a disjoint\n",
      "     |        union is performed.\n",
      "     |  \n",
      "     |  __coerce__(self, other)\n",
      "     |      Coercion rules.\n",
      "     |      \n",
      "     |      This method is needed to allow the graph to react to additions\n",
      "     |      with lists, tuples, integers, strings, vertices, edges and so on.\n",
      "     |  \n",
      "     |  __iadd__(self, other)\n",
      "     |      In-place addition (disjoint union).\n",
      "     |      \n",
      "     |      @see: L{__add__}\n",
      "     |  \n",
      "     |  __init__(self, *args, **kwds)\n",
      "     |      __init__(n=0, edges=None, directed=False, graph_attrs=None,\n",
      "     |      vertex_attrs=None, edge_attrs=None)\n",
      "     |      \n",
      "     |      Constructs a graph from scratch.\n",
      "     |      \n",
      "     |      @keyword n: the number of vertices. Can be omitted, the default is\n",
      "     |        zero. Note that if the edge list contains vertices with indexes\n",
      "     |        larger than or equal to M{m}, then the number of vertices will\n",
      "     |        be adjusted accordingly.\n",
      "     |      @keyword edges: the edge list where every list item is a pair of\n",
      "     |        integers. If any of the integers is larger than M{n-1}, the number\n",
      "     |        of vertices is adjusted accordingly. C{None} means no edges.\n",
      "     |      @keyword directed: whether the graph should be directed\n",
      "     |      @keyword graph_attrs: the attributes of the graph as a dictionary.\n",
      "     |      @keyword vertex_attrs: the attributes of the vertices as a dictionary.\n",
      "     |        Every dictionary value must be an iterable with exactly M{n} items.\n",
      "     |      @keyword edge_attrs: the attributes of the edges as a dictionary. Every\n",
      "     |        dictionary value must be an iterable with exactly M{m} items where\n",
      "     |        M{m} is the number of edges.\n",
      "     |  \n",
      "     |  __isub__(self, other)\n",
      "     |      In-place subtraction (difference).\n",
      "     |      \n",
      "     |      @see: L{__sub__}\n",
      "     |  \n",
      "     |  __mul__(self, other)\n",
      "     |      Copies exact replicas of the original graph an arbitrary number of\n",
      "     |      times.\n",
      "     |      \n",
      "     |      @param other: if it is an integer, multiplies the graph by creating the\n",
      "     |        given number of identical copies and taking the disjoint union of\n",
      "     |        them.\n",
      "     |  \n",
      "     |  __nonzero__(self)\n",
      "     |      Returns True if the graph has at least one vertex, False otherwise.\n",
      "     |  \n",
      "     |  __plot__(self, context, bbox, palette, *args, **kwds)\n",
      "     |      Plots the graph to the given Cairo context in the given bounding box\n",
      "     |      \n",
      "     |      The visual style of vertices and edges can be modified at three\n",
      "     |      places in the following order of precedence (lower indices override\n",
      "     |      higher indices):\n",
      "     |      \n",
      "     |        1. Keyword arguments of this function (or of L{plot()} which is\n",
      "     |           passed intact to C{Graph.__plot__()}.\n",
      "     |      \n",
      "     |        2. Vertex or edge attributes, specified later in the list of\n",
      "     |           keyword arguments.\n",
      "     |      \n",
      "     |        3. igraph-wide plotting defaults (see\n",
      "     |           L{igraph.config.Configuration})\n",
      "     |      \n",
      "     |        4. Built-in defaults.\n",
      "     |      \n",
      "     |      E.g., if the C{vertex_size} keyword attribute is not present,\n",
      "     |      but there exists a vertex attribute named C{size}, the sizes of\n",
      "     |      the vertices will be specified by that attribute.\n",
      "     |      \n",
      "     |      Besides the usual self-explanatory plotting parameters (C{context},\n",
      "     |      C{bbox}, C{palette}), it accepts the following keyword arguments:\n",
      "     |      \n",
      "     |        - C{autocurve}: whether to use curves instead of straight lines for\n",
      "     |          multiple edges on the graph plot. This argument may be C{True}\n",
      "     |          or C{False}; when omitted, C{True} is assumed for graphs with\n",
      "     |          less than 10.000 edges and C{False} otherwise.\n",
      "     |      \n",
      "     |        - C{drawer_factory}: a subclass of L{AbstractCairoGraphDrawer}\n",
      "     |          which will be used to draw the graph. You may also provide\n",
      "     |          a function here which takes two arguments: the Cairo context\n",
      "     |          to draw on and a bounding box (an instance of L{BoundingBox}).\n",
      "     |          If this keyword argument is missing, igraph will use the\n",
      "     |          default graph drawer which should be suitable for most purposes.\n",
      "     |          It is safe to omit this keyword argument unless you need to use\n",
      "     |          a specific graph drawer.\n",
      "     |      \n",
      "     |        - C{keep_aspect_ratio}: whether to keep the aspect ratio of the layout\n",
      "     |          that igraph calculates to place the nodes. C{True} means that the\n",
      "     |          layout will be scaled proportionally to fit into the bounding box\n",
      "     |          where the graph is to be drawn but the aspect ratio will be kept\n",
      "     |          the same (potentially leaving empty space next to, below or above\n",
      "     |          the graph). C{False} means that the layout will be scaled independently\n",
      "     |          along the X and Y axis in order to fill the entire bounding box.\n",
      "     |          The default is C{False}.\n",
      "     |      \n",
      "     |        - C{layout}: the layout to be used. If not an instance of\n",
      "     |          L{Layout}, it will be passed to L{Graph.layout} to calculate\n",
      "     |          the layout. Note that if you want a deterministic layout that\n",
      "     |          does not change with every plot, you must either use a\n",
      "     |          deterministic layout function (like L{Graph.layout_circle}) or\n",
      "     |          calculate the layout in advance and pass a L{Layout} object here.\n",
      "     |      \n",
      "     |        - C{margin}: the top, right, bottom, left margins as a 4-tuple.\n",
      "     |          If it has less than 4 elements or is a single float, the elements\n",
      "     |          will be re-used until the length is at least 4.\n",
      "     |      \n",
      "     |        - C{mark_groups}: whether to highlight some of the vertex groups by\n",
      "     |          colored polygons. This argument can be one of the following:\n",
      "     |      \n",
      "     |            - C{False}: no groups will be highlighted\n",
      "     |      \n",
      "     |            - A dict mapping tuples of vertex indices to color names.\n",
      "     |              The given vertex groups will be highlighted by the given\n",
      "     |              colors.\n",
      "     |      \n",
      "     |            - A list containing pairs or an iterable yielding pairs, where\n",
      "     |            the first element of each pair is a list of vertex indices and\n",
      "     |            the second element is a color.\n",
      "     |      \n",
      "     |          In place of lists of vertex indices, you may also use L{VertexSeq}\n",
      "     |          instances.\n",
      "     |      \n",
      "     |          In place of color names, you may also use color indices into the\n",
      "     |          current palette. C{None} as a color name will mean that the\n",
      "     |          corresponding group is ignored.\n",
      "     |      \n",
      "     |        - C{vertex_size}: size of the vertices. The corresponding vertex\n",
      "     |          attribute is called C{size}. The default is 10. Vertex sizes\n",
      "     |          are measured in the unit of the Cairo context on which igraph\n",
      "     |          is drawing.\n",
      "     |      \n",
      "     |        - C{vertex_color}: color of the vertices. The corresponding vertex\n",
      "     |          attribute is C{color}, the default is red.  Colors can be\n",
      "     |          specified either by common X11 color names (see the source\n",
      "     |          code of L{igraph.drawing.colors} for a list of known colors), by\n",
      "     |          3-tuples of floats (ranging between 0 and 255 for the R, G and\n",
      "     |          B components), by CSS-style string specifications (C{#rrggbb})\n",
      "     |          or by integer color indices of the specified palette.\n",
      "     |      \n",
      "     |        - C{vertex_frame_color}: color of the frame (i.e. stroke) of the\n",
      "     |          vertices. The corresponding vertex attribute is C{frame_color},\n",
      "     |          the default is black. See C{vertex_color} for the possible ways\n",
      "     |          of specifying a color.\n",
      "     |      \n",
      "     |        - C{vertex_frame_width}: the width of the frame (i.e. stroke) of the\n",
      "     |          vertices. The corresponding vertex attribute is C{frame_width}.\n",
      "     |          The default is 1. Vertex frame widths are measured in the unit of the\n",
      "     |          Cairo context on which igraph is drawing.\n",
      "     |      \n",
      "     |        - C{vertex_shape}: shape of the vertices. Alternatively it can\n",
      "     |          be specified by the C{shape} vertex attribute. Possibilities\n",
      "     |          are: C{square}, {circle}, {triangle}, {triangle-down} or\n",
      "     |          C{hidden}. See the source code of L{igraph.drawing} for a\n",
      "     |          list of alternative shape names that are also accepted and\n",
      "     |          mapped to these.\n",
      "     |      \n",
      "     |        - C{vertex_label}: labels drawn next to the vertices.\n",
      "     |          The corresponding vertex attribute is C{label}.\n",
      "     |      \n",
      "     |        - C{vertex_label_dist}: distance of the midpoint of the vertex\n",
      "     |          label from the center of the corresponding vertex.\n",
      "     |          The corresponding vertex attribute is C{label_dist}.\n",
      "     |      \n",
      "     |        - C{vertex_label_color}: color of the label. Corresponding\n",
      "     |          vertex attribute: C{label_color}. See C{vertex_color} for\n",
      "     |          color specification syntax.\n",
      "     |      \n",
      "     |        - C{vertex_label_size}: font size of the label, specified\n",
      "     |          in the unit of the Cairo context on which we are drawing.\n",
      "     |          Corresponding vertex attribute: C{label_size}.\n",
      "     |      \n",
      "     |        - C{vertex_label_angle}: the direction of the line connecting\n",
      "     |          the midpoint of the vertex with the midpoint of the label.\n",
      "     |          This can be used to position the labels relative to the\n",
      "     |          vertices themselves in conjunction with C{vertex_label_dist}.\n",
      "     |          Corresponding vertex attribute: C{label_angle}. The\n",
      "     |          default is C{-math.pi/2}.\n",
      "     |      \n",
      "     |        - C{vertex_order}: drawing order of the vertices. This must be\n",
      "     |          a list or tuple containing vertex indices; vertices are then\n",
      "     |          drawn according to this order.\n",
      "     |      \n",
      "     |        - C{vertex_order_by}: an alternative way to specify the drawing\n",
      "     |          order of the vertices; this attribute is interpreted as the name\n",
      "     |          of a vertex attribute, and vertices are drawn such that those\n",
      "     |          with a smaller attribute value are drawn first. You may also\n",
      "     |          reverse the order by passing a tuple here; the first element of\n",
      "     |          the tuple should be the name of the attribute, the second element\n",
      "     |          specifies whether the order is reversed (C{True}, C{False},\n",
      "     |          C{\"asc\"} and C{\"desc\"} are accepted values).\n",
      "     |      \n",
      "     |        - C{edge_color}: color of the edges. The corresponding edge\n",
      "     |          attribute is C{color}, the default is red. See C{vertex_color}\n",
      "     |          for color specification syntax.\n",
      "     |      \n",
      "     |        - C{edge_curved}: whether the edges should be curved. Positive\n",
      "     |          numbers correspond to edges curved in a counter-clockwise\n",
      "     |          direction, negative numbers correspond to edges curved in a\n",
      "     |          clockwise direction. Zero represents straight edges. C{True}\n",
      "     |          is interpreted as 0.5, C{False} is interpreted as 0. The\n",
      "     |          default is 0 which makes all the edges straight.\n",
      "     |      \n",
      "     |        - C{edge_width}: width of the edges in the default unit of the\n",
      "     |          Cairo context on which we are drawing. The corresponding\n",
      "     |          edge attribute is C{width}, the default is 1.\n",
      "     |      \n",
      "     |        - C{edge_arrow_size}: arrow size of the edges. The\n",
      "     |          corresponding edge attribute is C{arrow_size}, the default\n",
      "     |          is 1.\n",
      "     |      \n",
      "     |        - C{edge_arrow_width}: width of the arrowhead on the edge. The\n",
      "     |          corresponding edge attribute is C{arrow_width}, the default\n",
      "     |          is 1.\n",
      "     |      \n",
      "     |        - C{edge_order}: drawing order of the edges. This must be\n",
      "     |          a list or tuple containing edge indices; edges are then\n",
      "     |          drawn according to this order.\n",
      "     |      \n",
      "     |        - C{edge_order_by}: an alternative way to specify the drawing\n",
      "     |          order of the edges; this attribute is interpreted as the name\n",
      "     |          of an edge attribute, and edges are drawn such that those\n",
      "     |          with a smaller attribute value are drawn first. You may also\n",
      "     |          reverse the order by passing a tuple here; the first element of\n",
      "     |          the tuple should be the name of the attribute, the second element\n",
      "     |          specifies whether the order is reversed (C{True}, C{False},\n",
      "     |          C{\"asc\"} and C{\"desc\"} are accepted values).\n",
      "     |  \n",
      "     |  __reduce__(self)\n",
      "     |      Support for pickling.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Returns a string representation of the graph.\n",
      "     |      \n",
      "     |      Behind the scenes, this method constructs a L{GraphSummary}\n",
      "     |      instance and invokes its C{__str__} method with a verbosity of 1\n",
      "     |      and attribute printing turned off.\n",
      "     |      \n",
      "     |      See the documentation of L{GraphSummary} for more details about the\n",
      "     |      output.\n",
      "     |  \n",
      "     |  __sub__(self, other)\n",
      "     |      Removes the given object(s) from the graph\n",
      "     |      \n",
      "     |      @param other: if it is an integer, removes the vertex with the given\n",
      "     |        ID from the graph (note that the remaining vertices will get\n",
      "     |        re-indexed!). If it is a tuple, removes the given edge. If it is\n",
      "     |        a graph, takes the difference of the two graphs. Accepts\n",
      "     |        lists of integers or lists of tuples as well, but they can't be\n",
      "     |        mixed! Also accepts L{Edge} and L{EdgeSeq} objects.\n",
      "     |  \n",
      "     |  add_edge(self, source, target, **kwds)\n",
      "     |      add_edge(source, target, **kwds)\n",
      "     |      \n",
      "     |      Adds a single edge to the graph.\n",
      "     |      \n",
      "     |      Keyword arguments (except the source and target arguments) will be\n",
      "     |      assigned to the edge as attributes.\n",
      "     |      \n",
      "     |      @param source: the source vertex of the edge or its name.\n",
      "     |      @param target: the target vertex of the edge or its name.\n",
      "     |  \n",
      "     |  add_edges(self, es)\n",
      "     |      add_edges(es)\n",
      "     |      \n",
      "     |      Adds some edges to the graph.\n",
      "     |      \n",
      "     |      @param es: the list of edges to be added. Every edge is represented\n",
      "     |        with a tuple containing the vertex IDs or names of the two\n",
      "     |        endpoints. Vertices are enumerated from zero.\n",
      "     |  \n",
      "     |  add_vertex(self, name=None, **kwds)\n",
      "     |      add_vertex(name=None, **kwds)\n",
      "     |      \n",
      "     |      Adds a single vertex to the graph. Keyword arguments will be assigned\n",
      "     |      as vertex attributes. Note that C{name} as a keyword argument is treated\n",
      "     |      specially; if a graph has C{name} as a vertex attribute, it allows one\n",
      "     |      to refer to vertices by their names in most places where igraph expects\n",
      "     |      a vertex ID.\n",
      "     |  \n",
      "     |  add_vertices(self, n)\n",
      "     |      add_vertices(n)\n",
      "     |      \n",
      "     |      Adds some vertices to the graph.\n",
      "     |      \n",
      "     |      @param n: the number of vertices to be added, or the name of a single\n",
      "     |        vertex to be added, or an iterable of strings, each corresponding to the\n",
      "     |        name of a vertex to be added. Names will be assigned to the C{name}\n",
      "     |        vertex attribute.\n",
      "     |  \n",
      "     |  adhesion = edge_connectivity(...)\n",
      "     |      edge_connectivity(source=-1, target=-1, checks=True)\n",
      "     |      \n",
      "     |      Calculates the edge connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The edge connectivity between two given vertices is the number of edges\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of edge disjoint directed\n",
      "     |      paths between the vertices. The edge connectivity of the graph is the minimal\n",
      "     |      edge connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the edge connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall edge connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @return: the edge connectivity\n",
      "     |  \n",
      "     |  adjacent(self, *args, **kwds)\n",
      "     |      adjacent(vertex, mode=OUT)\n",
      "     |      \n",
      "     |      Returns the edges a given vertex is incident on.\n",
      "     |      \n",
      "     |      @deprecated: replaced by L{Graph.incident()} since igraph 0.6\n",
      "     |  \n",
      "     |  all_st_cuts(self, source, target)\n",
      "     |      Returns all the cuts between the source and target vertices in a\n",
      "     |      directed graph.\n",
      "     |      \n",
      "     |      This function lists all edge-cuts between a source and a target vertex.\n",
      "     |      Every cut is listed exactly once.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @return: a list of L{Cut} objects.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: JS Provan and DR Shier: A paradigm for listing (s,t)-cuts in\n",
      "     |        graphs. Algorithmica 15, 351--372, 1996.\n",
      "     |  \n",
      "     |  all_st_mincuts(self, source, target, capacity=None)\n",
      "     |      Returns all the mincuts between the source and target vertices in a\n",
      "     |      directed graph.\n",
      "     |      \n",
      "     |      This function lists all minimum edge-cuts between a source and a target\n",
      "     |      vertex.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @param capacity: the edge capacities (weights). If C{None}, all\n",
      "     |        edges have equal weight. May also be an attribute name.\n",
      "     |      @return: a list of L{Cut} objects.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: JS Provan and DR Shier: A paradigm for listing (s,t)-cuts in\n",
      "     |        graphs. Algorithmica 15, 351--372, 1996.\n",
      "     |  \n",
      "     |  alpha = independence_number(...)\n",
      "     |      independence_number()\n",
      "     |      \n",
      "     |      Returns the independence number of the graph.\n",
      "     |      \n",
      "     |      The independence number of the graph is the size of the largest\n",
      "     |      independent vertex set.\n",
      "     |      \n",
      "     |      @see: L{largest_independent_vertex_sets()} for the largest independent\n",
      "     |        vertex sets\n",
      "     |  \n",
      "     |  as_directed(self, *args, **kwds)\n",
      "     |      as_directed(*args, **kwds)\n",
      "     |      \n",
      "     |      Returns a directed copy of this graph. Arguments are passed on\n",
      "     |      to L{Graph.to_directed()} that is invoked on the copy.\n",
      "     |  \n",
      "     |  as_undirected(self, *args, **kwds)\n",
      "     |      as_undirected(*args, **kwds)\n",
      "     |      \n",
      "     |      Returns an undirected copy of this graph. Arguments are passed on\n",
      "     |      to L{Graph.to_undirected()} that is invoked on the copy.\n",
      "     |  \n",
      "     |  biconnected_components(self, return_articulation_points=False)\n",
      "     |      Calculates the biconnected components of the graph.\n",
      "     |      \n",
      "     |      @param return_articulation_points: whether to return the articulation\n",
      "     |        points as well\n",
      "     |      @return: a L{VertexCover} object describing the biconnected components,\n",
      "     |        and optionally the list of articulation points as well\n",
      "     |  \n",
      "     |  bipartite_projection(self, types='type', multiplicity=True, probe1=-1, which='both')\n",
      "     |      Projects a bipartite graph into two one-mode graphs. Edge directions\n",
      "     |      are ignored while projecting.\n",
      "     |      \n",
      "     |      Examples:\n",
      "     |      \n",
      "     |      >>> g = Graph.Full_Bipartite(10, 5)\n",
      "     |      >>> g1, g2 = g.bipartite_projection()\n",
      "     |      >>> g1.isomorphic(Graph.Full(10))\n",
      "     |      True\n",
      "     |      >>> g2.isomorphic(Graph.Full(5))\n",
      "     |      True\n",
      "     |      \n",
      "     |      @param types: an igraph vector containing the vertex types, or an\n",
      "     |        attribute name. Anything that evalulates to C{False} corresponds to\n",
      "     |        vertices of the first kind, everything else to the second kind.\n",
      "     |      @param multiplicity: if C{True}, then igraph keeps the multiplicity of\n",
      "     |        the edges in the projection in an edge attribute called C{\"weight\"}.\n",
      "     |        E.g., if there is an A-C-B and an A-D-B triplet in the bipartite\n",
      "     |        graph and there is no other X (apart from X=B and X=D) for which an\n",
      "     |        A-X-B triplet would exist in the bipartite graph, the multiplicity\n",
      "     |        of the A-B edge in the projection will be 2.\n",
      "     |      @param probe1: this argument can be used to specify the order of the\n",
      "     |        projections in the resulting list. If given and non-negative, then\n",
      "     |        it is considered as a vertex ID; the projection containing the\n",
      "     |        vertex will be the first one in the result.\n",
      "     |      @param which: this argument can be used to specify which of the two\n",
      "     |        projections should be returned if only one of them is needed. Passing\n",
      "     |        0 here means that only the first projection is returned, while 1 means\n",
      "     |        that only the second projection is returned. (Note that we use 0 and 1\n",
      "     |        because Python indexing is zero-based). C{False} is equivalent to 0 and\n",
      "     |        C{True} is equivalent to 1. Any other value means that both projections\n",
      "     |        will be returned in a tuple.\n",
      "     |      @return: a tuple containing the two projected one-mode graphs if C{which}\n",
      "     |        is not 1 or 2, or the projected one-mode graph specified by the\n",
      "     |        C{which} argument if its value is 0, 1, C{False} or C{True}.\n",
      "     |  \n",
      "     |  bipartite_projection_size(self, types='type', *args, **kwds)\n",
      "     |      bipartite_projection(types=\"type\")\n",
      "     |      \n",
      "     |      Calculates the number of vertices and edges in the bipartite\n",
      "     |      projections of this graph according to the specified vertex types.\n",
      "     |      This is useful if you have a bipartite graph and you want to estimate\n",
      "     |      the amount of memory you would need to calculate the projections\n",
      "     |      themselves.\n",
      "     |      \n",
      "     |      @param types: an igraph vector containing the vertex types, or an\n",
      "     |        attribute name. Anything that evalulates to C{False} corresponds to\n",
      "     |        vertices of the first kind, everything else to the second kind.\n",
      "     |      @return: a 4-tuple containing the number of vertices and edges in the\n",
      "     |        first projection, followed by the number of vertices and edges in the\n",
      "     |        second projection.\n",
      "     |  \n",
      "     |  blocks = biconnected_components(self, return_articulation_points=False)\n",
      "     |  \n",
      "     |  clusters(self, mode=2)\n",
      "     |      clusters(mode=STRONG)\n",
      "     |      \n",
      "     |      Calculates the (strong or weak) clusters (connected components) for\n",
      "     |      a given graph.\n",
      "     |      \n",
      "     |      @param mode: must be either C{STRONG} or C{WEAK}, depending on the\n",
      "     |        clusters being sought. Optional, defaults to C{STRONG}.\n",
      "     |      @return: a L{VertexClustering} object\n",
      "     |  \n",
      "     |  cohesion = vertex_connectivity(...)\n",
      "     |      vertex_connectivity(source=-1, target=-1, checks=True, neighbors=\"error\")\n",
      "     |      \n",
      "     |      Calculates the vertex connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The vertex connectivity between two given vertices is the number of vertices\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of vertex disjoint directed\n",
      "     |      paths between the vertices (apart from the source and target vertices of\n",
      "     |      course). The vertex connectivity of the graph is the minimal vertex\n",
      "     |      connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the vertex connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall vertex connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @param neighbors: tells igraph what to do when the two vertices are\n",
      "     |        connected. C{\"error\"} raises an exception, C{\"infinity\"} returns\n",
      "     |        infinity, C{\"ignore\"} ignores the edge.\n",
      "     |      @return: the vertex connectivity\n",
      "     |  \n",
      "     |  cohesive_blocks(self)\n",
      "     |      cohesive_blocks()\n",
      "     |      \n",
      "     |      Calculates the cohesive block structure of the graph.\n",
      "     |      \n",
      "     |      Cohesive blocking is a method of determining hierarchical subsets of graph\n",
      "     |      vertices based on their structural cohesion (i.e. vertex connectivity).\n",
      "     |      For a given graph G, a subset of its vertices S is said to be maximally\n",
      "     |      k-cohesive if there is no superset of S with vertex connectivity greater\n",
      "     |      than or equal to k. Cohesive blocking is a process through which, given a\n",
      "     |      k-cohesive set of vertices, maximally l-cohesive subsets are recursively\n",
      "     |      identified with l > k. Thus a hierarchy of vertex subsets is obtained in\n",
      "     |      the end, with the entire graph G at its root.\n",
      "     |      \n",
      "     |      @return: an instance of L{CohesiveBlocks}. See the documentation of\n",
      "     |        L{CohesiveBlocks} for more information.\n",
      "     |      @see: L{CohesiveBlocks}\n",
      "     |  \n",
      "     |  community_edge_betweenness(self, clusters=None, directed=True, weights=None)\n",
      "     |      Community structure based on the betweenness of the edges in the\n",
      "     |      network.\n",
      "     |      \n",
      "     |      The idea is that the betweenness of the edges connecting two\n",
      "     |      communities is typically high, as many of the shortest paths between\n",
      "     |      nodes in separate communities go through them. So we gradually remove\n",
      "     |      the edge with the highest betweenness and recalculate the betweennesses\n",
      "     |      after every removal. This way sooner or later the network falls of to\n",
      "     |      separate components. The result of the clustering will be represented\n",
      "     |      by a dendrogram.\n",
      "     |      \n",
      "     |      @param clusters: the number of clusters we would like to see. This\n",
      "     |        practically defines the \"level\" where we \"cut\" the dendrogram to\n",
      "     |        get the membership vector of the vertices. If C{None}, the dendrogram\n",
      "     |        is cut at the level which maximizes the modularity.\n",
      "     |      @param directed: whether the directionality of the edges should be\n",
      "     |        taken into account or not.\n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights.\n",
      "     |      @return: a L{VertexDendrogram} object, initally cut at the maximum\n",
      "     |        modularity or at the desired number of clusters.\n",
      "     |  \n",
      "     |  community_fastgreedy(self, weights=None)\n",
      "     |      Community structure based on the greedy optimization of modularity.\n",
      "     |      \n",
      "     |      This algorithm merges individual nodes into communities in a way that\n",
      "     |      greedily maximizes the modularity score of the graph. It can be proven\n",
      "     |      that if no merge can increase the current modularity score, the\n",
      "     |      algorithm can be stopped since no further increase can be achieved.\n",
      "     |      \n",
      "     |      This algorithm is said to run almost in linear time on sparse graphs.\n",
      "     |      \n",
      "     |      @param weights: edge attribute name or a list containing edge\n",
      "     |        weights\n",
      "     |      @return: an appropriate L{VertexDendrogram} object.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: A Clauset, MEJ Newman and C Moore: Finding community structure\n",
      "     |        in very large networks. Phys Rev E 70, 066111 (2004).\n",
      "     |  \n",
      "     |  community_infomap(self, edge_weights=None, vertex_weights=None, trials=10)\n",
      "     |      Finds the community structure of the network according to the Infomap\n",
      "     |      method of Martin Rosvall and Carl T. Bergstrom.\n",
      "     |      \n",
      "     |      @param edge_weights: name of an edge attribute or a list containing\n",
      "     |        edge weights.\n",
      "     |      @param vertex_weights: name of an vertex attribute or a list containing\n",
      "     |        vertex weights.\n",
      "     |      @param trials: the number of attempts to partition the network.\n",
      "     |      @return: an appropriate L{VertexClustering} object with an extra attribute\n",
      "     |        called C{codelength} that stores the code length determined by the\n",
      "     |        algorithm.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: M. Rosvall and C. T. Bergstrom: Maps of information flow reveal\n",
      "     |        community structure in complex networks, PNAS 105, 1118 (2008).\n",
      "     |        U{http://dx.doi.org/10.1073/pnas.0706851105},\n",
      "     |        U{http://arxiv.org/abs/0707.0609}.\n",
      "     |      @ref: M. Rosvall, D. Axelsson, and C. T. Bergstrom: The map equation,\n",
      "     |        Eur. Phys. J. Special Topics 178, 13 (2009).\n",
      "     |        U{http://dx.doi.org/10.1140/epjst/e2010-01179-1},\n",
      "     |        U{http://arxiv.org/abs/0906.1405}.\n",
      "     |  \n",
      "     |  community_label_propagation(self, weights=None, initial=None, fixed=None)\n",
      "     |      community_label_propagation(weights=None, initial=None, fixed=None)\n",
      "     |      \n",
      "     |      Finds the community structure of the graph according to the label\n",
      "     |      propagation method of Raghavan et al.\n",
      "     |      Initially, each vertex is assigned a different label. After that,\n",
      "     |      each vertex chooses the dominant label in its neighbourhood in each\n",
      "     |      iteration. Ties are broken randomly and the order in which the\n",
      "     |      vertices are updated is randomized before every iteration. The\n",
      "     |      algorithm ends when vertices reach a consensus.\n",
      "     |      Note that since ties are broken randomly, there is no guarantee that\n",
      "     |      the algorithm returns the same community structure after each run.\n",
      "     |      In fact, they frequently differ. See the paper of Raghavan et al\n",
      "     |      on how to come up with an aggregated community structure.\n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights\n",
      "     |      @param initial: name of a vertex attribute or a list containing\n",
      "     |        the initial vertex labels. Labels are identified by integers from\n",
      "     |        zero to M{n-1} where M{n} is the number of vertices. Negative\n",
      "     |        numbers may also be present in this vector, they represent unlabeled\n",
      "     |        vertices.\n",
      "     |      @param fixed: a list of booleans for each vertex. C{True} corresponds\n",
      "     |        to vertices whose labeling should not change during the algorithm.\n",
      "     |        It only makes sense if initial labels are also given. Unlabeled\n",
      "     |        vertices cannot be fixed.\n",
      "     |      @return: an appropriate L{VertexClustering} object.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Raghavan, U.N. and Albert, R. and Kumara, S. Near linear\n",
      "     |        time algorithm to detect community structures in large-scale\n",
      "     |        networks. Phys Rev E 76:036106, 2007.\n",
      "     |        U{http://arxiv.org/abs/0709.2938}.\n",
      "     |  \n",
      "     |  community_leading_eigenvector(self, clusters=None, weights=None, arpack_options=None)\n",
      "     |      community_leading_eigenvector(clusters=None, weights=None,\n",
      "     |        arpack_options=None)\n",
      "     |      \n",
      "     |      Newman's leading eigenvector method for detecting community structure.\n",
      "     |      This is the proper implementation of the recursive, divisive algorithm:\n",
      "     |      each split is done by maximizing the modularity regarding the\n",
      "     |      original network.\n",
      "     |      \n",
      "     |      @param clusters: the desired number of communities. If C{None}, the\n",
      "     |        algorithm tries to do as many splits as possible. Note that the\n",
      "     |        algorithm won't split a community further if the signs of the leading\n",
      "     |        eigenvector are all the same, so the actual number of discovered\n",
      "     |        communities can be less than the desired one.\n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @return: an appropriate L{VertexClustering} object.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: MEJ Newman: Finding community structure in networks using the\n",
      "     |      eigenvectors of matrices, arXiv:physics/0605087\n",
      "     |  \n",
      "     |  community_leading_eigenvector_naive(self, clusters=None, return_merges=False)\n",
      "     |      community_leading_eigenvector_naive(clusters=None,\n",
      "     |      return_merges=False)\n",
      "     |      \n",
      "     |      A naive implementation of Newman's eigenvector community structure\n",
      "     |      detection. This function splits the network into two components\n",
      "     |      according to the leading eigenvector of the modularity matrix and\n",
      "     |      then recursively takes the given number of steps by splitting the\n",
      "     |      communities as individual networks. This is not the correct way,\n",
      "     |      however, see the reference for explanation. Consider using the\n",
      "     |      correct L{community_leading_eigenvector} method instead.\n",
      "     |      \n",
      "     |      @param clusters: the desired number of communities. If C{None}, the\n",
      "     |        algorithm tries to do as many splits as possible. Note that the\n",
      "     |        algorithm won't split a community further if the signs of the leading\n",
      "     |        eigenvector are all the same, so the actual number of discovered\n",
      "     |        communities can be less than the desired one.\n",
      "     |      @param return_merges: whether the returned object should be a\n",
      "     |        dendrogram instead of a single clustering.\n",
      "     |      @return: an appropriate L{VertexClustering} or L{VertexDendrogram}\n",
      "     |        object.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: MEJ Newman: Finding community structure in networks using the\n",
      "     |      eigenvectors of matrices, arXiv:physics/0605087\n",
      "     |  \n",
      "     |  community_multilevel(self, weights=None, return_levels=False)\n",
      "     |      Community structure based on the multilevel algorithm of\n",
      "     |      Blondel et al.\n",
      "     |      \n",
      "     |      This is a bottom-up algorithm: initially every vertex belongs to a\n",
      "     |      separate community, and vertices are moved between communities\n",
      "     |      iteratively in a way that maximizes the vertices' local contribution\n",
      "     |      to the overall modularity score. When a consensus is reached (i.e. no\n",
      "     |      single move would increase the modularity score), every community in\n",
      "     |      the original graph is shrank to a single vertex (while keeping the\n",
      "     |      total weight of the adjacent edges) and the process continues on the\n",
      "     |      next level. The algorithm stops when it is not possible to increase\n",
      "     |      the modularity any more after shrinking the communities to vertices.\n",
      "     |      \n",
      "     |      This algorithm is said to run almost in linear time on sparse graphs.\n",
      "     |      \n",
      "     |      @param weights: edge attribute name or a list containing edge\n",
      "     |        weights\n",
      "     |      @param return_levels: if C{True}, the communities at each level are\n",
      "     |        returned in a list. If C{False}, only the community structure with\n",
      "     |        the best modularity is returned.\n",
      "     |      @return: a list of L{VertexClustering} objects, one corresponding to\n",
      "     |        each level (if C{return_levels} is C{True}), or a L{VertexClustering}\n",
      "     |        corresponding to the best modularity.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: VD Blondel, J-L Guillaume, R Lambiotte and E Lefebvre: Fast\n",
      "     |        unfolding of community hierarchies in large networks, J Stat Mech\n",
      "     |        P10008 (2008), http://arxiv.org/abs/0803.0476\n",
      "     |  \n",
      "     |  community_optimal_modularity(self, *args, **kwds)\n",
      "     |      Calculates the optimal modularity score of the graph and the\n",
      "     |      corresponding community structure.\n",
      "     |      \n",
      "     |      This function uses the GNU Linear Programming Kit to solve a large\n",
      "     |      integer optimization problem in order to find the optimal modularity\n",
      "     |      score and the corresponding community structure, therefore it is\n",
      "     |      unlikely to work for graphs larger than a few (less than a hundred)\n",
      "     |      vertices. Consider using one of the heuristic approaches instead if\n",
      "     |      you have such a large graph.\n",
      "     |      \n",
      "     |      @return: the calculated membership vector and the corresponding\n",
      "     |        modularity in a tuple.\n",
      "     |  \n",
      "     |  community_spinglass(self, *args, **kwds)\n",
      "     |      community_spinglass(weights=None, spins=25, parupdate=False,\n",
      "     |      start_temp=1, stop_temp=0.01, cool_fact=0.99, update_rule=\"config\",\n",
      "     |      gamma=1, implementation=\"orig\", lambda_=1)\n",
      "     |      \n",
      "     |      Finds the community structure of the graph according to the\n",
      "     |      spinglass community detection method of Reichardt & Bornholdt.\n",
      "     |      \n",
      "     |      @keyword weights: edge weights to be used. Can be a sequence or\n",
      "     |        iterable or even an edge attribute name.\n",
      "     |      @keyword spins: integer, the number of spins to use. This is the\n",
      "     |        upper limit for the number of communities. It is not a problem\n",
      "     |        to supply a (reasonably) big number here, in which case some\n",
      "     |        spin states will be unpopulated.\n",
      "     |      @keyword parupdate: whether to update the spins of the vertices in\n",
      "     |        parallel (synchronously) or not\n",
      "     |      @keyword start_temp: the starting temperature\n",
      "     |      @keyword stop_temp: the stop temperature\n",
      "     |      @keyword cool_fact: cooling factor for the simulated annealing\n",
      "     |      @keyword update_rule: specifies the null model of the simulation.\n",
      "     |        Possible values are C{\"config\"} (a random graph with the same\n",
      "     |        vertex degrees as the input graph) or C{\"simple\"} (a random\n",
      "     |        graph with the same number of edges)\n",
      "     |      @keyword gamma: the gamma argument of the algorithm, specifying the\n",
      "     |        balance between the importance of present and missing edges\n",
      "     |        within a community. The default value of 1.0 assigns equal\n",
      "     |        importance to both of them.\n",
      "     |      @keyword implementation: currently igraph contains two implementations\n",
      "     |        of the spinglass community detection algorithm. The faster\n",
      "     |        original implementation is the default. The other implementation\n",
      "     |        is able to take into account negative weights, this can be\n",
      "     |        chosen by setting C{implementation} to C{\"neg\"}\n",
      "     |      @keyword lambda_: the lambda argument of the algorithm, which\n",
      "     |        specifies the balance between the importance of present and missing\n",
      "     |        negatively weighted edges within a community. Smaller values of\n",
      "     |        lambda lead to communities with less negative intra-connectivity.\n",
      "     |        If the argument is zero, the algorithm reduces to a graph coloring\n",
      "     |        algorithm, using the number of spins as colors. This argument is\n",
      "     |        ignored if the original implementation is used. Note the underscore\n",
      "     |        at the end of the argument name; this is due to the fact that\n",
      "     |        lambda is a reserved keyword in Python.\n",
      "     |      @return: an appropriate L{VertexClustering} object.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Reichardt J and Bornholdt S: Statistical mechanics of\n",
      "     |        community detection. Phys Rev E 74:016110 (2006).\n",
      "     |        U{http://arxiv.org/abs/cond-mat/0603718}.\n",
      "     |      @ref: Traag VA and Bruggeman J: Community detection in networks\n",
      "     |        with positive and negative links. Phys Rev E 80:036115 (2009).\n",
      "     |        U{http://arxiv.org/abs/0811.2329}.\n",
      "     |  \n",
      "     |  community_walktrap(self, weights=None, steps=4)\n",
      "     |      Community detection algorithm of Latapy & Pons, based on random\n",
      "     |      walks.\n",
      "     |      \n",
      "     |      The basic idea of the algorithm is that short random walks tend to stay\n",
      "     |      in the same community. The result of the clustering will be represented\n",
      "     |      as a dendrogram.\n",
      "     |      \n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights\n",
      "     |      @param steps: length of random walks to perform\n",
      "     |      \n",
      "     |      @return: a L{VertexDendrogram} object, initially cut at the maximum\n",
      "     |        modularity.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Pascal Pons, Matthieu Latapy: Computing communities in large\n",
      "     |        networks using random walks, U{http://arxiv.org/abs/physics/0512106}.\n",
      "     |  \n",
      "     |  components = clusters(self, mode=2)\n",
      "     |  \n",
      "     |  count_automorphisms_vf2(self, color=None, edge_color=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      Returns the number of automorphisms of the graph.\n",
      "     |      \n",
      "     |      This function simply calls C{count_isomorphisms_vf2} with the graph\n",
      "     |      itself. See C{count_isomorphisms_vf2} for an explanation of the\n",
      "     |      parameters.\n",
      "     |      \n",
      "     |      @return: the number of automorphisms of the graph\n",
      "     |      @see: Graph.count_isomorphisms_vf2\n",
      "     |  \n",
      "     |  cut_vertices = articulation_points(...)\n",
      "     |      articulation_points()\n",
      "     |      \n",
      "     |      Returns the list of articulation points in the graph.\n",
      "     |      \n",
      "     |      A vertex is an articulation point if its removal increases the number of\n",
      "     |      connected components in the graph.\n",
      "     |  \n",
      "     |  degree_distribution(self, bin_width=1, *args, **kwds)\n",
      "     |      degree_distribution(bin_width=1, ...)\n",
      "     |      \n",
      "     |      Calculates the degree distribution of the graph.\n",
      "     |      \n",
      "     |      Unknown keyword arguments are directly passed to L{degree()}.\n",
      "     |      \n",
      "     |      @param bin_width: the bin width of the histogram\n",
      "     |      @return: a histogram representing the degree distribution of the\n",
      "     |        graph.\n",
      "     |  \n",
      "     |  delete_edges(self, *args, **kwds)\n",
      "     |      Deletes some edges from the graph.\n",
      "     |      \n",
      "     |      The set of edges to be deleted is determined by the positional and\n",
      "     |      keyword arguments. If any keyword argument is present, or the\n",
      "     |      first positional argument is callable, an edge\n",
      "     |      sequence is derived by calling L{EdgeSeq.select} with the same\n",
      "     |      positional and keyword arguments. Edges in the derived edge sequence\n",
      "     |      will be removed. Otherwise the first positional argument is considered\n",
      "     |      as follows:\n",
      "     |      \n",
      "     |        - C{None} - deletes all edges\n",
      "     |        - a single integer - deletes the edge with the given ID\n",
      "     |        - a list of integers - deletes the edges denoted by the given IDs\n",
      "     |        - a list of 2-tuples - deletes the edges denoted by the given\n",
      "     |          source-target vertex pairs. When multiple edges are present\n",
      "     |          between a given source-target vertex pair, only one is removed.\n",
      "     |  \n",
      "     |  dyad_census(self, *args, **kwds)\n",
      "     |      dyad_census()\n",
      "     |      \n",
      "     |      Calculates the dyad census of the graph.\n",
      "     |      \n",
      "     |      Dyad census means classifying each pair of vertices of a directed\n",
      "     |      graph into three categories: mutual (there is an edge from I{a} to\n",
      "     |      I{b} and also from I{b} to I{a}), asymmetric (there is an edge\n",
      "     |      from I{a} to I{b} or from I{b} to I{a} but not the other way round)\n",
      "     |      and null (there is no connection between I{a} and I{b}).\n",
      "     |      \n",
      "     |      @return: a L{DyadCensus} object.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Holland, P.W. and Leinhardt, S.  (1970).  A Method for Detecting\n",
      "     |        Structure in Sociometric Data.  American Journal of Sociology, 70,\n",
      "     |        492-513.\n",
      "     |  \n",
      "     |  edge_disjoint_paths = edge_connectivity(...)\n",
      "     |      edge_connectivity(source=-1, target=-1, checks=True)\n",
      "     |      \n",
      "     |      Calculates the edge connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The edge connectivity between two given vertices is the number of edges\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of edge disjoint directed\n",
      "     |      paths between the vertices. The edge connectivity of the graph is the minimal\n",
      "     |      edge connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the edge connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall edge connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @return: the edge connectivity\n",
      "     |  \n",
      "     |  evcent = eigenvector_centrality(...)\n",
      "     |      eigenvector_centrality(directed=True, scale=True, weights=None, return_eigenvalue=False, arpack_options=None)\n",
      "     |      \n",
      "     |      Calculates the eigenvector centralities of the vertices in a graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider edge directions in a directed\n",
      "     |        graph. Ignored for undirected graphs.\n",
      "     |      @param scale: whether to normalize the centralities so the largest\n",
      "     |        one will always be 1.\n",
      "     |      @param weights: edge weights given as a list or an edge attribute. If\n",
      "     |        C{None}, all edges have equal weight.\n",
      "     |      @param return_eigenvalue: whether to return the actual largest\n",
      "     |        eigenvalue along with the centralities\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object that can be used\n",
      "     |        to fine-tune the calculation. If it is omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @return: the eigenvector centralities in a list and optionally the\n",
      "     |        largest eigenvalue (as a second member of a tuple)\n",
      "     |  \n",
      "     |  get_adjacency(self, type=2, attribute=None, default=0, eids=False)\n",
      "     |      Returns the adjacency matrix of a graph.\n",
      "     |      \n",
      "     |      @param type: either C{GET_ADJACENCY_LOWER} (uses the lower\n",
      "     |        triangle of the matrix) or C{GET_ADJACENCY_UPPER}\n",
      "     |        (uses the upper triangle) or C{GET_ADJACENCY_BOTH}\n",
      "     |        (uses both parts). Ignored for directed graphs.\n",
      "     |      @param attribute: if C{None}, returns the ordinary adjacency\n",
      "     |        matrix. When the name of a valid edge attribute is given\n",
      "     |        here, the matrix returned will contain the default value\n",
      "     |        at the places where there is no edge or the value of the\n",
      "     |        given attribute where there is an edge. Multiple edges are\n",
      "     |        not supported, the value written in the matrix in this case\n",
      "     |        will be unpredictable. This parameter is ignored if\n",
      "     |        I{eids} is C{True}\n",
      "     |      @param default: the default value written to the cells in the\n",
      "     |        case of adjacency matrices with attributes.\n",
      "     |      @param eids: specifies whether the edge IDs should be returned\n",
      "     |        in the adjacency matrix. Since zero is a valid edge ID, the\n",
      "     |        cells in the matrix that correspond to unconnected vertex\n",
      "     |        pairs will contain -1 instead of 0 if I{eids} is C{True}.\n",
      "     |        If I{eids} is C{False}, the number of edges will be returned\n",
      "     |        in the matrix for each vertex pair.\n",
      "     |      @return: the adjacency matrix as a L{Matrix}.\n",
      "     |  \n",
      "     |  get_adjedgelist(self, *args, **kwds)\n",
      "     |      get_adjedgelist(mode=OUT)\n",
      "     |      \n",
      "     |      Returns the incidence list representation of the graph.\n",
      "     |      \n",
      "     |      @deprecated: replaced by L{Graph.get_inclist()} since igraph 0.6\n",
      "     |      @see: Graph.get_inclist()\n",
      "     |  \n",
      "     |  get_adjlist(self, mode=1)\n",
      "     |      get_adjlist(mode=OUT)\n",
      "     |      \n",
      "     |      Returns the adjacency list representation of the graph.\n",
      "     |      \n",
      "     |      The adjacency list representation is a list of lists. Each item of the\n",
      "     |      outer list belongs to a single vertex of the graph. The inner list\n",
      "     |      contains the neighbors of the given vertex.\n",
      "     |      \n",
      "     |      @param mode: if L{OUT}, returns the successors of the vertex. If\n",
      "     |        L{IN}, returns the predecessors of the vertex. If L{ALL}, both\n",
      "     |        the predecessors and the successors will be returned. Ignored\n",
      "     |        for undirected graphs.\n",
      "     |  \n",
      "     |  get_automorphisms_vf2(self, color=None, edge_color=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      Returns all the automorphisms of the graph\n",
      "     |      \n",
      "     |      This function simply calls C{get_isomorphisms_vf2} with the graph\n",
      "     |      itself. See C{get_isomorphisms_vf2} for an explanation of the\n",
      "     |      parameters.\n",
      "     |      \n",
      "     |      @return: a list of lists, each item containing a possible mapping\n",
      "     |        of the graph vertices to itself according to the automorphism\n",
      "     |      @see: Graph.get_isomorphisms_vf2\n",
      "     |  \n",
      "     |  get_incidence(self, types='type', *args, **kwds)\n",
      "     |      get_incidence(self, types=\"type\")\n",
      "     |      \n",
      "     |      Returns the incidence matrix of a bipartite graph. The incidence matrix\n",
      "     |      is an M{n} times M{m} matrix, where M{n} and M{m} are the number of\n",
      "     |      vertices in the two vertex classes.\n",
      "     |      \n",
      "     |      @param types: an igraph vector containing the vertex types, or an\n",
      "     |        attribute name. Anything that evalulates to C{False} corresponds to\n",
      "     |        vertices of the first kind, everything else to the second kind.\n",
      "     |      @return: the incidence matrix and two lists in a triplet. The first\n",
      "     |        list defines the mapping between row indices of the matrix and the\n",
      "     |        original vertex IDs. The second list is the same for the column\n",
      "     |        indices.\n",
      "     |  \n",
      "     |  get_inclist(self, mode=1)\n",
      "     |      get_inclist(mode=OUT)\n",
      "     |      \n",
      "     |      Returns the incidence list representation of the graph.\n",
      "     |      \n",
      "     |      The incidence list representation is a list of lists. Each\n",
      "     |      item of the outer list belongs to a single vertex of the graph.\n",
      "     |      The inner list contains the IDs of the incident edges of the\n",
      "     |      given vertex.\n",
      "     |      \n",
      "     |      @param mode: if L{OUT}, returns the successors of the vertex. If\n",
      "     |        L{IN}, returns the predecessors of the vertex. If L{ALL}, both\n",
      "     |        the predecessors and the successors will be returned. Ignored\n",
      "     |        for undirected graphs.\n",
      "     |  \n",
      "     |  gomory_hu_tree(self, capacity=None, flow='flow')\n",
      "     |      gomory_hu_tree(capacity=None, flow=\"flow\")\n",
      "     |      \n",
      "     |      Calculates the Gomory-Hu tree of an undirected graph with optional\n",
      "     |      edge capacities.\n",
      "     |      \n",
      "     |      The Gomory-Hu tree is a concise representation of the value of all the\n",
      "     |      maximum flows (or minimum cuts) in a graph. The vertices of the tree\n",
      "     |      correspond exactly to the vertices of the original graph in the same order.\n",
      "     |      Edges of the Gomory-Hu tree are annotated by flow values.  The value of\n",
      "     |      the maximum flow (or minimum cut) between an arbitrary (u,v) vertex\n",
      "     |      pair in the original graph is then given by the minimum flow value (i.e.\n",
      "     |      edge annotation) along the shortest path between u and v in the\n",
      "     |      Gomory-Hu tree.\n",
      "     |      \n",
      "     |      @param capacity: the edge capacities (weights). If C{None}, all\n",
      "     |        edges have equal weight. May also be an attribute name.\n",
      "     |      @param flow: the name of the edge attribute in the returned graph\n",
      "     |        in which the flow values will be stored.\n",
      "     |      @return: the Gomory-Hu tree as a L{Graph} object.\n",
      "     |  \n",
      "     |  indegree(self, *args, **kwds)\n",
      "     |      Returns the in-degrees in a list.\n",
      "     |      \n",
      "     |      See L{degree} for possible arguments.\n",
      "     |  \n",
      "     |  is_named(self)\n",
      "     |      is_named()\n",
      "     |      \n",
      "     |      Returns whether the graph is named, i.e. whether it has a \"name\"\n",
      "     |      vertex attribute.\n",
      "     |  \n",
      "     |  is_weighted(self)\n",
      "     |      is_weighted()\n",
      "     |      \n",
      "     |      Returns whether the graph is weighted, i.e. whether it has a \"weight\"\n",
      "     |      edge attribute.\n",
      "     |  \n",
      "     |  k_core(self, *args)\n",
      "     |      Returns some k-cores of the graph.\n",
      "     |      \n",
      "     |      The method accepts an arbitrary number of arguments representing\n",
      "     |      the desired indices of the M{k}-cores to be returned. The arguments\n",
      "     |      can also be lists or tuples. The result is a single L{Graph} object\n",
      "     |      if an only integer argument was given, otherwise the result is a\n",
      "     |      list of L{Graph} objects representing the desired k-cores in the\n",
      "     |      order the arguments were specified. If no argument is given, returns\n",
      "     |      all M{k}-cores in increasing order of M{k}.\n",
      "     |  \n",
      "     |  layout(self, layout=None, *args, **kwds)\n",
      "     |      Returns the layout of the graph according to a layout algorithm.\n",
      "     |      \n",
      "     |      Parameters and keyword arguments not specified here are passed to the\n",
      "     |      layout algorithm directly. See the documentation of the layout\n",
      "     |      algorithms for the explanation of these parameters.\n",
      "     |      \n",
      "     |      Registered layout names understood by this method are:\n",
      "     |      \n",
      "     |        - C{auto}, C{automatic}: automatic layout\n",
      "     |          (see L{Graph.layout_auto})\n",
      "     |      \n",
      "     |        - C{bipartite}: bipartite layout (see L{Graph.layout_bipartite})\n",
      "     |      \n",
      "     |        - C{circle}, C{circular}: circular layout\n",
      "     |          (see L{Graph.layout_circle})\n",
      "     |      \n",
      "     |        - C{drl}: DrL layout for large graphs (see L{Graph.layout_drl})\n",
      "     |      \n",
      "     |        - C{drl_3d}: 3D DrL layout for large graphs\n",
      "     |          (see L{Graph.layout_drl})\n",
      "     |      \n",
      "     |        - C{fr}, C{fruchterman_reingold}: Fruchterman-Reingold layout\n",
      "     |          (see L{Graph.layout_fruchterman_reingold}).\n",
      "     |      \n",
      "     |        - C{fr_3d}, C{fr3d}, C{fruchterman_reingold_3d}: 3D Fruchterman-\n",
      "     |          Reingold layout (see L{Graph.layout_fruchterman_reingold}).\n",
      "     |      \n",
      "     |        - C{grid}: regular grid layout in 2D (see L{Graph.layout_grid})\n",
      "     |      \n",
      "     |        - C{grid_3d}: regular grid layout in 3D (see L{Graph.layout_grid_3d})\n",
      "     |      \n",
      "     |        - C{graphopt}: the graphopt algorithm (see L{Graph.layout_graphopt})\n",
      "     |      \n",
      "     |        - C{kk}, C{kamada_kawai}: Kamada-Kawai layout\n",
      "     |          (see L{Graph.layout_kamada_kawai})\n",
      "     |      \n",
      "     |        - C{kk_3d}, C{kk3d}, C{kamada_kawai_3d}: 3D Kamada-Kawai layout\n",
      "     |          (see L{Graph.layout_kamada_kawai})\n",
      "     |      \n",
      "     |        - C{lgl}, C{large}, C{large_graph}: Large Graph Layout\n",
      "     |          (see L{Graph.layout_lgl})\n",
      "     |      \n",
      "     |        - C{mds}: multidimensional scaling layout (see L{Graph.layout_mds})\n",
      "     |      \n",
      "     |        - C{random}: random layout (see L{Graph.layout_random})\n",
      "     |      \n",
      "     |        - C{random_3d}: random 3D layout (see L{Graph.layout_random})\n",
      "     |      \n",
      "     |        - C{rt}, C{tree}, C{reingold_tilford}: Reingold-Tilford tree\n",
      "     |          layout (see L{Graph.layout_reingold_tilford})\n",
      "     |      \n",
      "     |        - C{rt_circular}, C{reingold_tilford_circular}: circular\n",
      "     |          Reingold-Tilford tree layout\n",
      "     |          (see L{Graph.layout_reingold_tilford_circular})\n",
      "     |      \n",
      "     |        - C{sphere}, C{spherical}, C{circle_3d}, C{circular_3d}: spherical\n",
      "     |          layout (see L{Graph.layout_circle})\n",
      "     |      \n",
      "     |        - C{star}: star layout (see L{Graph.layout_star})\n",
      "     |      \n",
      "     |        - C{sugiyama}: Sugiyama layout (see L{Graph.layout_sugiyama})\n",
      "     |      \n",
      "     |      @param layout: the layout to use. This can be one of the registered\n",
      "     |        layout names or a callable which returns either a L{Layout} object or\n",
      "     |        a list of lists containing the coordinates. If C{None}, uses the\n",
      "     |        value of the C{plotting.layout} configuration key.\n",
      "     |      @return: a L{Layout} object.\n",
      "     |  \n",
      "     |  layout_auto(self, *args, **kwds)\n",
      "     |      Chooses and runs a suitable layout function based on simple\n",
      "     |      topological properties of the graph.\n",
      "     |      \n",
      "     |      This function tries to choose an appropriate layout function for\n",
      "     |      the graph using the following rules:\n",
      "     |      \n",
      "     |        1. If the graph has an attribute called C{layout}, it will be\n",
      "     |           used. It may either be a L{Layout} instance, a list of\n",
      "     |           coordinate pairs, the name of a layout function, or a\n",
      "     |           callable function which generates the layout when called\n",
      "     |           with the graph as a parameter.\n",
      "     |      \n",
      "     |        2. Otherwise, if the graph has vertex attributes called C{x}\n",
      "     |           and C{y}, these will be used as coordinates in the layout.\n",
      "     |           When a 3D layout is requested (by setting C{dim} to 3),\n",
      "     |           a vertex attribute named C{z} will also be needed.\n",
      "     |      \n",
      "     |        3. Otherwise, if the graph is connected and has at most 100\n",
      "     |           vertices, the Kamada-Kawai layout will be used (see\n",
      "     |           L{Graph.layout_kamada_kawai()}).\n",
      "     |      \n",
      "     |        4. Otherwise, if the graph has at most 1000 vertices, the\n",
      "     |           Fruchterman-Reingold layout will be used (see\n",
      "     |           L{Graph.layout_fruchterman_reingold()}).\n",
      "     |      \n",
      "     |        5. If everything else above failed, the DrL layout algorithm\n",
      "     |           will be used (see L{Graph.layout_drl()}).\n",
      "     |      \n",
      "     |      All the arguments of this function except C{dim} are passed on\n",
      "     |      to the chosen layout function (in case we have to call some layout\n",
      "     |      function).\n",
      "     |      \n",
      "     |      @keyword dim: specifies whether we would like to obtain a 2D or a\n",
      "     |        3D layout.\n",
      "     |      @return: a L{Layout} object.\n",
      "     |  \n",
      "     |  layout_bipartite(*args, **kwds)\n",
      "     |      layout_bipartite(types=\"type\", hgap=1, vgap=1, maxiter=100)\n",
      "     |      \n",
      "     |      Place the vertices of a bipartite graph in two layers.\n",
      "     |      \n",
      "     |      The layout is created by placing the vertices in two rows, according\n",
      "     |      to their types. The positions of the vertices within the rows are\n",
      "     |      then optimized to minimize the number of edge crossings using the\n",
      "     |      heuristic used by the Sugiyama layout algorithm.\n",
      "     |      \n",
      "     |      @param types: an igraph vector containing the vertex types, or an\n",
      "     |        attribute name. Anything that evalulates to C{False} corresponds to\n",
      "     |        vertices of the first kind, everything else to the second kind.\n",
      "     |      @param hgap: minimum horizontal gap between vertices in the same layer.\n",
      "     |      @param vgap: vertical gap between the two layers.\n",
      "     |      @param maxiter: maximum number of iterations to take in the crossing\n",
      "     |        reduction step. Increase this if you feel that you are getting too many\n",
      "     |        edge crossings.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_circle(*args, **kwds)\n",
      "     |      layout_circle(dim=2)\n",
      "     |      \n",
      "     |      Places the vertices of the graph uniformly on a circle or a sphere.\n",
      "     |      \n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_drl(*args, **kwds)\n",
      "     |      layout_drl(weights=None, fixed=None, seed=None, options=None, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices on a 2D plane or in the 3D space ccording to the DrL\n",
      "     |      layout algorithm.\n",
      "     |      \n",
      "     |      This is an algorithm suitable for quite large graphs, but it can be\n",
      "     |      surprisingly slow for small ones (where the simpler force-based layouts\n",
      "     |      like C{layout_kamada_kawai()} or C{layout_fruchterman_reingold()} are\n",
      "     |      more useful.\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param seed: if C{None}, uses a random starting layout for the\n",
      "     |        algorithm. If a matrix (list of lists), uses the given matrix\n",
      "     |        as the starting position.\n",
      "     |      @param fixed: if a seed is given, you can specify some vertices to be\n",
      "     |        kept fixed at their original position in the seed by passing an\n",
      "     |        appropriate list here. The list must have exactly as many items as\n",
      "     |        the number of vertices in the graph. Items of the list that evaluate\n",
      "     |        to C{True} denote vertices that will not be moved.\n",
      "     |      @param options: if you give a string argument here, you can select from\n",
      "     |        five default preset parameterisations: C{default}, C{coarsen} for a\n",
      "     |        coarser layout, C{coarsest} for an even coarser layout, C{refine} for\n",
      "     |        refining an existing layout and C{final} for finalizing a layout. If\n",
      "     |        you supply an object that is not a string, the DrL layout parameters\n",
      "     |        are retrieved from the respective keys of the object (so it should\n",
      "     |        be a dict or something else that supports the mapping protocol).\n",
      "     |        The following keys can be used:\n",
      "     |        \n",
      "     |          - C{edge_cut}: edge cutting is done in the late stages of the\n",
      "     |            algorithm in order to achieve less dense layouts. Edges are\n",
      "     |            cut if there is a lot of stress on them (a large value in the\n",
      "     |            objective function sum). The edge cutting parameter is a value\n",
      "     |            between 0 and 1 with 0 representing no edge cutting and 1\n",
      "     |            representing maximal edge cutting.\n",
      "     |      \n",
      "     |          - C{init_iterations}: number of iterations in the initialization\n",
      "     |            phase\n",
      "     |      \n",
      "     |          - C{init_temperature}: start temperature during initialization\n",
      "     |      \n",
      "     |          - C{init_attraction}: attraction during initialization\n",
      "     |      \n",
      "     |          - C{init_damping_mult}: damping multiplier during initialization\n",
      "     |      \n",
      "     |          - C{liquid_iterations}, C{liquid_temperature}, C{liquid_attraction},\n",
      "     |            C{liquid_damping_mult}: same parameters for the liquid phase\n",
      "     |      \n",
      "     |          - C{expansion_iterations}, C{expansion_temperature},\n",
      "     |            C{expansion_attraction}, C{expansion_damping_mult}:\n",
      "     |            parameters for the expansion phase\n",
      "     |      \n",
      "     |          - C{cooldown_...}: parameters for the cooldown phase\n",
      "     |      \n",
      "     |          - C{crunch_...}: parameters for the crunch phase\n",
      "     |      \n",
      "     |          - C{simmer_...}: parameters for the simmer phase\n",
      "     |      \n",
      "     |        \n",
      "     |        Instead of a mapping, you can also use an arbitrary Python object\n",
      "     |        here: if the object does not support the mapping protocol, an\n",
      "     |        attribute of the object with the same name is looked up instead. If\n",
      "     |        a parameter cannot be found either as a key or an attribute, the\n",
      "     |        default from the C{default} preset will be used.\n",
      "     |      \n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_fruchterman_reingold(*args, **kwds)\n",
      "     |      layout_fruchterman_reingold(weights=None, maxiter=500, maxdelta=None, area=None,\n",
      "     |        coolexp=1.5, repulserad=None, seed=None, minx=None, maxx=None, miny=None, \n",
      "     |        maxy=None, minz=None, maxz=None, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices on a 2D plane or in the 3D space according to the\n",
      "     |      Fruchterman-Reingold algorithm.\n",
      "     |      \n",
      "     |      This is a force directed layout, see Fruchterman, T. M. J. and Reingold, E. M.:\n",
      "     |      Graph Drawing by Force-directed Placement.\n",
      "     |      Software -- Practice and Experience, 21/11, 1129--1164, 1991\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param maxiter: the number of iterations to perform. The default\n",
      "     |        is 500.\n",
      "     |      @param maxdelta: the maximum distance to move a vertex in\n",
      "     |        an iteration. The default is the number of vertices.\n",
      "     |      @param area: the area of the square on which the vertices\n",
      "     |        will be placed. The default is the square of the number of\n",
      "     |        vertices.\n",
      "     |      @param coolexp: the cooling exponent of the simulated annealing.\n",
      "     |        The default is 1.5.\n",
      "     |      @param repulserad: determines the radius at which vertex-vertex\n",
      "     |        repulsion cancels out attraction of adjacent vertices.\n",
      "     |        The default is the number of vertices^3.\n",
      "     |      @param minx: if not C{None}, it must be a vector with exactly as many\n",
      "     |        elements as there are vertices in the graph. Each element is a\n",
      "     |        minimum constraint on the X value of the vertex in the layout.\n",
      "     |      @param maxx: similar to I{minx}, but with maximum constraints\n",
      "     |      @param miny: similar to I{minx}, but with the Y coordinates\n",
      "     |      @param maxy: similar to I{maxx}, but with the Y coordinates\n",
      "     |      @param minz: similar to I{minx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param maxz: similar to I{maxx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param seed: if C{None}, uses a random starting layout for the\n",
      "     |        algorithm. If a matrix (list of lists), uses the given matrix\n",
      "     |        as the starting position.\n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_fruchterman_reingold_3d(*args, **kwds)\n",
      "     |      Alias for L{layout_fruchterman_reingold()} with dim=3.\n",
      "     |      \n",
      "     |      @see: Graph.layout_fruchterman_reingold()\n",
      "     |  \n",
      "     |  layout_graphopt(*args, **kwds)\n",
      "     |      layout_graphopt(niter=500, node_charge=0.001, node_mass=30, spring_length=0, spring_constant=1, max_sa_movement=5, seed=None)\n",
      "     |      \n",
      "     |      This is a port of the graphopt layout algorithm by Michael Schmuhl.\n",
      "     |      graphopt version 0.4.1 was rewritten in C and the support for layers\n",
      "     |      was removed.\n",
      "     |      \n",
      "     |      graphopt uses physical analogies for defining attracting and repelling\n",
      "     |      forces among the vertices and then the physical system is simulated\n",
      "     |      until it reaches an equilibrium or the maximal number of iterations is\n",
      "     |      reached.\n",
      "     |      \n",
      "     |      See U{http://www.schmuhl.org/graphopt/} for the original graphopt.\n",
      "     |      \n",
      "     |      @param niter: the number of iterations to perform. Should be a couple\n",
      "     |        of hundred in general.\n",
      "     |      \n",
      "     |      @param node_charge: the charge of the vertices, used to calculate electric\n",
      "     |        repulsion.\n",
      "     |      @param node_mass: the mass of the vertices, used for the spring forces\n",
      "     |      @param spring_length: the length of the springs\n",
      "     |      @param spring_constant: the spring constant\n",
      "     |      @param max_sa_movement: the maximum amount of movement allowed in a single\n",
      "     |        step along a single axis.\n",
      "     |      @param seed: a matrix containing a seed layout from which the algorithm\n",
      "     |        will be started. If C{None}, a random layout will be used.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_grid(*args, **kwds)\n",
      "     |      layout_grid(width=0, height=0, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices of a graph in a 2D or 3D grid.\n",
      "     |      \n",
      "     |      @param width: the number of vertices in a single row of the layout.\n",
      "     |        Zero or negative numbers mean that the width should be determined\n",
      "     |        automatically.\n",
      "     |      @param height: the number of vertices in a single column of the layout.\n",
      "     |        Zero or negative numbers mean that the height should be determined\n",
      "     |        automatically. It must not be given if the number of dimensions is 2.\n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_grid_3d(*args, **kwds)\n",
      "     |      Alias for L{layout_grid()} with dim=3.\n",
      "     |      \n",
      "     |      @see: Graph.layout_grid()\n",
      "     |  \n",
      "     |  layout_grid_fruchterman_reingold(*args, **kwds)\n",
      "     |      layout_grid_fruchterman_reingold(*args, **kwds)\n",
      "     |      \n",
      "     |      Compatibility alias to the Fruchterman-Reingold layout with the grid\n",
      "     |      option turned on.\n",
      "     |      \n",
      "     |      @see: Graph.layout_fruchterman_reingold()\n",
      "     |  \n",
      "     |  layout_kamada_kawai(*args, **kwds)\n",
      "     |      layout_kamada_kawai(maxiter=1000, sigma=None, initemp=10, coolexp=0.99,\n",
      "     |        kkconst=None, seed=None, minx=None, maxx=None, miny=None, maxy=None, \n",
      "     |        minz=None, maxz=None, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices on a plane according to the Kamada-Kawai algorithm.\n",
      "     |      \n",
      "     |      This is a force directed layout, see Kamada, T. and Kawai, S.:\n",
      "     |      An Algorithm for Drawing General Undirected Graphs.\n",
      "     |      Information Processing Letters, 31/1, 7--15, 1989.\n",
      "     |      \n",
      "     |      @param maxiter: the number of iterations to perform.\n",
      "     |      @param sigma: the standard base deviation of the position\n",
      "     |        change proposals. C{None} means the number of vertices / 4\n",
      "     |      @param initemp: initial temperature of the simulated annealing.\n",
      "     |      @param coolexp: cooling exponent of the simulated annealing.\n",
      "     |      @param kkconst: the Kamada-Kawai vertex attraction constant.\n",
      "     |        C{None} means the square of the number of vertices.\n",
      "     |      @param minx: if not C{None}, it must be a vector with exactly as many\n",
      "     |        elements as there are vertices in the graph. Each element is a\n",
      "     |        minimum constraint on the X value of the vertex in the layout.\n",
      "     |      @param maxx: similar to I{minx}, but with maximum constraints\n",
      "     |      @param miny: similar to I{minx}, but with the Y coordinates\n",
      "     |      @param maxy: similar to I{maxx}, but with the Y coordinates\n",
      "     |      @param minz: similar to I{minx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param maxz: similar to I{maxx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param seed: if C{None}, uses a random starting layout for the\n",
      "     |        algorithm. If a matrix (list of lists), uses the given matrix\n",
      "     |        as the starting position.\n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_kamada_kawai_3d(*args, **kwds)\n",
      "     |      Alias for L{layout_kamada_kawai()} with dim=3.\n",
      "     |      \n",
      "     |      @see: Graph.layout_kamada_kawai()\n",
      "     |  \n",
      "     |  layout_lgl(*args, **kwds)\n",
      "     |      layout_lgl(maxiter=150, maxdelta=-1, area=-1, coolexp=1.5, repulserad=-1, cellsize=-1, root=None)\n",
      "     |      \n",
      "     |      Places the vertices on a 2D plane according to the Large Graph Layout.\n",
      "     |      \n",
      "     |      @param maxiter: the number of iterations to perform.\n",
      "     |      @param maxdelta: the maximum distance to move a vertex in\n",
      "     |        an iteration. If negative, defaults to the number of vertices.\n",
      "     |      @param area: the area of the square on which the vertices\n",
      "     |        will be placed. If negative, defaults to the number of vertices\n",
      "     |        squared.\n",
      "     |      @param coolexp: the cooling exponent of the simulated annealing.\n",
      "     |      @param repulserad: determines the radius at which vertex-vertex\n",
      "     |        repulsion cancels out attraction of adjacent vertices.\n",
      "     |        If negative, defaults to M{area} times the number of vertices.\n",
      "     |      @param cellsize: the size of the grid cells. When calculating the\n",
      "     |        repulsion forces, only vertices in the same or neighboring\n",
      "     |        grid cells are taken into account. Defaults to the fourth\n",
      "     |        root of M{area}.\n",
      "     |      @param root: the root vertex, this is placed first, its neighbors\n",
      "     |        in the first iteration, second neighbors in the second,\n",
      "     |        etc. C{None} means that a random vertex will be chosen.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_mds(*args, **kwds)\n",
      "     |      layout_mds(dist=None, dim=2, arpack_options=None)\n",
      "     |      Places the vertices in an Euclidean space with the given number of\n",
      "     |      dimensions using multidimensional scaling.\n",
      "     |      \n",
      "     |      This layout requires a distance matrix, where the intersection of\n",
      "     |      row M{i} and column M{j} specifies the desired distance between\n",
      "     |      vertex M{i} and vertex M{j}. The algorithm will try to place the\n",
      "     |      vertices in a way that approximates the distance relations\n",
      "     |      prescribed in the distance matrix. igraph uses the classical\n",
      "     |      multidimensional scaling by Torgerson (see reference below).\n",
      "     |      \n",
      "     |      For unconnected graphs, the method will decompose the graph into\n",
      "     |      weakly connected components and then lay out the components\n",
      "     |      individually using the appropriate parts of the distance matrix.\n",
      "     |      \n",
      "     |      @param dist: the distance matrix. It must be symmetric and the\n",
      "     |        symmetry is not checked -- results are unspecified when a\n",
      "     |        non-symmetric distance matrix is used. If this parameter is\n",
      "     |        C{None}, the shortest path lengths will be used as distances.\n",
      "     |        Directed graphs are treated as undirected when calculating\n",
      "     |        the shortest path lengths to ensure symmetry.\n",
      "     |      @param dim: the number of dimensions. For 2D layouts, supply\n",
      "     |        2 here; for 3D layouts, supply 3.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @return: the calculated layout.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Cox & Cox: Multidimensional Scaling (1994), Chapman and\n",
      "     |        Hall, London.\n",
      "     |  \n",
      "     |  layout_random(*args, **kwds)\n",
      "     |      layout_random(dim=2)\n",
      "     |      Places the vertices of the graph randomly.\n",
      "     |      \n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the coordinate pairs in a list.\n",
      "     |  \n",
      "     |  layout_random_3d(*args, **kwds)\n",
      "     |      Alias for L{layout_random()} with dim=3.\n",
      "     |      \n",
      "     |      @see: Graph.layout_random()\n",
      "     |  \n",
      "     |  layout_reingold_tilford(*args, **kwds)\n",
      "     |      layout_reingold_tilford(mode=\"out\", root=None, rootlevel=None)\n",
      "     |      Places the vertices on a 2D plane according to the Reingold-Tilford\n",
      "     |      layout algorithm.\n",
      "     |      \n",
      "     |      This is a tree layout. If the given graph is not a tree, a breadth-first\n",
      "     |      search is executed first to obtain a possible spanning tree.\n",
      "     |      \n",
      "     |      @param mode: specifies which edges to consider when builing the tree.\n",
      "     |        If it is C{OUT} then only the outgoing, if it is C{IN} then only the\n",
      "     |        incoming edges of a parent are considered. If it is C{ALL} then all\n",
      "     |        edges are used (this was the behaviour in igraph 0.5 and before).\n",
      "     |        This parameter also influences how the root vertices are calculated\n",
      "     |        if they are not given. See the I{root} parameter.\n",
      "     |      @param root: the index of the root vertex or root vertices.\n",
      "     |        if this is a non-empty vector then the supplied vertex IDs are\n",
      "     |        used as the roots of the trees (or a single tree if the graph is\n",
      "     |        connected. If this is C{None} or an empty list, the root vertices\n",
      "     |        are automatically calculated based on topological sorting,\n",
      "     |        performed with the opposite of the I{mode} argument.\n",
      "     |      @param rootlevel: this argument is useful when drawing forests which are\n",
      "     |        not trees. It specifies the level of the root vertices for every tree\n",
      "     |        in the forest.\n",
      "     |      @return: the calculated layout.\n",
      "     |      \n",
      "     |      @see: layout_reingold_tilford_circular\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: EM Reingold, JS Tilford: I{Tidier Drawings of Trees.}\n",
      "     |      IEEE Transactions on Software Engineering 7:22, 223-228, 1981.\n",
      "     |  \n",
      "     |  layout_reingold_tilford_circular(*args, **kwds)\n",
      "     |      layout_reingold_tilford_circular(mode=\"out\", root=None, rootlevel=None)\n",
      "     |      Circular Reingold-Tilford layout for trees.\n",
      "     |      \n",
      "     |      This layout is similar to the Reingold-Tilford layout, but the vertices\n",
      "     |      are placed in a circular way, with the root vertex in the center.\n",
      "     |      \n",
      "     |      See L{layout_reingold_tilford} for the explanation of the parameters.\n",
      "     |      \n",
      "     |      @return: the calculated layout.\n",
      "     |      \n",
      "     |      @see: layout_reingold_tilford\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: EM Reingold, JS Tilford: I{Tidier Drawings of Trees.}\n",
      "     |      IEEE Transactions on Software Engineering 7:22, 223-228, 1981.\n",
      "     |  \n",
      "     |  layout_sphere = layout_circle_3d(*args, **kwds)\n",
      "     |      Alias for L{layout_circle()} with dim=3.\n",
      "     |      \n",
      "     |      @see: Graph.layout_circle()\n",
      "     |  \n",
      "     |  layout_star(*args, **kwds)\n",
      "     |      layout_star(center=0, order=None)\n",
      "     |      \n",
      "     |      Calculates a star-like layout for the graph.\n",
      "     |      \n",
      "     |      @param center: the ID of the vertex to put in the center\n",
      "     |      @param order: a numeric vector giving the order of the vertices\n",
      "     |        (including the center vertex!). If it is C{None}, the vertices\n",
      "     |        will be placed in increasing vertex ID order.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_sugiyama(self, layers=None, weights=None, hgap=1, vgap=1, maxiter=100, return_extended_graph=False)\n",
      "     |      layout_sugiyama(layers=None, weights=None, hgap=1, vgap=1, maxiter=100,\n",
      "     |                      return_extended_graph=False)\n",
      "     |      \n",
      "     |      Places the vertices using a layered Sugiyama layout.\n",
      "     |      \n",
      "     |      This is a layered layout that is most suitable for directed acyclic graphs,\n",
      "     |      although it works on undirected or cyclic graphs as well.\n",
      "     |      \n",
      "     |      Each vertex is assigned to a layer and each layer is placed on a horizontal\n",
      "     |      line. Vertices within the same layer are then permuted using the barycenter\n",
      "     |      heuristic that tries to minimize edge crossings.\n",
      "     |      \n",
      "     |      Dummy vertices will be added on edges that span more than one layer. The\n",
      "     |      returned layout therefore contains more rows than the number of nodes in\n",
      "     |      the original graph; the extra rows correspond to the dummy vertices.\n",
      "     |      \n",
      "     |      @param layers: a vector specifying a non-negative integer layer index for\n",
      "     |        each vertex, or the name of a numeric vertex attribute that contains\n",
      "     |        the layer indices. If C{None}, a layering will be determined\n",
      "     |        automatically. For undirected graphs, a spanning tree will be extracted\n",
      "     |        and vertices will be assigned to layers using a breadth first search from\n",
      "     |        the node with the largest degree. For directed graphs, cycles are broken\n",
      "     |        by reversing the direction of edges in an approximate feedback arc set\n",
      "     |        using the heuristic of Eades, Lin and Smyth, and then using longest path\n",
      "     |        layering to place the vertices in layers.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param hgap: minimum horizontal gap between vertices in the same layer.\n",
      "     |      @param vgap: vertical gap between layers. The layer index will be\n",
      "     |        multiplied by I{vgap} to obtain the Y coordinate.\n",
      "     |      @param maxiter: maximum number of iterations to take in the crossing\n",
      "     |        reduction step. Increase this if you feel that you are getting too many\n",
      "     |        edge crossings.\n",
      "     |      @param return_extended_graph: specifies that the extended graph with the\n",
      "     |        added dummy vertices should also be returned. When this is C{True}, the\n",
      "     |        result will be a tuple containing the layout and the extended graph. The\n",
      "     |        first |V| nodes of the extended graph will correspond to the nodes of the\n",
      "     |        original graph, the remaining ones are dummy nodes. Plotting the extended\n",
      "     |        graph with the returned layout and hidden dummy nodes will produce a layout\n",
      "     |        that is similar to the original graph, but with the added edge bends.\n",
      "     |        The extended graph also contains an edge attribute called C{_original_eid}\n",
      "     |        which specifies the ID of the edge in the original graph from which the\n",
      "     |        edge of the extended graph was created.\n",
      "     |      @return: the calculated layout, which may (and usually will) have more rows\n",
      "     |        than the number of vertices; the remaining rows correspond to the dummy\n",
      "     |        nodes introduced in the layering step. When C{return_extended_graph} is\n",
      "     |        C{True}, it will also contain the extended graph.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: K Sugiyama, S Tagawa, M Toda: Methods for visual understanding of\n",
      "     |        hierarchical system structures. IEEE Systems, Man and Cybernetics          11(2):109-125, 1981.\n",
      "     |      @ref: P Eades, X Lin and WF Smyth: A fast effective heuristic for the\n",
      "     |        feedback arc set problem. Information Processing Letters 47:319-323, 1993.\n",
      "     |  \n",
      "     |  maxflow(self, source, target, capacity=None)\n",
      "     |      maxflow(source, target, capacity=None)\n",
      "     |      \n",
      "     |      Returns a maximum flow between the given source and target vertices\n",
      "     |      in a graph.\n",
      "     |      \n",
      "     |      A maximum flow from I{source} to I{target} is an assignment of\n",
      "     |      non-negative real numbers to the edges of the graph, satisfying\n",
      "     |      two properties:\n",
      "     |      \n",
      "     |          1. For each edge, the flow (i.e. the assigned number) is not\n",
      "     |             more than the capacity of the edge (see the I{capacity}\n",
      "     |             argument)\n",
      "     |      \n",
      "     |          2. For every vertex except the source and the target, the\n",
      "     |             incoming flow is the same as the outgoing flow.\n",
      "     |      \n",
      "     |      The value of the flow is the incoming flow of the target or the\n",
      "     |      outgoing flow of the source (which are equal). The maximum flow\n",
      "     |      is the maximum possible such value.\n",
      "     |      \n",
      "     |      @param capacity: the edge capacities (weights). If C{None}, all\n",
      "     |        edges have equal weight. May also be an attribute name.\n",
      "     |      @return: a L{Flow} object describing the maximum flow\n",
      "     |  \n",
      "     |  maximum_bipartite_matching(self, types='type', weights=None, eps=None)\n",
      "     |      Finds a maximum matching in a bipartite graph.\n",
      "     |      \n",
      "     |      A maximum matching is a set of edges such that each vertex is incident on\n",
      "     |      at most one matched edge and the number (or weight) of such edges in the\n",
      "     |      set is as large as possible.\n",
      "     |      \n",
      "     |      @param types: vertex types in a list or the name of a vertex attribute\n",
      "     |        holding vertex types. Types should be denoted by zeros and ones (or\n",
      "     |        C{False} and C{True}) for the two sides of the bipartite graph.\n",
      "     |        If omitted, it defaults to C{type}, which is the default vertex type\n",
      "     |        attribute for bipartite graphs.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param eps: a small real number used in equality tests in the weighted\n",
      "     |        bipartite matching algorithm. Two real numbers are considered equal in\n",
      "     |        the algorithm if their difference is smaller than this value. This\n",
      "     |        is required to avoid the accumulation of numerical errors. If you\n",
      "     |        pass C{None} here, igraph will try to determine an appropriate value\n",
      "     |        automatically.\n",
      "     |      @return: an instance of L{Matching}.\n",
      "     |  \n",
      "     |  mincut(self, source=None, target=None, capacity=None)\n",
      "     |      mincut(source=None, target=None, capacity=None)\n",
      "     |      \n",
      "     |      Calculates the minimum cut between the given source and target vertices\n",
      "     |      or within the whole graph.\n",
      "     |      \n",
      "     |      The minimum cut is the minimum set of edges that needs to be removed to\n",
      "     |      separate the source and the target (if they are given) or to disconnect the\n",
      "     |      graph (if neither the source nor the target are given). The minimum is\n",
      "     |      calculated using the weights (capacities) of the edges, so the cut with\n",
      "     |      the minimum total capacity is calculated.\n",
      "     |      \n",
      "     |      For undirected graphs and no source and target, the method uses the\n",
      "     |      Stoer-Wagner algorithm. For a given source and target, the method uses the\n",
      "     |      push-relabel algorithm; see the references below.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID. If C{None}, the target must also be\n",
      "     |        C{None} and the calculation will be done for the entire graph (i.e.\n",
      "     |        all possible vertex pairs).\n",
      "     |      @param target: the target vertex ID. If C{None}, the source must also be\n",
      "     |        C{None} and the calculation will be done for the entire graph (i.e.\n",
      "     |        all possible vertex pairs).\n",
      "     |      @param capacity: the edge capacities (weights). If C{None}, all\n",
      "     |        edges have equal weight. May also be an attribute name.\n",
      "     |      @return: a L{Cut} object describing the minimum cut\n",
      "     |  \n",
      "     |  modularity(self, membership, weights=None)\n",
      "     |      modularity(membership, weights=None)\n",
      "     |      \n",
      "     |      Calculates the modularity score of the graph with respect to a given\n",
      "     |      clustering.\n",
      "     |      \n",
      "     |      The modularity of a graph w.r.t. some division measures how good the\n",
      "     |      division is, or how separated are the different vertex types from each\n",
      "     |      other. It's defined as M{Q=1/(2m)*sum(Aij-ki*kj/(2m)delta(ci,cj),i,j)}.\n",
      "     |      M{m} is the number of edges, M{Aij} is the element of the M{A}\n",
      "     |      adjacency matrix in row M{i} and column M{j}, M{ki} is the degree of\n",
      "     |      node M{i}, M{kj} is the degree of node M{j}, and M{Ci} and C{cj} are\n",
      "     |      the types of the two vertices (M{i} and M{j}). M{delta(x,y)} is one iff\n",
      "     |      M{x=y}, 0 otherwise.\n",
      "     |      \n",
      "     |      If edge weights are given, the definition of modularity is modified as\n",
      "     |      follows: M{Aij} becomes the weight of the corresponding edge, M{ki}\n",
      "     |      is the total weight of edges adjacent to vertex M{i}, M{kj} is the\n",
      "     |      total weight of edges adjacent to vertex M{j} and M{m} is the total\n",
      "     |      edge weight in the graph.\n",
      "     |      \n",
      "     |      @param membership: a membership list or a L{VertexClustering} object\n",
      "     |      @param weights: optional edge weights or C{None} if all edges are\n",
      "     |        weighed equally. Attribute names are also allowed.\n",
      "     |      @return: the modularity score\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: MEJ Newman and M Girvan: Finding and evaluating community\n",
      "     |        structure in networks. Phys Rev E 69 026113, 2004.\n",
      "     |  \n",
      "     |  omega = clique_number(...)\n",
      "     |      clique_number()\n",
      "     |      \n",
      "     |      Returns the clique number of the graph.\n",
      "     |      \n",
      "     |      The clique number of the graph is the size of the largest clique.\n",
      "     |      \n",
      "     |      @see: L{largest_cliques()} for the largest cliques.\n",
      "     |  \n",
      "     |  outdegree(self, *args, **kwds)\n",
      "     |      Returns the out-degrees in a list.\n",
      "     |      \n",
      "     |      See L{degree} for possible arguments.\n",
      "     |  \n",
      "     |  pagerank(self, vertices=None, directed=True, damping=0.85, weights=None, arpack_options=None, implementation='prpack', niter=1000, eps=0.001)\n",
      "     |      Calculates the Google PageRank values of a graph.\n",
      "     |      \n",
      "     |      @param vertices: the indices of the vertices being queried.\n",
      "     |        C{None} means all of the vertices.\n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param damping: the damping factor. M{1-damping} is the PageRank value\n",
      "     |        for nodes with no incoming links. It is also the probability of\n",
      "     |        resetting the random walk to a uniform distribution in each step.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable\n",
      "     |        or even an edge attribute name.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used. This argument is\n",
      "     |        ignored if not the ARPACK implementation is used, see the\n",
      "     |        I{implementation} argument.\n",
      "     |      @param implementation: which implementation to use to solve the\n",
      "     |        PageRank eigenproblem. Possible values are:\n",
      "     |          - C{\"prpack\"}: use the PRPACK library. This is a new\n",
      "     |            implementation in igraph 0.7\n",
      "     |          - C{\"arpack\"}: use the ARPACK library. This implementation\n",
      "     |            was used from version 0.5, until version 0.7.\n",
      "     |          - C{\"power\"}: use a simple power method. This is the\n",
      "     |            implementation that was used before igraph version 0.5.\n",
      "     |      @param niter: The number of iterations to use in the power method\n",
      "     |        implementation. It is ignored in the other implementations\n",
      "     |      @param eps: The power method implementation will consider the\n",
      "     |        calculation as complete if the difference of PageRank values between\n",
      "     |        iterations change less than this value for every node. It is\n",
      "     |        ignored by the other implementations.\n",
      "     |      @return: a list with the Google PageRank values of the specified\n",
      "     |        vertices.\n",
      "     |  \n",
      "     |  path_length_hist(self, directed=True)\n",
      "     |      path_length_hist(directed=True)\n",
      "     |      \n",
      "     |      Returns the path length histogram of the graph\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths. Ignored for\n",
      "     |        undirected graphs.\n",
      "     |      @return: a L{Histogram} object. The object will also have an\n",
      "     |        C{unconnected} attribute that stores the number of unconnected\n",
      "     |        vertex pairs (where the second vertex can not be reached from\n",
      "     |        the first one). The latter one will be of type long (and not\n",
      "     |        a simple integer), since this can be I{very} large.\n",
      "     |  \n",
      "     |  save = write(self, f, format=None, *args, **kwds)\n",
      "     |  \n",
      "     |  shell_index = coreness(...)\n",
      "     |      coreness(mode=ALL)\n",
      "     |      \n",
      "     |      Finds the coreness (shell index) of the vertices of the network.\n",
      "     |      \n",
      "     |      The M{k}-core of a graph is a maximal subgraph in which each vertex\n",
      "     |      has at least degree k. (Degree here means the degree in the\n",
      "     |      subgraph of course). The coreness of a vertex is M{k} if it\n",
      "     |      is a member of the M{k}-core but not a member of the M{k+1}-core.\n",
      "     |      \n",
      "     |      @param mode: whether to compute the in-corenesses (L{IN}), the\n",
      "     |        out-corenesses (L{OUT}) or the undirected corenesses (L{ALL}).\n",
      "     |        Ignored and assumed to be L{ALL} for undirected graphs.\n",
      "     |      @return: the corenesses for each vertex.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Vladimir Batagelj, Matjaz Zaversnik: I{An M{O(m)} Algorithm\n",
      "     |        for Core Decomposition of Networks.}\n",
      "     |  \n",
      "     |  shortest_paths_dijkstra = shortest_paths(...)\n",
      "     |      shortest_paths(source=None, target=None, weights=None, mode=OUT)\n",
      "     |      \n",
      "     |      Calculates shortest path lengths for given vertices in a graph.\n",
      "     |      \n",
      "     |      The algorithm used for the calculations is selected automatically:\n",
      "     |      a simple BFS is used for unweighted graphs, Dijkstra's algorithm is\n",
      "     |      used when all the weights are positive. Otherwise, the Bellman-Ford\n",
      "     |      algorithm is used if the number of requested source vertices is larger\n",
      "     |      than 100 and Johnson's algorithm is used otherwise.\n",
      "     |      \n",
      "     |      @param source: a list containing the source vertex IDs which should be\n",
      "     |        included in the result. If C{None}, all vertices will be considered.\n",
      "     |      @param target: a list containing the target vertex IDs which should be\n",
      "     |        included in the result. If C{None}, all vertices will be considered.\n",
      "     |      @param weights: a list containing the edge weights. It can also be\n",
      "     |        an attribute name (edge weights are retrieved from the given\n",
      "     |        attribute) or C{None} (all edges have equal weight).\n",
      "     |      @param mode: the type of shortest paths to be used for the\n",
      "     |        calculation in directed graphs. L{OUT} means only outgoing,\n",
      "     |        L{IN} means only incoming paths. L{ALL} means to consider\n",
      "     |        the directed graph as an undirected one.\n",
      "     |      @return: the shortest path lengths for given vertices in a matrix\n",
      "     |  \n",
      "     |  spanning_tree(self, weights=None, return_tree=True)\n",
      "     |      Calculates a minimum spanning tree for a graph.\n",
      "     |      \n",
      "     |      @param weights: a vector containing weights for every edge in\n",
      "     |        the graph. C{None} means that the graph is unweighted.\n",
      "     |      @param return_tree: whether to return the minimum spanning tree (when\n",
      "     |        C{return_tree} is C{True}) or to return the IDs of the edges in\n",
      "     |        the minimum spanning tree instead (when C{return_tree} is C{False}).\n",
      "     |        The default is C{True} for historical reasons as this argument was\n",
      "     |        introduced in igraph 0.6.\n",
      "     |      @return: the spanning tree as a L{Graph} object if C{return_tree}\n",
      "     |        is C{True} or the IDs of the edges that constitute the spanning\n",
      "     |        tree if C{return_tree} is C{False}.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Prim, R.C.: I{Shortest connection networks and some\n",
      "     |        generalizations}. Bell System Technical Journal 36:1389-1401, 1957.\n",
      "     |  \n",
      "     |  st_mincut(self, source, target, capacity=None)\n",
      "     |      st_mincut(source, target, capacity=None)\n",
      "     |      \n",
      "     |      Calculates the minimum cut between the source and target vertices in a\n",
      "     |      graph.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: the value of the minimum cut, the IDs of vertices in the\n",
      "     |        first and second partition, and the IDs of edges in the cut,\n",
      "     |        packed in a 4-tuple\n",
      "     |  \n",
      "     |  subgraph = induced_subgraph(...)\n",
      "     |      induced_subgraph(vertices, implementation=\"auto\")\n",
      "     |      \n",
      "     |      Returns a subgraph spanned by the given vertices.\n",
      "     |      \n",
      "     |      @param vertices: a list containing the vertex IDs which\n",
      "     |        should be included in the result.\n",
      "     |      @param implementation: the implementation to use when constructing\n",
      "     |        the new subgraph. igraph includes two implementations at the\n",
      "     |        moment. C{\"copy_and_delete\"} copies the original graph and\n",
      "     |        removes those vertices that are not in the given set. This is more\n",
      "     |        efficient if the size of the subgraph is comparable to the original\n",
      "     |        graph. The other implementation (C{\"create_from_scratch\"})\n",
      "     |        constructs the result graph from scratch and then copies the\n",
      "     |        attributes accordingly. This is a better solution if the subgraph\n",
      "     |        is relatively small, compared to the original graph. C{\"auto\"}\n",
      "     |        selects between the two implementations automatically, based on\n",
      "     |        the ratio of the size of the subgraph and the size of the original\n",
      "     |        graph.\n",
      "     |      @return: the subgraph\n",
      "     |  \n",
      "     |  summary(self, verbosity=0, width=None, *args, **kwds)\n",
      "     |      Returns the summary of the graph.\n",
      "     |      \n",
      "     |      The output of this method is similar to the output of the\n",
      "     |      C{__str__} method. If I{verbosity} is zero, only the header line\n",
      "     |      is returned (see C{__str__} for more details), otherwise the\n",
      "     |      header line and the edge list is printed.\n",
      "     |      \n",
      "     |      Behind the scenes, this method constructs a L{GraphSummary}\n",
      "     |      object and invokes its C{__str__} method.\n",
      "     |      \n",
      "     |      @param verbosity: if zero, only the header line is returned\n",
      "     |        (see C{__str__} for more details), otherwise the header line\n",
      "     |        and the full edge list is printed.\n",
      "     |      @param width: the number of characters to use in one line.\n",
      "     |        If C{None}, no limit will be enforced on the line lengths.\n",
      "     |      @return: the summary of the graph.\n",
      "     |  \n",
      "     |  transitivity_avglocal_undirected(self, mode='nan', weights=None)\n",
      "     |      Calculates the average of the vertex transitivities of the graph.\n",
      "     |      \n",
      "     |      In the unweighted case, the transitivity measures the probability that\n",
      "     |      two neighbors of a vertex are connected. In case of the average local\n",
      "     |      transitivity, this probability is calculated for each vertex and then\n",
      "     |      the average is taken. Vertices with less than two neighbors require\n",
      "     |      special treatment, they will either be left out from the calculation\n",
      "     |      or they will be considered as having zero transitivity, depending on\n",
      "     |      the I{mode} parameter. The calculation is slightly more involved for\n",
      "     |      weighted graphs; in this case, weights are taken into account according\n",
      "     |      to the formula of Barrat et al (see the references).\n",
      "     |      \n",
      "     |      Note that this measure is different from the global transitivity\n",
      "     |      measure (see L{transitivity_undirected()}) as it simply takes the\n",
      "     |      average local transitivity across the whole network.\n",
      "     |      \n",
      "     |      @param mode: defines how to treat vertices with degree less than two.\n",
      "     |        If C{TRANSITIVITY_ZERO} or C{\"zero\"}, these vertices will have zero\n",
      "     |        transitivity. If C{TRANSITIVITY_NAN} or C{\"nan\"}, these vertices\n",
      "     |        will be excluded from the average.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable\n",
      "     |        or even an edge attribute name.\n",
      "     |      \n",
      "     |      @see: L{transitivity_undirected()}, L{transitivity_local_undirected()}\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Watts DJ and Strogatz S: I{Collective dynamics of small-world\n",
      "     |        networks}. Nature 393(6884):440-442, 1998.\n",
      "     |      @ref: Barrat A, Barthelemy M, Pastor-Satorras R and Vespignani A:\n",
      "     |        I{The architecture of complex weighted networks}. PNAS 101, 3747 (2004).\n",
      "     |        U{http://arxiv.org/abs/cond-mat/0311416}.\n",
      "     |  \n",
      "     |  triad_census(self, *args, **kwds)\n",
      "     |      triad_census()\n",
      "     |      \n",
      "     |      Calculates the triad census of the graph.\n",
      "     |      \n",
      "     |      @return: a L{TriadCensus} object.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Davis, J.A. and Leinhardt, S.  (1972).  The Structure of\n",
      "     |        Positive Interpersonal Relations in Small Groups.  In:\n",
      "     |        J. Berger (Ed.), Sociological Theories in Progress, Volume 2,\n",
      "     |        218-251. Boston: Houghton Mifflin.\n",
      "     |  \n",
      "     |  vertex_disjoint_paths = vertex_connectivity(...)\n",
      "     |      vertex_connectivity(source=-1, target=-1, checks=True, neighbors=\"error\")\n",
      "     |      \n",
      "     |      Calculates the vertex connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The vertex connectivity between two given vertices is the number of vertices\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of vertex disjoint directed\n",
      "     |      paths between the vertices (apart from the source and target vertices of\n",
      "     |      course). The vertex connectivity of the graph is the minimal vertex\n",
      "     |      connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the vertex connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall vertex connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @param neighbors: tells igraph what to do when the two vertices are\n",
      "     |        connected. C{\"error\"} raises an exception, C{\"infinity\"} returns\n",
      "     |        infinity, C{\"ignore\"} ignores the edge.\n",
      "     |      @return: the vertex connectivity\n",
      "     |  \n",
      "     |  write(self, f, format=None, *args, **kwds)\n",
      "     |      Unified writing function for graphs.\n",
      "     |      \n",
      "     |      This method tries to identify the format of the graph given in\n",
      "     |      the first parameter (based on extension) and calls the corresponding\n",
      "     |      writer method.\n",
      "     |      \n",
      "     |      The remaining arguments are passed to the writer method without\n",
      "     |      any changes.\n",
      "     |      \n",
      "     |      @param f: the file containing the graph to be saved\n",
      "     |      @param format: the format of the file (if one wants to override the\n",
      "     |        format determined from the filename extension, or the filename itself\n",
      "     |        is a stream). C{None} means auto-detection. Possible values are:\n",
      "     |      \n",
      "     |          - C{\"adjacency\"}: adjacency matrix format\n",
      "     |      \n",
      "     |          - C{\"dimacs\"}: DIMACS format\n",
      "     |      \n",
      "     |          - C{\"dot\"}, C{\"graphviz\"}: GraphViz DOT format\n",
      "     |      \n",
      "     |          - C{\"edgelist\"}, C{\"edges\"} or C{\"edge\"}: numeric edge list format\n",
      "     |      \n",
      "     |          - C{\"gml\"}: GML format\n",
      "     |      \n",
      "     |          - C{\"graphml\"} and C{\"graphmlz\"}: standard and gzipped GraphML\n",
      "     |            format\n",
      "     |      \n",
      "     |          - C{\"gw\"}, C{\"leda\"}, C{\"lgr\"}: LEDA native format\n",
      "     |      \n",
      "     |          - C{\"lgl\"}: LGL format\n",
      "     |      \n",
      "     |          - C{\"ncol\"}: NCOL format\n",
      "     |      \n",
      "     |          - C{\"net\"}, C{\"pajek\"}: Pajek format\n",
      "     |      \n",
      "     |          - C{\"pickle\"}, C{\"picklez\"}: standard and gzipped Python pickled\n",
      "     |            format\n",
      "     |      \n",
      "     |          - C{\"svg\"}: SVG format\n",
      "     |      \n",
      "     |      @raises IOError: if the file format can't be identified and\n",
      "     |        none was given.\n",
      "     |  \n",
      "     |  write_adjacency(self, f, sep=' ', eol='\\n', *args, **kwds)\n",
      "     |      Writes the adjacency matrix of the graph to the given file\n",
      "     |      \n",
      "     |      All the remaining arguments not mentioned here are passed intact\n",
      "     |      to L{Graph.get_adjacency}.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written.\n",
      "     |      @param sep: the string that separates the matrix elements in a row\n",
      "     |      @param eol: the string that separates the rows of the matrix. Please\n",
      "     |        note that igraph is able to read back the written adjacency matrix\n",
      "     |        if and only if this is a single newline character\n",
      "     |  \n",
      "     |  write_dimacs(self, f, source=None, target=None, capacity='capacity')\n",
      "     |      Writes the graph in DIMACS format to the given file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle.\n",
      "     |      @param source: the source vertex ID. If C{None}, igraph will try to\n",
      "     |        infer it from the C{source} graph attribute.\n",
      "     |      @param target: the target vertex ID. If C{None}, igraph will try to\n",
      "     |        infer it from the C{target} graph attribute.\n",
      "     |      @param capacity: the capacities of the edges in a list or the name of\n",
      "     |        an edge attribute that holds the capacities. If there is no such\n",
      "     |        edge attribute, every edge will have a capacity of 1.\n",
      "     |  \n",
      "     |  write_graphmlz(self, f, compresslevel=9)\n",
      "     |      Writes the graph to a zipped GraphML file.\n",
      "     |      \n",
      "     |      The library uses the gzip compression algorithm, so the resulting\n",
      "     |      file can be unzipped with regular gzip uncompression (like\n",
      "     |      C{gunzip} or C{zcat} from Unix command line) or the Python C{gzip}\n",
      "     |      module.\n",
      "     |      \n",
      "     |      Uses a temporary file to store intermediate GraphML data, so\n",
      "     |      make sure you have enough free space to store the unzipped\n",
      "     |      GraphML file as well.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written.\n",
      "     |      @param compresslevel: the level of compression. 1 is fastest and\n",
      "     |        produces the least compression, and 9 is slowest and produces\n",
      "     |        the most compression.\n",
      "     |  \n",
      "     |  write_pickle(self, fname=None, version=-1)\n",
      "     |      Saves the graph in Python pickled format\n",
      "     |      \n",
      "     |      @param fname: the name of the file or a stream to save to. If\n",
      "     |        C{None}, saves the graph to a string and returns the string.\n",
      "     |      @param version: pickle protocol version to be used. If -1, uses\n",
      "     |        the highest protocol available\n",
      "     |      @return: C{None} if the graph was saved successfully to the\n",
      "     |        given file, or a string if C{fname} was C{None}.\n",
      "     |  \n",
      "     |  write_picklez(self, fname=None, version=-1)\n",
      "     |      Saves the graph in Python pickled format, compressed with\n",
      "     |      gzip.\n",
      "     |      \n",
      "     |      Saving in this format is a bit slower than saving in a Python pickle\n",
      "     |      without compression, but the final file takes up much less space on\n",
      "     |      the hard drive.\n",
      "     |      \n",
      "     |      @param fname: the name of the file or a stream to save to.\n",
      "     |      @param version: pickle protocol version to be used. If -1, uses\n",
      "     |        the highest protocol available\n",
      "     |      @return: C{None} if the graph was saved successfully to the\n",
      "     |        given file.\n",
      "     |  \n",
      "     |  write_svg(self, fname, layout='auto', width=None, height=None, labels='label', colors='color', shapes='shape', vertex_size=10, edge_colors='color', edge_stroke_widths='width', font_size=16, *args, **kwds)\n",
      "     |      Saves the graph as an SVG (Scalable Vector Graphics) file\n",
      "     |      \n",
      "     |      The file will be Inkscape (http://inkscape.org) compatible.\n",
      "     |      In Inkscape, as nodes are rearranged, the edges auto-update.\n",
      "     |      \n",
      "     |      @param fname: the name of the file or a Python file handle\n",
      "     |      @param layout: the layout of the graph. Can be either an\n",
      "     |        explicitly specified layout (using a list of coordinate\n",
      "     |        pairs) or the name of a layout algorithm (which should\n",
      "     |        refer to a method in the L{Graph} object, but without\n",
      "     |        the C{layout_} prefix.\n",
      "     |      @param width: the preferred width in pixels (default: 400)\n",
      "     |      @param height: the preferred height in pixels (default: 400)\n",
      "     |      @param labels: the vertex labels. Either it is the name of\n",
      "     |        a vertex attribute to use, or a list explicitly specifying\n",
      "     |        the labels. It can also be C{None}.\n",
      "     |      @param colors: the vertex colors. Either it is the name of\n",
      "     |        a vertex attribute to use, or a list explicitly specifying\n",
      "     |        the colors. A color can be anything acceptable in an SVG\n",
      "     |        file.\n",
      "     |      @param shapes: the vertex shapes. Either it is the name of\n",
      "     |        a vertex attribute to use, or a list explicitly specifying\n",
      "     |        the shapes as integers. Shape 0 means hidden (nothing is drawn),\n",
      "     |        shape 1 is a circle, shape 2 is a rectangle and shape 3 is a\n",
      "     |        rectangle that automatically sizes to the inner text.\n",
      "     |      @param vertex_size: vertex size in pixels\n",
      "     |      @param edge_colors: the edge colors. Either it is the name\n",
      "     |        of an edge attribute to use, or a list explicitly specifying\n",
      "     |        the colors. A color can be anything acceptable in an SVG\n",
      "     |        file.\n",
      "     |      @param edge_stroke_widths: the stroke widths of the edges. Either\n",
      "     |        it is the name of an edge attribute to use, or a list explicitly\n",
      "     |        specifying the stroke widths. The stroke width can be anything\n",
      "     |        acceptable in an SVG file.\n",
      "     |      @param font_size: font size. If it is a string, it is written into\n",
      "     |        the SVG file as-is (so you can specify anything which is valid\n",
      "     |        as the value of the C{font-size} style). If it is a number, it\n",
      "     |        is interpreted as pixel size and converted to the proper attribute\n",
      "     |        value accordingly.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  Bipartite(klass, types, *args, **kwds) from __builtin__.type\n",
      "     |      Bipartite(types, edges, directed=False)\n",
      "     |      \n",
      "     |      Creates a bipartite graph with the given vertex types and edges.\n",
      "     |      This is similar to the default constructor of the graph, the\n",
      "     |      only difference is that it checks whether all the edges go\n",
      "     |      between the two vertex classes and it assigns the type vector\n",
      "     |      to a C{type} attribute afterwards.\n",
      "     |      \n",
      "     |      Examples:\n",
      "     |      \n",
      "     |      >>> g = Graph.Bipartite([0, 1, 0, 1], [(0, 1), (2, 3), (0, 3)])\n",
      "     |      >>> g.is_bipartite()\n",
      "     |      True\n",
      "     |      >>> g.vs[\"type\"]\n",
      "     |      [False, True, False, True]\n",
      "     |      \n",
      "     |      @param types: the vertex types as a boolean list. Anything that\n",
      "     |        evaluates to C{False} will denote a vertex of the first kind,\n",
      "     |        anything that evaluates to C{True} will denote a vertex of the\n",
      "     |        second kind.\n",
      "     |      @param edges: the edges as a list of tuples.\n",
      "     |      @param directed: whether to create a directed graph. Bipartite\n",
      "     |        networks are usually undirected, so the default is C{False}\n",
      "     |      \n",
      "     |      @return: the graph with a binary vertex attribute named C{\"type\"} that\n",
      "     |        stores the vertex classes.\n",
      "     |  \n",
      "     |  DictList(klass, vertices, edges, directed=False, vertex_name_attr='name', edge_foreign_keys=('source', 'target'), iterative=False) from __builtin__.type\n",
      "     |      Constructs a graph from a list-of-dictionaries representation.\n",
      "     |      \n",
      "     |      This representation assumes that vertices and edges are encoded in\n",
      "     |      two lists, each list containing a Python dict for each vertex and\n",
      "     |      each edge, respectively. A distinguished element of the vertex dicts\n",
      "     |      contain a vertex ID which is used in the edge dicts to refer to\n",
      "     |      source and target vertices. All the remaining elements of the dict\n",
      "     |      are considered vertex and edge attributes. Note that the implementation\n",
      "     |      does not assume that the objects passed to this method are indeed\n",
      "     |      lists of dicts, but they should be iterable and they should yield\n",
      "     |      objects that behave as dicts. So, for instance, a database query\n",
      "     |      result is likely to be fit as long as it's iterable and yields\n",
      "     |      dict-like objects with every iteration.\n",
      "     |      \n",
      "     |      @param vertices: the data source for the vertices or C{None} if\n",
      "     |        there are no special attributes assigned to vertices and we\n",
      "     |        should simply use the edge list of dicts to infer vertex names.\n",
      "     |      @param edges: the data source for the edges.\n",
      "     |      @param directed: whether the constructed graph will be directed\n",
      "     |      @param vertex_name_attr: the name of the distinguished key in the\n",
      "     |        dicts in the vertex data source that contains the vertex names.\n",
      "     |        Ignored if C{vertices} is C{None}.\n",
      "     |      @param edge_foreign_keys: the name of the attributes in the dicts\n",
      "     |        in the edge data source that contain the source and target\n",
      "     |        vertex names.\n",
      "     |      @param iterative: whether to add the edges to the graph one by one,\n",
      "     |        iteratively, or to build a large edge list first and use that to\n",
      "     |        construct the graph. The latter approach is faster but it may\n",
      "     |        not be suitable if your dataset is large. The default is to\n",
      "     |        add the edges in a batch from an edge list.\n",
      "     |      @return: the graph that was constructed\n",
      "     |  \n",
      "     |  Formula = construct_graph_from_formula(cls, formula=None, attr='name', simplify=True) from __builtin__.type\n",
      "     |      Graph.Formula(formula = None, attr = \"name\", simplify = True)\n",
      "     |      \n",
      "     |      Generates a graph from a graph formula\n",
      "     |      \n",
      "     |      A graph formula is a simple string representation of a graph.\n",
      "     |      It is very handy for creating small graphs quickly. The string\n",
      "     |      consists of vertex names separated by edge operators. An edge\n",
      "     |      operator is a sequence of dashes (C{-}) that may or may not\n",
      "     |      start with an arrowhead (C{<} at the beginning of the sequence\n",
      "     |      or C{>} at the end of the sequence). The edge operators can\n",
      "     |      be arbitrarily long, i.e., you may use as many dashes to draw\n",
      "     |      them as you like. This makes a total of four different edge\n",
      "     |      operators:\n",
      "     |      \n",
      "     |        - C{-----} makes an undirected edge\n",
      "     |        - C{<----} makes a directed edge pointing from the vertex\n",
      "     |          on the right hand side of the operator to the vertex on\n",
      "     |          the left hand side\n",
      "     |        - C{---->} is the opposite of C{<----}\n",
      "     |        - C{<--->} creates a mutual directed edge pair between\n",
      "     |          the two vertices\n",
      "     |      \n",
      "     |      If you only use the undirected edge operator (C{-----}),\n",
      "     |      the graph will be undirected. Otherwise it will be directed.\n",
      "     |      Vertex names used in the formula will be assigned to the\n",
      "     |      C{name} vertex attribute of the graph.\n",
      "     |      \n",
      "     |      Some simple examples:\n",
      "     |      \n",
      "     |        >>> from igraph import Graph\n",
      "     |        >>> print Graph.Formula()           # empty graph\n",
      "     |        IGRAPH UN-- 0 0 --\n",
      "     |        + attr: name (v)\n",
      "     |        >>> g = Graph.Formula(\"A-B\")        # undirected graph\n",
      "     |        >>> g.vs[\"name\"]\n",
      "     |        ['A', 'B']\n",
      "     |        >>> print g\n",
      "     |        IGRAPH UN-- 2 1 --\n",
      "     |        + attr: name (v)\n",
      "     |        + edges (vertex names):\n",
      "     |        A--B\n",
      "     |        >>> g.get_edgelist()\n",
      "     |        [(0, 1)]\n",
      "     |        >>> g2 = Graph.Formula(\"A-----------B\")\n",
      "     |        >>> g2.isomorphic(g)\n",
      "     |        True\n",
      "     |        >>> g = Graph.Formula(\"A  --->  B\") # directed graph\n",
      "     |        >>> g.vs[\"name\"]\n",
      "     |        ['A', 'B']\n",
      "     |        >>> print g\n",
      "     |        IGRAPH DN-- 2 1 --\n",
      "     |        + attr: name (v)\n",
      "     |        + edges (vertex names):\n",
      "     |        A->B\n",
      "     |        \n",
      "     |      If you have may disconnected componnets, you can separate them\n",
      "     |      with commas. You can also specify isolated vertices:\n",
      "     |      \n",
      "     |        >>> g = Graph.Formula(\"A--B, C--D, E--F, G--H, I, J, K\")\n",
      "     |        >>> print \", \".join(g.vs[\"name\"])\n",
      "     |        A, B, C, D, E, F, G, H, I, J, K\n",
      "     |        >>> g.clusters().membership\n",
      "     |        [0, 0, 1, 1, 2, 2, 3, 3, 4, 5, 6]\n",
      "     |      \n",
      "     |      The colon (C{:}) operator can be used to specify vertex sets.\n",
      "     |      If an edge operator connects two vertex sets, then every vertex\n",
      "     |      from the first vertex set will be connected to every vertex in\n",
      "     |      the second set:\n",
      "     |      \n",
      "     |        >>> g = Graph.Formula(\"A:B:C:D --- E:F:G\")\n",
      "     |        >>> g.isomorphic(Graph.Full_Bipartite(4, 3))\n",
      "     |        True\n",
      "     |      \n",
      "     |      Note that you have to quote vertex names if they include spaces\n",
      "     |      or special characters:\n",
      "     |      \n",
      "     |        >>> g = Graph.Formula('\"this is\" +- \"a silly\" -+ \"graph here\"')\n",
      "     |        >>> g.vs[\"name\"]\n",
      "     |        ['this is', 'a silly', 'graph here']\n",
      "     |      \n",
      "     |      @param formula: the formula itself\n",
      "     |      @param attr: name of the vertex attribute where the vertex names\n",
      "     |                   will be stored\n",
      "     |      @param simplify: whether the simplify the constructed graph\n",
      "     |      @return: the constructed graph:\n",
      "     |  \n",
      "     |  Full_Bipartite(klass, *args, **kwds) from __builtin__.type\n",
      "     |      Full_Bipartite(n1, n2, directed=False, mode=ALL)\n",
      "     |      \n",
      "     |      Generates a full bipartite graph (directed or undirected, with or\n",
      "     |      without loops).\n",
      "     |      \n",
      "     |      >>> g = Graph.Full_Bipartite(2, 3)\n",
      "     |      >>> g.is_bipartite()\n",
      "     |      True\n",
      "     |      >>> g.vs[\"type\"]\n",
      "     |      [False, False, True, True, True]\n",
      "     |      \n",
      "     |      @param n1: the number of vertices of the first kind.\n",
      "     |      @param n2: the number of vertices of the second kind.\n",
      "     |      @param directed: whether tp generate a directed graph.\n",
      "     |      @param mode: if C{OUT}, then all vertices of the first kind are\n",
      "     |        connected to the others; C{IN} specifies the opposite direction,\n",
      "     |        C{ALL} creates mutual edges. Ignored for undirected graphs.\n",
      "     |      \n",
      "     |      @return: the graph with a binary vertex attribute named C{\"type\"} that\n",
      "     |        stores the vertex classes.\n",
      "     |  \n",
      "     |  GRG(klass, n, radius, torus=False) from __builtin__.type\n",
      "     |      GRG(n, radius, torus=False, return_coordinates=False)\n",
      "     |      \n",
      "     |      Generates a random geometric graph.\n",
      "     |      \n",
      "     |      The algorithm drops the vertices randomly on the 2D unit square and\n",
      "     |      connects them if they are closer to each other than the given radius.\n",
      "     |      The coordinates of the vertices are stored in the vertex attributes C{x}\n",
      "     |      and C{y}.\n",
      "     |      \n",
      "     |      @param n: The number of vertices in the graph\n",
      "     |      @param radius: The given radius\n",
      "     |      @param torus: This should be C{True} if we want to use a torus instead of a\n",
      "     |        square.\n",
      "     |  \n",
      "     |  Incidence(klass, *args, **kwds) from __builtin__.type\n",
      "     |      Incidence(matrix, directed=False, mode=ALL, multiple=False)\n",
      "     |      \n",
      "     |      Creates a bipartite graph from an incidence matrix.\n",
      "     |      \n",
      "     |      Example:\n",
      "     |      \n",
      "     |      >>> g = Graph.Incidence([[0, 1, 1], [1, 1, 0]])\n",
      "     |      \n",
      "     |      @param matrix: the incidence matrix.\n",
      "     |      @param directed: whether to create a directed graph.\n",
      "     |      @param mode: defines the direction of edges in the graph. If\n",
      "     |        C{OUT}, then edges go from vertices of the first kind\n",
      "     |        (corresponding to rows of the matrix) to vertices of the\n",
      "     |        second kind (the columns of the matrix). If C{IN}, the\n",
      "     |        opposite direction is used. C{ALL} creates mutual edges.\n",
      "     |        Ignored for undirected graphs.\n",
      "     |      @param multiple: defines what to do with non-zero entries in the\n",
      "     |        matrix. If C{False}, non-zero entries will create an edge no matter\n",
      "     |        what the value is. If C{True}, non-zero entries are rounded up to\n",
      "     |        the nearest integer and this will be the number of multiple edges\n",
      "     |        created.\n",
      "     |      \n",
      "     |      @return: the graph with a binary vertex attribute named C{\"type\"} that\n",
      "     |        stores the vertex classes.\n",
      "     |  \n",
      "     |  Load = Read(klass, f, format=None, *args, **kwds) from __builtin__.type\n",
      "     |      Unified reading function for graphs.\n",
      "     |      \n",
      "     |      This method tries to identify the format of the graph given in\n",
      "     |      the first parameter and calls the corresponding reader method.\n",
      "     |      \n",
      "     |      The remaining arguments are passed to the reader method without\n",
      "     |      any changes.\n",
      "     |      \n",
      "     |      @param f: the file containing the graph to be loaded\n",
      "     |      @param format: the format of the file (if known in advance).\n",
      "     |        C{None} means auto-detection. Possible values are: C{\"ncol\"}\n",
      "     |        (NCOL format), C{\"lgl\"} (LGL format), C{\"graphdb\"} (GraphDB\n",
      "     |        format), C{\"graphml\"}, C{\"graphmlz\"} (GraphML and gzipped\n",
      "     |        GraphML format), C{\"gml\"} (GML format), C{\"net\"}, C{\"pajek\"}\n",
      "     |        (Pajek format), C{\"dimacs\"} (DIMACS format), C{\"edgelist\"},\n",
      "     |        C{\"edges\"} or C{\"edge\"} (edge list), C{\"adjacency\"}\n",
      "     |        (adjacency matrix), C{\"dl\"} (DL format used by UCINET),\n",
      "     |        C{\"pickle\"} (Python pickled format),\n",
      "     |        C{\"picklez\"} (gzipped Python pickled format)\n",
      "     |      @raises IOError: if the file format can't be identified and\n",
      "     |        none was given.\n",
      "     |  \n",
      "     |  Random_Bipartite(klass, *args, **kwds) from __builtin__.type\n",
      "     |      Random_Bipartite(n1, n2, p=None, m=None, directed=False, neimode=ALL)\n",
      "     |      \n",
      "     |      Generates a random bipartite graph with the given number of vertices and\n",
      "     |      edges (if m is given), or with the given number of vertices and the given\n",
      "     |      connection probability (if p is given).\n",
      "     |      \n",
      "     |      If m is given but p is not, the generated graph will have n1 vertices of\n",
      "     |      type 1, n2 vertices of type 2 and m randomly selected edges between them. If\n",
      "     |      p is given but m is not, the generated graph will have n1 vertices of type 1\n",
      "     |      and n2 vertices of type 2, and each edge will exist between them with\n",
      "     |      probability p.\n",
      "     |      \n",
      "     |      @param n1: the number of vertices of type 1.\n",
      "     |      @param n2: the number of vertices of type 2.\n",
      "     |      @param p: the probability of edges. If given, C{m} must be missing.\n",
      "     |      @param m: the number of edges. If given, C{p} must be missing.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param neimode: if the graph is directed, specifies how the edges will be\n",
      "     |        generated. If it is C{\"all\"}, edges will be generated in both directions\n",
      "     |        (from type 1 to type 2 and vice versa) independently. If it is C{\"out\"}\n",
      "     |        edges will always point from type 1 to type 2. If it is C{\"in\"}, edges\n",
      "     |        will always point from type 2 to type 1. This argument is ignored for\n",
      "     |        undirected graphs.\n",
      "     |  \n",
      "     |  Read(klass, f, format=None, *args, **kwds) from __builtin__.type\n",
      "     |      Unified reading function for graphs.\n",
      "     |      \n",
      "     |      This method tries to identify the format of the graph given in\n",
      "     |      the first parameter and calls the corresponding reader method.\n",
      "     |      \n",
      "     |      The remaining arguments are passed to the reader method without\n",
      "     |      any changes.\n",
      "     |      \n",
      "     |      @param f: the file containing the graph to be loaded\n",
      "     |      @param format: the format of the file (if known in advance).\n",
      "     |        C{None} means auto-detection. Possible values are: C{\"ncol\"}\n",
      "     |        (NCOL format), C{\"lgl\"} (LGL format), C{\"graphdb\"} (GraphDB\n",
      "     |        format), C{\"graphml\"}, C{\"graphmlz\"} (GraphML and gzipped\n",
      "     |        GraphML format), C{\"gml\"} (GML format), C{\"net\"}, C{\"pajek\"}\n",
      "     |        (Pajek format), C{\"dimacs\"} (DIMACS format), C{\"edgelist\"},\n",
      "     |        C{\"edges\"} or C{\"edge\"} (edge list), C{\"adjacency\"}\n",
      "     |        (adjacency matrix), C{\"dl\"} (DL format used by UCINET),\n",
      "     |        C{\"pickle\"} (Python pickled format),\n",
      "     |        C{\"picklez\"} (gzipped Python pickled format)\n",
      "     |      @raises IOError: if the file format can't be identified and\n",
      "     |        none was given.\n",
      "     |  \n",
      "     |  Read_Adjacency(klass, f, sep=None, comment_char='#', attribute=None, *args, **kwds) from __builtin__.type\n",
      "     |      Constructs a graph based on an adjacency matrix from the given file\n",
      "     |      \n",
      "     |      Additional positional and keyword arguments not mentioned here are\n",
      "     |      passed intact to L{Graph.Adjacency}.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be read or a file object\n",
      "     |      @param sep: the string that separates the matrix elements in a row.\n",
      "     |        C{None} means an arbitrary sequence of whitespace characters.\n",
      "     |      @param comment_char: lines starting with this string are treated\n",
      "     |        as comments.\n",
      "     |      @param attribute: an edge attribute name where the edge weights are\n",
      "     |        stored in the case of a weighted adjacency matrix. If C{None},\n",
      "     |        no weights are stored, values larger than 1 are considered as\n",
      "     |        edge multiplicities.\n",
      "     |      @return: the created graph\n",
      "     |  \n",
      "     |  Read_DIMACS(cls, f, directed=False) from __builtin__.type\n",
      "     |      Read_DIMACS(f, directed=False)\n",
      "     |      \n",
      "     |      Reads a graph from a file conforming to the DIMACS minimum-cost flow\n",
      "     |      file format.\n",
      "     |      \n",
      "     |      For the exact definition of the format, see\n",
      "     |      U{http://lpsolve.sourceforge.net/5.5/DIMACS.htm}.\n",
      "     |      \n",
      "     |      Restrictions compared to the official description of the format are\n",
      "     |      as follows:\n",
      "     |      \n",
      "     |        - igraph's DIMACS reader requires only three fields in an arc\n",
      "     |          definition, describing the edge's source and target node and\n",
      "     |          its capacity.\n",
      "     |        - Source vertices are identified by 's' in the FLOW field, target\n",
      "     |          vertices are identified by 't'.\n",
      "     |        - Node indices start from 1. Only a single source and target node\n",
      "     |          is allowed.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |      @return: the generated graph. The indices of the source and target\n",
      "     |        vertices are attached as graph attributes C{source} and C{target},\n",
      "     |        the edge capacities are stored in the C{capacity} edge attribute.\n",
      "     |  \n",
      "     |  Read_GraphMLz(cls, f, *params, **kwds) from __builtin__.type\n",
      "     |      Read_GraphMLz(f, directed=True, index=0)\n",
      "     |      \n",
      "     |      Reads a graph from a zipped GraphML file.\n",
      "     |      \n",
      "     |      @param f: the name of the file\n",
      "     |      @param index: if the GraphML file contains multiple graphs,\n",
      "     |        specified the one that should be loaded. Graph indices\n",
      "     |        start from zero, so if you want to load the first graph,\n",
      "     |        specify 0 here.\n",
      "     |      @return: the loaded graph object\n",
      "     |  \n",
      "     |  Read_Pickle(klass, fname=None) from __builtin__.type\n",
      "     |      Reads a graph from Python pickled format\n",
      "     |      \n",
      "     |      @param fname: the name of the file, a stream to read from, or\n",
      "     |        a string containing the pickled data. The string is assumed to\n",
      "     |        hold pickled data if it is longer than 40 characters and\n",
      "     |        contains a substring that's peculiar to pickled versions\n",
      "     |        of an C{igraph} Graph object.\n",
      "     |      @return: the created graph object.\n",
      "     |  \n",
      "     |  Read_Picklez(klass, fname, *args, **kwds) from __builtin__.type\n",
      "     |      Reads a graph from compressed Python pickled format, uncompressing\n",
      "     |      it on-the-fly.\n",
      "     |      \n",
      "     |      @param fname: the name of the file or a stream to read from.\n",
      "     |      @return: the created graph object.\n",
      "     |  \n",
      "     |  TupleList(klass, edges, directed=False, vertex_name_attr='name', edge_attrs=None, weights=False) from __builtin__.type\n",
      "     |      Constructs a graph from a list-of-tuples representation.\n",
      "     |      \n",
      "     |      This representation assumes that the edges of the graph are encoded\n",
      "     |      in a list of tuples (or lists). Each item in the list must have at least\n",
      "     |      two elements, which specify the source and the target vertices of the edge.\n",
      "     |      The remaining elements (if any) specify the edge attributes of that edge,\n",
      "     |      where the names of the edge attributes originate from the C{edge_attrs}\n",
      "     |      list. The names of the vertices will be stored in the vertex attribute\n",
      "     |      given by C{vertex_name_attr}.\n",
      "     |      \n",
      "     |      The default parameters of this function are suitable for creating\n",
      "     |      unweighted graphs from lists where each item contains the source vertex\n",
      "     |      and the target vertex. If you have a weighted graph, you can use items\n",
      "     |      where the third item contains the weight of the edge by setting\n",
      "     |      C{edge_attrs} to C{\"weight\"} or C{[\"weight\"]}. If you have even more\n",
      "     |      edge attributes, add them to the end of each item in the C{edges}\n",
      "     |      list and also specify the corresponding edge attribute names in\n",
      "     |      C{edge_attrs} as a list.\n",
      "     |      \n",
      "     |      @param edges: the data source for the edges. This must be a list\n",
      "     |        where each item is a tuple (or list) containing at least two\n",
      "     |        items: the name of the source and the target vertex. Note that\n",
      "     |        names will be assigned to the C{name} vertex attribute (or another\n",
      "     |        vertex attribute if C{vertex_name_attr} is specified), even if\n",
      "     |        all the vertex names in the list are in fact numbers.\n",
      "     |      @param directed: whether the constructed graph will be directed\n",
      "     |      @param vertex_name_attr: the name of the vertex attribute that will\n",
      "     |        contain the vertex names.\n",
      "     |      @param edge_attrs: the names of the edge attributes that are filled\n",
      "     |        with the extra items in the edge list (starting from index 2, since\n",
      "     |        the first two items are the source and target vertices). C{None}\n",
      "     |        means that only the source and target vertices will be extracted\n",
      "     |        from each item. If you pass a string here, it will be wrapped in\n",
      "     |        a list for convenience.\n",
      "     |      @param weights: alternative way to specify that the graph is\n",
      "     |        weighted. If you set C{weights} to C{true} and C{edge_attrs} is\n",
      "     |        not given, it will be assumed that C{edge_attrs} is C{[\"weight\"]}\n",
      "     |        and igraph will parse the third element from each item into an\n",
      "     |        edge weight. If you set C{weights} to a string, it will be assumed\n",
      "     |        that C{edge_attrs} contains that string only, and igraph will\n",
      "     |        store the edge weights in that attribute.\n",
      "     |      @return: the graph that was constructed\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  es\n",
      "     |      The edge sequence of the graph\n",
      "     |  \n",
      "     |  vs\n",
      "     |      The vertex sequence of the graph\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from Graph:\n",
      "     |  \n",
      "     |  Adjacency(...)\n",
      "     |      Adjacency(matrix, mode=ADJ_DIRECTED)\n",
      "     |      \n",
      "     |      Generates a graph from its adjacency matrix.\n",
      "     |      \n",
      "     |      @param matrix: the adjacency matrix\n",
      "     |      @param mode: the mode to be used. Possible values are:\n",
      "     |      \n",
      "     |        - C{ADJ_DIRECTED} - the graph will be directed and a matrix\n",
      "     |          element gives the number of edges between two vertex.\n",
      "     |        - C{ADJ_UNDIRECTED} - alias to C{ADJ_MAX} for convenience.\n",
      "     |        - C{ADJ_MAX}   - undirected graph will be created and the number of\n",
      "     |          edges between vertex M{i} and M{j} is M{max(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_MIN}   - like C{ADJ_MAX}, but with M{min(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_PLUS}  - like C{ADJ_MAX}, but with M{A(i,j) + A(j,i)}\n",
      "     |        - C{ADJ_UPPER} - undirected graph with the upper right triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |        - C{ADJ_LOWER} - undirected graph with the lower left triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |      \n",
      "     |        These values can also be given as strings without the C{ADJ} prefix.\n",
      "     |  \n",
      "     |  Asymmetric_Preference(...)\n",
      "     |      Asymmetric_Preference(n, type_dist_matrix, pref_matrix, attribute=None, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on asymmetric vertex types and connection probabilities.\n",
      "     |      \n",
      "     |      This is the asymmetric variant of L{Graph.Preference}.\n",
      "     |      A given number of vertices are generated. Every vertex is assigned to an\n",
      "     |      \"incoming\" and an \"outgoing\" vertex type according to the given joint\n",
      "     |      type probabilities. Finally, every vertex pair is evaluated and a\n",
      "     |      directed edge is created between them with a probability depending on\n",
      "     |      the \"outgoing\" type of the source vertex and the \"incoming\" type of\n",
      "     |      the target vertex.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param type_dist_matrix: matrix giving the joint distribution of vertex\n",
      "     |        types\n",
      "     |      @param pref_matrix: matrix giving the connection probabilities for\n",
      "     |        different vertex types.\n",
      "     |      @param attribute: the vertex attribute name used to store the vertex\n",
      "     |        types. If C{None}, vertex types are not stored.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |  \n",
      "     |  Atlas(...)\n",
      "     |      Atlas(idx)\n",
      "     |      \n",
      "     |      Generates a graph from the Graph Atlas.\n",
      "     |      \n",
      "     |      @param idx: The index of the graph to be generated.\n",
      "     |        Indices start from zero, graphs are listed:\n",
      "     |      \n",
      "     |          1. in increasing order of number of vertices;\n",
      "     |          2. for a fixed number of vertices, in increasing order of the\n",
      "     |             number of edges;\n",
      "     |          3. for fixed numbers of vertices and edges, in increasing order\n",
      "     |             of the degree sequence, for example 111223 < 112222;\n",
      "     |          4. for fixed degree sequence, in increasing number of automorphisms.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: I{An Atlas of Graphs} by Ronald C. Read and Robin J. Wilson,\n",
      "     |        Oxford University Press, 1998.\n",
      "     |  \n",
      "     |  Barabasi(...)\n",
      "     |      Barabasi(n, m, outpref=False, directed=False, power=1,\n",
      "     |               zero_appeal=1, implementation=\"psumtree\", start_from=None)\n",
      "     |      \n",
      "     |      Generates a graph based on the Barabasi-Albert model.\n",
      "     |      \n",
      "     |      @param n: the number of vertices\n",
      "     |      @param m: either the number of outgoing edges generated for\n",
      "     |        each vertex or a list containing the number of outgoing\n",
      "     |        edges for each vertex explicitly.\n",
      "     |      @param outpref: C{True} if the out-degree of a given vertex\n",
      "     |        should also increase its citation probability (as well as\n",
      "     |        its in-degree), but it defaults to C{False}.\n",
      "     |      @param directed: C{True} if the generated graph should be\n",
      "     |        directed (default: C{False}).\n",
      "     |      @param power: the power constant of the nonlinear model.\n",
      "     |        It can be omitted, and in this case the usual linear model\n",
      "     |        will be used.\n",
      "     |      @param zero_appeal: the attractivity of vertices with degree\n",
      "     |        zero.\n",
      "     |      \n",
      "     |      @param implementation: the algorithm to use to generate the\n",
      "     |        network. Possible values are:\n",
      "     |      \n",
      "     |          - C{\"bag\"}: the algorithm that was the default in\n",
      "     |            igraph before 0.6. It works by putting the ids of the\n",
      "     |            vertices into a bag (multiset) exactly as many times\n",
      "     |            as their in-degree, plus once more. The required number\n",
      "     |            of cited vertices are then drawn from the bag with\n",
      "     |            replacement. It works only for I{power}=1 and\n",
      "     |            I{zero_appeal}=1.\n",
      "     |      \n",
      "     |          - C{\"psumtree\"}: this algorithm uses a partial prefix-sum\n",
      "     |            tree to generate the graph. It does not generate multiple\n",
      "     |            edges and it works for any values of I{power} and\n",
      "     |            I{zero_appeal}.\n",
      "     |      \n",
      "     |          - C{\"psumtree_multiple\"}: similar to C{\"psumtree\"}, but\n",
      "     |            it will generate multiple edges as well. igraph before\n",
      "     |            0.6 used this algorithm for I{power}s other than 1.\n",
      "     |      \n",
      "     |      @param start_from: if given and not C{None}, this must be another\n",
      "     |            L{Graph} object. igraph will use this graph as a starting\n",
      "     |            point for the preferential attachment model.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Barabasi, A-L and Albert, R. 1999. Emergence of scaling\n",
      "     |        in random networks. Science, 286 509-512.\n",
      "     |  \n",
      "     |  De_Bruijn(...)\n",
      "     |      De_Bruijn(m, n)\n",
      "     |      \n",
      "     |      Generates a de Bruijn graph with parameters (m, n)\n",
      "     |      \n",
      "     |      A de Bruijn graph represents relationships between strings. An alphabet\n",
      "     |      of M{m} letters are used and strings of length M{n} are considered.\n",
      "     |      A vertex corresponds to every possible string and there is a directed edge\n",
      "     |      from vertex M{v} to vertex M{w} if the string of M{v} can be transformed into\n",
      "     |      the string of M{w} by removing its first letter and appending a letter to it.\n",
      "     |      \n",
      "     |      Please note that the graph will have M{m^n} vertices and even more edges,\n",
      "     |      so probably you don't want to supply too big numbers for M{m} and M{n}.\n",
      "     |      \n",
      "     |      @param m: the size of the alphabet\n",
      "     |      @param n: the length of the strings\n",
      "     |  \n",
      "     |  Degree_Sequence(...)\n",
      "     |      Degree_Sequence(out, in=None, method=\"simple\")\n",
      "     |      \n",
      "     |      Generates a graph with a given degree sequence.\n",
      "     |      \n",
      "     |      @param out: the out-degree sequence for a directed graph. If the\n",
      "     |        in-degree sequence is omitted, the generated graph\n",
      "     |        will be undirected, so this will be the in-degree\n",
      "     |        sequence as well\n",
      "     |      @param in: the in-degree sequence for a directed graph.\n",
      "     |        If omitted, the generated graph will be undirected.\n",
      "     |      @param method: the generation method to be used. One of the following:\n",
      "     |        \n",
      "     |          - C{\"simple\"} -- simple generator that sometimes generates\n",
      "     |            loop edges and multiple edges. The generated graph is not\n",
      "     |            guaranteed to be connected.\n",
      "     |          - C{\"no_multiple\"} -- similar to C{\"simple\"} but avoids the\n",
      "     |            generation of multiple and loop edges at the expense of increased\n",
      "     |            time complexity. The method will re-start the generation every time\n",
      "     |            it gets stuck in a configuration where it is not possible to insert\n",
      "     |            any more edges without creating loops or multiple edges, and there\n",
      "     |            is no upper bound on the number of iterations, but it will succeed\n",
      "     |            eventually if the input degree sequence is graphical and throw an\n",
      "     |            exception if the input degree sequence is not graphical.\n",
      "     |          - C{\"vl\"} -- a more sophisticated generator that can sample\n",
      "     |            undirected, connected simple graphs uniformly. It uses\n",
      "     |            Monte-Carlo methods to randomize the graphs.\n",
      "     |            This generator should be favoured if undirected and connected\n",
      "     |            graphs are to be generated and execution time is not a concern.\n",
      "     |            igraph uses the original implementation of Fabien Viger; see the\n",
      "     |            following URL and the paper cited on it for the details of the\n",
      "     |            algorithm: U{http://www-rp.lip6.fr/~latapy/FV/generation.html}.\n",
      "     |  \n",
      "     |  Erdos_Renyi(...)\n",
      "     |      Erdos_Renyi(n, p, m, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on the Erdos-Renyi model.\n",
      "     |      \n",
      "     |      @param n: the number of vertices.\n",
      "     |      @param p: the probability of edges. If given, C{m} must be missing.\n",
      "     |      @param m: the number of edges. If given, C{p} must be missing.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether self-loops are allowed.\n",
      "     |  \n",
      "     |  Establishment(...)\n",
      "     |      Establishment(n, k, type_dist, pref_matrix, directed=False)\n",
      "     |      \n",
      "     |      Generates a graph based on a simple growing model with vertex types.\n",
      "     |      \n",
      "     |      A single vertex is added at each time step. This new vertex tries to\n",
      "     |      connect to k vertices in the graph. The probability that such a\n",
      "     |      connection is realized depends on the types of the vertices involved.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param k: the number of connections tried in each step\n",
      "     |      @param type_dist: list giving the distribution of vertex types\n",
      "     |      @param pref_matrix: matrix (list of lists) giving the connection\n",
      "     |        probabilities for different vertex types\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |  \n",
      "     |  Famous(...)\n",
      "     |      Famous(name)\n",
      "     |      \n",
      "     |      Generates a famous graph based on its name.\n",
      "     |      \n",
      "     |      Several famous graphs are known to C{igraph} including (but not limited to)\n",
      "     |      the Chvatal graph, the Petersen graph or the Tutte graph. This method\n",
      "     |      generates one of them based on its name (case insensitive). See the\n",
      "     |      documentation of the C interface of C{igraph} for the names available:\n",
      "     |      U{http://igraph.org/doc/c}.\n",
      "     |      \n",
      "     |      @param name: the name of the graph to be generated.\n",
      "     |  \n",
      "     |  Forest_Fire(...)\n",
      "     |      Forest_Fire(n, fw_prob, bw_factor=0.0, ambs=1, directed=False)\n",
      "     |      \n",
      "     |      Generates a graph based on the forest fire model\n",
      "     |      \n",
      "     |      The forest fire model is a growin graph model. In every time step, a new\n",
      "     |      vertex is added to the graph. The new vertex chooses an ambassador (or\n",
      "     |      more than one if M{ambs>1}) and starts a simulated forest fire at its\n",
      "     |      ambassador(s). The fire spreads through the edges. The spreading probability\n",
      "     |      along an edge is given by M{fw_prob}. The fire may also spread backwards\n",
      "     |      on an edge by probability M{fw_prob * bw_factor}. When the fire ended, the\n",
      "     |      newly added vertex connects to the vertices ``burned'' in the previous\n",
      "     |      fire.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param fw_prob: forward burning probability\n",
      "     |      @param bw_factor: ratio of backward and forward burning probability\n",
      "     |      @param ambs: number of ambassadors chosen in each step\n",
      "     |      @param directed: whether the graph will be directed\n",
      "     |  \n",
      "     |  Full(...)\n",
      "     |      Full(n, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a full graph (directed or undirected, with or without loops).\n",
      "     |      \n",
      "     |      @param n: the number of vertices.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether self-loops are allowed.\n",
      "     |  \n",
      "     |  Full_Citation(...)\n",
      "     |      Full_Citation(n, directed=False)\n",
      "     |      \n",
      "     |      Generates a full citation graph\n",
      "     |      \n",
      "     |      A full citation graph is a graph where the vertices are indexed from 0 to\n",
      "     |      M{n-1} and vertex M{i} has a directed edge towards all vertices with an\n",
      "     |      index less than M{i}.\n",
      "     |      \n",
      "     |      @param n: the number of vertices.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |  \n",
      "     |  Growing_Random(...)\n",
      "     |      Growing_Random(n, m, directed=False, citation=False)\n",
      "     |      \n",
      "     |      Generates a growing random graph.\n",
      "     |      \n",
      "     |      @param n: The number of vertices in the graph\n",
      "     |      @param m: The number of edges to add in each step (after adding a new vertex)\n",
      "     |      @param directed: whether the graph should be directed.\n",
      "     |      @param citation: whether the new edges should originate from the most\n",
      "     |         recently added vertex.\n",
      "     |  \n",
      "     |  Isoclass(...)\n",
      "     |      Isoclass(n, class, directed=False)\n",
      "     |      \n",
      "     |      Generates a graph with a given isomorphy class.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph (3 or 4)\n",
      "     |      @param class: the isomorphy class\n",
      "     |      @param directed: whether the graph should be directed.\n",
      "     |  \n",
      "     |  K_Regular(...)\n",
      "     |      K_Regular(n, k, directed=False, multiple=False)\n",
      "     |      \n",
      "     |      Generates a k-regular random graph\n",
      "     |      \n",
      "     |      A k-regular random graph is a random graph where each vertex has degree k.\n",
      "     |      If the graph is directed, both the in-degree and the out-degree of each vertex\n",
      "     |      will be k.\n",
      "     |      \n",
      "     |      @param n: The number of vertices in the graph\n",
      "     |      \n",
      "     |      @param k: The degree of each vertex if the graph is undirected, or the in-degree\n",
      "     |        and out-degree of each vertex if the graph is directed\n",
      "     |      @param directed: whether the graph should be directed.\n",
      "     |      @param multiple: whether it is allowed to create multiple edges.\n",
      "     |  \n",
      "     |  Kautz(...)\n",
      "     |      Kautz(m, n)\n",
      "     |      \n",
      "     |      Generates a Kautz graph with parameters (m, n)\n",
      "     |      \n",
      "     |      A Kautz graph is a labeled graph, vertices are labeled by strings\n",
      "     |      of length M{n+1} above an alphabet with M{m+1} letters, with\n",
      "     |      the restriction that every two consecutive letters in the string\n",
      "     |      must be different. There is a directed edge from a vertex M{v} to\n",
      "     |      another vertex M{w} if it is possible to transform the string of\n",
      "     |      M{v} into the string of M{w} by removing the first letter and\n",
      "     |      appending a letter to it.\n",
      "     |      \n",
      "     |      @param m: the size of the alphabet minus one\n",
      "     |      @param n: the length of the strings minus one\n",
      "     |  \n",
      "     |  LCF(...)\n",
      "     |      LCF(n, shifts, repeats)\n",
      "     |      \n",
      "     |      Generates a graph from LCF notation.\n",
      "     |      \n",
      "     |      LCF is short for Lederberg-Coxeter-Frucht, it is a concise notation\n",
      "     |      for 3-regular Hamiltonian graphs. It consists of three parameters,\n",
      "     |      the number of vertices in the graph, a list of shifts giving\n",
      "     |      additional edges to a cycle backbone and another integer giving how\n",
      "     |      many times the shifts should be performed. See\n",
      "     |      U{http://mathworld.wolfram.com/LCFNotation.html} for details.\n",
      "     |      \n",
      "     |      @param n: the number of vertices\n",
      "     |      @param shifts: the shifts in a list or tuple\n",
      "     |      @param repeats: the number of repeats\n",
      "     |  \n",
      "     |  Lattice(...)\n",
      "     |      Lattice(dim, nei=1, directed=False, mutual=True, circular=True)\n",
      "     |      \n",
      "     |      Generates a regular lattice.\n",
      "     |      \n",
      "     |      @param dim: list with the dimensions of the lattice\n",
      "     |      @param nei: value giving the distance (number of steps) within which\n",
      "     |         two vertices will be connected.\n",
      "     |      @param directed: whether to create a directed graph.\n",
      "     |      @param mutual: whether to create all connections as mutual\n",
      "     |          in case of a directed graph.\n",
      "     |      @param circular: whether the generated lattice is periodic.\n",
      "     |  \n",
      "     |  Preference(...)\n",
      "     |      Preference(n, type_dist, pref_matrix, attribute=None, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on vertex types and connection probabilities.\n",
      "     |      \n",
      "     |      This is practically the nongrowing variant of L{Graph.Establishment}.\n",
      "     |      A given number of vertices are generated. Every vertex is assigned to a\n",
      "     |      vertex type according to the given type probabilities. Finally, every\n",
      "     |      vertex pair is evaluated and an edge is created between them with a\n",
      "     |      probability depending on the types of the vertices involved.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param type_dist: list giving the distribution of vertex types\n",
      "     |      @param pref_matrix: matrix giving the connection probabilities for\n",
      "     |        different vertex types.\n",
      "     |      @param attribute: the vertex attribute name used to store the vertex\n",
      "     |        types. If C{None}, vertex types are not stored.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |  \n",
      "     |  Read_DL(...)\n",
      "     |      Read_DL(f, directed=True)\n",
      "     |      \n",
      "     |      Reads an UCINET DL file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |  \n",
      "     |  Read_Edgelist(...)\n",
      "     |      Read_Edgelist(f, directed=True)\n",
      "     |      \n",
      "     |      Reads an edge list from a file and creates a graph based on it.\n",
      "     |      \n",
      "     |      Please note that the vertex indices are zero-based.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |  \n",
      "     |  Read_GML(...)\n",
      "     |      Read_GML(f)\n",
      "     |      \n",
      "     |      Reads a GML file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |  \n",
      "     |  Read_GraphDB(...)\n",
      "     |      Read_GraphDB(f, directed=False)\n",
      "     |      \n",
      "     |      Reads a GraphDB format file and creates a graph based on it.\n",
      "     |      \n",
      "     |      GraphDB is a binary format, used in the graph database for\n",
      "     |      isomorphism testing (see U{http://amalfi.dis.unina.it/graph/}).\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |  \n",
      "     |  Read_GraphML(...)\n",
      "     |      Read_GraphML(f, directed=True, index=0)\n",
      "     |      \n",
      "     |      Reads a GraphML format file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param index: if the GraphML file contains multiple graphs,\n",
      "     |        specifies the one that should be loaded. Graph indices\n",
      "     |        start from zero, so if you want to load the first graph,\n",
      "     |        specify 0 here.\n",
      "     |  \n",
      "     |  Read_Lgl(...)\n",
      "     |      Read_Lgl(f, names=True, weights=\"if_present\", directed=True)\n",
      "     |      \n",
      "     |      Reads an .lgl file used by LGL.\n",
      "     |      \n",
      "     |      It is also useful for creating graphs from \"named\" (and\n",
      "     |      optionally weighted) edge lists.\n",
      "     |      \n",
      "     |      This format is used by the Large Graph Layout program. See the\n",
      "     |      U{documentation of LGL <http://bioinformatics.icmb.utexas.edu/lgl/>}\n",
      "     |      regarding the exact format description.\n",
      "     |      \n",
      "     |      LGL originally cannot deal with graphs containing multiple or loop\n",
      "     |      edges, but this condition is not checked here, as igraph is happy\n",
      "     |      with these.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param names: If C{True}, the vertex names are added as a\n",
      "     |        vertex attribute called 'name'.\n",
      "     |      @param weights: If True, the edge weights are added as an\n",
      "     |        edge attribute called 'weight', even if there are no\n",
      "     |        weights in the file. If False, the edge weights are never\n",
      "     |        added, even if they are present. C{\"auto\"} or C{\"if_present\"}\n",
      "     |        means that weights are added if there is at least one weighted\n",
      "     |        edge in the input file, but they are not added otherwise.\n",
      "     |      @param directed: whether the graph being created should be\n",
      "     |        directed\n",
      "     |  \n",
      "     |  Read_Ncol(...)\n",
      "     |      Read_Ncol(f, names=True, weights=\"if_present\", directed=True)\n",
      "     |      \n",
      "     |      Reads an .ncol file used by LGL.\n",
      "     |      \n",
      "     |      It is also useful for creating graphs from \"named\" (and\n",
      "     |      optionally weighted) edge lists.\n",
      "     |      \n",
      "     |      This format is used by the Large Graph Layout program. See the\n",
      "     |      U{documentation of LGL <http://bioinformatics.icmb.utexas.edu/lgl/>}\n",
      "     |      regarding the exact format description.\n",
      "     |      \n",
      "     |      LGL originally cannot deal with graphs containing multiple or loop\n",
      "     |      edges, but this condition is not checked here, as igraph is happy\n",
      "     |      with these.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param names: If C{True}, the vertex names are added as a\n",
      "     |        vertex attribute called 'name'.\n",
      "     |      @param weights: If True, the edge weights are added as an\n",
      "     |        edge attribute called 'weight', even if there are no\n",
      "     |        weights in the file. If False, the edge weights are never\n",
      "     |        added, even if they are present. C{\"auto\"} or C{\"if_present\"}\n",
      "     |        means that weights are added if there is at least one weighted\n",
      "     |        edge in the input file, but they are not added otherwise.\n",
      "     |      @param directed: whether the graph being created should be\n",
      "     |        directed\n",
      "     |  \n",
      "     |  Read_Pajek(...)\n",
      "     |      Read_Pajek(f)\n",
      "     |      \n",
      "     |      Reads a Pajek format file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |  \n",
      "     |  Recent_Degree(...)\n",
      "     |      Recent_Degree(n, m, window, outpref=False, directed=False, power=1)\n",
      "     |      \n",
      "     |      Generates a graph based on a stochastic model where the probability\n",
      "     |      of an edge gaining a new node is proportional to the edges gained in\n",
      "     |      a given time window.\n",
      "     |      \n",
      "     |      @param n: the number of vertices\n",
      "     |      @param m: either the number of outgoing edges generated for\n",
      "     |        each vertex or a list containing the number of outgoing\n",
      "     |        edges for each vertex explicitly.\n",
      "     |      @param window: size of the window in time steps\n",
      "     |      @param outpref: C{True} if the out-degree of a given vertex\n",
      "     |        should also increase its citation probability (as well as\n",
      "     |        its in-degree), but it defaults to C{False}.\n",
      "     |      @param directed: C{True} if the generated graph should be\n",
      "     |        directed (default: C{False}).\n",
      "     |      @param power: the power constant of the nonlinear model.\n",
      "     |        It can be omitted, and in this case the usual linear model\n",
      "     |        will be used.\n",
      "     |  \n",
      "     |  Ring(...)\n",
      "     |      Ring(n, directed=False, mutual=False, circular=True)\n",
      "     |      \n",
      "     |      Generates a ring graph.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the ring\n",
      "     |      @param directed: whether to create a directed ring.\n",
      "     |      @param mutual: whether to create mutual edges in a directed ring.\n",
      "     |      @param circular: whether to create a closed ring.\n",
      "     |  \n",
      "     |  SBM(...)\n",
      "     |      SBM(n, pref_matrix, block_sizes, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on a stochastic blockmodel.\n",
      "     |      \n",
      "     |      A given number of vertices are generated. Every vertex is assigned to a\n",
      "     |      vertex type according to the given block sizes. Vertices of the same\n",
      "     |      type will be assigned consecutive vertex IDs. Finally, every\n",
      "     |      vertex pair is evaluated and an edge is created between them with a\n",
      "     |      probability depending on the types of the vertices involved. The\n",
      "     |      probabilities are taken from the preference matrix.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param pref_matrix: matrix giving the connection probabilities for\n",
      "     |        different vertex types.\n",
      "     |      @param block_sizes: list giving the number of vertices in each block; must\n",
      "     |        sum up to I{n}.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |  \n",
      "     |  Star(...)\n",
      "     |      Star(n, mode=\"undirected\", center=0)\n",
      "     |      \n",
      "     |      Generates a star graph.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param mode: Gives the type of the star graph to create. Should be\n",
      "     |        either \"in\", \"out\", \"mutual\" or \"undirected\"\n",
      "     |      @param center: Vertex ID for the central vertex in the star.\n",
      "     |  \n",
      "     |  Static_Fitness(...)\n",
      "     |      Static_Fitness(m, fitness_out, fitness_in=None, loops=False, multiple=False)\n",
      "     |      \n",
      "     |      Generates a non-growing graph with edge probabilities proportional to node\n",
      "     |      fitnesses.\n",
      "     |      \n",
      "     |      The algorithm randomly selects vertex pairs and connects them until the given\n",
      "     |      number of edges are created. Each vertex is selected with a probability\n",
      "     |      proportional to its fitness; for directed graphs, a vertex is selected as a\n",
      "     |      source proportional to its out-fitness and as a target proportional to its\n",
      "     |      in-fitness.\n",
      "     |      \n",
      "     |      @param m: the number of edges in the graph\n",
      "     |      @param fitness_out: a numeric vector with non-negative entries, one for each\n",
      "     |        vertex. These values represent the fitness scores (out-fitness scores for\n",
      "     |        directed graphs). I{fitness} is an alias of this keyword argument.\n",
      "     |      @param fitness_in: a numeric vector with non-negative entries, one for each\n",
      "     |        vertex. These values represent the in-fitness scores for directed graphs.\n",
      "     |        For undirected graphs, this argument must be C{None}.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |      @param multiple: whether multiple edges are allowed.\n",
      "     |      @return: a directed or undirected graph with the prescribed power-law\n",
      "     |        degree distributions.\n",
      "     |  \n",
      "     |  Static_Power_Law(...)\n",
      "     |      Static_Power_Law(n, m, exponent_out, exponent_in=-1, loops=False,\n",
      "     |          multiple=False, finite_size_correction=True)\n",
      "     |      \n",
      "     |      Generates a non-growing graph with prescribed power-law degree distributions.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param m: the number of edges in the graph\n",
      "     |      @param exponent_out: the exponent of the out-degree distribution, which\n",
      "     |        must be between 2 and infinity (inclusive). When I{exponent_in} is\n",
      "     |        not given or negative, the graph will be undirected and this parameter\n",
      "     |        specifies the degree distribution. I{exponent} is an alias to this\n",
      "     |        keyword argument.\n",
      "     |      @param exponent_in: the exponent of the in-degree distribution, which\n",
      "     |        must be between 2 and infinity (inclusive) It can also be negative, in\n",
      "     |        which case an undirected graph will be generated.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |      @param multiple: whether multiple edges are allowed.\n",
      "     |      @param finite_size_correction: whether to apply a finite-size correction\n",
      "     |        to the generated fitness values for exponents less than 3. See the\n",
      "     |        paper of Cho et al for more details.\n",
      "     |      @return: a directed or undirected graph with the prescribed power-law\n",
      "     |        degree distributions.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Goh K-I, Kahng B, Kim D: Universal behaviour of load distribution\n",
      "     |        in scale-free networks. Phys Rev Lett 87(27):278701, 2001.\n",
      "     |      @ref: Cho YS, Kim JS, Park J, Kahng B, Kim D: Percolation transitions in\n",
      "     |        scale-free networks under the Achlioptas process. Phys Rev Lett\n",
      "     |        103:135702, 2009.\n",
      "     |  \n",
      "     |  Tree(...)\n",
      "     |      Tree(n, children, type=TREE_UNDIRECTED)\n",
      "     |      \n",
      "     |      Generates a tree in which almost all vertices have the same number of children.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param children: the number of children of a vertex in the graph\n",
      "     |      @param type: determines whether the tree should be directed, and if\n",
      "     |        this is the case, also its orientation. Must be one of\n",
      "     |        C{TREE_IN}, C{TREE_OUT} and C{TREE_UNDIRECTED}.\n",
      "     |  \n",
      "     |  Watts_Strogatz(...)\n",
      "     |      Watts_Strogatz(dim, size, nei, p, loops=False, multiple=False)\n",
      "     |      \n",
      "     |      @param dim: the dimension of the lattice\n",
      "     |      @param size: the size of the lattice along all dimensions\n",
      "     |      @param nei: value giving the distance (number of steps) within which\n",
      "     |         two vertices will be connected.\n",
      "     |      @param p: rewiring probability\n",
      "     |      \n",
      "     |      @param loops: specifies whether loop edges are allowed\n",
      "     |      @param multiple: specifies whether multiple edges are allowed\n",
      "     |      @see: L{Lattice()}, L{rewire()}, L{rewire_edges()} if more flexibility is\n",
      "     |        needed\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Duncan J Watts and Steven H Strogatz: I{Collective dynamics of\n",
      "     |        small world networks}, Nature 393, 440-442, 1998\n",
      "     |  \n",
      "     |  Weighted_Adjacency(...)\n",
      "     |      Weighted_Adjacency(matrix, mode=ADJ_DIRECTED, attr=\"weight\", loops=True)\n",
      "     |      \n",
      "     |      Generates a graph from its adjacency matrix.\n",
      "     |      \n",
      "     |      @param matrix: the adjacency matrix\n",
      "     |      @param mode: the mode to be used. Possible values are:\n",
      "     |      \n",
      "     |        - C{ADJ_DIRECTED} - the graph will be directed and a matrix\n",
      "     |          element gives the number of edges between two vertex.\n",
      "     |        - C{ADJ_UNDIRECTED} - alias to C{ADJ_MAX} for convenience.\n",
      "     |        - C{ADJ_MAX}   - undirected graph will be created and the number of\n",
      "     |          edges between vertex M{i} and M{j} is M{max(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_MIN}   - like C{ADJ_MAX}, but with M{min(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_PLUS}  - like C{ADJ_MAX}, but with M{A(i,j) + A(j,i)}\n",
      "     |        - C{ADJ_UPPER} - undirected graph with the upper right triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |        - C{ADJ_LOWER} - undirected graph with the lower left triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |      \n",
      "     |        These values can also be given as strings without the C{ADJ} prefix.\n",
      "     |      @param attr: the name of the edge attribute that stores the edge\n",
      "     |        weights.\n",
      "     |      @param loops: whether to include loop edges. When C{False}, the diagonal\n",
      "     |        of the adjacency matrix will be ignored.\n",
      "     |  \n",
      "     |  __and__(...)\n",
      "     |      x.__and__(y) <==> x&y\n",
      "     |  \n",
      "     |  __delitem__(...)\n",
      "     |      x.__delitem__(y) <==> del x[y]\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __invert__(...)\n",
      "     |      x.__invert__() <==> ~x\n",
      "     |  \n",
      "     |  __or__(...)\n",
      "     |      x.__or__(y) <==> x|y\n",
      "     |  \n",
      "     |  __rand__(...)\n",
      "     |      x.__rand__(y) <==> y&x\n",
      "     |  \n",
      "     |  __ror__(...)\n",
      "     |      x.__ror__(y) <==> y|x\n",
      "     |  \n",
      "     |  __setitem__(...)\n",
      "     |      x.__setitem__(i, y) <==> x[i]=y\n",
      "     |  \n",
      "     |  all_minimal_st_separators(...)\n",
      "     |      all_minimal_st_separators()\n",
      "     |      \n",
      "     |      Returns a list containing all the minimal s-t separators of a graph.\n",
      "     |      \n",
      "     |      A minimal separator is a set of vertices whose removal disconnects the graph,\n",
      "     |      while the removal of any subset of the set keeps the graph connected.\n",
      "     |      \n",
      "     |      @return: a list where each item lists the vertex indices of a given\n",
      "     |        minimal s-t separator.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Anne Berry, Jean-Paul Bordat and Olivier Cogis: Generating all the\n",
      "     |        minimal separators of a graph. In: Peter Widmayer, Gabriele Neyer and\n",
      "     |        Stephan Eidenbenz (eds.): Graph-theoretic concepts in computer science,\n",
      "     |        1665, 167--172, 1999. Springer.\n",
      "     |  \n",
      "     |  are_connected(...)\n",
      "     |      are_connected(v1, v2)\n",
      "     |      \n",
      "     |      Decides whether two given vertices are directly connected.\n",
      "     |      \n",
      "     |      @param v1: the ID or name of the first vertex\n",
      "     |      @param v2: the ID or name of the second vertex\n",
      "     |      @return: C{True} if there exists an edge from v1 to v2, C{False}\n",
      "     |        otherwise.\n",
      "     |  \n",
      "     |  articulation_points(...)\n",
      "     |      articulation_points()\n",
      "     |      \n",
      "     |      Returns the list of articulation points in the graph.\n",
      "     |      \n",
      "     |      A vertex is an articulation point if its removal increases the number of\n",
      "     |      connected components in the graph.\n",
      "     |  \n",
      "     |  assortativity(...)\n",
      "     |      assortativity(types1, types2=None, directed=True)\n",
      "     |      \n",
      "     |      Returns the assortativity of the graph based on numeric properties\n",
      "     |      of the vertices.\n",
      "     |      \n",
      "     |      This coefficient is basically the correlation between the actual\n",
      "     |      connectivity patterns of the vertices and the pattern expected from the\n",
      "     |      disribution of the vertex types.\n",
      "     |      \n",
      "     |      See equation (21) in Newman MEJ: Mixing patterns in networks, Phys Rev E\n",
      "     |      67:026126 (2003) for the proper definition. The actual calculation is\n",
      "     |      performed using equation (26) in the same paper for directed graphs, and\n",
      "     |      equation (4) in Newman MEJ: Assortative mixing in networks, Phys Rev Lett\n",
      "     |      89:208701 (2002) for undirected graphs.\n",
      "     |      \n",
      "     |      @param types1: vertex types in a list or the name of a vertex attribute\n",
      "     |        holding vertex types. Types are ideally denoted by numeric values.\n",
      "     |      @param types2: in directed assortativity calculations, each vertex can\n",
      "     |        have an out-type and an in-type. In this case, I{types1} contains the\n",
      "     |        out-types and this parameter contains the in-types in a list or the\n",
      "     |        name of a vertex attribute. If C{None}, it is assumed to be equal\n",
      "     |        to I{types1}.\n",
      "     |      \n",
      "     |      @param directed: whether to consider edge directions or not.\n",
      "     |      @return: the assortativity coefficient\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Newman MEJ: Mixing patterns in networks, Phys Rev E 67:026126, 2003.\n",
      "     |      @ref: Newman MEJ: Assortative mixing in networks, Phys Rev Lett 89:208701,\n",
      "     |        2002.\n",
      "     |      @see: L{assortativity_degree()} when the types are the vertex degrees\n",
      "     |  \n",
      "     |  assortativity_degree(...)\n",
      "     |      assortativity_degree(directed=True)\n",
      "     |      \n",
      "     |      Returns the assortativity of a graph based on vertex degrees.\n",
      "     |      \n",
      "     |      See L{assortativity()} for the details. L{assortativity_degree()} simply\n",
      "     |      calls L{assortativity()} with the vertex degrees as types.\n",
      "     |      \n",
      "     |      @param directed: whether to consider edge directions for directed graphs\n",
      "     |        or not. This argument is ignored for undirected graphs.\n",
      "     |      @return: the assortativity coefficient\n",
      "     |      \n",
      "     |      @see: L{assortativity()}\n",
      "     |  \n",
      "     |  assortativity_nominal(...)\n",
      "     |      assortativity_nominal(types, directed=True)\n",
      "     |      \n",
      "     |      Returns the assortativity of the graph based on vertex categories.\n",
      "     |      \n",
      "     |      Assuming that the vertices belong to different categories, this\n",
      "     |      function calculates the assortativity coefficient, which specifies\n",
      "     |      the extent to which the connections stay within categories. The\n",
      "     |      assortativity coefficient is one if all the connections stay within\n",
      "     |      categories and minus one if all the connections join vertices of\n",
      "     |      different categories. For a randomly connected network, it is\n",
      "     |      asymptotically zero.\n",
      "     |      \n",
      "     |      See equation (2) in Newman MEJ: Mixing patterns in networks, Phys Rev E\n",
      "     |      67:026126 (2003) for the proper definition.\n",
      "     |      \n",
      "     |      @param types: vertex types in a list or the name of a vertex attribute\n",
      "     |        holding vertex types. Types should be denoted by numeric values.\n",
      "     |      @param directed: whether to consider edge directions or not.\n",
      "     |      @return: the assortativity coefficient\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Newman MEJ: Mixing patterns in networks, Phys Rev E 67:026126, 2003.\n",
      "     |  \n",
      "     |  attributes(...)\n",
      "     |      attributes()\n",
      "     |      \n",
      "     |      @return: the attribute name list of the graph\n",
      "     |  \n",
      "     |  authority_score(...)\n",
      "     |      authority_score(weights=None, scale=True, arpack_options=None, return_eigenvalue=False)\n",
      "     |      \n",
      "     |      Calculates Kleinberg's authority score for the vertices of the graph\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param scale: whether to normalize the scores so that the largest one\n",
      "     |        is 1.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @param return_eigenvalue: whether to return the largest eigenvalue\n",
      "     |      @return: the authority scores in a list and optionally the largest eigenvalue\n",
      "     |        as a second member of a tuple\n",
      "     |      \n",
      "     |      @see: hub_score()\n",
      "     |  \n",
      "     |  average_path_length(...)\n",
      "     |      average_path_length(directed=True, unconn=True)\n",
      "     |      \n",
      "     |      Calculates the average path length in a graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths in case of a\n",
      "     |        directed graph. Ignored for undirected graphs.\n",
      "     |      @param unconn: what to do when the graph is unconnected. If C{True},\n",
      "     |        the average of the geodesic lengths in the components is\n",
      "     |        calculated. Otherwise for all unconnected vertex pairs,\n",
      "     |        a path length equal to the number of vertices is used.\n",
      "     |      @return: the average path length in the graph\n",
      "     |  \n",
      "     |  betweenness(...)\n",
      "     |      betweenness(vertices=None, directed=True, cutoff=None, weights=None, nobigint=True)\n",
      "     |      \n",
      "     |      Calculates or estimates the betweenness of vertices in a graph.\n",
      "     |      \n",
      "     |      Keyword arguments:\n",
      "     |      @param vertices: the vertices for which the betweennesses must be returned.\n",
      "     |        If C{None}, assumes all of the vertices in the graph.\n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param cutoff: if it is an integer, only paths less than or equal to this\n",
      "     |        length are considered, effectively resulting in an estimation of the\n",
      "     |        betweenness for the given vertices. If C{None}, the exact betweenness is\n",
      "     |        returned.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param nobigint: if C{True}, igraph uses the longest available integer\n",
      "     |        type on the current platform to count shortest paths. For some large\n",
      "     |        networks that have a specific structure, the counters may overflow.\n",
      "     |        To prevent this, use C{nobigint=False}, which forces igraph to use\n",
      "     |        arbitrary precision integers at the expense of increased computation\n",
      "     |        time.\n",
      "     |      @return: the (possibly estimated) betweenness of the given vertices in a list\n",
      "     |  \n",
      "     |  bfs(...)\n",
      "     |      bfs(vid, mode=OUT)\n",
      "     |      \n",
      "     |      Conducts a breadth first search (BFS) on the graph.\n",
      "     |      \n",
      "     |      @param vid: the root vertex ID\n",
      "     |      @param mode: either L{IN} or L{OUT} or L{ALL}, ignored\n",
      "     |        for undirected graphs.\n",
      "     |      @return: a tuple with the following items:\n",
      "     |         - The vertex IDs visited (in order)\n",
      "     |         - The start indices of the layers in the vertex list\n",
      "     |         - The parent of every vertex in the BFS\n",
      "     |  \n",
      "     |  bfsiter(...)\n",
      "     |      bfsiter(vid, mode=OUT, advanced=False)\n",
      "     |      \n",
      "     |      Constructs a breadth first search (BFS) iterator of the graph.\n",
      "     |      \n",
      "     |      @param vid: the root vertex ID\n",
      "     |      @param mode: either L{IN} or L{OUT} or L{ALL}.\n",
      "     |      @param advanced: if C{False}, the iterator returns the next\n",
      "     |        vertex in BFS order in every step. If C{True}, the iterator\n",
      "     |        returns the distance of the vertex from the root and the\n",
      "     |        parent of the vertex in the BFS tree as well.\n",
      "     |      @return: the BFS iterator as an L{igraph.BFSIter} object.\n",
      "     |  \n",
      "     |  bibcoupling(...)\n",
      "     |      bibcoupling(vertices=None)\n",
      "     |      \n",
      "     |      Calculates bibliographic coupling scores for given vertices in a graph.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None}, all vertices\n",
      "     |        will be considered.\n",
      "     |      @return: bibliographic coupling scores for all given vertices in a matrix.\n",
      "     |  \n",
      "     |  canonical_permutation(...)\n",
      "     |      canonical_permutation(sh=\"fm\")\n",
      "     |      \n",
      "     |      Calculates the canonical permutation of a graph using the BLISS isomorphism\n",
      "     |      algorithm.\n",
      "     |      \n",
      "     |      Passing the permutation returned here to L{Graph.permute_vertices()} will\n",
      "     |      transform the graph into its canonical form.\n",
      "     |      \n",
      "     |      See U{http://www.tcs.hut.fi/Software/bliss/index.html} for more information\n",
      "     |      about the BLISS algorithm and canonical permutations.\n",
      "     |      \n",
      "     |      @param sh: splitting heuristics for graph as a case-insensitive string,\n",
      "     |        with the following possible values:\n",
      "     |      \n",
      "     |          - C{\"f\"}: first non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fl\"}: first largest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fs\"}: first smallest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fm\"}: first maximally non-trivially connected non-singleton\n",
      "     |            cell\n",
      "     |      \n",
      "     |          - C{\"flm\"}: largest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fsm\"}: smallest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |      @return: a permutation vector containing vertex IDs. Vertex 0 in the original\n",
      "     |        graph will be mapped to an ID contained in the first element of this\n",
      "     |        vector; vertex 1 will be mapped to the second and so on.\n",
      "     |  \n",
      "     |  clique_number(...)\n",
      "     |      clique_number()\n",
      "     |      \n",
      "     |      Returns the clique number of the graph.\n",
      "     |      \n",
      "     |      The clique number of the graph is the size of the largest clique.\n",
      "     |      \n",
      "     |      @see: L{largest_cliques()} for the largest cliques.\n",
      "     |  \n",
      "     |  cliques(...)\n",
      "     |      cliques(min=0, max=0)\n",
      "     |      \n",
      "     |      Returns some or all cliques of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      A clique is a complete subgraph -- a set of vertices where an edge\n",
      "     |      is present between any two of them (excluding loops)\n",
      "     |      \n",
      "     |      @param min: the minimum size of cliques to be returned. If zero or\n",
      "     |        negative, no lower bound will be used.\n",
      "     |      @param max: the maximum size of cliques to be returned. If zero or\n",
      "     |        negative, no upper bound will be used.\n",
      "     |  \n",
      "     |  closeness(...)\n",
      "     |      closeness(vertices=None, mode=ALL, cutoff=None, weights=None,\n",
      "     |                normalized=True)\n",
      "     |      \n",
      "     |      Calculates the closeness centralities of given vertices in a graph.\n",
      "     |      \n",
      "     |      The closeness centerality of a vertex measures how easily other\n",
      "     |      vertices can be reached from it (or the other way: how easily it\n",
      "     |      can be reached from the other vertices). It is defined as the\n",
      "     |      number of the number of vertices minus one divided by the sum of\n",
      "     |      the lengths of all geodesics from/to the given vertex.\n",
      "     |      \n",
      "     |      If the graph is not connected, and there is no path between two\n",
      "     |      vertices, the number of vertices is used instead the length of\n",
      "     |      the geodesic. This is always longer than the longest possible\n",
      "     |      geodesic.\n",
      "     |      \n",
      "     |      @param vertices: the vertices for which the closenesses must\n",
      "     |        be returned. If C{None}, uses all of the vertices in the graph.\n",
      "     |      @param mode: must be one of L{IN}, L{OUT} and L{ALL}. L{IN} means\n",
      "     |        that the length of the incoming paths, L{OUT} means that the\n",
      "     |        length of the outgoing paths must be calculated. L{ALL} means\n",
      "     |        that both of them must be calculated.\n",
      "     |      @param cutoff: if it is an integer, only paths less than or equal to this\n",
      "     |        length are considered, effectively resulting in an estimation of the\n",
      "     |        closeness for the given vertices (which is always an underestimation of the\n",
      "     |        real closeness, since some vertex pairs will appear as disconnected even\n",
      "     |        though they are connected).. If C{None}, the exact closeness is\n",
      "     |        returned.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param normalized: Whether to normalize the raw closeness scores by\n",
      "     |        multiplying by the number of vertices minus one.\n",
      "     |      @return: the calculated closenesses in a list\n",
      "     |  \n",
      "     |  cocitation(...)\n",
      "     |      cocitation(vertices=None)\n",
      "     |      \n",
      "     |      Calculates cocitation scores for given vertices in a graph.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None}, all vertices\n",
      "     |        will be considered.\n",
      "     |      @return: cocitation scores for all given vertices in a matrix.\n",
      "     |  \n",
      "     |  complementer(...)\n",
      "     |      complementer(loops=False)\n",
      "     |      \n",
      "     |      Returns the complementer of the graph\n",
      "     |      \n",
      "     |      @param loops: whether to include loop edges in the complementer.\n",
      "     |      @return: the complementer of the graph\n",
      "     |  \n",
      "     |  compose(...)\n",
      "     |      compose(other)\n",
      "     |      \n",
      "     |      Returns the composition of two graphs.\n",
      "     |  \n",
      "     |  constraint(...)\n",
      "     |      constraint(vertices=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates Burt's constraint scores for given vertices in a graph.\n",
      "     |      \n",
      "     |      Burt's constraint is higher if ego has less, or mutually stronger\n",
      "     |      related (i.e. more redundant) contacts. Burt's measure of\n",
      "     |      constraint, C[i], of vertex i's ego network V[i], is defined for\n",
      "     |      directed and valued graphs as follows:\n",
      "     |      \n",
      "     |      C[i] = sum( sum( (p[i,q] p[q,j])^2, q in V[i], q != i,j ), j in V[], j != i)\n",
      "     |      \n",
      "     |      for a graph of order (ie. number od vertices) N, where proportional\n",
      "     |      tie strengths are defined as follows:\n",
      "     |      \n",
      "     |      p[i,j]=(a[i,j]+a[j,i]) / sum(a[i,k]+a[k,i], k in V[i], k != i),\n",
      "     |      a[i,j] are elements of A and the latter being the graph adjacency matrix.\n",
      "     |      \n",
      "     |      For isolated vertices, constraint is undefined.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed or C{None} for all vertices.\n",
      "     |      @param weights: weights associated to the edges. Can be an attribute name\n",
      "     |        as well. If C{None}, every edge will have the same weight.\n",
      "     |      @return: constraint scores for all given vertices in a matrix.\n",
      "     |  \n",
      "     |  contract_vertices(...)\n",
      "     |      contract_vertices(mapping, combine_attrs=None)\n",
      "     |      \n",
      "     |      Contracts some vertices in the graph, i.e. replaces groups of vertices\n",
      "     |      with single vertices. Edges are not affected.\n",
      "     |      \n",
      "     |      @param mapping: numeric vector which gives the mapping between old and\n",
      "     |        new vertex IDs. Vertices having the same new vertex ID in this vector\n",
      "     |        will be remapped into a single new vertex. It is safe to pass the\n",
      "     |        membership vector of a L{VertexClustering} object here.\n",
      "     |      @param combine_attrs: specifies how to combine the attributes of\n",
      "     |        the vertices being collapsed into a single one. If it is C{None},\n",
      "     |        all the attributes will be lost. If it is a function, the\n",
      "     |        attributes of the vertices will be collected and passed on to\n",
      "     |        that function which will return the new attribute value that has to\n",
      "     |        be assigned to the single collapsed vertex. It can also be one of\n",
      "     |        the following string constants which define built-in collapsing\n",
      "     |        functions: C{sum}, C{prod}, C{mean}, C{median}, C{max}, C{min},\n",
      "     |        C{first}, C{last}, C{random}. You can also specify different\n",
      "     |        combination functions for different attributes by passing a dict\n",
      "     |        here which maps attribute names to functions. See\n",
      "     |        L{Graph.simplify()} for more details.\n",
      "     |      @return: C{None}.\n",
      "     |      @see: L{Graph.simplify()}\n",
      "     |  \n",
      "     |  convergence_degree(...)\n",
      "     |      convergence_degree()\n",
      "     |      \n",
      "     |      Undocumented (yet).\n",
      "     |  \n",
      "     |  convergence_field_size(...)\n",
      "     |      convergence_field_size()\n",
      "     |      \n",
      "     |      Undocumented (yet).\n",
      "     |  \n",
      "     |  copy(...)\n",
      "     |      copy()\n",
      "     |      \n",
      "     |      Creates an exact deep copy of the graph.\n",
      "     |  \n",
      "     |  coreness(...)\n",
      "     |      coreness(mode=ALL)\n",
      "     |      \n",
      "     |      Finds the coreness (shell index) of the vertices of the network.\n",
      "     |      \n",
      "     |      The M{k}-core of a graph is a maximal subgraph in which each vertex\n",
      "     |      has at least degree k. (Degree here means the degree in the\n",
      "     |      subgraph of course). The coreness of a vertex is M{k} if it\n",
      "     |      is a member of the M{k}-core but not a member of the M{k+1}-core.\n",
      "     |      \n",
      "     |      @param mode: whether to compute the in-corenesses (L{IN}), the\n",
      "     |        out-corenesses (L{OUT}) or the undirected corenesses (L{ALL}).\n",
      "     |        Ignored and assumed to be L{ALL} for undirected graphs.\n",
      "     |      @return: the corenesses for each vertex.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Vladimir Batagelj, Matjaz Zaversnik: I{An M{O(m)} Algorithm\n",
      "     |        for Core Decomposition of Networks.}\n",
      "     |  \n",
      "     |  count_isomorphisms_vf2(...)\n",
      "     |      count_isomorphisms_vf2(other=None, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Determines the number of isomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph. If C{None}, the number of automorphisms\n",
      "     |        will be returned.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: the number of isomorphisms between the two given graphs (or the\n",
      "     |        number of automorphisms if C{other} is C{None}.\n",
      "     |  \n",
      "     |  count_multiple(...)\n",
      "     |      count_multiple(edges=None)\n",
      "     |      \n",
      "     |      Counts the multiplicities of the given edges.\n",
      "     |      \n",
      "     |      @param edges: edge indices for which we want to count their\n",
      "     |        multiplicity. If C{None}, all edges are counted.\n",
      "     |      @return: the multiplicities of the given edges as a list.\n",
      "     |  \n",
      "     |  count_subisomorphisms_vf2(...)\n",
      "     |      count_subisomorphisms_vf2(other, color1=None, color2=None,\n",
      "     |        edge_color1=None, edge_color2=None, node_compat_fn=None,\n",
      "     |        edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Determines the number of subisomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: the number of subisomorphisms between the two given graphs\n",
      "     |  \n",
      "     |  decompose(...)\n",
      "     |      decompose(mode=STRONG, maxcompno=None, minelements=1)\n",
      "     |      \n",
      "     |      Decomposes the graph into subgraphs.\n",
      "     |      \n",
      "     |      @param mode: must be either STRONG or WEAK, depending on the\n",
      "     |        clusters being sought.\n",
      "     |      @param maxcompno: maximum number of components to return.\n",
      "     |        C{None} means all possible components.\n",
      "     |      @param minelements: minimum number of vertices in a component.\n",
      "     |        By setting this to 2, isolated vertices are not returned\n",
      "     |        as separate components.\n",
      "     |      @return: a list of the subgraphs. Every returned subgraph is a\n",
      "     |        copy of the original.\n",
      "     |  \n",
      "     |  degree(...)\n",
      "     |      degree(vertices, mode=ALL, loops=True)\n",
      "     |      \n",
      "     |      Returns some vertex degrees from the graph.\n",
      "     |      \n",
      "     |      This method accepts a single vertex ID or a list of vertex IDs as a\n",
      "     |      parameter, and returns the degree of the given vertices (in the\n",
      "     |      form of a single integer or a list, depending on the input\n",
      "     |      parameter).\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @param mode: the type of degree to be returned (L{OUT} for\n",
      "     |        out-degrees, L{IN} IN for in-degrees or L{ALL} for the sum of\n",
      "     |        them).\n",
      "     |      @param loops: whether self-loops should be counted.\n",
      "     |  \n",
      "     |  delete_vertices(...)\n",
      "     |      delete_vertices(vs)\n",
      "     |      \n",
      "     |      Deletes vertices and all its edges from the graph.\n",
      "     |      \n",
      "     |      @param vs: a single vertex ID or the list of vertex IDs\n",
      "     |        to be deleted.\n",
      "     |  \n",
      "     |  density(...)\n",
      "     |      density(loops=False)\n",
      "     |      \n",
      "     |      Calculates the density of the graph.\n",
      "     |      \n",
      "     |      @param loops: whether to take loops into consideration. If C{True},\n",
      "     |        the algorithm assumes that there might be some loops in the graph\n",
      "     |        and calculates the density accordingly. If C{False}, the algorithm\n",
      "     |        assumes that there can't be any loops.\n",
      "     |      @return: the reciprocity of the graph.\n",
      "     |  \n",
      "     |  diameter(...)\n",
      "     |      diameter(directed=True, unconn=True, weights=None)\n",
      "     |      \n",
      "     |      Calculates the diameter of the graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param unconn: if C{True} and the graph is unconnected, the\n",
      "     |        longest geodesic within a component will be returned. If\n",
      "     |        C{False} and the graph is unconnected, the result is the\n",
      "     |        number of vertices if there are no weights or infinity\n",
      "     |        if there are weights.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the diameter\n",
      "     |  \n",
      "     |  difference(...)\n",
      "     |      difference(other)\n",
      "     |      \n",
      "     |      Subtracts the given graph from the original\n",
      "     |  \n",
      "     |  disjoint_union(...)\n",
      "     |      disjoint_union(graphs)\n",
      "     |      \n",
      "     |      Creates the disjoint union of two (or more) graphs.\n",
      "     |      \n",
      "     |      @param graphs: the list of graphs to be united with the current one.\n",
      "     |  \n",
      "     |  diversity(...)\n",
      "     |      diversity(vertices=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates the structural diversity index of the vertices.\n",
      "     |      \n",
      "     |      The structural diversity index of a vertex is simply the (normalized)\n",
      "     |      Shannon entropy of the weights of the edges incident on the vertex.\n",
      "     |      \n",
      "     |      The measure is defined for undirected graphs only; edge directions are\n",
      "     |      ignored.\n",
      "     |      \n",
      "     |      @param vertices: the vertices for which the diversity indices must\n",
      "     |        be returned. If C{None}, uses all of the vertices in the graph.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the calculated diversity indices in a list, or a single number if\n",
      "     |        a single vertex was supplied.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Eagle N, Macy M and Claxton R: Network diversity and economic\n",
      "     |        development, Science 328, 1029--1031, 2010.\n",
      "     |  \n",
      "     |  eccentricity(...)\n",
      "     |      eccentricity(vertices=None, mode=ALL)\n",
      "     |      \n",
      "     |      Calculates the eccentricities of given vertices in a graph.\n",
      "     |      \n",
      "     |      The eccentricity of a vertex is calculated by measuring the\n",
      "     |      shortest distance from (or to) the vertex, to (or from) all other\n",
      "     |      vertices in the graph, and taking the maximum.\n",
      "     |      \n",
      "     |      @param vertices: the vertices for which the eccentricity scores must\n",
      "     |        be returned. If C{None}, uses all of the vertices in the graph.\n",
      "     |      @param mode: must be one of L{IN}, L{OUT} and L{ALL}. L{IN} means\n",
      "     |        that edge directions are followed; C{OUT} means that edge directions\n",
      "     |        are followed the opposite direction; C{ALL} means that directions are\n",
      "     |        ignored. The argument has no effect for undirected graphs.\n",
      "     |      @return: the calculated eccentricities in a list, or a single number if\n",
      "     |        a single vertex was supplied.\n",
      "     |  \n",
      "     |  ecount(...)\n",
      "     |      ecount()\n",
      "     |      \n",
      "     |      Counts the number of edges.\n",
      "     |      @return: the number of edges in the graph.\n",
      "     |      @rtype: integer\n",
      "     |  \n",
      "     |  edge_attributes(...)\n",
      "     |      edge_attributes()\n",
      "     |      \n",
      "     |      @return: the attribute name list of the graph's edges\n",
      "     |  \n",
      "     |  edge_betweenness(...)\n",
      "     |      edge_betweenness(directed=True, cutoff=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates or estimates the edge betweennesses in a graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param cutoff: if it is an integer, only paths less than or equal to this\n",
      "     |        length are considered, effectively resulting in an estimation of the\n",
      "     |        betweenness values. If C{None}, the exact betweennesses are\n",
      "     |        returned.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: a list with the (exact or estimated) edge betweennesses of all\n",
      "     |        edges.\n",
      "     |  \n",
      "     |  edge_connectivity(...)\n",
      "     |      edge_connectivity(source=-1, target=-1, checks=True)\n",
      "     |      \n",
      "     |      Calculates the edge connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The edge connectivity between two given vertices is the number of edges\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of edge disjoint directed\n",
      "     |      paths between the vertices. The edge connectivity of the graph is the minimal\n",
      "     |      edge connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the edge connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall edge connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @return: the edge connectivity\n",
      "     |  \n",
      "     |  eigen_adjacency(...)\n",
      "     |  \n",
      "     |  eigenvector_centrality(...)\n",
      "     |      eigenvector_centrality(directed=True, scale=True, weights=None, return_eigenvalue=False, arpack_options=None)\n",
      "     |      \n",
      "     |      Calculates the eigenvector centralities of the vertices in a graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider edge directions in a directed\n",
      "     |        graph. Ignored for undirected graphs.\n",
      "     |      @param scale: whether to normalize the centralities so the largest\n",
      "     |        one will always be 1.\n",
      "     |      @param weights: edge weights given as a list or an edge attribute. If\n",
      "     |        C{None}, all edges have equal weight.\n",
      "     |      @param return_eigenvalue: whether to return the actual largest\n",
      "     |        eigenvalue along with the centralities\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object that can be used\n",
      "     |        to fine-tune the calculation. If it is omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @return: the eigenvector centralities in a list and optionally the\n",
      "     |        largest eigenvalue (as a second member of a tuple)\n",
      "     |  \n",
      "     |  farthest_points(...)\n",
      "     |      farthest_points(directed=True, unconn=True, weights=None)\n",
      "     |      \n",
      "     |      Returns two vertex IDs whose distance equals the actual diameter\n",
      "     |      of the graph.\n",
      "     |      \n",
      "     |      If there are many shortest paths with the length of the diameter,\n",
      "     |      it returns the first one it found.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param unconn: if C{True} and the graph is unconnected, the\n",
      "     |        longest geodesic within a component will be returned. If\n",
      "     |        C{False} and the graph is unconnected, the result contains the\n",
      "     |        number of vertices if there are no weights or infinity\n",
      "     |        if there are weights.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: a triplet containing the two vertex IDs and their distance.\n",
      "     |        The IDs are C{None} if the graph is unconnected and C{unconn}\n",
      "     |        is C{False}.\n",
      "     |  \n",
      "     |  feedback_arc_set(...)\n",
      "     |      feedback_arc_set(weights=None, method=\"eades\")\n",
      "     |      \n",
      "     |      Calculates an approximately or exactly minimal feedback arc set.\n",
      "     |      \n",
      "     |      A feedback arc set is a set of edges whose removal makes the graph acyclic.\n",
      "     |      Since this is always possible by removing all the edges, we are in general\n",
      "     |      interested in removing the smallest possible number of edges, or an edge set\n",
      "     |      with as small total weight as possible. This method calculates one such edge\n",
      "     |      set. Note that the task is trivial for an undirected graph as it is enough\n",
      "     |      to find a spanning tree and then remove all the edges not in the spanning\n",
      "     |      tree. Of course it is more complicated for directed graphs.\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name. When given, the algorithm will strive to\n",
      "     |        remove lightweight edges in order to minimize the total weight of the\n",
      "     |        feedback arc set.\n",
      "     |      @param method: the algorithm to use. C{\"eades\"} uses the greedy cycle\n",
      "     |        breaking heuristic of Eades, Lin and Smyth, which is linear in the number\n",
      "     |        of edges but not necessarily optimal; however, it guarantees that the\n",
      "     |        number of edges to be removed is smaller than |E|/2 - |V|/6. C{\"ip\"} uses\n",
      "     |        an integer programming formulation which is guaranteed to yield an optimal\n",
      "     |        result, but is too slow for large graphs.\n",
      "     |      @return: the IDs of the edges to be removed, in a list.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Eades P, Lin X and Smyth WF: A fast and effective heuristic for the\n",
      "     |        feedback arc set problem. In: Proc Inf Process Lett 319-323, 1993.\n",
      "     |  \n",
      "     |  get_all_shortest_paths(...)\n",
      "     |      get_all_shortest_paths(v, to=None, weights=None, mode=OUT)\n",
      "     |      \n",
      "     |      Calculates all of the shortest paths from/to a given node in a graph.\n",
      "     |      \n",
      "     |      @param v: the source for the calculated paths\n",
      "     |      @param to: a vertex selector describing the destination for\n",
      "     |        the calculated paths. This can be a single vertex ID, a list of\n",
      "     |        vertex IDs, a single vertex name, a list of vertex names or a\n",
      "     |        L{VertexSeq} object. C{None} means all the vertices.\n",
      "     |      @param weights: edge weights in a list or the name of an edge attribute\n",
      "     |        holding edge weights. If C{None}, all edges are assumed to have\n",
      "     |        equal weight.\n",
      "     |      @param mode: the directionality of the paths. L{IN} means to\n",
      "     |        calculate incoming paths, L{OUT} means to calculate outgoing\n",
      "     |        paths, L{ALL} means to calculate both ones.\n",
      "     |      @return: all of the shortest path from the given node to every other\n",
      "     |        reachable node in the graph in a list. Note that in case of mode=L{IN},\n",
      "     |        the vertices in a path are returned in reversed order!\n",
      "     |  \n",
      "     |  get_diameter(...)\n",
      "     |      get_diameter(directed=True, unconn=True, weights=None)\n",
      "     |      \n",
      "     |      Returns a path with the actual diameter of the graph.\n",
      "     |      \n",
      "     |      If there are many shortest paths with the length of the diameter,\n",
      "     |      it returns the first one it founds.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param unconn: if C{True} and the graph is unconnected, the\n",
      "     |        longest geodesic within a component will be returned. If\n",
      "     |        C{False} and the graph is unconnected, the result is the\n",
      "     |        number of vertices if there are no weights or infinity\n",
      "     |        if there are weights.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the vertices in the path in order.\n",
      "     |  \n",
      "     |  get_edgelist(...)\n",
      "     |      get_edgelist()\n",
      "     |      \n",
      "     |      Returns the edge list of a graph.\n",
      "     |  \n",
      "     |  get_eid(...)\n",
      "     |      get_eid(v1, v2, directed=True, error=True)\n",
      "     |      \n",
      "     |      Returns the edge ID of an arbitrary edge between vertices v1 and v2\n",
      "     |      \n",
      "     |      @param v1: the ID or name of the first vertex\n",
      "     |      @param v2: the ID or name of the second vertex\n",
      "     |      @param directed: whether edge directions should be considered in\n",
      "     |        directed graphs. The default is C{True}. Ignored for undirected\n",
      "     |        graphs.\n",
      "     |      @param error: if C{True}, an exception will be raised when the\n",
      "     |        given edge does not exist. If C{False}, -1 will be returned in\n",
      "     |        that case.\n",
      "     |      @return: the edge ID of an arbitrary edge between vertices v1 and v2\n",
      "     |  \n",
      "     |  get_eids(...)\n",
      "     |      get_eids(pairs=None, path=None, directed=True, error=True)\n",
      "     |      \n",
      "     |      Returns the edge IDs of some edges between some vertices.\n",
      "     |      \n",
      "     |      This method can operate in two different modes, depending on which\n",
      "     |      of the keyword arguments C{pairs} and C{path} are given.\n",
      "     |      \n",
      "     |      The method does not consider multiple edges; if there are multiple\n",
      "     |      edges between a pair of vertices, only the ID of one of the edges\n",
      "     |      is returned.\n",
      "     |      \n",
      "     |      @param pairs: a list of integer pairs. Each integer pair is considered\n",
      "     |        as a source-target vertex pair; the corresponding edge is looked up\n",
      "     |        in the graph and the edge ID is returned for each pair.\n",
      "     |      @param path: a list of vertex IDs. The list is considered as a\n",
      "     |        continuous path from the first vertex to the last, passing\n",
      "     |        through the intermediate vertices. The corresponding edge IDs\n",
      "     |        between the first and the second, the second and the third and\n",
      "     |        so on are looked up in the graph and the edge IDs are returned.\n",
      "     |        If both C{path} and C{pairs} are given, the two lists are\n",
      "     |        concatenated.\n",
      "     |      @param directed: whether edge directions should be considered in\n",
      "     |        directed graphs. The default is C{True}. Ignored for undirected\n",
      "     |        graphs.\n",
      "     |      @param error: if C{True}, an exception will be raised if a given\n",
      "     |        edge does not exist. If C{False}, -1 will be returned in\n",
      "     |        that case.\n",
      "     |      @return: the edge IDs in a list\n",
      "     |  \n",
      "     |  get_isomorphisms_vf2(...)\n",
      "     |      get_isomorphisms_vf2(other=None, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Returns all isomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph. If C{None}, the automorphisms\n",
      "     |        will be returned.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: a list of lists, each item of the list containing the mapping\n",
      "     |        from vertices of the second graph to the vertices of the first one\n",
      "     |  \n",
      "     |  get_shortest_paths(...)\n",
      "     |      get_shortest_paths(v, to=None, weights=None, mode=OUT, output=\"vpath\")\n",
      "     |      \n",
      "     |      Calculates the shortest paths from/to a given node in a graph.\n",
      "     |      \n",
      "     |      @param v: the source/destination for the calculated paths\n",
      "     |      @param to: a vertex selector describing the destination/source for\n",
      "     |        the calculated paths. This can be a single vertex ID, a list of\n",
      "     |        vertex IDs, a single vertex name, a list of vertex names or a\n",
      "     |        L{VertexSeq} object. C{None} means all the vertices.\n",
      "     |      @param weights: edge weights in a list or the name of an edge attribute\n",
      "     |        holding edge weights. If C{None}, all edges are assumed to have\n",
      "     |        equal weight.\n",
      "     |      @param mode: the directionality of the paths. L{IN} means to\n",
      "     |        calculate incoming paths, L{OUT} means to calculate outgoing\n",
      "     |        paths, L{ALL} means to calculate both ones.\n",
      "     |      @param output: determines what should be returned. If this is\n",
      "     |        C{\"vpath\"}, a list of vertex IDs will be returned, one path\n",
      "     |        for each target vertex. For unconnected graphs, some of the list\n",
      "     |        elements may be empty. Note that in case of mode=L{IN}, the vertices\n",
      "     |        in a path are returned in reversed order. If C{output=\"epath\"},\n",
      "     |        edge IDs are returned instead of vertex IDs.\n",
      "     |      @return: see the documentation of the C{output} parameter.\n",
      "     |  \n",
      "     |  get_subisomorphisms_lad(...)\n",
      "     |      get_subisomorphisms_lad(other, domains=None, induced=False, time_limit=0)\n",
      "     |      \n",
      "     |      Returns all subisomorphisms between the graph and another one using the LAD\n",
      "     |      algorithm.\n",
      "     |      \n",
      "     |      The optional C{domains} argument may be used to restrict vertices that\n",
      "     |      may match each other. You can also specify whether you are interested\n",
      "     |      in induced subgraphs only or not.\n",
      "     |      \n",
      "     |      @param other: the pattern graph we are looking for in the graph.\n",
      "     |      @param domains: a list of lists, one sublist belonging to each vertex in\n",
      "     |        the template graph. Sublist M{i} contains the indices of the vertices in\n",
      "     |        the original graph that may match vertex M{i} in the template graph.\n",
      "     |        C{None} means that every vertex may match every other vertex.\n",
      "     |      @param induced: whether to consider induced subgraphs only.\n",
      "     |      @param time_limit: an optimal time limit in seconds. Only the integral\n",
      "     |        part of this number is taken into account. If the time limit is\n",
      "     |        exceeded, the method will throw an exception.\n",
      "     |      @return: a list of lists, each item of the list containing the mapping\n",
      "     |        from vertices of the second graph to the vertices of the first one\n",
      "     |  \n",
      "     |  get_subisomorphisms_vf2(...)\n",
      "     |      get_subisomorphisms_vf2(other, color1=None, color2=None,\n",
      "     |        edge_color1=None, edge_color2=None, node_compat_fn=None,\n",
      "     |        edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Returns all subisomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: a list of lists, each item of the list containing the mapping\n",
      "     |        from vertices of the second graph to the vertices of the first one\n",
      "     |  \n",
      "     |  girth(...)\n",
      "     |      girth(return_shortest_circle=False)\n",
      "     |      \n",
      "     |      Returns the girth of the graph.\n",
      "     |      \n",
      "     |      The girth of a graph is the length of the shortest circle in it.\n",
      "     |      \n",
      "     |      @param return_shortest_circle: whether to return one of the shortest\n",
      "     |        circles found in the graph.\n",
      "     |      @return: the length of the shortest circle or (if C{return_shortest_circle})\n",
      "     |        is true, the shortest circle itself as a list\n",
      "     |  \n",
      "     |  has_multiple(...)\n",
      "     |      has_multiple()\n",
      "     |      \n",
      "     |      Checks whether the graph has multiple edges.\n",
      "     |      \n",
      "     |      @return: C{True} if the graph has at least one multiple edge,\n",
      "     |               C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  hub_score(...)\n",
      "     |      hub_score(weights=None, scale=True, arpack_options=None, return_eigenvalue=False)\n",
      "     |      \n",
      "     |      Calculates Kleinberg's hub score for the vertices of the graph\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param scale: whether to normalize the scores so that the largest one\n",
      "     |        is 1.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @param return_eigenvalue: whether to return the largest eigenvalue\n",
      "     |      @return: the hub scores in a list and optionally the largest eigenvalue\n",
      "     |        as a second member of a tuple\n",
      "     |      \n",
      "     |      @see: authority_score()\n",
      "     |  \n",
      "     |  incident(...)\n",
      "     |      incident(vertex, mode=OUT)\n",
      "     |      \n",
      "     |      Returns the edges a given vertex is incident on.\n",
      "     |      \n",
      "     |      @param vertex: a vertex ID\n",
      "     |      @param mode: whether to return only successors (L{OUT}),\n",
      "     |        predecessors (L{IN}) or both (L{ALL}). Ignored for undirected\n",
      "     |        graphs.\n",
      "     |  \n",
      "     |  independence_number(...)\n",
      "     |      independence_number()\n",
      "     |      \n",
      "     |      Returns the independence number of the graph.\n",
      "     |      \n",
      "     |      The independence number of the graph is the size of the largest\n",
      "     |      independent vertex set.\n",
      "     |      \n",
      "     |      @see: L{largest_independent_vertex_sets()} for the largest independent\n",
      "     |        vertex sets\n",
      "     |  \n",
      "     |  independent_vertex_sets(...)\n",
      "     |      independent_vertex_sets(min=0, max=0)\n",
      "     |      \n",
      "     |      Returns some or all independent vertex sets of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      Two vertices are independent if there is no edge between them. Members\n",
      "     |      of an independent vertex set are mutually independent.\n",
      "     |      \n",
      "     |      @param min: the minimum size of sets to be returned. If zero or\n",
      "     |        negative, no lower bound will be used.\n",
      "     |      @param max: the maximum size of sets to be returned. If zero or\n",
      "     |        negative, no upper bound will be used.\n",
      "     |  \n",
      "     |  induced_subgraph(...)\n",
      "     |      induced_subgraph(vertices, implementation=\"auto\")\n",
      "     |      \n",
      "     |      Returns a subgraph spanned by the given vertices.\n",
      "     |      \n",
      "     |      @param vertices: a list containing the vertex IDs which\n",
      "     |        should be included in the result.\n",
      "     |      @param implementation: the implementation to use when constructing\n",
      "     |        the new subgraph. igraph includes two implementations at the\n",
      "     |        moment. C{\"copy_and_delete\"} copies the original graph and\n",
      "     |        removes those vertices that are not in the given set. This is more\n",
      "     |        efficient if the size of the subgraph is comparable to the original\n",
      "     |        graph. The other implementation (C{\"create_from_scratch\"})\n",
      "     |        constructs the result graph from scratch and then copies the\n",
      "     |        attributes accordingly. This is a better solution if the subgraph\n",
      "     |        is relatively small, compared to the original graph. C{\"auto\"}\n",
      "     |        selects between the two implementations automatically, based on\n",
      "     |        the ratio of the size of the subgraph and the size of the original\n",
      "     |        graph.\n",
      "     |      @return: the subgraph\n",
      "     |  \n",
      "     |  intersection(...)\n",
      "     |      intersection(graphs)\n",
      "     |      \n",
      "     |      Creates the intersection of two (or more) graphs.\n",
      "     |      \n",
      "     |      @param graphs: the list of graphs to be intersected with\n",
      "     |        the current one.\n",
      "     |  \n",
      "     |  is_bipartite(...)\n",
      "     |      is_bipartite(return_types=False)\n",
      "     |      \n",
      "     |      Decides whether the graph is bipartite or not.\n",
      "     |      \n",
      "     |      Vertices of a bipartite graph can be partitioned into two groups A\n",
      "     |      and B in a way that all edges go between the two groups.\n",
      "     |      \n",
      "     |      @param return_types: if C{False}, the method will simply\n",
      "     |        return C{True} or C{False} depending on whether the graph is\n",
      "     |        bipartite or not. If C{True}, the actual group assignments\n",
      "     |        are also returned as a list of boolean values. (Note that\n",
      "     |        the group assignment is not unique, especially if the graph\n",
      "     |        consists of multiple components, since the assignments of\n",
      "     |        components are independent from each other).\n",
      "     |      @return: C{True} if the graph is bipartite, C{False} if not.\n",
      "     |        If C{return_types} is C{True}, the group assignment is also\n",
      "     |        returned.\n",
      "     |  \n",
      "     |  is_connected(...)\n",
      "     |      is_connected(mode=STRONG)\n",
      "     |      \n",
      "     |      Decides whether the graph is connected.\n",
      "     |      \n",
      "     |      @param mode: whether we should calculate strong or weak connectivity.\n",
      "     |      @return: C{True} if the graph is connected, C{False} otherwise.\n",
      "     |  \n",
      "     |  is_dag(...)\n",
      "     |      is_dag()\n",
      "     |      \n",
      "     |      Checks whether the graph is a DAG (directed acyclic graph).\n",
      "     |      \n",
      "     |      A DAG is a directed graph with no directed cycles.\n",
      "     |      \n",
      "     |      @return: C{True} if it is a DAG, C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  is_directed(...)\n",
      "     |      is_directed()\n",
      "     |      \n",
      "     |      Checks whether the graph is directed.\n",
      "     |      @return: C{True} if it is directed, C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  is_loop(...)\n",
      "     |      is_loop(edges=None)\n",
      "     |      \n",
      "     |      Checks whether a specific set of edges contain loop edges\n",
      "     |      \n",
      "     |      @param edges: edge indices which we want to check. If C{None}, all\n",
      "     |        edges are checked.\n",
      "     |      @return: a list of booleans, one for every edge given\n",
      "     |  \n",
      "     |  is_minimal_separator(...)\n",
      "     |      is_minimal_separator(vertices)\n",
      "     |      \n",
      "     |      Decides whether the given vertex set is a minimal separator.\n",
      "     |      \n",
      "     |      A minimal separator is a set of vertices whose removal disconnects the graph,\n",
      "     |      while the removal of any subset of the set keeps the graph connected.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @return: C{True} is the given vertex set is a minimal separator, C{False}\n",
      "     |        otherwise.\n",
      "     |  \n",
      "     |  is_multiple(...)\n",
      "     |      is_multiple(edges=None)\n",
      "     |      \n",
      "     |      Checks whether an edge is a multiple edge.\n",
      "     |      \n",
      "     |      Also works for a set of edges -- in this case, every edge is checked\n",
      "     |      one by one. Note that if there are multiple edges going between a\n",
      "     |      pair of vertices, there is always one of them that is I{not}\n",
      "     |      reported as multiple (only the others). This allows one to easily\n",
      "     |      detect the edges that have to be deleted in order to make the graph\n",
      "     |      free of multiple edges.\n",
      "     |      \n",
      "     |      @param edges: edge indices which we want to check. If C{None}, all\n",
      "     |        edges are checked.\n",
      "     |      @return: a list of booleans, one for every edge given\n",
      "     |  \n",
      "     |  is_mutual(...)\n",
      "     |      is_mutual(edges=None)\n",
      "     |      \n",
      "     |      Checks whether an edge has an opposite pair.\n",
      "     |      \n",
      "     |      Also works for a set of edges -- in this case, every edge is checked\n",
      "     |      one by one. The result will be a list of booleans (or a single boolean\n",
      "     |      if only an edge index is supplied), every boolean corresponding to an\n",
      "     |      edge in the edge set supplied. C{True} is returned for a given edge\n",
      "     |      M{a} --> M{b} if there exists another edge M{b} --> M{a} in the\n",
      "     |      original graph (not the given edge set!). All edges in an undirected\n",
      "     |      graph are mutual. In case there are multiple edges between M{a}\n",
      "     |      and M{b}, it is enough to have at least one edge in either direction\n",
      "     |      to report all edges between them as mutual, so the multiplicity\n",
      "     |      of edges do not matter.\n",
      "     |      \n",
      "     |      @param edges: edge indices which we want to check. If C{None}, all\n",
      "     |        edges are checked.\n",
      "     |      @return: a list of booleans, one for every edge given\n",
      "     |  \n",
      "     |  is_separator(...)\n",
      "     |      is_separator(vertices)\n",
      "     |      \n",
      "     |      Decides whether the removal of the given vertices disconnects the graph.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @return: C{True} is the given vertex set is a separator, C{False} if not.\n",
      "     |  \n",
      "     |  is_simple(...)\n",
      "     |      is_simple()\n",
      "     |      \n",
      "     |      Checks whether the graph is simple (no loop or multiple edges).\n",
      "     |      \n",
      "     |      @return: C{True} if it is simple, C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  isoclass(...)\n",
      "     |      isoclass(vertices)\n",
      "     |      \n",
      "     |      Returns the isomorphy class of the graph or its subgraph.\n",
      "     |      \n",
      "     |      Isomorphy class calculations are implemented only for graphs with\n",
      "     |      3 or 4 vertices.\n",
      "     |      \n",
      "     |      @param vertices: a list of vertices if we want to calculate the\n",
      "     |        isomorphy class for only a subset of vertices. C{None} means to\n",
      "     |        use the full graph.\n",
      "     |      @return: the isomorphy class of the (sub)graph\n",
      "     |  \n",
      "     |  isomorphic(...)\n",
      "     |      isomorphic(other)\n",
      "     |      \n",
      "     |      Checks whether the graph is isomorphic to another graph.\n",
      "     |      \n",
      "     |      The algorithm being used is selected using a simple heuristic:\n",
      "     |      \n",
      "     |        - If one graph is directed and the other undirected, an exception\n",
      "     |          is thrown.\n",
      "     |      \n",
      "     |        - If the two graphs does not have the same number of vertices and\n",
      "     |          edges, it returns with C{False}\n",
      "     |      \n",
      "     |        - If the graphs have three or four vertices, then an O(1) algorithm\n",
      "     |          is used with precomputed data.\n",
      "     |      \n",
      "     |        - Otherwise if the graphs are directed, then the VF2 isomorphism\n",
      "     |          algorithm is used (see L{Graph.isomorphic_vf2}).\n",
      "     |      \n",
      "     |        - Otherwise the BLISS isomorphism algorithm is used, see\n",
      "     |          L{Graph.isomorphic_bliss}.\n",
      "     |      \n",
      "     |      @return: C{True} if the graphs are isomorphic, C{False} otherwise.\n",
      "     |  \n",
      "     |  isomorphic_bliss(...)\n",
      "     |      isomorphic_bliss(other, return_mapping_12=False, return_mapping_21=False,\n",
      "     |        sh1=\"fm\", sh2=\"fm\")\n",
      "     |      \n",
      "     |      Checks whether the graph is isomorphic to another graph, using the\n",
      "     |      BLISS isomorphism algorithm.\n",
      "     |      \n",
      "     |      See U{http://www.tcs.hut.fi/Software/bliss/index.html} for more information\n",
      "     |      about the BLISS algorithm.\n",
      "     |      \n",
      "     |      @param other: the other graph with which we want to compare the graph.\n",
      "     |      @param return_mapping_12: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the first graph to the second.\n",
      "     |      @param return_mapping_21: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the second graph to the first.\n",
      "     |      @param sh1: splitting heuristics for the first graph as a\n",
      "     |        case-insensitive string, with the following possible values:\n",
      "     |      \n",
      "     |          - C{\"f\"}: first non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fl\"}: first largest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fs\"}: first smallest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fm\"}: first maximally non-trivially connected non-singleton\n",
      "     |            cell\n",
      "     |      \n",
      "     |          - C{\"flm\"}: largest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fsm\"}: smallest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |      @param sh2: splitting heuristics to be used for the second graph.\n",
      "     |        Accepted values are as above.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graphs\n",
      "     |        are isomorphic, C{False} otherwise. If any or both mappings are\n",
      "     |        calculated, the result is a 3-tuple, the first element being the\n",
      "     |        above mentioned boolean, the second element being the 1 -> 2 mapping\n",
      "     |        and the third element being the 2 -> 1 mapping. If the corresponding\n",
      "     |        mapping was not calculated, C{None} is returned in the appropriate\n",
      "     |        element of the 3-tuple.\n",
      "     |  \n",
      "     |  isomorphic_vf2(...)\n",
      "     |      isomorphic_vf2(other=None, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, return_mapping_12=False, return_mapping_21=False,\n",
      "     |        node_compat_fn=None, edge_compat_fn=None, callback=None)\n",
      "     |      \n",
      "     |      Checks whether the graph is isomorphic to another graph, using the\n",
      "     |      VF2 isomorphism algorithm.\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph with which we want to compare the graph.\n",
      "     |        If C{None}, the automorphjisms of the graph will be tested.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param return_mapping_12: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the first graph to the second.\n",
      "     |      @param return_mapping_21: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the second graph to the first.\n",
      "     |      @param callback: if not C{None}, the isomorphism search will not stop at\n",
      "     |        the first match; it will call this callback function instead for every\n",
      "     |        isomorphism found. The callback function must accept four arguments:\n",
      "     |        the first graph, the second graph, a mapping from the nodes of the\n",
      "     |        first graph to the second, and a mapping from the nodes of the second\n",
      "     |        graph to the first. The function must return C{True} if the search\n",
      "     |        should continue or C{False} otherwise.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graphs\n",
      "     |        are isomorphic, C{False} otherwise. If any or both mappings are\n",
      "     |        calculated, the result is a 3-tuple, the first element being the\n",
      "     |        above mentioned boolean, the second element being the 1 -> 2 mapping\n",
      "     |        and the third element being the 2 -> 1 mapping. If the corresponding\n",
      "     |        mapping was not calculated, C{None} is returned in the appropriate\n",
      "     |        element of the 3-tuple.\n",
      "     |  \n",
      "     |  knn(...)\n",
      "     |      knn(vids=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates the average degree of the neighbors for each vertex, and\n",
      "     |      the same quantity as the function of vertex degree.\n",
      "     |      \n",
      "     |      @param vids: the vertices for which the calculation is performed.\n",
      "     |        C{None} means all vertices.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name. If this is given, the vertex strength\n",
      "     |        will be used instead of the vertex degree in the calculations, but\n",
      "     |        the \"ordinary\" vertex degree will be used for the second (degree-\n",
      "     |        dependent) list in the result.\n",
      "     |      @return: two lists in a tuple. The first list contains the average\n",
      "     |        degree of neighbors for each vertex, the second contains the average\n",
      "     |        degree of neighbors as a function of vertex degree. The zeroth element\n",
      "     |        of this list corresponds to vertices of degree 1.\n",
      "     |  \n",
      "     |  laplacian(...)\n",
      "     |      laplacian(weights=None, normalized=False)\n",
      "     |      \n",
      "     |      Returns the Laplacian matrix of a graph.\n",
      "     |      \n",
      "     |      The Laplacian matrix is similar to the adjacency matrix, but the edges\n",
      "     |      are denoted with -1 and the diagonal contains the node degrees.\n",
      "     |      \n",
      "     |      Normalized Laplacian matrices have 1 or 0 in their diagonals (0 for vertices\n",
      "     |      with no edges), edges are denoted by 1 / sqrt(d_i * d_j) where d_i is the\n",
      "     |      degree of node i.\n",
      "     |      \n",
      "     |      Multiple edges and self-loops are silently ignored. Although it is\n",
      "     |      possible to calculate the Laplacian matrix of a directed graph, it does\n",
      "     |      not make much sense.\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name. When edge weights are used, the degree\n",
      "     |        of a node is considered to be the weight of its incident edges.\n",
      "     |      @param normalized: whether to return the normalized Laplacian matrix.\n",
      "     |      @return: the Laplacian matrix.\n",
      "     |  \n",
      "     |  largest_cliques(...)\n",
      "     |      largest_cliques()\n",
      "     |      \n",
      "     |      Returns the largest cliques of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      Quite intuitively a clique is considered largest if there is no clique\n",
      "     |      with more vertices in the whole graph. All largest cliques are maximal\n",
      "     |      (i.e. nonextendable) but not all maximal cliques are largest.\n",
      "     |      \n",
      "     |      @see: L{clique_number()} for the size of the largest cliques or\n",
      "     |        L{maximal_cliques()} for the maximal cliques\n",
      "     |  \n",
      "     |  largest_independent_vertex_sets(...)\n",
      "     |      largest_independent_vertex_sets()\n",
      "     |      \n",
      "     |      Returns the largest independent vertex sets of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      Quite intuitively an independent vertex set is considered largest if\n",
      "     |      there is no other set with more vertices in the whole graph. All largest\n",
      "     |      sets are maximal (i.e. nonextendable) but not all maximal sets\n",
      "     |      are largest.\n",
      "     |      \n",
      "     |      @see: L{independence_number()} for the size of the largest independent\n",
      "     |        vertex sets or L{maximal_independent_vertex_sets()} for the maximal\n",
      "     |        (nonextendable) independent vertex sets\n",
      "     |  \n",
      "     |  linegraph(...)\n",
      "     |      linegraph()\n",
      "     |      \n",
      "     |      Returns the line graph of the graph.\n",
      "     |      \n",
      "     |      The line graph M{L(G)} of an undirected graph is defined as follows:\n",
      "     |      M{L(G)} has one vertex for each edge in G and two vertices in M{L(G)}\n",
      "     |      are connected iff their corresponding edges in the original graph\n",
      "     |      share an end point.\n",
      "     |      \n",
      "     |      The line graph of a directed graph is slightly different: two vertices\n",
      "     |      are connected by a directed edge iff the target of the first vertex's\n",
      "     |      corresponding edge is the same as the source of the second vertex's\n",
      "     |      corresponding edge.\n",
      "     |  \n",
      "     |  maxdegree(...)\n",
      "     |      maxdegree(vertices=None, mode=ALL, loops=False)\n",
      "     |      \n",
      "     |      Returns the maximum degree of a vertex set in the graph.\n",
      "     |      \n",
      "     |      This method accepts a single vertex ID or a list of vertex IDs as a\n",
      "     |      parameter, and returns the degree of the given vertices (in the\n",
      "     |      form of a single integer or a list, depending on the input\n",
      "     |      parameter).\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs, or\n",
      "     |        C{None} meaning all the vertices in the graph.\n",
      "     |      @param mode: the type of degree to be returned (L{OUT} for\n",
      "     |        out-degrees, L{IN} IN for in-degrees or L{ALL} for the sum of\n",
      "     |        them).\n",
      "     |      @param loops: whether self-loops should be counted.\n",
      "     |  \n",
      "     |  maxflow_value(...)\n",
      "     |      maxflow_value(source, target, capacity=None)\n",
      "     |      \n",
      "     |      Returns the value of the maximum flow between the source and target vertices.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: the value of the maximum flow between the given vertices\n",
      "     |  \n",
      "     |  maximal_cliques(...)\n",
      "     |      maximal_cliques(min=0, max=0, file=None)\n",
      "     |      \n",
      "     |      Returns the maximal cliques of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      A maximal clique is a clique which can't be extended by adding any other\n",
      "     |      vertex to it. A maximal clique is not necessarily one of the largest\n",
      "     |      cliques in the graph.\n",
      "     |      \n",
      "     |      @param min: the minimum size of maximal cliques to be returned. If zero\n",
      "     |        or negative, no lower bound will be used.\n",
      "     |      \n",
      "     |      @param max: the maximum size of maximal cliques to be returned. If zero\n",
      "     |        or negative, no upper bound will be used. If nonzero, the size of every\n",
      "     |        maximal clique found will be compared to this value and a clique will\n",
      "     |        be returned only if its size is smaller than this limit.\n",
      "     |      \n",
      "     |      @param file: a file object or the name of the file to write the results\n",
      "     |        to. When this argument is C{None}, the maximal cliques will be returned\n",
      "     |        as a list of lists.\n",
      "     |      @return: the maximal cliques of the graph as a list of lists, or C{None}\n",
      "     |        if the C{file} argument was given.@see: L{largest_cliques()} for the largest cliques.\n",
      "     |  \n",
      "     |  maximal_independent_vertex_sets(...)\n",
      "     |      maximal_independent_vertex_sets()\n",
      "     |      \n",
      "     |      Returns the maximal independent vertex sets of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      A maximal independent vertex set is an independent vertex set\n",
      "     |      which can't be extended by adding any other vertex to it. A maximal\n",
      "     |      independent vertex set is not necessarily one of the largest\n",
      "     |      independent vertex sets in the graph.\n",
      "     |      \n",
      "     |      @see: L{largest_independent_vertex_sets()} for the largest independent\n",
      "     |        vertex sets\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Tsukiyama, M. Ide, H. Ariyoshi and I. Shirawaka: I{A new\n",
      "     |        algorithm for generating all the maximal independent sets}.\n",
      "     |        SIAM J Computing, 6:505--517, 1977.\n",
      "     |  \n",
      "     |  mincut_value(...)\n",
      "     |      mincut_value(source=-1, target=-1, capacity=None)\n",
      "     |      \n",
      "     |      Returns the minimum cut between the source and target vertices or within\n",
      "     |      the whole graph.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID. If negative, the calculation is\n",
      "     |        done for every vertex except the target and the minimum is returned.\n",
      "     |      @param target: the target vertex ID. If negative, the calculation is\n",
      "     |        done for every vertex except the source and the minimum is returned.\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: the value of the minimum cut between the given vertices\n",
      "     |  \n",
      "     |  minimum_size_separators(...)\n",
      "     |      minimum_size_separators()\n",
      "     |      \n",
      "     |      Returns a list containing all separator vertex sets of minimum size.\n",
      "     |      \n",
      "     |      A vertex set is a separator if its removal disconnects the graph. This method\n",
      "     |      lists all the separators for which no smaller separator set exists in the\n",
      "     |      given graph.\n",
      "     |      \n",
      "     |      @return: a list where each item lists the vertex indices of a given\n",
      "     |        separator of minimum size.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Arkady Kanevsky: Finding all minimum-size separating vertex sets\n",
      "     |        in a graph. Networks 23:533--541, 1993.\n",
      "     |  \n",
      "     |  motifs_randesu(...)\n",
      "     |      motifs_randesu(size=3, cut_prob=None, callback=None)\n",
      "     |      \n",
      "     |      Counts the number of motifs in the graph\n",
      "     |      \n",
      "     |      Motifs are small subgraphs of a given structure in a graph. It is\n",
      "     |      argued that the motif profile (ie. the number of different motifs in\n",
      "     |      the graph) is characteristic for different types of networks and\n",
      "     |      network function is related to the motifs in the graph.\n",
      "     |      \n",
      "     |      This function is able to find the different motifs of size three\n",
      "     |      and four (ie. the number of different subgraphs with three and four\n",
      "     |      vertices) in the network.\n",
      "     |      \n",
      "     |      In a big network the total number of motifs can be very large, so\n",
      "     |      it takes a lot of time to find all of them. In such cases, a sampling\n",
      "     |      method can be used. This function is capable of doing sampling via\n",
      "     |      the I{cut_prob} argument. This argument gives the probability that\n",
      "     |      a branch of the motif search tree will not be explored.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wernicke and F. Rasche: FANMOD: a tool for fast network\n",
      "     |        motif detection, Bioinformatics 22(9), 1152--1153, 2006.\n",
      "     |      \n",
      "     |      @param size: the size of the motifs (3 or 4).\n",
      "     |      @param cut_prob: the cut probabilities for different levels of the search\n",
      "     |        tree. This must be a list of length I{size} or C{None} to find all\n",
      "     |        motifs.\n",
      "     |      @param callback: C{None} or a callable that will be called for every motif\n",
      "     |        found in the graph. The callable must accept three parameters: the graph\n",
      "     |        itself, the list of vertices in the motif and the isomorphy class of the\n",
      "     |        motif (see L{Graph.isoclass()}). The search will stop when the callback\n",
      "     |        returns an object with a non-zero truth value or raises an exception.\n",
      "     |      @return: the list of motifs if I{callback} is C{None}, or C{None} otherwise\n",
      "     |      @see: Graph.motifs_randesu_no()\n",
      "     |  \n",
      "     |  motifs_randesu_estimate(...)\n",
      "     |      motifs_randesu_estimate(size=3, cut_prob=None, sample)\n",
      "     |      \n",
      "     |      Counts the total number of motifs in the graph\n",
      "     |      \n",
      "     |      Motifs are small subgraphs of a given structure in a graph.\n",
      "     |      This function estimates the total number of motifs in a graph without\n",
      "     |      assigning isomorphism classes to them by extrapolating from a random\n",
      "     |      sample of vertices.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wernicke and F. Rasche: FANMOD: a tool for fast network\n",
      "     |        motif detection, Bioinformatics 22(9), 1152--1153, 2006.\n",
      "     |      \n",
      "     |      @param size: the size of the motifs (3 or 4).\n",
      "     |      @param cut_prob: the cut probabilities for different levels of the search\n",
      "     |        tree. This must be a list of length I{size} or C{None} to find all\n",
      "     |        motifs.\n",
      "     |      @param sample: the size of the sample or the vertex IDs of the vertices\n",
      "     |        to be used for sampling.\n",
      "     |      @see: Graph.motifs_randesu()\n",
      "     |  \n",
      "     |  motifs_randesu_no(...)\n",
      "     |      motifs_randesu_no(size=3, cut_prob=None)\n",
      "     |      \n",
      "     |      Counts the total number of motifs in the graph\n",
      "     |      \n",
      "     |      Motifs are small subgraphs of a given structure in a graph.\n",
      "     |      This function counts the total number of motifs in a graph without\n",
      "     |      assigning isomorphism classes to them.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wernicke and F. Rasche: FANMOD: a tool for fast network\n",
      "     |        motif detection, Bioinformatics 22(9), 1152--1153, 2006.\n",
      "     |      \n",
      "     |      @param size: the size of the motifs (3 or 4).\n",
      "     |      @param cut_prob: the cut probabilities for different levels of the search\n",
      "     |        tree. This must be a list of length I{size} or C{None} to find all\n",
      "     |        motifs.\n",
      "     |      @see: Graph.motifs_randesu()\n",
      "     |  \n",
      "     |  neighborhood(...)\n",
      "     |      neighborhood(vertices=None, order=1, mode=ALL)\n",
      "     |      \n",
      "     |      For each vertex specified by I{vertices}, returns the\n",
      "     |      vertices reachable from that vertex in at most I{order} steps.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs, or\n",
      "     |        C{None} meaning all the vertices in the graph.\n",
      "     |      @param order: the order of the neighborhood, i.e. the maximum number of\n",
      "     |        steps to take from the seed vertex.\n",
      "     |      @param mode: specifies how to take into account the direction of\n",
      "     |        the edges if a directed graph is analyzed. C{\"out\"} means that\n",
      "     |        only the outgoing edges are followed, so all vertices reachable\n",
      "     |        from the source vertex in at most I{order} steps are counted.\n",
      "     |        C{\"in\"} means that only the incoming edges are followed (in\n",
      "     |        reverse direction of course), so all vertices from which the source\n",
      "     |        vertex is reachable in at most I{order} steps are counted. C{\"all\"}\n",
      "     |        treats directed edges as undirected.\n",
      "     |      @return: a single list specifying the neighborhood if I{vertices}\n",
      "     |        was an integer specifying a single vertex index, or a list of lists\n",
      "     |        if I{vertices} was a list or C{None}.\n",
      "     |  \n",
      "     |  neighborhood_size(...)\n",
      "     |      neighborhood_size(vertices=None, order=1, mode=ALL)\n",
      "     |      \n",
      "     |      For each vertex specified by I{vertices}, returns the number of\n",
      "     |      vertices reachable from that vertex in at most I{order} steps.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs, or\n",
      "     |        C{None} meaning all the vertices in the graph.\n",
      "     |      @param order: the order of the neighborhood, i.e. the maximum number of\n",
      "     |        steps to take from the seed vertex.\n",
      "     |      @param mode: specifies how to take into account the direction of\n",
      "     |        the edges if a directed graph is analyzed. C{\"out\"} means that\n",
      "     |        only the outgoing edges are followed, so all vertices reachable\n",
      "     |        from the source vertex in at most I{order} steps are counted.\n",
      "     |        C{\"in\"} means that only the incoming edges are followed (in\n",
      "     |        reverse direction of course), so all vertices from which the source\n",
      "     |        vertex is reachable in at most I{order} steps are counted. C{\"all\"}\n",
      "     |        treats directed edges as undirected.\n",
      "     |      @return: a single number specifying the neighborhood size if I{vertices}\n",
      "     |        was an integer specifying a single vertex index, or a list of sizes\n",
      "     |        if I{vertices} was a list or C{None}.\n",
      "     |  \n",
      "     |  neighbors(...)\n",
      "     |      neighbors(vertex, mode=ALL)\n",
      "     |      \n",
      "     |      Returns adjacent vertices to a given vertex.\n",
      "     |      \n",
      "     |      @param vertex: a vertex ID\n",
      "     |      @param mode: whether to return only successors (L{OUT}),\n",
      "     |        predecessors (L{IN}) or both (L{ALL}). Ignored for undirected\n",
      "     |        graphs.\n",
      "     |  \n",
      "     |  permute_vertices(...)\n",
      "     |      permute_vertices(permutation)\n",
      "     |      \n",
      "     |      Permutes the vertices of the graph according to the given permutation\n",
      "     |      and returns the new graph.\n",
      "     |      \n",
      "     |      Vertex M{k} of the original graph will become vertex M{permutation[k]}\n",
      "     |      in the new graph. No validity checks are performed on the permutation\n",
      "     |      vector.\n",
      "     |      \n",
      "     |      @return: the new graph\n",
      "     |  \n",
      "     |  personalized_pagerank(...)\n",
      "     |      personalized_pagerank(vertices=None, directed=True, damping=0.85,\n",
      "     |              reset=None, reset_vertices=None, weights=None, \n",
      "     |              arpack_options=None, implementation=\"prpack\", niter=1000,\n",
      "     |              eps=0.001)\n",
      "     |      \n",
      "     |      Calculates the personalized PageRank values of a graph.\n",
      "     |      \n",
      "     |      The personalized PageRank calculation is similar to the PageRank\n",
      "     |      calculation, but the random walk is reset to a non-uniform distribution\n",
      "     |      over the vertices in every step with probability M{1-damping} instead of a\n",
      "     |      uniform distribution.\n",
      "     |      \n",
      "     |      @param vertices: the indices of the vertices being queried.\n",
      "     |        C{None} means all of the vertices.\n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param damping: the damping factor.\n",
      "     |        M{1-damping} is the PageRank value for vertices with no\n",
      "     |        incoming links.\n",
      "     |      @param reset: the distribution over the vertices to be used when resetting\n",
      "     |        the random walk. Can be a sequence, an iterable or a vertex attribute\n",
      "     |        name as long as they return a list of floats whose length is equal to\n",
      "     |        the number of vertices. If C{None}, a uniform distribution is assumed,\n",
      "     |        which makes the method equivalent to the original PageRank algorithm.\n",
      "     |      @param reset_vertices: an alternative way to specify the distribution\n",
      "     |        over the vertices to be used when resetting the random walk. Simply\n",
      "     |        supply a list of vertex IDs here, or a L{VertexSeq} or a L{Vertex}.\n",
      "     |        Resetting will take place using a uniform distribution over the specified\n",
      "     |        vertices.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used. This argument is\n",
      "     |        ignored if not the ARPACK implementation is used, see the \n",
      "     |        I{implementation} argument.\n",
      "     |      @param implementation: which implementation to use to solve the \n",
      "     |        PageRank eigenproblem. Possible values are:\n",
      "     |      \n",
      "     |          - C{\"prpack\"}: use the PRPACK library. This is a new \n",
      "     |            implementation in igraph 0.7\n",
      "     |      \n",
      "     |          - C{\"arpack\"}: use the ARPACK library. This implementation\n",
      "     |            was used from version 0.5, until version 0.7.\n",
      "     |      \n",
      "     |          - C{\"power\"}: use a simple power method. This is the\n",
      "     |            implementation that was used before igraph version 0.5.\n",
      "     |      \n",
      "     |      @param niter: The number of iterations to use in the power method\n",
      "     |        implementation. It is ignored in the other implementations.\n",
      "     |      @param eps: The power method implementation will consider the\n",
      "     |        calculation as complete if the difference of PageRank values between\n",
      "     |        iterations change less than this value for every node. It is \n",
      "     |        ignored by the other implementations.\n",
      "     |      @return: a list with the personalized PageRank values of the specified\n",
      "     |        vertices.\n",
      "     |  \n",
      "     |  predecessors(...)\n",
      "     |      predecessors(vertex)\n",
      "     |      \n",
      "     |      Returns the predecessors of a given vertex.\n",
      "     |      \n",
      "     |      Equivalent to calling the L{Graph.neighbors} method with type=L{IN}.\n",
      "     |  \n",
      "     |  radius(...)\n",
      "     |      radius(mode=OUT)\n",
      "     |      \n",
      "     |      Calculates the radius of the graph.\n",
      "     |      \n",
      "     |      The radius of a graph is defined as the minimum eccentricity of\n",
      "     |      its vertices (see L{eccentricity()}).\n",
      "     |      @param mode: what kind of paths to consider for the calculation\n",
      "     |        in case of directed graphs. C{OUT} considers paths that follow\n",
      "     |        edge directions, C{IN} considers paths that follow the opposite\n",
      "     |        edge directions, C{ALL} ignores edge directions. The argument is\n",
      "     |        ignored for undirected graphs.\n",
      "     |      @return: the radius\n",
      "     |      @see: L{Graph.eccentricity()}\n",
      "     |  \n",
      "     |  reciprocity(...)\n",
      "     |      reciprocity(ignore_loops=True, mode=\"default\")\n",
      "     |      \n",
      "     |      Reciprocity defines the proportion of mutual connections in a\n",
      "     |      directed graph. It is most commonly defined as the probability\n",
      "     |      that the opposite counterpart of a directed edge is also included\n",
      "     |      in the graph. This measure is calculated if C{mode} is C{\"default\"}.\n",
      "     |      \n",
      "     |      Prior to igraph 0.6, another measure was implemented, defined as\n",
      "     |      the probability of mutual connection between a vertex pair if we\n",
      "     |      know that there is a (possibly non-mutual) connection between them.\n",
      "     |      In other words, (unordered) vertex pairs are classified into three\n",
      "     |      groups: (1) disconnected, (2) non-reciprocally connected and (3)\n",
      "     |      reciprocally connected. The result is the size of group (3), divided\n",
      "     |      by the sum of sizes of groups (2) and (3). This measure is calculated\n",
      "     |      if C{mode} is C{\"ratio\"}.\n",
      "     |      \n",
      "     |      @param ignore_loops: whether loop edges should be ignored.\n",
      "     |      @param mode: the algorithm to use to calculate the reciprocity; see\n",
      "     |        above for more details.\n",
      "     |      @return: the reciprocity of the graph\n",
      "     |  \n",
      "     |  rewire(...)\n",
      "     |      rewire(n=1000, mode=\"simple\")\n",
      "     |      \n",
      "     |      Randomly rewires the graph while preserving the degree distribution.\n",
      "     |      \n",
      "     |      Please note that the rewiring is done \"in-place\", so the original\n",
      "     |      graph will be modified. If you want to preserve the original graph,\n",
      "     |      use the L{copy} method before.\n",
      "     |      \n",
      "     |      @param n: the number of rewiring trials.\n",
      "     |      @param mode: the rewiring algorithm to use. It can either be C{\"simple\"} or\n",
      "     |        C{\"loops\"}; the former does not create or destroy loop edges while the\n",
      "     |        latter does.\n",
      "     |  \n",
      "     |  rewire_edges(...)\n",
      "     |      rewire_edges(prob, loops=False, multiple=False)\n",
      "     |      \n",
      "     |      Rewires the edges of a graph with constant probability.\n",
      "     |      \n",
      "     |      Each endpoint of each edge of the graph will be rewired with a constant\n",
      "     |      probability, given in the first argument.\n",
      "     |      \n",
      "     |      Please note that the rewiring is done \"in-place\", so the original\n",
      "     |      graph will be modified. If you want to preserve the original graph,\n",
      "     |      use the L{copy} method before.\n",
      "     |      \n",
      "     |      @param prob: rewiring probability\n",
      "     |      @param loops: whether the algorithm is allowed to create loop edges\n",
      "     |      @param multiple: whether the algorithm is allowed to create multiple\n",
      "     |        edges.\n",
      "     |  \n",
      "     |  shortest_paths(...)\n",
      "     |      shortest_paths(source=None, target=None, weights=None, mode=OUT)\n",
      "     |      \n",
      "     |      Calculates shortest path lengths for given vertices in a graph.\n",
      "     |      \n",
      "     |      The algorithm used for the calculations is selected automatically:\n",
      "     |      a simple BFS is used for unweighted graphs, Dijkstra's algorithm is\n",
      "     |      used when all the weights are positive. Otherwise, the Bellman-Ford\n",
      "     |      algorithm is used if the number of requested source vertices is larger\n",
      "     |      than 100 and Johnson's algorithm is used otherwise.\n",
      "     |      \n",
      "     |      @param source: a list containing the source vertex IDs which should be\n",
      "     |        included in the result. If C{None}, all vertices will be considered.\n",
      "     |      @param target: a list containing the target vertex IDs which should be\n",
      "     |        included in the result. If C{None}, all vertices will be considered.\n",
      "     |      @param weights: a list containing the edge weights. It can also be\n",
      "     |        an attribute name (edge weights are retrieved from the given\n",
      "     |        attribute) or C{None} (all edges have equal weight).\n",
      "     |      @param mode: the type of shortest paths to be used for the\n",
      "     |        calculation in directed graphs. L{OUT} means only outgoing,\n",
      "     |        L{IN} means only incoming paths. L{ALL} means to consider\n",
      "     |        the directed graph as an undirected one.\n",
      "     |      @return: the shortest path lengths for given vertices in a matrix\n",
      "     |  \n",
      "     |  similarity_dice(...)\n",
      "     |      similarity_dice(vertices=None, pairs=None, mode=IGRAPH_ALL, loops=True)\n",
      "     |      \n",
      "     |      Dice similarity coefficient of vertices.\n",
      "     |      \n",
      "     |      The Dice similarity coefficient of two vertices is twice the number of\n",
      "     |      their common neighbors divided by the sum of their degrees. This\n",
      "     |      coefficient is very similar to the Jaccard coefficient, but usually\n",
      "     |      gives higher similarities than its counterpart.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None} and I{pairs} is also\n",
      "     |        C{None}, all vertices will be considered.\n",
      "     |      @param pairs: the vertex pairs to be analysed. If this is given, I{vertices}\n",
      "     |        must be C{None}, and the similarity values will be calculated only for the\n",
      "     |        given pairs. Vertex pairs must be specified as tuples of vertex IDs.\n",
      "     |      @param mode: which neighbors should be considered for directed graphs.\n",
      "     |        Can be L{ALL}, L{IN} or L{OUT}, ignored for undirected graphs.\n",
      "     |      @param loops: whether vertices should be considered adjacent to\n",
      "     |        themselves. Setting this to C{True} assumes a loop edge for all vertices\n",
      "     |        even if none is present in the graph. Setting this to C{False} may\n",
      "     |        result in strange results: nonadjacent vertices may have larger\n",
      "     |        similarities compared to the case when an edge is added between them --\n",
      "     |        however, this might be exactly the result you want to get.\n",
      "     |      @return: the pairwise similarity coefficients for the vertices specified,\n",
      "     |        in the form of a matrix if C{pairs} is C{None} or in the form of a list\n",
      "     |        if C{pairs} is not C{None}.\n",
      "     |  \n",
      "     |  similarity_inverse_log_weighted(...)\n",
      "     |      similarity_inverse_log_weighted(vertices=None, mode=IGRAPH_ALL)\n",
      "     |      \n",
      "     |      Inverse log-weighted similarity coefficient of vertices.\n",
      "     |      \n",
      "     |      Each vertex is assigned a weight which is 1 / log(degree). The\n",
      "     |      log-weighted similarity of two vertices is the sum of the weights\n",
      "     |      of their common neighbors.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None}, all vertices\n",
      "     |        will be considered.\n",
      "     |      @param mode: which neighbors should be considered for directed graphs.\n",
      "     |        Can be L{ALL}, L{IN} or L{OUT}, ignored for undirected graphs.\n",
      "     |        L{IN} means that the weights are determined by the out-degrees, L{OUT}\n",
      "     |        means that the weights are determined by the in-degrees.\n",
      "     |      @return: the pairwise similarity coefficients for the vertices specified,\n",
      "     |        in the form of a matrix (list of lists).\n",
      "     |  \n",
      "     |  similarity_jaccard(...)\n",
      "     |      similarity_jaccard(vertices=None, pairs=None, mode=IGRAPH_ALL, loops=True)\n",
      "     |      \n",
      "     |      Jaccard similarity coefficient of vertices.\n",
      "     |      \n",
      "     |      The Jaccard similarity coefficient of two vertices is the number of their\n",
      "     |      common neighbors divided by the number of vertices that are adjacent to\n",
      "     |      at least one of them.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None} and I{pairs} is also\n",
      "     |        C{None}, all vertices will be considered.\n",
      "     |      @param pairs: the vertex pairs to be analysed. If this is given, I{vertices}\n",
      "     |        must be C{None}, and the similarity values will be calculated only for the\n",
      "     |        given pairs. Vertex pairs must be specified as tuples of vertex IDs.\n",
      "     |      @param mode: which neighbors should be considered for directed graphs.\n",
      "     |        Can be L{ALL}, L{IN} or L{OUT}, ignored for undirected graphs.\n",
      "     |      @param loops: whether vertices should be considered adjacent to\n",
      "     |        themselves. Setting this to C{True} assumes a loop edge for all vertices\n",
      "     |        even if none is present in the graph. Setting this to C{False} may\n",
      "     |        result in strange results: nonadjacent vertices may have larger\n",
      "     |        similarities compared to the case when an edge is added between them --\n",
      "     |        however, this might be exactly the result you want to get.\n",
      "     |      @return: the pairwise similarity coefficients for the vertices specified,\n",
      "     |        in the form of a matrix if C{pairs} is C{None} or in the form of a list\n",
      "     |        if C{pairs} is not C{None}.\n",
      "     |  \n",
      "     |  simplify(...)\n",
      "     |      simplify(multiple=True, loops=True, combine_edges=None)\n",
      "     |      \n",
      "     |      Simplifies a graph by removing self-loops and/or multiple edges.\n",
      "     |      \n",
      "     |      \n",
      "     |      For example, suppose you have a graph with an edge attribute named\n",
      "     |      C{weight}. C{graph.simplify(combine_edges=max)} will take the\n",
      "     |      maximum of the weights of multiple edges and assign that weight to\n",
      "     |      the collapsed edge. C{graph.simplify(combine_edges=sum)} will\n",
      "     |      take the sum of the weights. You can also write\n",
      "     |      C{graph.simplify(combine_edges=dict(weight=\"sum\"))} or\n",
      "     |      C{graph.simplify(combine_edges=dict(weight=sum))}, since\n",
      "     |      C{sum} is recognised both as a Python built-in function and as\n",
      "     |      a string constant.\n",
      "     |      \n",
      "     |      @param multiple: whether to remove multiple edges.\n",
      "     |      @param loops: whether to remove loops.\n",
      "     |      @param combine_edges: specifies how to combine the attributes of\n",
      "     |        multiple edges between the same pair of vertices into a single\n",
      "     |        attribute. If it is C{None}, only one of the edges will be kept\n",
      "     |        and all the attributes will be lost. If it is a function, the\n",
      "     |        attributes of multiple edges will be collected and passed on to\n",
      "     |        that function which will return the new attribute value that has to\n",
      "     |        be assigned to the single collapsed edge. It can also be one of\n",
      "     |        the following string constants:\n",
      "     |      \n",
      "     |          - C{\"ignore\"}: all the edge attributes will be ignored.\n",
      "     |      \n",
      "     |          - C{\"sum\"}: the sum of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"product\"}: the product of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |          - C{\"mean\"}: the mean of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"median\"}: the median of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"min\"}: the minimum of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"max\"}: the maximum of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"first\"}: the attribute value of the first edge in the collapsed set\n",
      "     |            will be used for the new edge.\n",
      "     |      \n",
      "     |          - C{\"last\"}: the attribute value of the last edge in the collapsed set\n",
      "     |            will be used for the new edge.\n",
      "     |      \n",
      "     |          - C{\"random\"}: a randomly selected value will be used for the new edge\n",
      "     |      \n",
      "     |          - C{\"concat\"}: the attribute values will be concatenated for the new\n",
      "     |            edge.\n",
      "     |      \n",
      "     |        You can also use a dict mapping edge attribute names to functions or\n",
      "     |        the above string constants if you want to make the behaviour of the\n",
      "     |        simplification process depend on the name of the attribute.\n",
      "     |        C{None} is a special key in this dict, its value will be used for all\n",
      "     |        the attributes not specified explicitly in the dictionary.\n",
      "     |  \n",
      "     |  strength(...)\n",
      "     |      strength(vertices, mode=ALL, loops=True, weights=None)\n",
      "     |      \n",
      "     |      Returns the strength (weighted degree) of some vertices from the graph\n",
      "     |      \n",
      "     |      This method accepts a single vertex ID or a list of vertex IDs as a\n",
      "     |      parameter, and returns the strength (that is, the sum of the weights\n",
      "     |      of all incident edges) of the given vertices (in the\n",
      "     |      form of a single integer or a list, depending on the input\n",
      "     |      parameter).\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @param mode: the type of degree to be returned (L{OUT} for\n",
      "     |        out-degrees, L{IN} IN for in-degrees or L{ALL} for the sum of\n",
      "     |        them).\n",
      "     |      @param loops: whether self-loops should be counted.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |  \n",
      "     |  subcomponent(...)\n",
      "     |      subcomponent(v, mode=ALL)\n",
      "     |      \n",
      "     |      Determines the indices of vertices which are in the same component as a given vertex.\n",
      "     |      \n",
      "     |      @param v: the index of the vertex used as the source/destination\n",
      "     |      @param mode: if equals to L{IN}, returns the vertex IDs from\n",
      "     |        where the given vertex can be reached. If equals to L{OUT},\n",
      "     |        returns the vertex IDs which are reachable from the given\n",
      "     |        vertex. If equals to L{ALL}, returns all vertices within the\n",
      "     |        same component as the given vertex, ignoring edge directions.\n",
      "     |        Note that this is not equal to calculating the union of the \n",
      "     |        results of L{IN} and L{OUT}.\n",
      "     |      @return: the indices of vertices which are in the same component as a given vertex.\n",
      "     |  \n",
      "     |  subgraph_edges(...)\n",
      "     |      subgraph_edges(edges, delete_vertices=True)\n",
      "     |      \n",
      "     |      Returns a subgraph spanned by the given edges.\n",
      "     |      \n",
      "     |      @param edges: a list containing the edge IDs which should\n",
      "     |        be included in the result.\n",
      "     |      @param delete_vertices: if C{True}, vertices not incident on\n",
      "     |        any of the specified edges will be deleted from the result.\n",
      "     |        If C{False}, all vertices will be kept.\n",
      "     |      @return: the subgraph\n",
      "     |  \n",
      "     |  subisomorphic_lad(...)\n",
      "     |      subisomorphic_lad(other, domains=None, induced=False, time_limit=0, \n",
      "     |        return_mapping=False)\n",
      "     |      \n",
      "     |      Checks whether a subgraph of the graph is isomorphic to another graph.\n",
      "     |      \n",
      "     |      The optional C{domains} argument may be used to restrict vertices that\n",
      "     |      may match each other. You can also specify whether you are interested\n",
      "     |      in induced subgraphs only or not.\n",
      "     |      \n",
      "     |      @param other: the pattern graph we are looking for in the graph.\n",
      "     |      @param domains: a list of lists, one sublist belonging to each vertex in\n",
      "     |        the template graph. Sublist M{i} contains the indices of the vertices in\n",
      "     |        the original graph that may match vertex M{i} in the template graph.\n",
      "     |        C{None} means that every vertex may match every other vertex.\n",
      "     |      @param induced: whether to consider induced subgraphs only.\n",
      "     |      @param time_limit: an optimal time limit in seconds. Only the integral\n",
      "     |        part of this number is taken into account. If the time limit is\n",
      "     |        exceeded, the method will throw an exception.\n",
      "     |      @param return_mapping: when C{True}, the function will return a tuple,\n",
      "     |        where the first element is a boolean denoting whether a subisomorphism\n",
      "     |        has been found or not, and the second element describes the mapping\n",
      "     |        of the vertices from the template graph to the original graph. When\n",
      "     |        C{False}, only the boolean is returned.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graph\n",
      "     |        contains a subgraph that is isomorphic to the given template, C{False}\n",
      "     |        otherwise. If the mapping is calculated, the result is a tuple, the first\n",
      "     |        element being the above mentioned boolean, and the second element being\n",
      "     |        the mapping from the target to the original graph.\n",
      "     |  \n",
      "     |  subisomorphic_vf2(...)\n",
      "     |      subisomorphic_vf2(other, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, return_mapping_12=False, return_mapping_21=False,\n",
      "     |        callback=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Checks whether a subgraph of the graph is isomorphic to another graph.\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph with which we want to compare the graph.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param return_mapping_12: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the first graph to the second. The mapping can contain\n",
      "     |        -1 if a given node is not mapped.\n",
      "     |      @param return_mapping_21: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the second graph to the first. The mapping can contain\n",
      "     |        -1 if a given node is not mapped.\n",
      "     |      @param callback: if not C{None}, the subisomorphism search will not stop at\n",
      "     |        the first match; it will call this callback function instead for every\n",
      "     |        subisomorphism found. The callback function must accept four arguments:\n",
      "     |        the first graph, the second graph, a mapping from the nodes of the\n",
      "     |        first graph to the second, and a mapping from the nodes of the second\n",
      "     |        graph to the first. The function must return C{True} if the search\n",
      "     |        should continue or C{False} otherwise.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graph\n",
      "     |        contains a subgraph that's isomorphic to the given one, C{False}\n",
      "     |        otherwise. If any or both mappings are calculated, the result is a\n",
      "     |        3-tuple, the first element being the above mentioned boolean, the\n",
      "     |        second element being the 1 -> 2 mapping and the third element being\n",
      "     |        the 2 -> 1 mapping. If the corresponding mapping was not calculated,\n",
      "     |        C{None} is returned in the appropriate element of the 3-tuple.\n",
      "     |  \n",
      "     |  successors(...)\n",
      "     |      successors(vertex)\n",
      "     |      \n",
      "     |      Returns the successors of a given vertex.\n",
      "     |      \n",
      "     |      Equivalent to calling the L{Graph.neighbors} method with type=L{OUT}.\n",
      "     |  \n",
      "     |  to_directed(...)\n",
      "     |      to_directed(mutual=True)\n",
      "     |      \n",
      "     |      Converts an undirected graph to directed.\n",
      "     |      \n",
      "     |      @param mutual: C{True} if mutual directed edges should be\n",
      "     |        created for every undirected edge. If C{False}, a directed\n",
      "     |        edge with arbitrary direction is created.\n",
      "     |  \n",
      "     |  to_undirected(...)\n",
      "     |      to_undirected(mode=\"collapse\", combine_edges=None)\n",
      "     |      \n",
      "     |      Converts a directed graph to undirected.\n",
      "     |      \n",
      "     |      @param mode: specifies what to do with multiple directed edges\n",
      "     |        going between the same vertex pair. C{True} or C{\"collapse\"}\n",
      "     |        means that only a single edge should be created from multiple\n",
      "     |        directed edges. C{False} or C{\"each\"} means that every edge\n",
      "     |        will be kept (with the arrowheads removed). C{\"mutual\"}\n",
      "     |        creates one undirected edge for each mutual directed edge pair.\n",
      "     |      @param combine_edges: specifies how to combine the attributes of\n",
      "     |        multiple edges between the same pair of vertices into a single\n",
      "     |        attribute. See L{Graph.simplify()} for more details.\n",
      "     |  \n",
      "     |  topological_sorting(...)\n",
      "     |      topological_sorting(mode=OUT)\n",
      "     |      \n",
      "     |      Calculates a possible topological sorting of the graph.\n",
      "     |      \n",
      "     |      Returns a partial sorting and issues a warning if the graph is not\n",
      "     |      a directed acyclic graph.\n",
      "     |      \n",
      "     |      @param mode: if L{OUT}, vertices are returned according to the\n",
      "     |        forward topological order -- all vertices come before their\n",
      "     |        successors. If L{IN}, all vertices come before their ancestors.\n",
      "     |      @return: a possible topological ordering as a list\n",
      "     |  \n",
      "     |  transitivity_local_undirected(...)\n",
      "     |      transitivity_local_undirected(vertices=None, mode=\"nan\", weights=None)\n",
      "     |      \n",
      "     |      Calculates the local transitivity (clustering coefficient) of the\n",
      "     |      given vertices in the graph.\n",
      "     |      \n",
      "     |      The transitivity measures the probability that two neighbors of a\n",
      "     |      vertex are connected. In case of the local transitivity, this\n",
      "     |      probability is calculated separately for each vertex.\n",
      "     |      \n",
      "     |      Note that this measure is different from the global transitivity\n",
      "     |      measure (see L{transitivity_undirected()}) as it calculates\n",
      "     |      a transitivity value for each vertex individually.\n",
      "     |      \n",
      "     |      The traditional local transitivity measure applies for unweighted graphs\n",
      "     |      only. When the C{weights} argument is given, this function calculates\n",
      "     |      the weighted local transitivity proposed by Barrat et al (see references).\n",
      "     |      \n",
      "     |      @param vertices: a list containing the vertex IDs which should be\n",
      "     |        included in the result. C{None} means all of the vertices.\n",
      "     |      @param mode: defines how to treat vertices with degree less than two.\n",
      "     |        If C{TRANSITIVITT_ZERO} or C{\"zero\"}, these vertices will have\n",
      "     |        zero transitivity. If C{TRANSITIVITY_NAN} or C{\"nan\"}, these\n",
      "     |        vertices will have C{NaN} (not a number) as their transitivity.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the transitivities for the given vertices in a list\n",
      "     |      @see: L{transitivity_undirected()}, L{transitivity_avglocal_undirected()}\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Watts DJ and Strogatz S: I{Collective dynamics of small-world\n",
      "     |        networks}. Nature 393(6884):440-442, 1998.\n",
      "     |      @ref: Barrat A, Barthelemy M, Pastor-Satorras R and Vespignani A:\n",
      "     |        I{The architecture of complex weighted networks}. PNAS 101, 3747 (2004).\n",
      "     |        U{http://arxiv.org/abs/cond-mat/0311416}.\n",
      "     |  \n",
      "     |  transitivity_undirected(...)\n",
      "     |      transitivity_undirected(mode=\"nan\")\n",
      "     |      \n",
      "     |      Calculates the global transitivity (clustering coefficient) of the\n",
      "     |      graph.\n",
      "     |      \n",
      "     |      The transitivity measures the probability that two neighbors of a\n",
      "     |      vertex are connected. More precisely, this is the ratio of the\n",
      "     |      triangles and connected triplets in the graph. The result is a\n",
      "     |      single real number. Directed graphs are considered as undirected\n",
      "     |      ones.\n",
      "     |      \n",
      "     |      Note that this measure is different from the local transitivity\n",
      "     |      measure (see L{transitivity_local_undirected()}) as it calculates\n",
      "     |      a single value for the whole graph.\n",
      "     |      \n",
      "     |      @param mode: if C{TRANSITIVITY_ZERO} or C{\"zero\"}, the result will\n",
      "     |        be zero if the graph does not have any triplets. If C{\"nan\"} or\n",
      "     |        C{TRANSITIVITY_NAN}, the result will be C{NaN} (not a number).\n",
      "     |      @return: the transitivity\n",
      "     |      @see: L{transitivity_local_undirected()}, L{transitivity_avglocal_undirected()}\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wasserman and K. Faust: I{Social Network Analysis: Methods and\n",
      "     |        Applications}. Cambridge: Cambridge University Press, 1994.\n",
      "     |  \n",
      "     |  unfold_tree(...)\n",
      "     |      unfold_tree(sources=None, mode=OUT)\n",
      "     |      \n",
      "     |      Unfolds the graph using a BFS to a tree by duplicating vertices as necessary.\n",
      "     |      \n",
      "     |      @param sources: the source vertices to start the unfolding from. It should be a\n",
      "     |        list of vertex indices, preferably one vertex from each connected component.\n",
      "     |        You can use L{Graph.topological_sorting()} to determine a suitable set. A single\n",
      "     |        vertex index is also accepted.\n",
      "     |      @param mode: which edges to follow during the BFS. C{OUT} follows outgoing edges,\n",
      "     |        C{IN} follows incoming edges, C{ALL} follows both. Ignored for undirected\n",
      "     |        graphs.\n",
      "     |      @return: the unfolded tree graph and a mapping from the new vertex indices to the\n",
      "     |        old ones.\n",
      "     |  \n",
      "     |  union(...)\n",
      "     |      union(graphs)\n",
      "     |      \n",
      "     |      Creates the union of two (or more) graphs.\n",
      "     |      \n",
      "     |      @param graphs: the list of graphs to be united with\n",
      "     |        the current one.\n",
      "     |  \n",
      "     |  vcount(...)\n",
      "     |      vcount()\n",
      "     |      \n",
      "     |      Counts the number of vertices.\n",
      "     |      @return: the number of vertices in the graph.\n",
      "     |      @rtype: integer\n",
      "     |  \n",
      "     |  vertex_attributes(...)\n",
      "     |      vertex_attributes()\n",
      "     |      \n",
      "     |      @return: the attribute name list of the graph's vertices\n",
      "     |  \n",
      "     |  vertex_connectivity(...)\n",
      "     |      vertex_connectivity(source=-1, target=-1, checks=True, neighbors=\"error\")\n",
      "     |      \n",
      "     |      Calculates the vertex connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The vertex connectivity between two given vertices is the number of vertices\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of vertex disjoint directed\n",
      "     |      paths between the vertices (apart from the source and target vertices of\n",
      "     |      course). The vertex connectivity of the graph is the minimal vertex\n",
      "     |      connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the vertex connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall vertex connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @param neighbors: tells igraph what to do when the two vertices are\n",
      "     |        connected. C{\"error\"} raises an exception, C{\"infinity\"} returns\n",
      "     |        infinity, C{\"ignore\"} ignores the edge.\n",
      "     |      @return: the vertex connectivity\n",
      "     |  \n",
      "     |  write_dot(...)\n",
      "     |      write_dot(f)\n",
      "     |      \n",
      "     |      Writes the graph in DOT format to the given file.\n",
      "     |      \n",
      "     |      DOT is the format used by the U{GraphViz <http://www.graphviz.org>}\n",
      "     |      software package.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  write_edgelist(...)\n",
      "     |      write_edgelist(f)\n",
      "     |      \n",
      "     |      Writes the edge list of a graph to a file.\n",
      "     |      \n",
      "     |      Directed edges are written in (from, to) order.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  write_gml(...)\n",
      "     |      write_gml(f, creator=None, ids=None)\n",
      "     |      \n",
      "     |      Writes the graph in GML format to the given file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param creator: optional creator information to be written to the file.\n",
      "     |        If C{None}, the current date and time is added.\n",
      "     |      @param ids: optional numeric vertex IDs to use in the file. This must\n",
      "     |        be a list of integers or C{None}. If C{None}, the C{id} attribute of\n",
      "     |        the vertices are used, or if they don't exist, numeric vertex IDs\n",
      "     |        will be generated automatically.\n",
      "     |  \n",
      "     |  write_graphml(...)\n",
      "     |      write_graphml(f)\n",
      "     |      \n",
      "     |      Writes the graph to a GraphML file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  write_leda(...)\n",
      "     |      write_leda(f, names=\"name\", weights=\"weights\")\n",
      "     |      \n",
      "     |      Writes the graph to a file in LEDA native format.\n",
      "     |      \n",
      "     |      The LEDA format supports at most one attribute per vertex and edge. You can\n",
      "     |      specify which vertex and edge attribute you want to use. Note that the\n",
      "     |      name of the attribute is not saved in the LEDA file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param names: the name of the vertex attribute to be stored along with\n",
      "     |        the vertices. It is usually used to store the vertex names (hence the\n",
      "     |        name of the keyword argument), but you may also use a numeric attribute.\n",
      "     |        If you don't want to store any vertex attributes, supply C{None} here.\n",
      "     |      @param weights: the name of the edge attribute to be stored along with\n",
      "     |        the edges. It is usually used to store the edge weights (hence the\n",
      "     |        name of the keyword argument), but you may also use a string attribute.\n",
      "     |        If you don't want to store any edge attributes, supply C{None} here.\n",
      "     |  \n",
      "     |  write_lgl(...)\n",
      "     |      write_lgl(f, names=\"name\", weights=\"weights\", isolates=True)\n",
      "     |      \n",
      "     |      Writes the edge list of a graph to a file in .lgl format.\n",
      "     |      \n",
      "     |      Note that multiple edges and/or loops break the LGL software,\n",
      "     |      but igraph does not check for this condition. Unless you know\n",
      "     |      that the graph does not have multiple edges and/or loops, it\n",
      "     |      is wise to call L{simplify()} before saving.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param names: the name of the vertex attribute containing the name\n",
      "     |        of the vertices. If you don't want to store vertex names,\n",
      "     |        supply C{None} here.\n",
      "     |      @param weights: the name of the edge attribute containing the weight\n",
      "     |        of the vertices. If you don't want to store weights,\n",
      "     |        supply C{None} here.\n",
      "     |      @param isolates: whether to include isolated vertices in the output.\n",
      "     |  \n",
      "     |  write_ncol(...)\n",
      "     |      write_ncol(f, names=\"name\", weights=\"weights\")\n",
      "     |      \n",
      "     |      Writes the edge list of a graph to a file in .ncol format.\n",
      "     |      \n",
      "     |      Note that multiple edges and/or loops break the LGL software,\n",
      "     |      but igraph does not check for this condition. Unless you know\n",
      "     |      that the graph does not have multiple edges and/or loops, it\n",
      "     |      is wise to call L{simplify()} before saving.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param names: the name of the vertex attribute containing the name\n",
      "     |        of the vertices. If you don't want to store vertex names,\n",
      "     |        supply C{None} here.\n",
      "     |      @param weights: the name of the edge attribute containing the weight\n",
      "     |        of the vertices. If you don't want to store weights,\n",
      "     |        supply C{None} here.\n",
      "     |  \n",
      "     |  write_pajek(...)\n",
      "     |      write_pajek(f)\n",
      "     |      \n",
      "     |      Writes the graph in Pajek format to the given file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from Graph:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "    \n",
      "    GraphBase = class Graph(__builtin__.object)\n",
      "     |  Low-level representation of a graph.\n",
      "     |  \n",
      "     |  Don't use it directly, use L{igraph.Graph} instead.\n",
      "     |  \n",
      "     |  @undocumented: _Bipartite, _Full_Bipartite, _GRG, _Incidence, _is_matching,\n",
      "     |    _is_maximal_matching, _layout_sugiyama, _maximum_bipartite_matching,\n",
      "     |    _spanning_tree\n",
      "     |  @deffield ref: Reference\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  Adjacency(...)\n",
      "     |      Adjacency(matrix, mode=ADJ_DIRECTED)\n",
      "     |      \n",
      "     |      Generates a graph from its adjacency matrix.\n",
      "     |      \n",
      "     |      @param matrix: the adjacency matrix\n",
      "     |      @param mode: the mode to be used. Possible values are:\n",
      "     |      \n",
      "     |        - C{ADJ_DIRECTED} - the graph will be directed and a matrix\n",
      "     |          element gives the number of edges between two vertex.\n",
      "     |        - C{ADJ_UNDIRECTED} - alias to C{ADJ_MAX} for convenience.\n",
      "     |        - C{ADJ_MAX}   - undirected graph will be created and the number of\n",
      "     |          edges between vertex M{i} and M{j} is M{max(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_MIN}   - like C{ADJ_MAX}, but with M{min(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_PLUS}  - like C{ADJ_MAX}, but with M{A(i,j) + A(j,i)}\n",
      "     |        - C{ADJ_UPPER} - undirected graph with the upper right triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |        - C{ADJ_LOWER} - undirected graph with the lower left triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |      \n",
      "     |        These values can also be given as strings without the C{ADJ} prefix.\n",
      "     |  \n",
      "     |  Asymmetric_Preference(...)\n",
      "     |      Asymmetric_Preference(n, type_dist_matrix, pref_matrix, attribute=None, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on asymmetric vertex types and connection probabilities.\n",
      "     |      \n",
      "     |      This is the asymmetric variant of L{Graph.Preference}.\n",
      "     |      A given number of vertices are generated. Every vertex is assigned to an\n",
      "     |      \"incoming\" and an \"outgoing\" vertex type according to the given joint\n",
      "     |      type probabilities. Finally, every vertex pair is evaluated and a\n",
      "     |      directed edge is created between them with a probability depending on\n",
      "     |      the \"outgoing\" type of the source vertex and the \"incoming\" type of\n",
      "     |      the target vertex.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param type_dist_matrix: matrix giving the joint distribution of vertex\n",
      "     |        types\n",
      "     |      @param pref_matrix: matrix giving the connection probabilities for\n",
      "     |        different vertex types.\n",
      "     |      @param attribute: the vertex attribute name used to store the vertex\n",
      "     |        types. If C{None}, vertex types are not stored.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |  \n",
      "     |  Atlas(...)\n",
      "     |      Atlas(idx)\n",
      "     |      \n",
      "     |      Generates a graph from the Graph Atlas.\n",
      "     |      \n",
      "     |      @param idx: The index of the graph to be generated.\n",
      "     |        Indices start from zero, graphs are listed:\n",
      "     |      \n",
      "     |          1. in increasing order of number of vertices;\n",
      "     |          2. for a fixed number of vertices, in increasing order of the\n",
      "     |             number of edges;\n",
      "     |          3. for fixed numbers of vertices and edges, in increasing order\n",
      "     |             of the degree sequence, for example 111223 < 112222;\n",
      "     |          4. for fixed degree sequence, in increasing number of automorphisms.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: I{An Atlas of Graphs} by Ronald C. Read and Robin J. Wilson,\n",
      "     |        Oxford University Press, 1998.\n",
      "     |  \n",
      "     |  Barabasi(...)\n",
      "     |      Barabasi(n, m, outpref=False, directed=False, power=1,\n",
      "     |               zero_appeal=1, implementation=\"psumtree\", start_from=None)\n",
      "     |      \n",
      "     |      Generates a graph based on the Barabasi-Albert model.\n",
      "     |      \n",
      "     |      @param n: the number of vertices\n",
      "     |      @param m: either the number of outgoing edges generated for\n",
      "     |        each vertex or a list containing the number of outgoing\n",
      "     |        edges for each vertex explicitly.\n",
      "     |      @param outpref: C{True} if the out-degree of a given vertex\n",
      "     |        should also increase its citation probability (as well as\n",
      "     |        its in-degree), but it defaults to C{False}.\n",
      "     |      @param directed: C{True} if the generated graph should be\n",
      "     |        directed (default: C{False}).\n",
      "     |      @param power: the power constant of the nonlinear model.\n",
      "     |        It can be omitted, and in this case the usual linear model\n",
      "     |        will be used.\n",
      "     |      @param zero_appeal: the attractivity of vertices with degree\n",
      "     |        zero.\n",
      "     |      \n",
      "     |      @param implementation: the algorithm to use to generate the\n",
      "     |        network. Possible values are:\n",
      "     |      \n",
      "     |          - C{\"bag\"}: the algorithm that was the default in\n",
      "     |            igraph before 0.6. It works by putting the ids of the\n",
      "     |            vertices into a bag (multiset) exactly as many times\n",
      "     |            as their in-degree, plus once more. The required number\n",
      "     |            of cited vertices are then drawn from the bag with\n",
      "     |            replacement. It works only for I{power}=1 and\n",
      "     |            I{zero_appeal}=1.\n",
      "     |      \n",
      "     |          - C{\"psumtree\"}: this algorithm uses a partial prefix-sum\n",
      "     |            tree to generate the graph. It does not generate multiple\n",
      "     |            edges and it works for any values of I{power} and\n",
      "     |            I{zero_appeal}.\n",
      "     |      \n",
      "     |          - C{\"psumtree_multiple\"}: similar to C{\"psumtree\"}, but\n",
      "     |            it will generate multiple edges as well. igraph before\n",
      "     |            0.6 used this algorithm for I{power}s other than 1.\n",
      "     |      \n",
      "     |      @param start_from: if given and not C{None}, this must be another\n",
      "     |            L{Graph} object. igraph will use this graph as a starting\n",
      "     |            point for the preferential attachment model.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Barabasi, A-L and Albert, R. 1999. Emergence of scaling\n",
      "     |        in random networks. Science, 286 509-512.\n",
      "     |  \n",
      "     |  De_Bruijn(...)\n",
      "     |      De_Bruijn(m, n)\n",
      "     |      \n",
      "     |      Generates a de Bruijn graph with parameters (m, n)\n",
      "     |      \n",
      "     |      A de Bruijn graph represents relationships between strings. An alphabet\n",
      "     |      of M{m} letters are used and strings of length M{n} are considered.\n",
      "     |      A vertex corresponds to every possible string and there is a directed edge\n",
      "     |      from vertex M{v} to vertex M{w} if the string of M{v} can be transformed into\n",
      "     |      the string of M{w} by removing its first letter and appending a letter to it.\n",
      "     |      \n",
      "     |      Please note that the graph will have M{m^n} vertices and even more edges,\n",
      "     |      so probably you don't want to supply too big numbers for M{m} and M{n}.\n",
      "     |      \n",
      "     |      @param m: the size of the alphabet\n",
      "     |      @param n: the length of the strings\n",
      "     |  \n",
      "     |  Degree_Sequence(...)\n",
      "     |      Degree_Sequence(out, in=None, method=\"simple\")\n",
      "     |      \n",
      "     |      Generates a graph with a given degree sequence.\n",
      "     |      \n",
      "     |      @param out: the out-degree sequence for a directed graph. If the\n",
      "     |        in-degree sequence is omitted, the generated graph\n",
      "     |        will be undirected, so this will be the in-degree\n",
      "     |        sequence as well\n",
      "     |      @param in: the in-degree sequence for a directed graph.\n",
      "     |        If omitted, the generated graph will be undirected.\n",
      "     |      @param method: the generation method to be used. One of the following:\n",
      "     |        \n",
      "     |          - C{\"simple\"} -- simple generator that sometimes generates\n",
      "     |            loop edges and multiple edges. The generated graph is not\n",
      "     |            guaranteed to be connected.\n",
      "     |          - C{\"no_multiple\"} -- similar to C{\"simple\"} but avoids the\n",
      "     |            generation of multiple and loop edges at the expense of increased\n",
      "     |            time complexity. The method will re-start the generation every time\n",
      "     |            it gets stuck in a configuration where it is not possible to insert\n",
      "     |            any more edges without creating loops or multiple edges, and there\n",
      "     |            is no upper bound on the number of iterations, but it will succeed\n",
      "     |            eventually if the input degree sequence is graphical and throw an\n",
      "     |            exception if the input degree sequence is not graphical.\n",
      "     |          - C{\"vl\"} -- a more sophisticated generator that can sample\n",
      "     |            undirected, connected simple graphs uniformly. It uses\n",
      "     |            Monte-Carlo methods to randomize the graphs.\n",
      "     |            This generator should be favoured if undirected and connected\n",
      "     |            graphs are to be generated and execution time is not a concern.\n",
      "     |            igraph uses the original implementation of Fabien Viger; see the\n",
      "     |            following URL and the paper cited on it for the details of the\n",
      "     |            algorithm: U{http://www-rp.lip6.fr/~latapy/FV/generation.html}.\n",
      "     |  \n",
      "     |  Erdos_Renyi(...)\n",
      "     |      Erdos_Renyi(n, p, m, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on the Erdos-Renyi model.\n",
      "     |      \n",
      "     |      @param n: the number of vertices.\n",
      "     |      @param p: the probability of edges. If given, C{m} must be missing.\n",
      "     |      @param m: the number of edges. If given, C{p} must be missing.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether self-loops are allowed.\n",
      "     |  \n",
      "     |  Establishment(...)\n",
      "     |      Establishment(n, k, type_dist, pref_matrix, directed=False)\n",
      "     |      \n",
      "     |      Generates a graph based on a simple growing model with vertex types.\n",
      "     |      \n",
      "     |      A single vertex is added at each time step. This new vertex tries to\n",
      "     |      connect to k vertices in the graph. The probability that such a\n",
      "     |      connection is realized depends on the types of the vertices involved.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param k: the number of connections tried in each step\n",
      "     |      @param type_dist: list giving the distribution of vertex types\n",
      "     |      @param pref_matrix: matrix (list of lists) giving the connection\n",
      "     |        probabilities for different vertex types\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |  \n",
      "     |  Famous(...)\n",
      "     |      Famous(name)\n",
      "     |      \n",
      "     |      Generates a famous graph based on its name.\n",
      "     |      \n",
      "     |      Several famous graphs are known to C{igraph} including (but not limited to)\n",
      "     |      the Chvatal graph, the Petersen graph or the Tutte graph. This method\n",
      "     |      generates one of them based on its name (case insensitive). See the\n",
      "     |      documentation of the C interface of C{igraph} for the names available:\n",
      "     |      U{http://igraph.org/doc/c}.\n",
      "     |      \n",
      "     |      @param name: the name of the graph to be generated.\n",
      "     |  \n",
      "     |  Forest_Fire(...)\n",
      "     |      Forest_Fire(n, fw_prob, bw_factor=0.0, ambs=1, directed=False)\n",
      "     |      \n",
      "     |      Generates a graph based on the forest fire model\n",
      "     |      \n",
      "     |      The forest fire model is a growin graph model. In every time step, a new\n",
      "     |      vertex is added to the graph. The new vertex chooses an ambassador (or\n",
      "     |      more than one if M{ambs>1}) and starts a simulated forest fire at its\n",
      "     |      ambassador(s). The fire spreads through the edges. The spreading probability\n",
      "     |      along an edge is given by M{fw_prob}. The fire may also spread backwards\n",
      "     |      on an edge by probability M{fw_prob * bw_factor}. When the fire ended, the\n",
      "     |      newly added vertex connects to the vertices ``burned'' in the previous\n",
      "     |      fire.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param fw_prob: forward burning probability\n",
      "     |      @param bw_factor: ratio of backward and forward burning probability\n",
      "     |      @param ambs: number of ambassadors chosen in each step\n",
      "     |      @param directed: whether the graph will be directed\n",
      "     |  \n",
      "     |  Full(...)\n",
      "     |      Full(n, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a full graph (directed or undirected, with or without loops).\n",
      "     |      \n",
      "     |      @param n: the number of vertices.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether self-loops are allowed.\n",
      "     |  \n",
      "     |  Full_Citation(...)\n",
      "     |      Full_Citation(n, directed=False)\n",
      "     |      \n",
      "     |      Generates a full citation graph\n",
      "     |      \n",
      "     |      A full citation graph is a graph where the vertices are indexed from 0 to\n",
      "     |      M{n-1} and vertex M{i} has a directed edge towards all vertices with an\n",
      "     |      index less than M{i}.\n",
      "     |      \n",
      "     |      @param n: the number of vertices.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |  \n",
      "     |  Growing_Random(...)\n",
      "     |      Growing_Random(n, m, directed=False, citation=False)\n",
      "     |      \n",
      "     |      Generates a growing random graph.\n",
      "     |      \n",
      "     |      @param n: The number of vertices in the graph\n",
      "     |      @param m: The number of edges to add in each step (after adding a new vertex)\n",
      "     |      @param directed: whether the graph should be directed.\n",
      "     |      @param citation: whether the new edges should originate from the most\n",
      "     |         recently added vertex.\n",
      "     |  \n",
      "     |  Isoclass(...)\n",
      "     |      Isoclass(n, class, directed=False)\n",
      "     |      \n",
      "     |      Generates a graph with a given isomorphy class.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph (3 or 4)\n",
      "     |      @param class: the isomorphy class\n",
      "     |      @param directed: whether the graph should be directed.\n",
      "     |  \n",
      "     |  K_Regular(...)\n",
      "     |      K_Regular(n, k, directed=False, multiple=False)\n",
      "     |      \n",
      "     |      Generates a k-regular random graph\n",
      "     |      \n",
      "     |      A k-regular random graph is a random graph where each vertex has degree k.\n",
      "     |      If the graph is directed, both the in-degree and the out-degree of each vertex\n",
      "     |      will be k.\n",
      "     |      \n",
      "     |      @param n: The number of vertices in the graph\n",
      "     |      \n",
      "     |      @param k: The degree of each vertex if the graph is undirected, or the in-degree\n",
      "     |        and out-degree of each vertex if the graph is directed\n",
      "     |      @param directed: whether the graph should be directed.\n",
      "     |      @param multiple: whether it is allowed to create multiple edges.\n",
      "     |  \n",
      "     |  Kautz(...)\n",
      "     |      Kautz(m, n)\n",
      "     |      \n",
      "     |      Generates a Kautz graph with parameters (m, n)\n",
      "     |      \n",
      "     |      A Kautz graph is a labeled graph, vertices are labeled by strings\n",
      "     |      of length M{n+1} above an alphabet with M{m+1} letters, with\n",
      "     |      the restriction that every two consecutive letters in the string\n",
      "     |      must be different. There is a directed edge from a vertex M{v} to\n",
      "     |      another vertex M{w} if it is possible to transform the string of\n",
      "     |      M{v} into the string of M{w} by removing the first letter and\n",
      "     |      appending a letter to it.\n",
      "     |      \n",
      "     |      @param m: the size of the alphabet minus one\n",
      "     |      @param n: the length of the strings minus one\n",
      "     |  \n",
      "     |  LCF(...)\n",
      "     |      LCF(n, shifts, repeats)\n",
      "     |      \n",
      "     |      Generates a graph from LCF notation.\n",
      "     |      \n",
      "     |      LCF is short for Lederberg-Coxeter-Frucht, it is a concise notation\n",
      "     |      for 3-regular Hamiltonian graphs. It consists of three parameters,\n",
      "     |      the number of vertices in the graph, a list of shifts giving\n",
      "     |      additional edges to a cycle backbone and another integer giving how\n",
      "     |      many times the shifts should be performed. See\n",
      "     |      U{http://mathworld.wolfram.com/LCFNotation.html} for details.\n",
      "     |      \n",
      "     |      @param n: the number of vertices\n",
      "     |      @param shifts: the shifts in a list or tuple\n",
      "     |      @param repeats: the number of repeats\n",
      "     |  \n",
      "     |  Lattice(...)\n",
      "     |      Lattice(dim, nei=1, directed=False, mutual=True, circular=True)\n",
      "     |      \n",
      "     |      Generates a regular lattice.\n",
      "     |      \n",
      "     |      @param dim: list with the dimensions of the lattice\n",
      "     |      @param nei: value giving the distance (number of steps) within which\n",
      "     |         two vertices will be connected.\n",
      "     |      @param directed: whether to create a directed graph.\n",
      "     |      @param mutual: whether to create all connections as mutual\n",
      "     |          in case of a directed graph.\n",
      "     |      @param circular: whether the generated lattice is periodic.\n",
      "     |  \n",
      "     |  Preference(...)\n",
      "     |      Preference(n, type_dist, pref_matrix, attribute=None, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on vertex types and connection probabilities.\n",
      "     |      \n",
      "     |      This is practically the nongrowing variant of L{Graph.Establishment}.\n",
      "     |      A given number of vertices are generated. Every vertex is assigned to a\n",
      "     |      vertex type according to the given type probabilities. Finally, every\n",
      "     |      vertex pair is evaluated and an edge is created between them with a\n",
      "     |      probability depending on the types of the vertices involved.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param type_dist: list giving the distribution of vertex types\n",
      "     |      @param pref_matrix: matrix giving the connection probabilities for\n",
      "     |        different vertex types.\n",
      "     |      @param attribute: the vertex attribute name used to store the vertex\n",
      "     |        types. If C{None}, vertex types are not stored.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |  \n",
      "     |  Read_DIMACS(...)\n",
      "     |      Read_DIMACS(f, directed=False)\n",
      "     |      \n",
      "     |      Reads a graph from a file conforming to the DIMACS minimum-cost flow file format.\n",
      "     |      \n",
      "     |      For the exact description of the format, see\n",
      "     |      U{http://lpsolve.sourceforge.net/5.5/DIMACS.htm}\n",
      "     |      \n",
      "     |      Restrictions compared to the official description of the format:\n",
      "     |      \n",
      "     |        - igraph's DIMACS reader requires only three fields in an arc definition,\n",
      "     |          describing the edge's source and target node and its capacity.\n",
      "     |        - Source vertices are identified by 's' in the FLOW field, target vertices are\n",
      "     |          identified by 't'.\n",
      "     |        - Node indices start from 1. Only a single source and target node is allowed.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |      @return: the generated graph, the source and the target of the flow and the edge\n",
      "     |        capacities in a tuple\n",
      "     |  \n",
      "     |  Read_DL(...)\n",
      "     |      Read_DL(f, directed=True)\n",
      "     |      \n",
      "     |      Reads an UCINET DL file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |  \n",
      "     |  Read_Edgelist(...)\n",
      "     |      Read_Edgelist(f, directed=True)\n",
      "     |      \n",
      "     |      Reads an edge list from a file and creates a graph based on it.\n",
      "     |      \n",
      "     |      Please note that the vertex indices are zero-based.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |  \n",
      "     |  Read_GML(...)\n",
      "     |      Read_GML(f)\n",
      "     |      \n",
      "     |      Reads a GML file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |  \n",
      "     |  Read_GraphDB(...)\n",
      "     |      Read_GraphDB(f, directed=False)\n",
      "     |      \n",
      "     |      Reads a GraphDB format file and creates a graph based on it.\n",
      "     |      \n",
      "     |      GraphDB is a binary format, used in the graph database for\n",
      "     |      isomorphism testing (see U{http://amalfi.dis.unina.it/graph/}).\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param directed: whether the generated graph should be directed.\n",
      "     |  \n",
      "     |  Read_GraphML(...)\n",
      "     |      Read_GraphML(f, directed=True, index=0)\n",
      "     |      \n",
      "     |      Reads a GraphML format file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param index: if the GraphML file contains multiple graphs,\n",
      "     |        specifies the one that should be loaded. Graph indices\n",
      "     |        start from zero, so if you want to load the first graph,\n",
      "     |        specify 0 here.\n",
      "     |  \n",
      "     |  Read_Lgl(...)\n",
      "     |      Read_Lgl(f, names=True, weights=\"if_present\", directed=True)\n",
      "     |      \n",
      "     |      Reads an .lgl file used by LGL.\n",
      "     |      \n",
      "     |      It is also useful for creating graphs from \"named\" (and\n",
      "     |      optionally weighted) edge lists.\n",
      "     |      \n",
      "     |      This format is used by the Large Graph Layout program. See the\n",
      "     |      U{documentation of LGL <http://bioinformatics.icmb.utexas.edu/lgl/>}\n",
      "     |      regarding the exact format description.\n",
      "     |      \n",
      "     |      LGL originally cannot deal with graphs containing multiple or loop\n",
      "     |      edges, but this condition is not checked here, as igraph is happy\n",
      "     |      with these.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param names: If C{True}, the vertex names are added as a\n",
      "     |        vertex attribute called 'name'.\n",
      "     |      @param weights: If True, the edge weights are added as an\n",
      "     |        edge attribute called 'weight', even if there are no\n",
      "     |        weights in the file. If False, the edge weights are never\n",
      "     |        added, even if they are present. C{\"auto\"} or C{\"if_present\"}\n",
      "     |        means that weights are added if there is at least one weighted\n",
      "     |        edge in the input file, but they are not added otherwise.\n",
      "     |      @param directed: whether the graph being created should be\n",
      "     |        directed\n",
      "     |  \n",
      "     |  Read_Ncol(...)\n",
      "     |      Read_Ncol(f, names=True, weights=\"if_present\", directed=True)\n",
      "     |      \n",
      "     |      Reads an .ncol file used by LGL.\n",
      "     |      \n",
      "     |      It is also useful for creating graphs from \"named\" (and\n",
      "     |      optionally weighted) edge lists.\n",
      "     |      \n",
      "     |      This format is used by the Large Graph Layout program. See the\n",
      "     |      U{documentation of LGL <http://bioinformatics.icmb.utexas.edu/lgl/>}\n",
      "     |      regarding the exact format description.\n",
      "     |      \n",
      "     |      LGL originally cannot deal with graphs containing multiple or loop\n",
      "     |      edges, but this condition is not checked here, as igraph is happy\n",
      "     |      with these.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |      @param names: If C{True}, the vertex names are added as a\n",
      "     |        vertex attribute called 'name'.\n",
      "     |      @param weights: If True, the edge weights are added as an\n",
      "     |        edge attribute called 'weight', even if there are no\n",
      "     |        weights in the file. If False, the edge weights are never\n",
      "     |        added, even if they are present. C{\"auto\"} or C{\"if_present\"}\n",
      "     |        means that weights are added if there is at least one weighted\n",
      "     |        edge in the input file, but they are not added otherwise.\n",
      "     |      @param directed: whether the graph being created should be\n",
      "     |        directed\n",
      "     |  \n",
      "     |  Read_Pajek(...)\n",
      "     |      Read_Pajek(f)\n",
      "     |      \n",
      "     |      Reads a Pajek format file and creates a graph based on it.\n",
      "     |      \n",
      "     |      @param f: the name of the file or a Python file handle\n",
      "     |  \n",
      "     |  Recent_Degree(...)\n",
      "     |      Recent_Degree(n, m, window, outpref=False, directed=False, power=1)\n",
      "     |      \n",
      "     |      Generates a graph based on a stochastic model where the probability\n",
      "     |      of an edge gaining a new node is proportional to the edges gained in\n",
      "     |      a given time window.\n",
      "     |      \n",
      "     |      @param n: the number of vertices\n",
      "     |      @param m: either the number of outgoing edges generated for\n",
      "     |        each vertex or a list containing the number of outgoing\n",
      "     |        edges for each vertex explicitly.\n",
      "     |      @param window: size of the window in time steps\n",
      "     |      @param outpref: C{True} if the out-degree of a given vertex\n",
      "     |        should also increase its citation probability (as well as\n",
      "     |        its in-degree), but it defaults to C{False}.\n",
      "     |      @param directed: C{True} if the generated graph should be\n",
      "     |        directed (default: C{False}).\n",
      "     |      @param power: the power constant of the nonlinear model.\n",
      "     |        It can be omitted, and in this case the usual linear model\n",
      "     |        will be used.\n",
      "     |  \n",
      "     |  Ring(...)\n",
      "     |      Ring(n, directed=False, mutual=False, circular=True)\n",
      "     |      \n",
      "     |      Generates a ring graph.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the ring\n",
      "     |      @param directed: whether to create a directed ring.\n",
      "     |      @param mutual: whether to create mutual edges in a directed ring.\n",
      "     |      @param circular: whether to create a closed ring.\n",
      "     |  \n",
      "     |  SBM(...)\n",
      "     |      SBM(n, pref_matrix, block_sizes, directed=False, loops=False)\n",
      "     |      \n",
      "     |      Generates a graph based on a stochastic blockmodel.\n",
      "     |      \n",
      "     |      A given number of vertices are generated. Every vertex is assigned to a\n",
      "     |      vertex type according to the given block sizes. Vertices of the same\n",
      "     |      type will be assigned consecutive vertex IDs. Finally, every\n",
      "     |      vertex pair is evaluated and an edge is created between them with a\n",
      "     |      probability depending on the types of the vertices involved. The\n",
      "     |      probabilities are taken from the preference matrix.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param pref_matrix: matrix giving the connection probabilities for\n",
      "     |        different vertex types.\n",
      "     |      @param block_sizes: list giving the number of vertices in each block; must\n",
      "     |        sum up to I{n}.\n",
      "     |      @param directed: whether to generate a directed graph.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |  \n",
      "     |  Star(...)\n",
      "     |      Star(n, mode=\"undirected\", center=0)\n",
      "     |      \n",
      "     |      Generates a star graph.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param mode: Gives the type of the star graph to create. Should be\n",
      "     |        either \"in\", \"out\", \"mutual\" or \"undirected\"\n",
      "     |      @param center: Vertex ID for the central vertex in the star.\n",
      "     |  \n",
      "     |  Static_Fitness(...)\n",
      "     |      Static_Fitness(m, fitness_out, fitness_in=None, loops=False, multiple=False)\n",
      "     |      \n",
      "     |      Generates a non-growing graph with edge probabilities proportional to node\n",
      "     |      fitnesses.\n",
      "     |      \n",
      "     |      The algorithm randomly selects vertex pairs and connects them until the given\n",
      "     |      number of edges are created. Each vertex is selected with a probability\n",
      "     |      proportional to its fitness; for directed graphs, a vertex is selected as a\n",
      "     |      source proportional to its out-fitness and as a target proportional to its\n",
      "     |      in-fitness.\n",
      "     |      \n",
      "     |      @param m: the number of edges in the graph\n",
      "     |      @param fitness_out: a numeric vector with non-negative entries, one for each\n",
      "     |        vertex. These values represent the fitness scores (out-fitness scores for\n",
      "     |        directed graphs). I{fitness} is an alias of this keyword argument.\n",
      "     |      @param fitness_in: a numeric vector with non-negative entries, one for each\n",
      "     |        vertex. These values represent the in-fitness scores for directed graphs.\n",
      "     |        For undirected graphs, this argument must be C{None}.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |      @param multiple: whether multiple edges are allowed.\n",
      "     |      @return: a directed or undirected graph with the prescribed power-law\n",
      "     |        degree distributions.\n",
      "     |  \n",
      "     |  Static_Power_Law(...)\n",
      "     |      Static_Power_Law(n, m, exponent_out, exponent_in=-1, loops=False,\n",
      "     |          multiple=False, finite_size_correction=True)\n",
      "     |      \n",
      "     |      Generates a non-growing graph with prescribed power-law degree distributions.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param m: the number of edges in the graph\n",
      "     |      @param exponent_out: the exponent of the out-degree distribution, which\n",
      "     |        must be between 2 and infinity (inclusive). When I{exponent_in} is\n",
      "     |        not given or negative, the graph will be undirected and this parameter\n",
      "     |        specifies the degree distribution. I{exponent} is an alias to this\n",
      "     |        keyword argument.\n",
      "     |      @param exponent_in: the exponent of the in-degree distribution, which\n",
      "     |        must be between 2 and infinity (inclusive) It can also be negative, in\n",
      "     |        which case an undirected graph will be generated.\n",
      "     |      @param loops: whether loop edges are allowed.\n",
      "     |      @param multiple: whether multiple edges are allowed.\n",
      "     |      @param finite_size_correction: whether to apply a finite-size correction\n",
      "     |        to the generated fitness values for exponents less than 3. See the\n",
      "     |        paper of Cho et al for more details.\n",
      "     |      @return: a directed or undirected graph with the prescribed power-law\n",
      "     |        degree distributions.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Goh K-I, Kahng B, Kim D: Universal behaviour of load distribution\n",
      "     |        in scale-free networks. Phys Rev Lett 87(27):278701, 2001.\n",
      "     |      @ref: Cho YS, Kim JS, Park J, Kahng B, Kim D: Percolation transitions in\n",
      "     |        scale-free networks under the Achlioptas process. Phys Rev Lett\n",
      "     |        103:135702, 2009.\n",
      "     |  \n",
      "     |  Tree(...)\n",
      "     |      Tree(n, children, type=TREE_UNDIRECTED)\n",
      "     |      \n",
      "     |      Generates a tree in which almost all vertices have the same number of children.\n",
      "     |      \n",
      "     |      @param n: the number of vertices in the graph\n",
      "     |      @param children: the number of children of a vertex in the graph\n",
      "     |      @param type: determines whether the tree should be directed, and if\n",
      "     |        this is the case, also its orientation. Must be one of\n",
      "     |        C{TREE_IN}, C{TREE_OUT} and C{TREE_UNDIRECTED}.\n",
      "     |  \n",
      "     |  Watts_Strogatz(...)\n",
      "     |      Watts_Strogatz(dim, size, nei, p, loops=False, multiple=False)\n",
      "     |      \n",
      "     |      @param dim: the dimension of the lattice\n",
      "     |      @param size: the size of the lattice along all dimensions\n",
      "     |      @param nei: value giving the distance (number of steps) within which\n",
      "     |         two vertices will be connected.\n",
      "     |      @param p: rewiring probability\n",
      "     |      \n",
      "     |      @param loops: specifies whether loop edges are allowed\n",
      "     |      @param multiple: specifies whether multiple edges are allowed\n",
      "     |      @see: L{Lattice()}, L{rewire()}, L{rewire_edges()} if more flexibility is\n",
      "     |        needed\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Duncan J Watts and Steven H Strogatz: I{Collective dynamics of\n",
      "     |        small world networks}, Nature 393, 440-442, 1998\n",
      "     |  \n",
      "     |  Weighted_Adjacency(...)\n",
      "     |      Weighted_Adjacency(matrix, mode=ADJ_DIRECTED, attr=\"weight\", loops=True)\n",
      "     |      \n",
      "     |      Generates a graph from its adjacency matrix.\n",
      "     |      \n",
      "     |      @param matrix: the adjacency matrix\n",
      "     |      @param mode: the mode to be used. Possible values are:\n",
      "     |      \n",
      "     |        - C{ADJ_DIRECTED} - the graph will be directed and a matrix\n",
      "     |          element gives the number of edges between two vertex.\n",
      "     |        - C{ADJ_UNDIRECTED} - alias to C{ADJ_MAX} for convenience.\n",
      "     |        - C{ADJ_MAX}   - undirected graph will be created and the number of\n",
      "     |          edges between vertex M{i} and M{j} is M{max(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_MIN}   - like C{ADJ_MAX}, but with M{min(A(i,j), A(j,i))}\n",
      "     |        - C{ADJ_PLUS}  - like C{ADJ_MAX}, but with M{A(i,j) + A(j,i)}\n",
      "     |        - C{ADJ_UPPER} - undirected graph with the upper right triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |        - C{ADJ_LOWER} - undirected graph with the lower left triangle of\n",
      "     |          the matrix (including the diagonal)\n",
      "     |      \n",
      "     |        These values can also be given as strings without the C{ADJ} prefix.\n",
      "     |      @param attr: the name of the edge attribute that stores the edge\n",
      "     |        weights.\n",
      "     |      @param loops: whether to include loop edges. When C{False}, the diagonal\n",
      "     |        of the adjacency matrix will be ignored.\n",
      "     |  \n",
      "     |  __and__(...)\n",
      "     |      x.__and__(y) <==> x&y\n",
      "     |  \n",
      "     |  __delitem__(...)\n",
      "     |      x.__delitem__(y) <==> del x[y]\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __init__(...)\n",
      "     |      x.__init__(...) initializes x; see help(type(x)) for signature\n",
      "     |  \n",
      "     |  __invert__(...)\n",
      "     |      x.__invert__() <==> ~x\n",
      "     |  \n",
      "     |  __or__(...)\n",
      "     |      x.__or__(y) <==> x|y\n",
      "     |  \n",
      "     |  __rand__(...)\n",
      "     |      x.__rand__(y) <==> y&x\n",
      "     |  \n",
      "     |  __ror__(...)\n",
      "     |      x.__ror__(y) <==> y|x\n",
      "     |  \n",
      "     |  __setitem__(...)\n",
      "     |      x.__setitem__(i, y) <==> x[i]=y\n",
      "     |  \n",
      "     |  __str__(...)\n",
      "     |      x.__str__() <==> str(x)\n",
      "     |  \n",
      "     |  add_edges(...)\n",
      "     |      add_edges(es)\n",
      "     |      \n",
      "     |      Adds edges to the graph.\n",
      "     |      \n",
      "     |      @param es: the list of edges to be added. Every edge is\n",
      "     |        represented with a tuple, containing the vertex IDs of the\n",
      "     |        two endpoints. Vertices are enumerated from zero.\n",
      "     |  \n",
      "     |  add_vertices(...)\n",
      "     |      add_vertices(n)\n",
      "     |      \n",
      "     |      Adds vertices to the graph.\n",
      "     |      \n",
      "     |      @param n: the number of vertices to be added\n",
      "     |  \n",
      "     |  all_minimal_st_separators(...)\n",
      "     |      all_minimal_st_separators()\n",
      "     |      \n",
      "     |      Returns a list containing all the minimal s-t separators of a graph.\n",
      "     |      \n",
      "     |      A minimal separator is a set of vertices whose removal disconnects the graph,\n",
      "     |      while the removal of any subset of the set keeps the graph connected.\n",
      "     |      \n",
      "     |      @return: a list where each item lists the vertex indices of a given\n",
      "     |        minimal s-t separator.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Anne Berry, Jean-Paul Bordat and Olivier Cogis: Generating all the\n",
      "     |        minimal separators of a graph. In: Peter Widmayer, Gabriele Neyer and\n",
      "     |        Stephan Eidenbenz (eds.): Graph-theoretic concepts in computer science,\n",
      "     |        1665, 167--172, 1999. Springer.\n",
      "     |  \n",
      "     |  all_st_cuts(...)\n",
      "     |      all_st_cuts(source, target)\n",
      "     |      \n",
      "     |      Returns all the cuts between the source and target vertices in a\n",
      "     |      directed graph.\n",
      "     |      \n",
      "     |      This function lists all edge-cuts between a source and a target vertex.\n",
      "     |      Every cut is listed exactly once.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a list of L{Cut} objects. It is\n",
      "     |        advised to use that.\n",
      "     |      @return: a tuple where the first element is a list of lists of edge IDs\n",
      "     |        representing a cut and the second element is a list of lists of vertex\n",
      "     |        IDs representing the sets of vertices that were separated by the cuts.\n",
      "     |  \n",
      "     |  all_st_mincuts(...)\n",
      "     |      all_st_mincuts(source, target)\n",
      "     |      \n",
      "     |      Returns all minimum cuts between the source and target vertices in a\n",
      "     |      directed graph.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a list of L{Cut} objects. It is\n",
      "     |        advised to use that.\n",
      "     |  \n",
      "     |  are_connected(...)\n",
      "     |      are_connected(v1, v2)\n",
      "     |      \n",
      "     |      Decides whether two given vertices are directly connected.\n",
      "     |      \n",
      "     |      @param v1: the ID or name of the first vertex\n",
      "     |      @param v2: the ID or name of the second vertex\n",
      "     |      @return: C{True} if there exists an edge from v1 to v2, C{False}\n",
      "     |        otherwise.\n",
      "     |  \n",
      "     |  articulation_points(...)\n",
      "     |      articulation_points()\n",
      "     |      \n",
      "     |      Returns the list of articulation points in the graph.\n",
      "     |      \n",
      "     |      A vertex is an articulation point if its removal increases the number of\n",
      "     |      connected components in the graph.\n",
      "     |  \n",
      "     |  assortativity(...)\n",
      "     |      assortativity(types1, types2=None, directed=True)\n",
      "     |      \n",
      "     |      Returns the assortativity of the graph based on numeric properties\n",
      "     |      of the vertices.\n",
      "     |      \n",
      "     |      This coefficient is basically the correlation between the actual\n",
      "     |      connectivity patterns of the vertices and the pattern expected from the\n",
      "     |      disribution of the vertex types.\n",
      "     |      \n",
      "     |      See equation (21) in Newman MEJ: Mixing patterns in networks, Phys Rev E\n",
      "     |      67:026126 (2003) for the proper definition. The actual calculation is\n",
      "     |      performed using equation (26) in the same paper for directed graphs, and\n",
      "     |      equation (4) in Newman MEJ: Assortative mixing in networks, Phys Rev Lett\n",
      "     |      89:208701 (2002) for undirected graphs.\n",
      "     |      \n",
      "     |      @param types1: vertex types in a list or the name of a vertex attribute\n",
      "     |        holding vertex types. Types are ideally denoted by numeric values.\n",
      "     |      @param types2: in directed assortativity calculations, each vertex can\n",
      "     |        have an out-type and an in-type. In this case, I{types1} contains the\n",
      "     |        out-types and this parameter contains the in-types in a list or the\n",
      "     |        name of a vertex attribute. If C{None}, it is assumed to be equal\n",
      "     |        to I{types1}.\n",
      "     |      \n",
      "     |      @param directed: whether to consider edge directions or not.\n",
      "     |      @return: the assortativity coefficient\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Newman MEJ: Mixing patterns in networks, Phys Rev E 67:026126, 2003.\n",
      "     |      @ref: Newman MEJ: Assortative mixing in networks, Phys Rev Lett 89:208701,\n",
      "     |        2002.\n",
      "     |      @see: L{assortativity_degree()} when the types are the vertex degrees\n",
      "     |  \n",
      "     |  assortativity_degree(...)\n",
      "     |      assortativity_degree(directed=True)\n",
      "     |      \n",
      "     |      Returns the assortativity of a graph based on vertex degrees.\n",
      "     |      \n",
      "     |      See L{assortativity()} for the details. L{assortativity_degree()} simply\n",
      "     |      calls L{assortativity()} with the vertex degrees as types.\n",
      "     |      \n",
      "     |      @param directed: whether to consider edge directions for directed graphs\n",
      "     |        or not. This argument is ignored for undirected graphs.\n",
      "     |      @return: the assortativity coefficient\n",
      "     |      \n",
      "     |      @see: L{assortativity()}\n",
      "     |  \n",
      "     |  assortativity_nominal(...)\n",
      "     |      assortativity_nominal(types, directed=True)\n",
      "     |      \n",
      "     |      Returns the assortativity of the graph based on vertex categories.\n",
      "     |      \n",
      "     |      Assuming that the vertices belong to different categories, this\n",
      "     |      function calculates the assortativity coefficient, which specifies\n",
      "     |      the extent to which the connections stay within categories. The\n",
      "     |      assortativity coefficient is one if all the connections stay within\n",
      "     |      categories and minus one if all the connections join vertices of\n",
      "     |      different categories. For a randomly connected network, it is\n",
      "     |      asymptotically zero.\n",
      "     |      \n",
      "     |      See equation (2) in Newman MEJ: Mixing patterns in networks, Phys Rev E\n",
      "     |      67:026126 (2003) for the proper definition.\n",
      "     |      \n",
      "     |      @param types: vertex types in a list or the name of a vertex attribute\n",
      "     |        holding vertex types. Types should be denoted by numeric values.\n",
      "     |      @param directed: whether to consider edge directions or not.\n",
      "     |      @return: the assortativity coefficient\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Newman MEJ: Mixing patterns in networks, Phys Rev E 67:026126, 2003.\n",
      "     |  \n",
      "     |  attributes(...)\n",
      "     |      attributes()\n",
      "     |      \n",
      "     |      @return: the attribute name list of the graph\n",
      "     |  \n",
      "     |  authority_score(...)\n",
      "     |      authority_score(weights=None, scale=True, arpack_options=None, return_eigenvalue=False)\n",
      "     |      \n",
      "     |      Calculates Kleinberg's authority score for the vertices of the graph\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param scale: whether to normalize the scores so that the largest one\n",
      "     |        is 1.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @param return_eigenvalue: whether to return the largest eigenvalue\n",
      "     |      @return: the authority scores in a list and optionally the largest eigenvalue\n",
      "     |        as a second member of a tuple\n",
      "     |      \n",
      "     |      @see: hub_score()\n",
      "     |  \n",
      "     |  average_path_length(...)\n",
      "     |      average_path_length(directed=True, unconn=True)\n",
      "     |      \n",
      "     |      Calculates the average path length in a graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths in case of a\n",
      "     |        directed graph. Ignored for undirected graphs.\n",
      "     |      @param unconn: what to do when the graph is unconnected. If C{True},\n",
      "     |        the average of the geodesic lengths in the components is\n",
      "     |        calculated. Otherwise for all unconnected vertex pairs,\n",
      "     |        a path length equal to the number of vertices is used.\n",
      "     |      @return: the average path length in the graph\n",
      "     |  \n",
      "     |  betweenness(...)\n",
      "     |      betweenness(vertices=None, directed=True, cutoff=None, weights=None, nobigint=True)\n",
      "     |      \n",
      "     |      Calculates or estimates the betweenness of vertices in a graph.\n",
      "     |      \n",
      "     |      Keyword arguments:\n",
      "     |      @param vertices: the vertices for which the betweennesses must be returned.\n",
      "     |        If C{None}, assumes all of the vertices in the graph.\n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param cutoff: if it is an integer, only paths less than or equal to this\n",
      "     |        length are considered, effectively resulting in an estimation of the\n",
      "     |        betweenness for the given vertices. If C{None}, the exact betweenness is\n",
      "     |        returned.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param nobigint: if C{True}, igraph uses the longest available integer\n",
      "     |        type on the current platform to count shortest paths. For some large\n",
      "     |        networks that have a specific structure, the counters may overflow.\n",
      "     |        To prevent this, use C{nobigint=False}, which forces igraph to use\n",
      "     |        arbitrary precision integers at the expense of increased computation\n",
      "     |        time.\n",
      "     |      @return: the (possibly estimated) betweenness of the given vertices in a list\n",
      "     |  \n",
      "     |  bfs(...)\n",
      "     |      bfs(vid, mode=OUT)\n",
      "     |      \n",
      "     |      Conducts a breadth first search (BFS) on the graph.\n",
      "     |      \n",
      "     |      @param vid: the root vertex ID\n",
      "     |      @param mode: either L{IN} or L{OUT} or L{ALL}, ignored\n",
      "     |        for undirected graphs.\n",
      "     |      @return: a tuple with the following items:\n",
      "     |         - The vertex IDs visited (in order)\n",
      "     |         - The start indices of the layers in the vertex list\n",
      "     |         - The parent of every vertex in the BFS\n",
      "     |  \n",
      "     |  bfsiter(...)\n",
      "     |      bfsiter(vid, mode=OUT, advanced=False)\n",
      "     |      \n",
      "     |      Constructs a breadth first search (BFS) iterator of the graph.\n",
      "     |      \n",
      "     |      @param vid: the root vertex ID\n",
      "     |      @param mode: either L{IN} or L{OUT} or L{ALL}.\n",
      "     |      @param advanced: if C{False}, the iterator returns the next\n",
      "     |        vertex in BFS order in every step. If C{True}, the iterator\n",
      "     |        returns the distance of the vertex from the root and the\n",
      "     |        parent of the vertex in the BFS tree as well.\n",
      "     |      @return: the BFS iterator as an L{igraph.BFSIter} object.\n",
      "     |  \n",
      "     |  bibcoupling(...)\n",
      "     |      bibcoupling(vertices=None)\n",
      "     |      \n",
      "     |      Calculates bibliographic coupling scores for given vertices in a graph.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None}, all vertices\n",
      "     |        will be considered.\n",
      "     |      @return: bibliographic coupling scores for all given vertices in a matrix.\n",
      "     |  \n",
      "     |  biconnected_components(...)\n",
      "     |      biconnected_components(return_articulation_points=True)\n",
      "     |      \n",
      "     |      Calculates the biconnected components of the graph.\n",
      "     |      \n",
      "     |      @param return_articulation_points: whether to return the articulation\n",
      "     |        points as well\n",
      "     |      @return: a list of lists containing edge indices making up spanning trees\n",
      "     |        of the biconnected components (one spanning tree for each component)\n",
      "     |        and optionally the list of articulation points\n",
      "     |  \n",
      "     |  bipartite_projection(...)\n",
      "     |      bipartite_projection(types, multiplicity=True, probe1=-1, which=-1)\n",
      "     |      \n",
      "     |      Internal function, undocumented.\n",
      "     |      \n",
      "     |      @see: Graph.bipartite_projection()\n",
      "     |  \n",
      "     |  bipartite_projection_size(...)\n",
      "     |      bipartite_projection_size(types)\n",
      "     |      \n",
      "     |      Internal function, undocumented.\n",
      "     |      \n",
      "     |      @see: Graph.bipartite_projection_size()\n",
      "     |  \n",
      "     |  canonical_permutation(...)\n",
      "     |      canonical_permutation(sh=\"fm\")\n",
      "     |      \n",
      "     |      Calculates the canonical permutation of a graph using the BLISS isomorphism\n",
      "     |      algorithm.\n",
      "     |      \n",
      "     |      Passing the permutation returned here to L{Graph.permute_vertices()} will\n",
      "     |      transform the graph into its canonical form.\n",
      "     |      \n",
      "     |      See U{http://www.tcs.hut.fi/Software/bliss/index.html} for more information\n",
      "     |      about the BLISS algorithm and canonical permutations.\n",
      "     |      \n",
      "     |      @param sh: splitting heuristics for graph as a case-insensitive string,\n",
      "     |        with the following possible values:\n",
      "     |      \n",
      "     |          - C{\"f\"}: first non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fl\"}: first largest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fs\"}: first smallest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fm\"}: first maximally non-trivially connected non-singleton\n",
      "     |            cell\n",
      "     |      \n",
      "     |          - C{\"flm\"}: largest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fsm\"}: smallest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |      @return: a permutation vector containing vertex IDs. Vertex 0 in the original\n",
      "     |        graph will be mapped to an ID contained in the first element of this\n",
      "     |        vector; vertex 1 will be mapped to the second and so on.\n",
      "     |  \n",
      "     |  clique_number(...)\n",
      "     |      clique_number()\n",
      "     |      \n",
      "     |      Returns the clique number of the graph.\n",
      "     |      \n",
      "     |      The clique number of the graph is the size of the largest clique.\n",
      "     |      \n",
      "     |      @see: L{largest_cliques()} for the largest cliques.\n",
      "     |  \n",
      "     |  cliques(...)\n",
      "     |      cliques(min=0, max=0)\n",
      "     |      \n",
      "     |      Returns some or all cliques of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      A clique is a complete subgraph -- a set of vertices where an edge\n",
      "     |      is present between any two of them (excluding loops)\n",
      "     |      \n",
      "     |      @param min: the minimum size of cliques to be returned. If zero or\n",
      "     |        negative, no lower bound will be used.\n",
      "     |      @param max: the maximum size of cliques to be returned. If zero or\n",
      "     |        negative, no upper bound will be used.\n",
      "     |  \n",
      "     |  closeness(...)\n",
      "     |      closeness(vertices=None, mode=ALL, cutoff=None, weights=None,\n",
      "     |                normalized=True)\n",
      "     |      \n",
      "     |      Calculates the closeness centralities of given vertices in a graph.\n",
      "     |      \n",
      "     |      The closeness centerality of a vertex measures how easily other\n",
      "     |      vertices can be reached from it (or the other way: how easily it\n",
      "     |      can be reached from the other vertices). It is defined as the\n",
      "     |      number of the number of vertices minus one divided by the sum of\n",
      "     |      the lengths of all geodesics from/to the given vertex.\n",
      "     |      \n",
      "     |      If the graph is not connected, and there is no path between two\n",
      "     |      vertices, the number of vertices is used instead the length of\n",
      "     |      the geodesic. This is always longer than the longest possible\n",
      "     |      geodesic.\n",
      "     |      \n",
      "     |      @param vertices: the vertices for which the closenesses must\n",
      "     |        be returned. If C{None}, uses all of the vertices in the graph.\n",
      "     |      @param mode: must be one of L{IN}, L{OUT} and L{ALL}. L{IN} means\n",
      "     |        that the length of the incoming paths, L{OUT} means that the\n",
      "     |        length of the outgoing paths must be calculated. L{ALL} means\n",
      "     |        that both of them must be calculated.\n",
      "     |      @param cutoff: if it is an integer, only paths less than or equal to this\n",
      "     |        length are considered, effectively resulting in an estimation of the\n",
      "     |        closeness for the given vertices (which is always an underestimation of the\n",
      "     |        real closeness, since some vertex pairs will appear as disconnected even\n",
      "     |        though they are connected).. If C{None}, the exact closeness is\n",
      "     |        returned.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param normalized: Whether to normalize the raw closeness scores by\n",
      "     |        multiplying by the number of vertices minus one.\n",
      "     |      @return: the calculated closenesses in a list\n",
      "     |  \n",
      "     |  clusters(...)\n",
      "     |      clusters(mode=STRONG)\n",
      "     |      \n",
      "     |      Calculates the (strong or weak) clusters for a given graph.\n",
      "     |      \n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a L{VertexClustering} object.\n",
      "     |        It is advised to use that.\n",
      "     |      @param mode: must be either C{STRONG} or C{WEAK}, depending on\n",
      "     |        the clusters being sought. Optional, defaults to C{STRONG}.\n",
      "     |      @return: the component index for every node in the graph.\n",
      "     |  \n",
      "     |  cocitation(...)\n",
      "     |      cocitation(vertices=None)\n",
      "     |      \n",
      "     |      Calculates cocitation scores for given vertices in a graph.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None}, all vertices\n",
      "     |        will be considered.\n",
      "     |      @return: cocitation scores for all given vertices in a matrix.\n",
      "     |  \n",
      "     |  cohesive_blocks(...)\n",
      "     |      cohesive_blocks()\n",
      "     |      \n",
      "     |      Calculates the cohesive block structure of the graph.\n",
      "     |      \n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a L{CohesiveBlocks} object.\n",
      "     |        It is advised to use that.\n",
      "     |  \n",
      "     |  community_edge_betweenness(...)\n",
      "     |      community_edge_betweenness(directed=True, weights=None)\n",
      "     |      \n",
      "     |      Community structure detection based on the betweenness of the edges in\n",
      "     |      the network. This algorithm was invented by M Girvan and MEJ Newman,\n",
      "     |      see: M Girvan and MEJ Newman: Community structure in social and biological\n",
      "     |      networks, Proc. Nat. Acad. Sci. USA 99, 7821-7826 (2002).\n",
      "     |      \n",
      "     |      The idea is that the betweenness of the edges connecting two communities\n",
      "     |      is typically high. So we gradually remove the edge with the highest\n",
      "     |      betweenness from the network and recalculate edge betweenness after every\n",
      "     |      removal, as long as all edges are removed.\n",
      "     |      \n",
      "     |      @attention: this function is wrapped in a more convenient syntax in the\n",
      "     |        derived class L{Graph}. It is advised to use that instead of this version.\n",
      "     |      \n",
      "     |      @param directed: whether to take into account the directedness of the edges\n",
      "     |        when we calculate the betweenness values.\n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights.\n",
      "     |      \n",
      "     |      @return: a tuple with the merge matrix that describes the dendrogram\n",
      "     |        and the modularity scores before each merge. The modularity scores\n",
      "     |        use the weights if the original graph was weighted.\n",
      "     |  \n",
      "     |  community_fastgreedy(...)\n",
      "     |      community_fastgreedy(weights=None)\n",
      "     |      \n",
      "     |      Finds the community structure of the graph according to the algorithm of\n",
      "     |      Clauset et al based on the greedy optimization of modularity.\n",
      "     |      \n",
      "     |      This is a bottom-up algorithm: initially every vertex belongs to a separate\n",
      "     |      community, and communities are merged one by one. In every step, the two\n",
      "     |      communities being merged are the ones which result in the maximal increase\n",
      "     |      in modularity.\n",
      "     |      \n",
      "     |      @attention: this function is wrapped in a more convenient syntax in the\n",
      "     |        derived class L{Graph}. It is advised to use that instead of this version.\n",
      "     |      \n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights\n",
      "     |      @return: a tuple with the following elements:\n",
      "     |        1. The list of merges\n",
      "     |        2. The modularity scores before each merge\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: A. Clauset, M. E. J. Newman and C. Moore: I{Finding community\n",
      "     |        structure in very large networks.} Phys Rev E 70, 066111 (2004).\n",
      "     |      @see: modularity()\n",
      "     |  \n",
      "     |  community_infomap(...)\n",
      "     |      community_infomap(edge_weights=None, vertex_weights=None, trials=10)\n",
      "     |      \n",
      "     |      Finds the community structure of the network according to the Infomap\n",
      "     |      method of Martin Rosvall and Carl T. Bergstrom.\n",
      "     |      \n",
      "     |      See U{http://www.mapequation.org} for a visualization of the algorithm\n",
      "     |      or one of the references provided below.\n",
      "     |      \n",
      "     |      @param edge_weights: name of an edge attribute or a list containing\n",
      "     |        edge weights.\n",
      "     |      @param vertex_weights: name of an vertex attribute or a list containing\n",
      "     |        vertex weights.\n",
      "     |      @param trials: the number of attempts to partition the network.\n",
      "     |      @return: the calculated membership vector and the corresponding\n",
      "     |        codelength in a tuple.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: M. Rosvall and C. T. Bergstrom: I{Maps of information flow reveal\n",
      "     |        community structure in complex networks}. PNAS 105, 1118 (2008).\n",
      "     |        U{http://arxiv.org/abs/0707.0609}\n",
      "     |      @ref: M. Rosvall, D. Axelsson and C. T. Bergstrom: I{The map equation}.\n",
      "     |        Eur Phys J Special Topics 178, 13 (2009). U{http://arxiv.org/abs/0906.1405}\n",
      "     |  \n",
      "     |  community_label_propagation(...)\n",
      "     |      community_label_propagation(weights=None, initial=None, fixed=None)\n",
      "     |      \n",
      "     |      Finds the community structure of the graph according to the label\n",
      "     |      propagation method of Raghavan et al.\n",
      "     |      \n",
      "     |      Initially, each vertex is assigned a different label. After that,\n",
      "     |      each vertex chooses the dominant label in its neighbourhood in each\n",
      "     |      iteration. Ties are broken randomly and the order in which the\n",
      "     |      vertices are updated is randomized before every iteration. The algorithm\n",
      "     |      ends when vertices reach a consensus.\n",
      "     |      \n",
      "     |      Note that since ties are broken randomly, there is no guarantee that\n",
      "     |      the algorithm returns the same community structure after each run.\n",
      "     |      In fact, they frequently differ. See the paper of Raghavan et al\n",
      "     |      on how to come up with an aggregated community structure.\n",
      "     |      \n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights\n",
      "     |      @param initial: name of a vertex attribute or a list containing\n",
      "     |        the initial vertex labels. Labels are identified by integers from\n",
      "     |        zero to M{n-1} where M{n} is the number of vertices. Negative\n",
      "     |        numbers may also be present in this vector, they represent unlabeled\n",
      "     |        vertices.\n",
      "     |      @param fixed: a list of booleans for each vertex. C{True} corresponds\n",
      "     |        to vertices whose labeling should not change during the algorithm.\n",
      "     |        It only makes sense if initial labels are also given. Unlabeled\n",
      "     |        vertices cannot be fixed. Note that vertex attribute names are not\n",
      "     |        accepted here.\n",
      "     |      @return: the resulting membership vector\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Raghavan, U.N. and Albert, R. and Kumara, S. Near linear\n",
      "     |        time algorithm to detect community structures in large-scale\n",
      "     |        networks. Phys Rev E 76:036106, 2007. U{http://arxiv.org/abs/0709.2938}.\n",
      "     |  \n",
      "     |  community_leading_eigenvector(...)\n",
      "     |      community_leading_eigenvector(n=-1, arpack_options=None, weights=None)\n",
      "     |      \n",
      "     |      A proper implementation of Newman's eigenvector community structure\n",
      "     |      detection. Each split is done by maximizing the modularity regarding\n",
      "     |      the original network. See the reference for details.\n",
      "     |      \n",
      "     |      @attention: this function is wrapped in a more convenient syntax in the\n",
      "     |        derived class L{Graph}. It is advised to use that instead of this version.\n",
      "     |      \n",
      "     |      @param n: the desired number of communities. If negative, the algorithm\n",
      "     |        tries to do as many splits as possible. Note that the algorithm\n",
      "     |        won't split a community further if the signs of the leading eigenvector\n",
      "     |        are all the same.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights\n",
      "     |      @return: a tuple where the first element is the membership vector of the\n",
      "     |        clustering and the second element is the merge matrix.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: MEJ Newman: Finding community structure in networks using the\n",
      "     |        eigenvectors of matrices, arXiv:physics/0605087\n",
      "     |  \n",
      "     |  community_multilevel(...)\n",
      "     |      community_multilevel(weights=None, return_levels=True)\n",
      "     |      \n",
      "     |      Finds the community structure of the graph according to the multilevel\n",
      "     |      algorithm of Blondel et al. This is a bottom-up algorithm: initially\n",
      "     |      every vertex belongs to a separate community, and vertices are moved\n",
      "     |      between communities iteratively in a way that maximizes the vertices'\n",
      "     |      local contribution to the overall modularity score. When a consensus is\n",
      "     |      reached (i.e. no single move would increase the modularity score), every\n",
      "     |      community in the original graph is shrank to a single vertex (while\n",
      "     |      keeping the total weight of the incident edges) and the process continues\n",
      "     |      on the next level. The algorithm stops when it is not possible to increase\n",
      "     |      the modularity any more after shrinking the communities to vertices.\n",
      "     |      \n",
      "     |      @attention: this function is wrapped in a more convenient syntax in the\n",
      "     |        derived class L{Graph}. It is advised to use that instead of this version.\n",
      "     |      \n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights\n",
      "     |      @param return_levels: if C{True}, returns the multilevel result. If\n",
      "     |        C{False}, only the best level (corresponding to the best modularity)\n",
      "     |        is returned.\n",
      "     |      @return: either a single list describing the community membership of each\n",
      "     |        vertex (if C{return_levels} is C{False}), or a list of community membership\n",
      "     |        vectors, one corresponding to each level and a list of corresponding\n",
      "     |        modularities (if C{return_levels} is C{True}).\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: VD Blondel, J-L Guillaume, R Lambiotte and E Lefebvre: Fast\n",
      "     |        unfolding of community hierarchies in large networks. J Stat Mech\n",
      "     |        P10008 (2008), http://arxiv.org/abs/0803.0476\n",
      "     |      @see: modularity()\n",
      "     |  \n",
      "     |  community_optimal_modularity(...)\n",
      "     |      community_optimal_modularity(weights=None)\n",
      "     |      \n",
      "     |      Calculates the optimal modularity score of the graph and the\n",
      "     |      corresponding community structure.\n",
      "     |      \n",
      "     |      This function uses the GNU Linear Programming Kit to solve a large\n",
      "     |      integer optimization problem in order to find the optimal modularity\n",
      "     |      score and the corresponding community structure, therefore it is\n",
      "     |      unlikely to work for graphs larger than a few (less than a hundred)\n",
      "     |      vertices. Consider using one of the heuristic approaches instead if\n",
      "     |      you have such a large graph.\n",
      "     |      \n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights.\n",
      "     |      \n",
      "     |      @return: the calculated membership vector and the corresponding\n",
      "     |        modularity in a tuple.\n",
      "     |  \n",
      "     |  community_spinglass(...)\n",
      "     |      community_spinglass(weights=None, spins=25, parupdate=False, start_temp=1, stop_temp=0.01, cool_fact=0.99, update_rule=\"config\", gamma=1, implementation=\"orig\", lambda=1)\n",
      "     |      \n",
      "     |      Finds the community structure of the graph according to the spinglass\n",
      "     |      community detection method of Reichardt & Bornholdt.\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param spins: integer, the number of spins to use. This is the upper limit\n",
      "     |        for the number of communities. It is not a problem to supply a\n",
      "     |        (reasonably) big number here, in which case some spin states will be\n",
      "     |        unpopulated.\n",
      "     |      @param parupdate: whether to update the spins of the vertices in parallel\n",
      "     |        (synchronously) or not\n",
      "     |      @param start_temp: the starting temperature\n",
      "     |      @param stop_temp: the stop temperature\n",
      "     |      @param cool_fact: cooling factor for the simulated annealing\n",
      "     |      @param update_rule: specifies the null model of the simulation. Possible\n",
      "     |        values are C{\"config\"} (a random graph with the same vertex degrees\n",
      "     |        as the input graph) or C{\"simple\"} (a random graph with the same number\n",
      "     |        of edges)\n",
      "     |      @param gamma: the gamma argument of the algorithm, specifying the balance\n",
      "     |        between the importance of present and missing edges within a community.\n",
      "     |        The default value of 1.0 assigns equal importance to both of them.\n",
      "     |      @param implementation: currently igraph contains two implementations for\n",
      "     |        the spinglass community detection algorithm. The faster original\n",
      "     |        implementation is the default. The other implementation is able to take\n",
      "     |        into account negative weights, this can be chosen by setting\n",
      "     |        C{implementation} to C{\"neg\"}.\n",
      "     |      @param lambda: the lambda argument of the algorithm, which specifies the\n",
      "     |        balance between the importance of present and missing negatively\n",
      "     |        weighted edges within a community. Smaller values of lambda lead\n",
      "     |        to communities with less negative intra-connectivity. If the argument\n",
      "     |        is zero, the algorithm reduces to a graph coloring algorithm, using\n",
      "     |        the number of spins as colors. This argument is ignored if the\n",
      "     |        original implementation is used.\n",
      "     |      @return: the community membership vector.\n",
      "     |  \n",
      "     |  community_walktrap(...)\n",
      "     |      community_walktrap(weights=None, steps=None)\n",
      "     |      \n",
      "     |      Finds the community structure of the graph according to the random walk\n",
      "     |      method of Latapy & Pons.\n",
      "     |      \n",
      "     |      The basic idea of the algorithm is that short random walks tend to stay\n",
      "     |      in the same community. The method provides a dendrogram.\n",
      "     |      \n",
      "     |      @attention: this function is wrapped in a more convenient syntax in the\n",
      "     |        derived class L{Graph}. It is advised to use that instead of this version.\n",
      "     |      \n",
      "     |      @param weights: name of an edge attribute or a list containing\n",
      "     |        edge weights\n",
      "     |      @return: a tuple with the list of merges and the modularity scores corresponding\n",
      "     |        to each merge\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Pascal Pons, Matthieu Latapy: Computing communities in large networks\n",
      "     |        using random walks, U{http://arxiv.org/abs/physics/0512106}.\n",
      "     |      @see: modularity()\n",
      "     |  \n",
      "     |  complementer(...)\n",
      "     |      complementer(loops=False)\n",
      "     |      \n",
      "     |      Returns the complementer of the graph\n",
      "     |      \n",
      "     |      @param loops: whether to include loop edges in the complementer.\n",
      "     |      @return: the complementer of the graph\n",
      "     |  \n",
      "     |  compose(...)\n",
      "     |      compose(other)\n",
      "     |      \n",
      "     |      Returns the composition of two graphs.\n",
      "     |  \n",
      "     |  constraint(...)\n",
      "     |      constraint(vertices=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates Burt's constraint scores for given vertices in a graph.\n",
      "     |      \n",
      "     |      Burt's constraint is higher if ego has less, or mutually stronger\n",
      "     |      related (i.e. more redundant) contacts. Burt's measure of\n",
      "     |      constraint, C[i], of vertex i's ego network V[i], is defined for\n",
      "     |      directed and valued graphs as follows:\n",
      "     |      \n",
      "     |      C[i] = sum( sum( (p[i,q] p[q,j])^2, q in V[i], q != i,j ), j in V[], j != i)\n",
      "     |      \n",
      "     |      for a graph of order (ie. number od vertices) N, where proportional\n",
      "     |      tie strengths are defined as follows:\n",
      "     |      \n",
      "     |      p[i,j]=(a[i,j]+a[j,i]) / sum(a[i,k]+a[k,i], k in V[i], k != i),\n",
      "     |      a[i,j] are elements of A and the latter being the graph adjacency matrix.\n",
      "     |      \n",
      "     |      For isolated vertices, constraint is undefined.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed or C{None} for all vertices.\n",
      "     |      @param weights: weights associated to the edges. Can be an attribute name\n",
      "     |        as well. If C{None}, every edge will have the same weight.\n",
      "     |      @return: constraint scores for all given vertices in a matrix.\n",
      "     |  \n",
      "     |  contract_vertices(...)\n",
      "     |      contract_vertices(mapping, combine_attrs=None)\n",
      "     |      \n",
      "     |      Contracts some vertices in the graph, i.e. replaces groups of vertices\n",
      "     |      with single vertices. Edges are not affected.\n",
      "     |      \n",
      "     |      @param mapping: numeric vector which gives the mapping between old and\n",
      "     |        new vertex IDs. Vertices having the same new vertex ID in this vector\n",
      "     |        will be remapped into a single new vertex. It is safe to pass the\n",
      "     |        membership vector of a L{VertexClustering} object here.\n",
      "     |      @param combine_attrs: specifies how to combine the attributes of\n",
      "     |        the vertices being collapsed into a single one. If it is C{None},\n",
      "     |        all the attributes will be lost. If it is a function, the\n",
      "     |        attributes of the vertices will be collected and passed on to\n",
      "     |        that function which will return the new attribute value that has to\n",
      "     |        be assigned to the single collapsed vertex. It can also be one of\n",
      "     |        the following string constants which define built-in collapsing\n",
      "     |        functions: C{sum}, C{prod}, C{mean}, C{median}, C{max}, C{min},\n",
      "     |        C{first}, C{last}, C{random}. You can also specify different\n",
      "     |        combination functions for different attributes by passing a dict\n",
      "     |        here which maps attribute names to functions. See\n",
      "     |        L{Graph.simplify()} for more details.\n",
      "     |      @return: C{None}.\n",
      "     |      @see: L{Graph.simplify()}\n",
      "     |  \n",
      "     |  convergence_degree(...)\n",
      "     |      convergence_degree()\n",
      "     |      \n",
      "     |      Undocumented (yet).\n",
      "     |  \n",
      "     |  convergence_field_size(...)\n",
      "     |      convergence_field_size()\n",
      "     |      \n",
      "     |      Undocumented (yet).\n",
      "     |  \n",
      "     |  copy(...)\n",
      "     |      copy()\n",
      "     |      \n",
      "     |      Creates an exact deep copy of the graph.\n",
      "     |  \n",
      "     |  coreness(...)\n",
      "     |      coreness(mode=ALL)\n",
      "     |      \n",
      "     |      Finds the coreness (shell index) of the vertices of the network.\n",
      "     |      \n",
      "     |      The M{k}-core of a graph is a maximal subgraph in which each vertex\n",
      "     |      has at least degree k. (Degree here means the degree in the\n",
      "     |      subgraph of course). The coreness of a vertex is M{k} if it\n",
      "     |      is a member of the M{k}-core but not a member of the M{k+1}-core.\n",
      "     |      \n",
      "     |      @param mode: whether to compute the in-corenesses (L{IN}), the\n",
      "     |        out-corenesses (L{OUT}) or the undirected corenesses (L{ALL}).\n",
      "     |        Ignored and assumed to be L{ALL} for undirected graphs.\n",
      "     |      @return: the corenesses for each vertex.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Vladimir Batagelj, Matjaz Zaversnik: I{An M{O(m)} Algorithm\n",
      "     |        for Core Decomposition of Networks.}\n",
      "     |  \n",
      "     |  count_isomorphisms_vf2(...)\n",
      "     |      count_isomorphisms_vf2(other=None, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Determines the number of isomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph. If C{None}, the number of automorphisms\n",
      "     |        will be returned.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: the number of isomorphisms between the two given graphs (or the\n",
      "     |        number of automorphisms if C{other} is C{None}.\n",
      "     |  \n",
      "     |  count_multiple(...)\n",
      "     |      count_multiple(edges=None)\n",
      "     |      \n",
      "     |      Counts the multiplicities of the given edges.\n",
      "     |      \n",
      "     |      @param edges: edge indices for which we want to count their\n",
      "     |        multiplicity. If C{None}, all edges are counted.\n",
      "     |      @return: the multiplicities of the given edges as a list.\n",
      "     |  \n",
      "     |  count_subisomorphisms_vf2(...)\n",
      "     |      count_subisomorphisms_vf2(other, color1=None, color2=None,\n",
      "     |        edge_color1=None, edge_color2=None, node_compat_fn=None,\n",
      "     |        edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Determines the number of subisomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: the number of subisomorphisms between the two given graphs\n",
      "     |  \n",
      "     |  decompose(...)\n",
      "     |      decompose(mode=STRONG, maxcompno=None, minelements=1)\n",
      "     |      \n",
      "     |      Decomposes the graph into subgraphs.\n",
      "     |      \n",
      "     |      @param mode: must be either STRONG or WEAK, depending on the\n",
      "     |        clusters being sought.\n",
      "     |      @param maxcompno: maximum number of components to return.\n",
      "     |        C{None} means all possible components.\n",
      "     |      @param minelements: minimum number of vertices in a component.\n",
      "     |        By setting this to 2, isolated vertices are not returned\n",
      "     |        as separate components.\n",
      "     |      @return: a list of the subgraphs. Every returned subgraph is a\n",
      "     |        copy of the original.\n",
      "     |  \n",
      "     |  degree(...)\n",
      "     |      degree(vertices, mode=ALL, loops=True)\n",
      "     |      \n",
      "     |      Returns some vertex degrees from the graph.\n",
      "     |      \n",
      "     |      This method accepts a single vertex ID or a list of vertex IDs as a\n",
      "     |      parameter, and returns the degree of the given vertices (in the\n",
      "     |      form of a single integer or a list, depending on the input\n",
      "     |      parameter).\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @param mode: the type of degree to be returned (L{OUT} for\n",
      "     |        out-degrees, L{IN} IN for in-degrees or L{ALL} for the sum of\n",
      "     |        them).\n",
      "     |      @param loops: whether self-loops should be counted.\n",
      "     |  \n",
      "     |  delete_edges(...)\n",
      "     |      delete_edges(es)\n",
      "     |      \n",
      "     |      Removes edges from the graph.\n",
      "     |      \n",
      "     |      All vertices will be kept, even if they lose all their edges.\n",
      "     |      Nonexistent edges will be silently ignored.\n",
      "     |      \n",
      "     |      @param es: the list of edges to be removed. Edges are identifed by\n",
      "     |        edge IDs. L{EdgeSeq} objects are also accepted here.\n",
      "     |  \n",
      "     |  delete_vertices(...)\n",
      "     |      delete_vertices(vs)\n",
      "     |      \n",
      "     |      Deletes vertices and all its edges from the graph.\n",
      "     |      \n",
      "     |      @param vs: a single vertex ID or the list of vertex IDs\n",
      "     |        to be deleted.\n",
      "     |  \n",
      "     |  density(...)\n",
      "     |      density(loops=False)\n",
      "     |      \n",
      "     |      Calculates the density of the graph.\n",
      "     |      \n",
      "     |      @param loops: whether to take loops into consideration. If C{True},\n",
      "     |        the algorithm assumes that there might be some loops in the graph\n",
      "     |        and calculates the density accordingly. If C{False}, the algorithm\n",
      "     |        assumes that there can't be any loops.\n",
      "     |      @return: the reciprocity of the graph.\n",
      "     |  \n",
      "     |  diameter(...)\n",
      "     |      diameter(directed=True, unconn=True, weights=None)\n",
      "     |      \n",
      "     |      Calculates the diameter of the graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param unconn: if C{True} and the graph is unconnected, the\n",
      "     |        longest geodesic within a component will be returned. If\n",
      "     |        C{False} and the graph is unconnected, the result is the\n",
      "     |        number of vertices if there are no weights or infinity\n",
      "     |        if there are weights.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the diameter\n",
      "     |  \n",
      "     |  difference(...)\n",
      "     |      difference(other)\n",
      "     |      \n",
      "     |      Subtracts the given graph from the original\n",
      "     |  \n",
      "     |  disjoint_union(...)\n",
      "     |      disjoint_union(graphs)\n",
      "     |      \n",
      "     |      Creates the disjoint union of two (or more) graphs.\n",
      "     |      \n",
      "     |      @param graphs: the list of graphs to be united with the current one.\n",
      "     |  \n",
      "     |  diversity(...)\n",
      "     |      diversity(vertices=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates the structural diversity index of the vertices.\n",
      "     |      \n",
      "     |      The structural diversity index of a vertex is simply the (normalized)\n",
      "     |      Shannon entropy of the weights of the edges incident on the vertex.\n",
      "     |      \n",
      "     |      The measure is defined for undirected graphs only; edge directions are\n",
      "     |      ignored.\n",
      "     |      \n",
      "     |      @param vertices: the vertices for which the diversity indices must\n",
      "     |        be returned. If C{None}, uses all of the vertices in the graph.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the calculated diversity indices in a list, or a single number if\n",
      "     |        a single vertex was supplied.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Eagle N, Macy M and Claxton R: Network diversity and economic\n",
      "     |        development, Science 328, 1029--1031, 2010.\n",
      "     |  \n",
      "     |  dyad_census(...)\n",
      "     |      dyad_census()\n",
      "     |      \n",
      "     |      Dyad census, as defined by Holland and Leinhardt\n",
      "     |      \n",
      "     |      Dyad census means classifying each pair of vertices of a directed\n",
      "     |      graph into three categories: mutual, there is an edge from I{a} to\n",
      "     |      I{b} and also from I{b} to I{a}; asymmetric, there is an edge\n",
      "     |      either from I{a} to I{b} or from I{b} to I{a} but not the other way\n",
      "     |      and null, no edges between I{a} and I{b}.\n",
      "     |      \n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a L{DyadCensus} object.\n",
      "     |        It is advised to use that.\n",
      "     |      \n",
      "     |      @return: the number of mutual, asymmetric and null connections in a\n",
      "     |        3-tuple.\n",
      "     |  \n",
      "     |  eccentricity(...)\n",
      "     |      eccentricity(vertices=None, mode=ALL)\n",
      "     |      \n",
      "     |      Calculates the eccentricities of given vertices in a graph.\n",
      "     |      \n",
      "     |      The eccentricity of a vertex is calculated by measuring the\n",
      "     |      shortest distance from (or to) the vertex, to (or from) all other\n",
      "     |      vertices in the graph, and taking the maximum.\n",
      "     |      \n",
      "     |      @param vertices: the vertices for which the eccentricity scores must\n",
      "     |        be returned. If C{None}, uses all of the vertices in the graph.\n",
      "     |      @param mode: must be one of L{IN}, L{OUT} and L{ALL}. L{IN} means\n",
      "     |        that edge directions are followed; C{OUT} means that edge directions\n",
      "     |        are followed the opposite direction; C{ALL} means that directions are\n",
      "     |        ignored. The argument has no effect for undirected graphs.\n",
      "     |      @return: the calculated eccentricities in a list, or a single number if\n",
      "     |        a single vertex was supplied.\n",
      "     |  \n",
      "     |  ecount(...)\n",
      "     |      ecount()\n",
      "     |      \n",
      "     |      Counts the number of edges.\n",
      "     |      @return: the number of edges in the graph.\n",
      "     |      @rtype: integer\n",
      "     |  \n",
      "     |  edge_attributes(...)\n",
      "     |      edge_attributes()\n",
      "     |      \n",
      "     |      @return: the attribute name list of the graph's edges\n",
      "     |  \n",
      "     |  edge_betweenness(...)\n",
      "     |      edge_betweenness(directed=True, cutoff=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates or estimates the edge betweennesses in a graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param cutoff: if it is an integer, only paths less than or equal to this\n",
      "     |        length are considered, effectively resulting in an estimation of the\n",
      "     |        betweenness values. If C{None}, the exact betweennesses are\n",
      "     |        returned.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: a list with the (exact or estimated) edge betweennesses of all\n",
      "     |        edges.\n",
      "     |  \n",
      "     |  edge_connectivity(...)\n",
      "     |      edge_connectivity(source=-1, target=-1, checks=True)\n",
      "     |      \n",
      "     |      Calculates the edge connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The edge connectivity between two given vertices is the number of edges\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of edge disjoint directed\n",
      "     |      paths between the vertices. The edge connectivity of the graph is the minimal\n",
      "     |      edge connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the edge connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall edge connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @return: the edge connectivity\n",
      "     |  \n",
      "     |  eigen_adjacency(...)\n",
      "     |  \n",
      "     |  eigenvector_centrality(...)\n",
      "     |      eigenvector_centrality(directed=True, scale=True, weights=None, return_eigenvalue=False, arpack_options=None)\n",
      "     |      \n",
      "     |      Calculates the eigenvector centralities of the vertices in a graph.\n",
      "     |      \n",
      "     |      @param directed: whether to consider edge directions in a directed\n",
      "     |        graph. Ignored for undirected graphs.\n",
      "     |      @param scale: whether to normalize the centralities so the largest\n",
      "     |        one will always be 1.\n",
      "     |      @param weights: edge weights given as a list or an edge attribute. If\n",
      "     |        C{None}, all edges have equal weight.\n",
      "     |      @param return_eigenvalue: whether to return the actual largest\n",
      "     |        eigenvalue along with the centralities\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object that can be used\n",
      "     |        to fine-tune the calculation. If it is omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @return: the eigenvector centralities in a list and optionally the\n",
      "     |        largest eigenvalue (as a second member of a tuple)\n",
      "     |  \n",
      "     |  farthest_points(...)\n",
      "     |      farthest_points(directed=True, unconn=True, weights=None)\n",
      "     |      \n",
      "     |      Returns two vertex IDs whose distance equals the actual diameter\n",
      "     |      of the graph.\n",
      "     |      \n",
      "     |      If there are many shortest paths with the length of the diameter,\n",
      "     |      it returns the first one it found.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param unconn: if C{True} and the graph is unconnected, the\n",
      "     |        longest geodesic within a component will be returned. If\n",
      "     |        C{False} and the graph is unconnected, the result contains the\n",
      "     |        number of vertices if there are no weights or infinity\n",
      "     |        if there are weights.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: a triplet containing the two vertex IDs and their distance.\n",
      "     |        The IDs are C{None} if the graph is unconnected and C{unconn}\n",
      "     |        is C{False}.\n",
      "     |  \n",
      "     |  feedback_arc_set(...)\n",
      "     |      feedback_arc_set(weights=None, method=\"eades\")\n",
      "     |      \n",
      "     |      Calculates an approximately or exactly minimal feedback arc set.\n",
      "     |      \n",
      "     |      A feedback arc set is a set of edges whose removal makes the graph acyclic.\n",
      "     |      Since this is always possible by removing all the edges, we are in general\n",
      "     |      interested in removing the smallest possible number of edges, or an edge set\n",
      "     |      with as small total weight as possible. This method calculates one such edge\n",
      "     |      set. Note that the task is trivial for an undirected graph as it is enough\n",
      "     |      to find a spanning tree and then remove all the edges not in the spanning\n",
      "     |      tree. Of course it is more complicated for directed graphs.\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name. When given, the algorithm will strive to\n",
      "     |        remove lightweight edges in order to minimize the total weight of the\n",
      "     |        feedback arc set.\n",
      "     |      @param method: the algorithm to use. C{\"eades\"} uses the greedy cycle\n",
      "     |        breaking heuristic of Eades, Lin and Smyth, which is linear in the number\n",
      "     |        of edges but not necessarily optimal; however, it guarantees that the\n",
      "     |        number of edges to be removed is smaller than |E|/2 - |V|/6. C{\"ip\"} uses\n",
      "     |        an integer programming formulation which is guaranteed to yield an optimal\n",
      "     |        result, but is too slow for large graphs.\n",
      "     |      @return: the IDs of the edges to be removed, in a list.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Eades P, Lin X and Smyth WF: A fast and effective heuristic for the\n",
      "     |        feedback arc set problem. In: Proc Inf Process Lett 319-323, 1993.\n",
      "     |  \n",
      "     |  get_adjacency(...)\n",
      "     |      get_adjacency(type=GET_ADJACENCY_BOTH, eids=False)\n",
      "     |      \n",
      "     |      Returns the adjacency matrix of a graph.\n",
      "     |      \n",
      "     |      @param type: either C{GET_ADJACENCY_LOWER} (uses the\n",
      "     |        lower triangle of the matrix) or C{GET_ADJACENCY_UPPER}\n",
      "     |        (uses the upper triangle) or C{GET_ADJACENCY_BOTH}\n",
      "     |        (uses both parts). Ignored for directed graphs.\n",
      "     |      @param eids: if C{True}, the result matrix will contain\n",
      "     |        zeros for non-edges and the ID of the edge plus one\n",
      "     |        for edges in the appropriate cell. If C{False}, the\n",
      "     |        result matrix will contain the number of edges for\n",
      "     |        each vertex pair.\n",
      "     |      @return: the adjacency matrix.\n",
      "     |  \n",
      "     |  get_all_shortest_paths(...)\n",
      "     |      get_all_shortest_paths(v, to=None, weights=None, mode=OUT)\n",
      "     |      \n",
      "     |      Calculates all of the shortest paths from/to a given node in a graph.\n",
      "     |      \n",
      "     |      @param v: the source for the calculated paths\n",
      "     |      @param to: a vertex selector describing the destination for\n",
      "     |        the calculated paths. This can be a single vertex ID, a list of\n",
      "     |        vertex IDs, a single vertex name, a list of vertex names or a\n",
      "     |        L{VertexSeq} object. C{None} means all the vertices.\n",
      "     |      @param weights: edge weights in a list or the name of an edge attribute\n",
      "     |        holding edge weights. If C{None}, all edges are assumed to have\n",
      "     |        equal weight.\n",
      "     |      @param mode: the directionality of the paths. L{IN} means to\n",
      "     |        calculate incoming paths, L{OUT} means to calculate outgoing\n",
      "     |        paths, L{ALL} means to calculate both ones.\n",
      "     |      @return: all of the shortest path from the given node to every other\n",
      "     |        reachable node in the graph in a list. Note that in case of mode=L{IN},\n",
      "     |        the vertices in a path are returned in reversed order!\n",
      "     |  \n",
      "     |  get_diameter(...)\n",
      "     |      get_diameter(directed=True, unconn=True, weights=None)\n",
      "     |      \n",
      "     |      Returns a path with the actual diameter of the graph.\n",
      "     |      \n",
      "     |      If there are many shortest paths with the length of the diameter,\n",
      "     |      it returns the first one it founds.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param unconn: if C{True} and the graph is unconnected, the\n",
      "     |        longest geodesic within a component will be returned. If\n",
      "     |        C{False} and the graph is unconnected, the result is the\n",
      "     |        number of vertices if there are no weights or infinity\n",
      "     |        if there are weights.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the vertices in the path in order.\n",
      "     |  \n",
      "     |  get_edgelist(...)\n",
      "     |      get_edgelist()\n",
      "     |      \n",
      "     |      Returns the edge list of a graph.\n",
      "     |  \n",
      "     |  get_eid(...)\n",
      "     |      get_eid(v1, v2, directed=True, error=True)\n",
      "     |      \n",
      "     |      Returns the edge ID of an arbitrary edge between vertices v1 and v2\n",
      "     |      \n",
      "     |      @param v1: the ID or name of the first vertex\n",
      "     |      @param v2: the ID or name of the second vertex\n",
      "     |      @param directed: whether edge directions should be considered in\n",
      "     |        directed graphs. The default is C{True}. Ignored for undirected\n",
      "     |        graphs.\n",
      "     |      @param error: if C{True}, an exception will be raised when the\n",
      "     |        given edge does not exist. If C{False}, -1 will be returned in\n",
      "     |        that case.\n",
      "     |      @return: the edge ID of an arbitrary edge between vertices v1 and v2\n",
      "     |  \n",
      "     |  get_eids(...)\n",
      "     |      get_eids(pairs=None, path=None, directed=True, error=True)\n",
      "     |      \n",
      "     |      Returns the edge IDs of some edges between some vertices.\n",
      "     |      \n",
      "     |      This method can operate in two different modes, depending on which\n",
      "     |      of the keyword arguments C{pairs} and C{path} are given.\n",
      "     |      \n",
      "     |      The method does not consider multiple edges; if there are multiple\n",
      "     |      edges between a pair of vertices, only the ID of one of the edges\n",
      "     |      is returned.\n",
      "     |      \n",
      "     |      @param pairs: a list of integer pairs. Each integer pair is considered\n",
      "     |        as a source-target vertex pair; the corresponding edge is looked up\n",
      "     |        in the graph and the edge ID is returned for each pair.\n",
      "     |      @param path: a list of vertex IDs. The list is considered as a\n",
      "     |        continuous path from the first vertex to the last, passing\n",
      "     |        through the intermediate vertices. The corresponding edge IDs\n",
      "     |        between the first and the second, the second and the third and\n",
      "     |        so on are looked up in the graph and the edge IDs are returned.\n",
      "     |        If both C{path} and C{pairs} are given, the two lists are\n",
      "     |        concatenated.\n",
      "     |      @param directed: whether edge directions should be considered in\n",
      "     |        directed graphs. The default is C{True}. Ignored for undirected\n",
      "     |        graphs.\n",
      "     |      @param error: if C{True}, an exception will be raised if a given\n",
      "     |        edge does not exist. If C{False}, -1 will be returned in\n",
      "     |        that case.\n",
      "     |      @return: the edge IDs in a list\n",
      "     |  \n",
      "     |  get_incidence(...)\n",
      "     |      get_incidence(types)\n",
      "     |      \n",
      "     |      Internal function, undocumented.\n",
      "     |      \n",
      "     |      @see: Graph.get_incidence()\n",
      "     |  \n",
      "     |  get_isomorphisms_vf2(...)\n",
      "     |      get_isomorphisms_vf2(other=None, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Returns all isomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph. If C{None}, the automorphisms\n",
      "     |        will be returned.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: a list of lists, each item of the list containing the mapping\n",
      "     |        from vertices of the second graph to the vertices of the first one\n",
      "     |  \n",
      "     |  get_shortest_paths(...)\n",
      "     |      get_shortest_paths(v, to=None, weights=None, mode=OUT, output=\"vpath\")\n",
      "     |      \n",
      "     |      Calculates the shortest paths from/to a given node in a graph.\n",
      "     |      \n",
      "     |      @param v: the source/destination for the calculated paths\n",
      "     |      @param to: a vertex selector describing the destination/source for\n",
      "     |        the calculated paths. This can be a single vertex ID, a list of\n",
      "     |        vertex IDs, a single vertex name, a list of vertex names or a\n",
      "     |        L{VertexSeq} object. C{None} means all the vertices.\n",
      "     |      @param weights: edge weights in a list or the name of an edge attribute\n",
      "     |        holding edge weights. If C{None}, all edges are assumed to have\n",
      "     |        equal weight.\n",
      "     |      @param mode: the directionality of the paths. L{IN} means to\n",
      "     |        calculate incoming paths, L{OUT} means to calculate outgoing\n",
      "     |        paths, L{ALL} means to calculate both ones.\n",
      "     |      @param output: determines what should be returned. If this is\n",
      "     |        C{\"vpath\"}, a list of vertex IDs will be returned, one path\n",
      "     |        for each target vertex. For unconnected graphs, some of the list\n",
      "     |        elements may be empty. Note that in case of mode=L{IN}, the vertices\n",
      "     |        in a path are returned in reversed order. If C{output=\"epath\"},\n",
      "     |        edge IDs are returned instead of vertex IDs.\n",
      "     |      @return: see the documentation of the C{output} parameter.\n",
      "     |  \n",
      "     |  get_subisomorphisms_lad(...)\n",
      "     |      get_subisomorphisms_lad(other, domains=None, induced=False, time_limit=0)\n",
      "     |      \n",
      "     |      Returns all subisomorphisms between the graph and another one using the LAD\n",
      "     |      algorithm.\n",
      "     |      \n",
      "     |      The optional C{domains} argument may be used to restrict vertices that\n",
      "     |      may match each other. You can also specify whether you are interested\n",
      "     |      in induced subgraphs only or not.\n",
      "     |      \n",
      "     |      @param other: the pattern graph we are looking for in the graph.\n",
      "     |      @param domains: a list of lists, one sublist belonging to each vertex in\n",
      "     |        the template graph. Sublist M{i} contains the indices of the vertices in\n",
      "     |        the original graph that may match vertex M{i} in the template graph.\n",
      "     |        C{None} means that every vertex may match every other vertex.\n",
      "     |      @param induced: whether to consider induced subgraphs only.\n",
      "     |      @param time_limit: an optimal time limit in seconds. Only the integral\n",
      "     |        part of this number is taken into account. If the time limit is\n",
      "     |        exceeded, the method will throw an exception.\n",
      "     |      @return: a list of lists, each item of the list containing the mapping\n",
      "     |        from vertices of the second graph to the vertices of the first one\n",
      "     |  \n",
      "     |  get_subisomorphisms_vf2(...)\n",
      "     |      get_subisomorphisms_vf2(other, color1=None, color2=None,\n",
      "     |        edge_color1=None, edge_color2=None, node_compat_fn=None,\n",
      "     |        edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Returns all subisomorphisms between the graph and another one\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: a list of lists, each item of the list containing the mapping\n",
      "     |        from vertices of the second graph to the vertices of the first one\n",
      "     |  \n",
      "     |  girth(...)\n",
      "     |      girth(return_shortest_circle=False)\n",
      "     |      \n",
      "     |      Returns the girth of the graph.\n",
      "     |      \n",
      "     |      The girth of a graph is the length of the shortest circle in it.\n",
      "     |      \n",
      "     |      @param return_shortest_circle: whether to return one of the shortest\n",
      "     |        circles found in the graph.\n",
      "     |      @return: the length of the shortest circle or (if C{return_shortest_circle})\n",
      "     |        is true, the shortest circle itself as a list\n",
      "     |  \n",
      "     |  gomory_hu_tree(...)\n",
      "     |      gomory_hu_tree(capacity=None)\n",
      "     |      \n",
      "     |      Internal function, undocumented.\n",
      "     |      \n",
      "     |      @see: Graph.gomory_hu_tree()\n",
      "     |  \n",
      "     |  has_multiple(...)\n",
      "     |      has_multiple()\n",
      "     |      \n",
      "     |      Checks whether the graph has multiple edges.\n",
      "     |      \n",
      "     |      @return: C{True} if the graph has at least one multiple edge,\n",
      "     |               C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  hub_score(...)\n",
      "     |      hub_score(weights=None, scale=True, arpack_options=None, return_eigenvalue=False)\n",
      "     |      \n",
      "     |      Calculates Kleinberg's hub score for the vertices of the graph\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param scale: whether to normalize the scores so that the largest one\n",
      "     |        is 1.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @param return_eigenvalue: whether to return the largest eigenvalue\n",
      "     |      @return: the hub scores in a list and optionally the largest eigenvalue\n",
      "     |        as a second member of a tuple\n",
      "     |      \n",
      "     |      @see: authority_score()\n",
      "     |  \n",
      "     |  incident(...)\n",
      "     |      incident(vertex, mode=OUT)\n",
      "     |      \n",
      "     |      Returns the edges a given vertex is incident on.\n",
      "     |      \n",
      "     |      @param vertex: a vertex ID\n",
      "     |      @param mode: whether to return only successors (L{OUT}),\n",
      "     |        predecessors (L{IN}) or both (L{ALL}). Ignored for undirected\n",
      "     |        graphs.\n",
      "     |  \n",
      "     |  independence_number(...)\n",
      "     |      independence_number()\n",
      "     |      \n",
      "     |      Returns the independence number of the graph.\n",
      "     |      \n",
      "     |      The independence number of the graph is the size of the largest\n",
      "     |      independent vertex set.\n",
      "     |      \n",
      "     |      @see: L{largest_independent_vertex_sets()} for the largest independent\n",
      "     |        vertex sets\n",
      "     |  \n",
      "     |  independent_vertex_sets(...)\n",
      "     |      independent_vertex_sets(min=0, max=0)\n",
      "     |      \n",
      "     |      Returns some or all independent vertex sets of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      Two vertices are independent if there is no edge between them. Members\n",
      "     |      of an independent vertex set are mutually independent.\n",
      "     |      \n",
      "     |      @param min: the minimum size of sets to be returned. If zero or\n",
      "     |        negative, no lower bound will be used.\n",
      "     |      @param max: the maximum size of sets to be returned. If zero or\n",
      "     |        negative, no upper bound will be used.\n",
      "     |  \n",
      "     |  induced_subgraph(...)\n",
      "     |      induced_subgraph(vertices, implementation=\"auto\")\n",
      "     |      \n",
      "     |      Returns a subgraph spanned by the given vertices.\n",
      "     |      \n",
      "     |      @param vertices: a list containing the vertex IDs which\n",
      "     |        should be included in the result.\n",
      "     |      @param implementation: the implementation to use when constructing\n",
      "     |        the new subgraph. igraph includes two implementations at the\n",
      "     |        moment. C{\"copy_and_delete\"} copies the original graph and\n",
      "     |        removes those vertices that are not in the given set. This is more\n",
      "     |        efficient if the size of the subgraph is comparable to the original\n",
      "     |        graph. The other implementation (C{\"create_from_scratch\"})\n",
      "     |        constructs the result graph from scratch and then copies the\n",
      "     |        attributes accordingly. This is a better solution if the subgraph\n",
      "     |        is relatively small, compared to the original graph. C{\"auto\"}\n",
      "     |        selects between the two implementations automatically, based on\n",
      "     |        the ratio of the size of the subgraph and the size of the original\n",
      "     |        graph.\n",
      "     |      @return: the subgraph\n",
      "     |  \n",
      "     |  intersection(...)\n",
      "     |      intersection(graphs)\n",
      "     |      \n",
      "     |      Creates the intersection of two (or more) graphs.\n",
      "     |      \n",
      "     |      @param graphs: the list of graphs to be intersected with\n",
      "     |        the current one.\n",
      "     |  \n",
      "     |  is_bipartite(...)\n",
      "     |      is_bipartite(return_types=False)\n",
      "     |      \n",
      "     |      Decides whether the graph is bipartite or not.\n",
      "     |      \n",
      "     |      Vertices of a bipartite graph can be partitioned into two groups A\n",
      "     |      and B in a way that all edges go between the two groups.\n",
      "     |      \n",
      "     |      @param return_types: if C{False}, the method will simply\n",
      "     |        return C{True} or C{False} depending on whether the graph is\n",
      "     |        bipartite or not. If C{True}, the actual group assignments\n",
      "     |        are also returned as a list of boolean values. (Note that\n",
      "     |        the group assignment is not unique, especially if the graph\n",
      "     |        consists of multiple components, since the assignments of\n",
      "     |        components are independent from each other).\n",
      "     |      @return: C{True} if the graph is bipartite, C{False} if not.\n",
      "     |        If C{return_types} is C{True}, the group assignment is also\n",
      "     |        returned.\n",
      "     |  \n",
      "     |  is_connected(...)\n",
      "     |      is_connected(mode=STRONG)\n",
      "     |      \n",
      "     |      Decides whether the graph is connected.\n",
      "     |      \n",
      "     |      @param mode: whether we should calculate strong or weak connectivity.\n",
      "     |      @return: C{True} if the graph is connected, C{False} otherwise.\n",
      "     |  \n",
      "     |  is_dag(...)\n",
      "     |      is_dag()\n",
      "     |      \n",
      "     |      Checks whether the graph is a DAG (directed acyclic graph).\n",
      "     |      \n",
      "     |      A DAG is a directed graph with no directed cycles.\n",
      "     |      \n",
      "     |      @return: C{True} if it is a DAG, C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  is_directed(...)\n",
      "     |      is_directed()\n",
      "     |      \n",
      "     |      Checks whether the graph is directed.\n",
      "     |      @return: C{True} if it is directed, C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  is_loop(...)\n",
      "     |      is_loop(edges=None)\n",
      "     |      \n",
      "     |      Checks whether a specific set of edges contain loop edges\n",
      "     |      \n",
      "     |      @param edges: edge indices which we want to check. If C{None}, all\n",
      "     |        edges are checked.\n",
      "     |      @return: a list of booleans, one for every edge given\n",
      "     |  \n",
      "     |  is_minimal_separator(...)\n",
      "     |      is_minimal_separator(vertices)\n",
      "     |      \n",
      "     |      Decides whether the given vertex set is a minimal separator.\n",
      "     |      \n",
      "     |      A minimal separator is a set of vertices whose removal disconnects the graph,\n",
      "     |      while the removal of any subset of the set keeps the graph connected.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @return: C{True} is the given vertex set is a minimal separator, C{False}\n",
      "     |        otherwise.\n",
      "     |  \n",
      "     |  is_multiple(...)\n",
      "     |      is_multiple(edges=None)\n",
      "     |      \n",
      "     |      Checks whether an edge is a multiple edge.\n",
      "     |      \n",
      "     |      Also works for a set of edges -- in this case, every edge is checked\n",
      "     |      one by one. Note that if there are multiple edges going between a\n",
      "     |      pair of vertices, there is always one of them that is I{not}\n",
      "     |      reported as multiple (only the others). This allows one to easily\n",
      "     |      detect the edges that have to be deleted in order to make the graph\n",
      "     |      free of multiple edges.\n",
      "     |      \n",
      "     |      @param edges: edge indices which we want to check. If C{None}, all\n",
      "     |        edges are checked.\n",
      "     |      @return: a list of booleans, one for every edge given\n",
      "     |  \n",
      "     |  is_mutual(...)\n",
      "     |      is_mutual(edges=None)\n",
      "     |      \n",
      "     |      Checks whether an edge has an opposite pair.\n",
      "     |      \n",
      "     |      Also works for a set of edges -- in this case, every edge is checked\n",
      "     |      one by one. The result will be a list of booleans (or a single boolean\n",
      "     |      if only an edge index is supplied), every boolean corresponding to an\n",
      "     |      edge in the edge set supplied. C{True} is returned for a given edge\n",
      "     |      M{a} --> M{b} if there exists another edge M{b} --> M{a} in the\n",
      "     |      original graph (not the given edge set!). All edges in an undirected\n",
      "     |      graph are mutual. In case there are multiple edges between M{a}\n",
      "     |      and M{b}, it is enough to have at least one edge in either direction\n",
      "     |      to report all edges between them as mutual, so the multiplicity\n",
      "     |      of edges do not matter.\n",
      "     |      \n",
      "     |      @param edges: edge indices which we want to check. If C{None}, all\n",
      "     |        edges are checked.\n",
      "     |      @return: a list of booleans, one for every edge given\n",
      "     |  \n",
      "     |  is_separator(...)\n",
      "     |      is_separator(vertices)\n",
      "     |      \n",
      "     |      Decides whether the removal of the given vertices disconnects the graph.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @return: C{True} is the given vertex set is a separator, C{False} if not.\n",
      "     |  \n",
      "     |  is_simple(...)\n",
      "     |      is_simple()\n",
      "     |      \n",
      "     |      Checks whether the graph is simple (no loop or multiple edges).\n",
      "     |      \n",
      "     |      @return: C{True} if it is simple, C{False} otherwise.\n",
      "     |      @rtype: boolean\n",
      "     |  \n",
      "     |  isoclass(...)\n",
      "     |      isoclass(vertices)\n",
      "     |      \n",
      "     |      Returns the isomorphy class of the graph or its subgraph.\n",
      "     |      \n",
      "     |      Isomorphy class calculations are implemented only for graphs with\n",
      "     |      3 or 4 vertices.\n",
      "     |      \n",
      "     |      @param vertices: a list of vertices if we want to calculate the\n",
      "     |        isomorphy class for only a subset of vertices. C{None} means to\n",
      "     |        use the full graph.\n",
      "     |      @return: the isomorphy class of the (sub)graph\n",
      "     |  \n",
      "     |  isomorphic(...)\n",
      "     |      isomorphic(other)\n",
      "     |      \n",
      "     |      Checks whether the graph is isomorphic to another graph.\n",
      "     |      \n",
      "     |      The algorithm being used is selected using a simple heuristic:\n",
      "     |      \n",
      "     |        - If one graph is directed and the other undirected, an exception\n",
      "     |          is thrown.\n",
      "     |      \n",
      "     |        - If the two graphs does not have the same number of vertices and\n",
      "     |          edges, it returns with C{False}\n",
      "     |      \n",
      "     |        - If the graphs have three or four vertices, then an O(1) algorithm\n",
      "     |          is used with precomputed data.\n",
      "     |      \n",
      "     |        - Otherwise if the graphs are directed, then the VF2 isomorphism\n",
      "     |          algorithm is used (see L{Graph.isomorphic_vf2}).\n",
      "     |      \n",
      "     |        - Otherwise the BLISS isomorphism algorithm is used, see\n",
      "     |          L{Graph.isomorphic_bliss}.\n",
      "     |      \n",
      "     |      @return: C{True} if the graphs are isomorphic, C{False} otherwise.\n",
      "     |  \n",
      "     |  isomorphic_bliss(...)\n",
      "     |      isomorphic_bliss(other, return_mapping_12=False, return_mapping_21=False,\n",
      "     |        sh1=\"fm\", sh2=\"fm\")\n",
      "     |      \n",
      "     |      Checks whether the graph is isomorphic to another graph, using the\n",
      "     |      BLISS isomorphism algorithm.\n",
      "     |      \n",
      "     |      See U{http://www.tcs.hut.fi/Software/bliss/index.html} for more information\n",
      "     |      about the BLISS algorithm.\n",
      "     |      \n",
      "     |      @param other: the other graph with which we want to compare the graph.\n",
      "     |      @param return_mapping_12: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the first graph to the second.\n",
      "     |      @param return_mapping_21: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the second graph to the first.\n",
      "     |      @param sh1: splitting heuristics for the first graph as a\n",
      "     |        case-insensitive string, with the following possible values:\n",
      "     |      \n",
      "     |          - C{\"f\"}: first non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fl\"}: first largest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fs\"}: first smallest non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fm\"}: first maximally non-trivially connected non-singleton\n",
      "     |            cell\n",
      "     |      \n",
      "     |          - C{\"flm\"}: largest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |          - C{\"fsm\"}: smallest maximally non-trivially connected\n",
      "     |            non-singleton cell\n",
      "     |      \n",
      "     |      @param sh2: splitting heuristics to be used for the second graph.\n",
      "     |        Accepted values are as above.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graphs\n",
      "     |        are isomorphic, C{False} otherwise. If any or both mappings are\n",
      "     |        calculated, the result is a 3-tuple, the first element being the\n",
      "     |        above mentioned boolean, the second element being the 1 -> 2 mapping\n",
      "     |        and the third element being the 2 -> 1 mapping. If the corresponding\n",
      "     |        mapping was not calculated, C{None} is returned in the appropriate\n",
      "     |        element of the 3-tuple.\n",
      "     |  \n",
      "     |  isomorphic_vf2(...)\n",
      "     |      isomorphic_vf2(other=None, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, return_mapping_12=False, return_mapping_21=False,\n",
      "     |        node_compat_fn=None, edge_compat_fn=None, callback=None)\n",
      "     |      \n",
      "     |      Checks whether the graph is isomorphic to another graph, using the\n",
      "     |      VF2 isomorphism algorithm.\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph with which we want to compare the graph.\n",
      "     |        If C{None}, the automorphjisms of the graph will be tested.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param return_mapping_12: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the first graph to the second.\n",
      "     |      @param return_mapping_21: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the second graph to the first.\n",
      "     |      @param callback: if not C{None}, the isomorphism search will not stop at\n",
      "     |        the first match; it will call this callback function instead for every\n",
      "     |        isomorphism found. The callback function must accept four arguments:\n",
      "     |        the first graph, the second graph, a mapping from the nodes of the\n",
      "     |        first graph to the second, and a mapping from the nodes of the second\n",
      "     |        graph to the first. The function must return C{True} if the search\n",
      "     |        should continue or C{False} otherwise.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graphs\n",
      "     |        are isomorphic, C{False} otherwise. If any or both mappings are\n",
      "     |        calculated, the result is a 3-tuple, the first element being the\n",
      "     |        above mentioned boolean, the second element being the 1 -> 2 mapping\n",
      "     |        and the third element being the 2 -> 1 mapping. If the corresponding\n",
      "     |        mapping was not calculated, C{None} is returned in the appropriate\n",
      "     |        element of the 3-tuple.\n",
      "     |  \n",
      "     |  knn(...)\n",
      "     |      knn(vids=None, weights=None)\n",
      "     |      \n",
      "     |      Calculates the average degree of the neighbors for each vertex, and\n",
      "     |      the same quantity as the function of vertex degree.\n",
      "     |      \n",
      "     |      @param vids: the vertices for which the calculation is performed.\n",
      "     |        C{None} means all vertices.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name. If this is given, the vertex strength\n",
      "     |        will be used instead of the vertex degree in the calculations, but\n",
      "     |        the \"ordinary\" vertex degree will be used for the second (degree-\n",
      "     |        dependent) list in the result.\n",
      "     |      @return: two lists in a tuple. The first list contains the average\n",
      "     |        degree of neighbors for each vertex, the second contains the average\n",
      "     |        degree of neighbors as a function of vertex degree. The zeroth element\n",
      "     |        of this list corresponds to vertices of degree 1.\n",
      "     |  \n",
      "     |  laplacian(...)\n",
      "     |      laplacian(weights=None, normalized=False)\n",
      "     |      \n",
      "     |      Returns the Laplacian matrix of a graph.\n",
      "     |      \n",
      "     |      The Laplacian matrix is similar to the adjacency matrix, but the edges\n",
      "     |      are denoted with -1 and the diagonal contains the node degrees.\n",
      "     |      \n",
      "     |      Normalized Laplacian matrices have 1 or 0 in their diagonals (0 for vertices\n",
      "     |      with no edges), edges are denoted by 1 / sqrt(d_i * d_j) where d_i is the\n",
      "     |      degree of node i.\n",
      "     |      \n",
      "     |      Multiple edges and self-loops are silently ignored. Although it is\n",
      "     |      possible to calculate the Laplacian matrix of a directed graph, it does\n",
      "     |      not make much sense.\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name. When edge weights are used, the degree\n",
      "     |        of a node is considered to be the weight of its incident edges.\n",
      "     |      @param normalized: whether to return the normalized Laplacian matrix.\n",
      "     |      @return: the Laplacian matrix.\n",
      "     |  \n",
      "     |  largest_cliques(...)\n",
      "     |      largest_cliques()\n",
      "     |      \n",
      "     |      Returns the largest cliques of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      Quite intuitively a clique is considered largest if there is no clique\n",
      "     |      with more vertices in the whole graph. All largest cliques are maximal\n",
      "     |      (i.e. nonextendable) but not all maximal cliques are largest.\n",
      "     |      \n",
      "     |      @see: L{clique_number()} for the size of the largest cliques or\n",
      "     |        L{maximal_cliques()} for the maximal cliques\n",
      "     |  \n",
      "     |  largest_independent_vertex_sets(...)\n",
      "     |      largest_independent_vertex_sets()\n",
      "     |      \n",
      "     |      Returns the largest independent vertex sets of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      Quite intuitively an independent vertex set is considered largest if\n",
      "     |      there is no other set with more vertices in the whole graph. All largest\n",
      "     |      sets are maximal (i.e. nonextendable) but not all maximal sets\n",
      "     |      are largest.\n",
      "     |      \n",
      "     |      @see: L{independence_number()} for the size of the largest independent\n",
      "     |        vertex sets or L{maximal_independent_vertex_sets()} for the maximal\n",
      "     |        (nonextendable) independent vertex sets\n",
      "     |  \n",
      "     |  layout_bipartite(...)\n",
      "     |      layout_bipartite(types=\"type\", hgap=1, vgap=1, maxiter=100)\n",
      "     |      \n",
      "     |      Place the vertices of a bipartite graph in two layers.\n",
      "     |      \n",
      "     |      The layout is created by placing the vertices in two rows, according\n",
      "     |      to their types. The positions of the vertices within the rows are\n",
      "     |      then optimized to minimize the number of edge crossings using the\n",
      "     |      heuristic used by the Sugiyama layout algorithm.\n",
      "     |      \n",
      "     |      @param types: an igraph vector containing the vertex types, or an\n",
      "     |        attribute name. Anything that evalulates to C{False} corresponds to\n",
      "     |        vertices of the first kind, everything else to the second kind.\n",
      "     |      @param hgap: minimum horizontal gap between vertices in the same layer.\n",
      "     |      @param vgap: vertical gap between the two layers.\n",
      "     |      @param maxiter: maximum number of iterations to take in the crossing\n",
      "     |        reduction step. Increase this if you feel that you are getting too many\n",
      "     |        edge crossings.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_circle(...)\n",
      "     |      layout_circle(dim=2)\n",
      "     |      \n",
      "     |      Places the vertices of the graph uniformly on a circle or a sphere.\n",
      "     |      \n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_drl(...)\n",
      "     |      layout_drl(weights=None, fixed=None, seed=None, options=None, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices on a 2D plane or in the 3D space ccording to the DrL\n",
      "     |      layout algorithm.\n",
      "     |      \n",
      "     |      This is an algorithm suitable for quite large graphs, but it can be\n",
      "     |      surprisingly slow for small ones (where the simpler force-based layouts\n",
      "     |      like C{layout_kamada_kawai()} or C{layout_fruchterman_reingold()} are\n",
      "     |      more useful.\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param seed: if C{None}, uses a random starting layout for the\n",
      "     |        algorithm. If a matrix (list of lists), uses the given matrix\n",
      "     |        as the starting position.\n",
      "     |      @param fixed: if a seed is given, you can specify some vertices to be\n",
      "     |        kept fixed at their original position in the seed by passing an\n",
      "     |        appropriate list here. The list must have exactly as many items as\n",
      "     |        the number of vertices in the graph. Items of the list that evaluate\n",
      "     |        to C{True} denote vertices that will not be moved.\n",
      "     |      @param options: if you give a string argument here, you can select from\n",
      "     |        five default preset parameterisations: C{default}, C{coarsen} for a\n",
      "     |        coarser layout, C{coarsest} for an even coarser layout, C{refine} for\n",
      "     |        refining an existing layout and C{final} for finalizing a layout. If\n",
      "     |        you supply an object that is not a string, the DrL layout parameters\n",
      "     |        are retrieved from the respective keys of the object (so it should\n",
      "     |        be a dict or something else that supports the mapping protocol).\n",
      "     |        The following keys can be used:\n",
      "     |        \n",
      "     |          - C{edge_cut}: edge cutting is done in the late stages of the\n",
      "     |            algorithm in order to achieve less dense layouts. Edges are\n",
      "     |            cut if there is a lot of stress on them (a large value in the\n",
      "     |            objective function sum). The edge cutting parameter is a value\n",
      "     |            between 0 and 1 with 0 representing no edge cutting and 1\n",
      "     |            representing maximal edge cutting.\n",
      "     |      \n",
      "     |          - C{init_iterations}: number of iterations in the initialization\n",
      "     |            phase\n",
      "     |      \n",
      "     |          - C{init_temperature}: start temperature during initialization\n",
      "     |      \n",
      "     |          - C{init_attraction}: attraction during initialization\n",
      "     |      \n",
      "     |          - C{init_damping_mult}: damping multiplier during initialization\n",
      "     |      \n",
      "     |          - C{liquid_iterations}, C{liquid_temperature}, C{liquid_attraction},\n",
      "     |            C{liquid_damping_mult}: same parameters for the liquid phase\n",
      "     |      \n",
      "     |          - C{expansion_iterations}, C{expansion_temperature},\n",
      "     |            C{expansion_attraction}, C{expansion_damping_mult}:\n",
      "     |            parameters for the expansion phase\n",
      "     |      \n",
      "     |          - C{cooldown_...}: parameters for the cooldown phase\n",
      "     |      \n",
      "     |          - C{crunch_...}: parameters for the crunch phase\n",
      "     |      \n",
      "     |          - C{simmer_...}: parameters for the simmer phase\n",
      "     |      \n",
      "     |        \n",
      "     |        Instead of a mapping, you can also use an arbitrary Python object\n",
      "     |        here: if the object does not support the mapping protocol, an\n",
      "     |        attribute of the object with the same name is looked up instead. If\n",
      "     |        a parameter cannot be found either as a key or an attribute, the\n",
      "     |        default from the C{default} preset will be used.\n",
      "     |      \n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_fruchterman_reingold(...)\n",
      "     |      layout_fruchterman_reingold(weights=None, maxiter=500, maxdelta=None, area=None,\n",
      "     |        coolexp=1.5, repulserad=None, seed=None, minx=None, maxx=None, miny=None, \n",
      "     |        maxy=None, minz=None, maxz=None, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices on a 2D plane or in the 3D space according to the\n",
      "     |      Fruchterman-Reingold algorithm.\n",
      "     |      \n",
      "     |      This is a force directed layout, see Fruchterman, T. M. J. and Reingold, E. M.:\n",
      "     |      Graph Drawing by Force-directed Placement.\n",
      "     |      Software -- Practice and Experience, 21/11, 1129--1164, 1991\n",
      "     |      \n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param maxiter: the number of iterations to perform. The default\n",
      "     |        is 500.\n",
      "     |      @param maxdelta: the maximum distance to move a vertex in\n",
      "     |        an iteration. The default is the number of vertices.\n",
      "     |      @param area: the area of the square on which the vertices\n",
      "     |        will be placed. The default is the square of the number of\n",
      "     |        vertices.\n",
      "     |      @param coolexp: the cooling exponent of the simulated annealing.\n",
      "     |        The default is 1.5.\n",
      "     |      @param repulserad: determines the radius at which vertex-vertex\n",
      "     |        repulsion cancels out attraction of adjacent vertices.\n",
      "     |        The default is the number of vertices^3.\n",
      "     |      @param minx: if not C{None}, it must be a vector with exactly as many\n",
      "     |        elements as there are vertices in the graph. Each element is a\n",
      "     |        minimum constraint on the X value of the vertex in the layout.\n",
      "     |      @param maxx: similar to I{minx}, but with maximum constraints\n",
      "     |      @param miny: similar to I{minx}, but with the Y coordinates\n",
      "     |      @param maxy: similar to I{maxx}, but with the Y coordinates\n",
      "     |      @param minz: similar to I{minx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param maxz: similar to I{maxx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param seed: if C{None}, uses a random starting layout for the\n",
      "     |        algorithm. If a matrix (list of lists), uses the given matrix\n",
      "     |        as the starting position.\n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_graphopt(...)\n",
      "     |      layout_graphopt(niter=500, node_charge=0.001, node_mass=30, spring_length=0, spring_constant=1, max_sa_movement=5, seed=None)\n",
      "     |      \n",
      "     |      This is a port of the graphopt layout algorithm by Michael Schmuhl.\n",
      "     |      graphopt version 0.4.1 was rewritten in C and the support for layers\n",
      "     |      was removed.\n",
      "     |      \n",
      "     |      graphopt uses physical analogies for defining attracting and repelling\n",
      "     |      forces among the vertices and then the physical system is simulated\n",
      "     |      until it reaches an equilibrium or the maximal number of iterations is\n",
      "     |      reached.\n",
      "     |      \n",
      "     |      See U{http://www.schmuhl.org/graphopt/} for the original graphopt.\n",
      "     |      \n",
      "     |      @param niter: the number of iterations to perform. Should be a couple\n",
      "     |        of hundred in general.\n",
      "     |      \n",
      "     |      @param node_charge: the charge of the vertices, used to calculate electric\n",
      "     |        repulsion.\n",
      "     |      @param node_mass: the mass of the vertices, used for the spring forces\n",
      "     |      @param spring_length: the length of the springs\n",
      "     |      @param spring_constant: the spring constant\n",
      "     |      @param max_sa_movement: the maximum amount of movement allowed in a single\n",
      "     |        step along a single axis.\n",
      "     |      @param seed: a matrix containing a seed layout from which the algorithm\n",
      "     |        will be started. If C{None}, a random layout will be used.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_grid(...)\n",
      "     |      layout_grid(width=0, height=0, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices of a graph in a 2D or 3D grid.\n",
      "     |      \n",
      "     |      @param width: the number of vertices in a single row of the layout.\n",
      "     |        Zero or negative numbers mean that the width should be determined\n",
      "     |        automatically.\n",
      "     |      @param height: the number of vertices in a single column of the layout.\n",
      "     |        Zero or negative numbers mean that the height should be determined\n",
      "     |        automatically. It must not be given if the number of dimensions is 2.\n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_kamada_kawai(...)\n",
      "     |      layout_kamada_kawai(maxiter=1000, sigma=None, initemp=10, coolexp=0.99,\n",
      "     |        kkconst=None, seed=None, minx=None, maxx=None, miny=None, maxy=None, \n",
      "     |        minz=None, maxz=None, dim=2)\n",
      "     |      \n",
      "     |      Places the vertices on a plane according to the Kamada-Kawai algorithm.\n",
      "     |      \n",
      "     |      This is a force directed layout, see Kamada, T. and Kawai, S.:\n",
      "     |      An Algorithm for Drawing General Undirected Graphs.\n",
      "     |      Information Processing Letters, 31/1, 7--15, 1989.\n",
      "     |      \n",
      "     |      @param maxiter: the number of iterations to perform.\n",
      "     |      @param sigma: the standard base deviation of the position\n",
      "     |        change proposals. C{None} means the number of vertices / 4\n",
      "     |      @param initemp: initial temperature of the simulated annealing.\n",
      "     |      @param coolexp: cooling exponent of the simulated annealing.\n",
      "     |      @param kkconst: the Kamada-Kawai vertex attraction constant.\n",
      "     |        C{None} means the square of the number of vertices.\n",
      "     |      @param minx: if not C{None}, it must be a vector with exactly as many\n",
      "     |        elements as there are vertices in the graph. Each element is a\n",
      "     |        minimum constraint on the X value of the vertex in the layout.\n",
      "     |      @param maxx: similar to I{minx}, but with maximum constraints\n",
      "     |      @param miny: similar to I{minx}, but with the Y coordinates\n",
      "     |      @param maxy: similar to I{maxx}, but with the Y coordinates\n",
      "     |      @param minz: similar to I{minx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param maxz: similar to I{maxx}, but with the Z coordinates. Use only\n",
      "     |        for 3D layouts (C{dim}=3).\n",
      "     |      @param seed: if C{None}, uses a random starting layout for the\n",
      "     |        algorithm. If a matrix (list of lists), uses the given matrix\n",
      "     |        as the starting position.\n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_lgl(...)\n",
      "     |      layout_lgl(maxiter=150, maxdelta=-1, area=-1, coolexp=1.5, repulserad=-1, cellsize=-1, root=None)\n",
      "     |      \n",
      "     |      Places the vertices on a 2D plane according to the Large Graph Layout.\n",
      "     |      \n",
      "     |      @param maxiter: the number of iterations to perform.\n",
      "     |      @param maxdelta: the maximum distance to move a vertex in\n",
      "     |        an iteration. If negative, defaults to the number of vertices.\n",
      "     |      @param area: the area of the square on which the vertices\n",
      "     |        will be placed. If negative, defaults to the number of vertices\n",
      "     |        squared.\n",
      "     |      @param coolexp: the cooling exponent of the simulated annealing.\n",
      "     |      @param repulserad: determines the radius at which vertex-vertex\n",
      "     |        repulsion cancels out attraction of adjacent vertices.\n",
      "     |        If negative, defaults to M{area} times the number of vertices.\n",
      "     |      @param cellsize: the size of the grid cells. When calculating the\n",
      "     |        repulsion forces, only vertices in the same or neighboring\n",
      "     |        grid cells are taken into account. Defaults to the fourth\n",
      "     |        root of M{area}.\n",
      "     |      @param root: the root vertex, this is placed first, its neighbors\n",
      "     |        in the first iteration, second neighbors in the second,\n",
      "     |        etc. C{None} means that a random vertex will be chosen.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  layout_mds(...)\n",
      "     |      layout_mds(dist=None, dim=2, arpack_options=None)\n",
      "     |      Places the vertices in an Euclidean space with the given number of\n",
      "     |      dimensions using multidimensional scaling.\n",
      "     |      \n",
      "     |      This layout requires a distance matrix, where the intersection of\n",
      "     |      row M{i} and column M{j} specifies the desired distance between\n",
      "     |      vertex M{i} and vertex M{j}. The algorithm will try to place the\n",
      "     |      vertices in a way that approximates the distance relations\n",
      "     |      prescribed in the distance matrix. igraph uses the classical\n",
      "     |      multidimensional scaling by Torgerson (see reference below).\n",
      "     |      \n",
      "     |      For unconnected graphs, the method will decompose the graph into\n",
      "     |      weakly connected components and then lay out the components\n",
      "     |      individually using the appropriate parts of the distance matrix.\n",
      "     |      \n",
      "     |      @param dist: the distance matrix. It must be symmetric and the\n",
      "     |        symmetry is not checked -- results are unspecified when a\n",
      "     |        non-symmetric distance matrix is used. If this parameter is\n",
      "     |        C{None}, the shortest path lengths will be used as distances.\n",
      "     |        Directed graphs are treated as undirected when calculating\n",
      "     |        the shortest path lengths to ensure symmetry.\n",
      "     |      @param dim: the number of dimensions. For 2D layouts, supply\n",
      "     |        2 here; for 3D layouts, supply 3.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used.\n",
      "     |      @return: the calculated layout.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Cox & Cox: Multidimensional Scaling (1994), Chapman and\n",
      "     |        Hall, London.\n",
      "     |  \n",
      "     |  layout_random(...)\n",
      "     |      layout_random(dim=2)\n",
      "     |      Places the vertices of the graph randomly.\n",
      "     |      \n",
      "     |      @param dim: the desired number of dimensions for the layout. dim=2\n",
      "     |        means a 2D layout, dim=3 means a 3D layout.\n",
      "     |      @return: the coordinate pairs in a list.\n",
      "     |  \n",
      "     |  layout_reingold_tilford(...)\n",
      "     |      layout_reingold_tilford(mode=\"out\", root=None, rootlevel=None)\n",
      "     |      Places the vertices on a 2D plane according to the Reingold-Tilford\n",
      "     |      layout algorithm.\n",
      "     |      \n",
      "     |      This is a tree layout. If the given graph is not a tree, a breadth-first\n",
      "     |      search is executed first to obtain a possible spanning tree.\n",
      "     |      \n",
      "     |      @param mode: specifies which edges to consider when builing the tree.\n",
      "     |        If it is C{OUT} then only the outgoing, if it is C{IN} then only the\n",
      "     |        incoming edges of a parent are considered. If it is C{ALL} then all\n",
      "     |        edges are used (this was the behaviour in igraph 0.5 and before).\n",
      "     |        This parameter also influences how the root vertices are calculated\n",
      "     |        if they are not given. See the I{root} parameter.\n",
      "     |      @param root: the index of the root vertex or root vertices.\n",
      "     |        if this is a non-empty vector then the supplied vertex IDs are\n",
      "     |        used as the roots of the trees (or a single tree if the graph is\n",
      "     |        connected. If this is C{None} or an empty list, the root vertices\n",
      "     |        are automatically calculated based on topological sorting,\n",
      "     |        performed with the opposite of the I{mode} argument.\n",
      "     |      @param rootlevel: this argument is useful when drawing forests which are\n",
      "     |        not trees. It specifies the level of the root vertices for every tree\n",
      "     |        in the forest.\n",
      "     |      @return: the calculated layout.\n",
      "     |      \n",
      "     |      @see: layout_reingold_tilford_circular\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: EM Reingold, JS Tilford: I{Tidier Drawings of Trees.}\n",
      "     |      IEEE Transactions on Software Engineering 7:22, 223-228, 1981.\n",
      "     |  \n",
      "     |  layout_reingold_tilford_circular(...)\n",
      "     |      layout_reingold_tilford_circular(mode=\"out\", root=None, rootlevel=None)\n",
      "     |      Circular Reingold-Tilford layout for trees.\n",
      "     |      \n",
      "     |      This layout is similar to the Reingold-Tilford layout, but the vertices\n",
      "     |      are placed in a circular way, with the root vertex in the center.\n",
      "     |      \n",
      "     |      See L{layout_reingold_tilford} for the explanation of the parameters.\n",
      "     |      \n",
      "     |      @return: the calculated layout.\n",
      "     |      \n",
      "     |      @see: layout_reingold_tilford\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: EM Reingold, JS Tilford: I{Tidier Drawings of Trees.}\n",
      "     |      IEEE Transactions on Software Engineering 7:22, 223-228, 1981.\n",
      "     |  \n",
      "     |  layout_star(...)\n",
      "     |      layout_star(center=0, order=None)\n",
      "     |      \n",
      "     |      Calculates a star-like layout for the graph.\n",
      "     |      \n",
      "     |      @param center: the ID of the vertex to put in the center\n",
      "     |      @param order: a numeric vector giving the order of the vertices\n",
      "     |        (including the center vertex!). If it is C{None}, the vertices\n",
      "     |        will be placed in increasing vertex ID order.\n",
      "     |      @return: the calculated layout.\n",
      "     |  \n",
      "     |  linegraph(...)\n",
      "     |      linegraph()\n",
      "     |      \n",
      "     |      Returns the line graph of the graph.\n",
      "     |      \n",
      "     |      The line graph M{L(G)} of an undirected graph is defined as follows:\n",
      "     |      M{L(G)} has one vertex for each edge in G and two vertices in M{L(G)}\n",
      "     |      are connected iff their corresponding edges in the original graph\n",
      "     |      share an end point.\n",
      "     |      \n",
      "     |      The line graph of a directed graph is slightly different: two vertices\n",
      "     |      are connected by a directed edge iff the target of the first vertex's\n",
      "     |      corresponding edge is the same as the source of the second vertex's\n",
      "     |      corresponding edge.\n",
      "     |  \n",
      "     |  maxdegree(...)\n",
      "     |      maxdegree(vertices=None, mode=ALL, loops=False)\n",
      "     |      \n",
      "     |      Returns the maximum degree of a vertex set in the graph.\n",
      "     |      \n",
      "     |      This method accepts a single vertex ID or a list of vertex IDs as a\n",
      "     |      parameter, and returns the degree of the given vertices (in the\n",
      "     |      form of a single integer or a list, depending on the input\n",
      "     |      parameter).\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs, or\n",
      "     |        C{None} meaning all the vertices in the graph.\n",
      "     |      @param mode: the type of degree to be returned (L{OUT} for\n",
      "     |        out-degrees, L{IN} IN for in-degrees or L{ALL} for the sum of\n",
      "     |        them).\n",
      "     |      @param loops: whether self-loops should be counted.\n",
      "     |  \n",
      "     |  maxflow(...)\n",
      "     |      maxflow(source, target, capacity=None)\n",
      "     |      \n",
      "     |      Returns the maximum flow between the source and target vertices.\n",
      "     |      \n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a L{Flow} object. It is advised\n",
      "     |        to use that.\n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: a tuple containing the following: the value of the maximum flow\n",
      "     |        between the given vertices, the flow value on all the edges, the edge\n",
      "     |        IDs that are part of the corresponding minimum cut, and the vertex IDs\n",
      "     |        on one side of the cut. For directed graphs, the flow value vector gives\n",
      "     |        the flow value on each edge. For undirected graphs, the flow value is\n",
      "     |        positive if the flow goes from the smaller vertex ID to the bigger one\n",
      "     |        and negative if the flow goes from the bigger vertex ID to the smaller.\n",
      "     |  \n",
      "     |  maxflow_value(...)\n",
      "     |      maxflow_value(source, target, capacity=None)\n",
      "     |      \n",
      "     |      Returns the value of the maximum flow between the source and target vertices.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: the value of the maximum flow between the given vertices\n",
      "     |  \n",
      "     |  maximal_cliques(...)\n",
      "     |      maximal_cliques(min=0, max=0, file=None)\n",
      "     |      \n",
      "     |      Returns the maximal cliques of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      A maximal clique is a clique which can't be extended by adding any other\n",
      "     |      vertex to it. A maximal clique is not necessarily one of the largest\n",
      "     |      cliques in the graph.\n",
      "     |      \n",
      "     |      @param min: the minimum size of maximal cliques to be returned. If zero\n",
      "     |        or negative, no lower bound will be used.\n",
      "     |      \n",
      "     |      @param max: the maximum size of maximal cliques to be returned. If zero\n",
      "     |        or negative, no upper bound will be used. If nonzero, the size of every\n",
      "     |        maximal clique found will be compared to this value and a clique will\n",
      "     |        be returned only if its size is smaller than this limit.\n",
      "     |      \n",
      "     |      @param file: a file object or the name of the file to write the results\n",
      "     |        to. When this argument is C{None}, the maximal cliques will be returned\n",
      "     |        as a list of lists.\n",
      "     |      @return: the maximal cliques of the graph as a list of lists, or C{None}\n",
      "     |        if the C{file} argument was given.@see: L{largest_cliques()} for the largest cliques.\n",
      "     |  \n",
      "     |  maximal_independent_vertex_sets(...)\n",
      "     |      maximal_independent_vertex_sets()\n",
      "     |      \n",
      "     |      Returns the maximal independent vertex sets of the graph as a list of tuples.\n",
      "     |      \n",
      "     |      A maximal independent vertex set is an independent vertex set\n",
      "     |      which can't be extended by adding any other vertex to it. A maximal\n",
      "     |      independent vertex set is not necessarily one of the largest\n",
      "     |      independent vertex sets in the graph.\n",
      "     |      \n",
      "     |      @see: L{largest_independent_vertex_sets()} for the largest independent\n",
      "     |        vertex sets\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Tsukiyama, M. Ide, H. Ariyoshi and I. Shirawaka: I{A new\n",
      "     |        algorithm for generating all the maximal independent sets}.\n",
      "     |        SIAM J Computing, 6:505--517, 1977.\n",
      "     |  \n",
      "     |  mincut(...)\n",
      "     |      mincut(source=None, target=None, capacity=None)\n",
      "     |      \n",
      "     |      Calculates the minimum cut between the source and target vertices or\n",
      "     |      within the whole graph.\n",
      "     |      \n",
      "     |      The minimum cut is the minimum set of edges that needs to be removed\n",
      "     |      to separate the source and the target (if they are given) or to disconnect\n",
      "     |      the graph (if the source and target are not given). The minimum is\n",
      "     |      calculated using the weights (capacities) of the edges, so the cut with\n",
      "     |      the minimum total capacity is calculated.\n",
      "     |      For undirected graphs and no source and target, the method uses the Stoer-Wagner\n",
      "     |      algorithm. For a given source and target, the method uses the push-relabel\n",
      "     |      algorithm; see the references below.\n",
      "     |      \n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a L{Cut} object. It is advised\n",
      "     |        to use that.\n",
      "     |      @param source: the source vertex ID. If C{None}, target must also be\n",
      "     |        {None} and the calculation will be done for the entire graph (i.e. all\n",
      "     |        possible vertex pairs).\n",
      "     |      @param target: the target vertex ID. If C{None}, source must also be\n",
      "     |        {None} and the calculation will be done for the entire graph (i.e. all\n",
      "     |        possible vertex pairs).\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: the value of the minimum cut, the IDs of vertices in the\n",
      "     |        first and second partition, and the IDs of edges in the cut,\n",
      "     |        packed in a 4-tuple\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: M. Stoer, F. Wagner: A simple min-cut algorithm. Journal of\n",
      "     |        the ACM 44(4):585-591, 1997.\n",
      "     |      @ref: A. V. Goldberg, R. E. Tarjan: A new approach to the maximum-flow problem.\n",
      "     |        Journal of the ACM 35(4):921-940, 1988.\n",
      "     |  \n",
      "     |  mincut_value(...)\n",
      "     |      mincut_value(source=-1, target=-1, capacity=None)\n",
      "     |      \n",
      "     |      Returns the minimum cut between the source and target vertices or within\n",
      "     |      the whole graph.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID. If negative, the calculation is\n",
      "     |        done for every vertex except the target and the minimum is returned.\n",
      "     |      @param target: the target vertex ID. If negative, the calculation is\n",
      "     |        done for every vertex except the source and the minimum is returned.\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: the value of the minimum cut between the given vertices\n",
      "     |  \n",
      "     |  minimum_size_separators(...)\n",
      "     |      minimum_size_separators()\n",
      "     |      \n",
      "     |      Returns a list containing all separator vertex sets of minimum size.\n",
      "     |      \n",
      "     |      A vertex set is a separator if its removal disconnects the graph. This method\n",
      "     |      lists all the separators for which no smaller separator set exists in the\n",
      "     |      given graph.\n",
      "     |      \n",
      "     |      @return: a list where each item lists the vertex indices of a given\n",
      "     |        separator of minimum size.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Arkady Kanevsky: Finding all minimum-size separating vertex sets\n",
      "     |        in a graph. Networks 23:533--541, 1993.\n",
      "     |  \n",
      "     |  modularity(...)\n",
      "     |      modularity(membership, weights=None)\n",
      "     |      \n",
      "     |      Calculates the modularity of the graph with respect to some vertex types.\n",
      "     |      \n",
      "     |      The modularity of a graph w.r.t. some division measures how good the\n",
      "     |      division is, or how separated are the different vertex types from each\n",
      "     |      other. It is defined as M{Q=1/(2m) * sum(Aij-ki*kj/(2m)delta(ci,cj),i,j)}.\n",
      "     |      M{m} is the number of edges, M{Aij} is the element of the M{A} adjacency\n",
      "     |      matrix in row M{i} and column M{j}, M{ki} is the degree of node M{i},\n",
      "     |      M{kj} is the degree of node M{j}, and M{Ci} and C{cj} are the types of\n",
      "     |      the two vertices (M{i} and M{j}). M{delta(x,y)} is one iff M{x=y}, 0\n",
      "     |      otherwise.\n",
      "     |      \n",
      "     |      If edge weights are given, the definition of modularity is modified as\n",
      "     |      follows: M{Aij} becomes the weight of the corresponding edge, M{ki}\n",
      "     |      is the total weight of edges incident on vertex M{i}, M{kj} is the\n",
      "     |      total weight of edges incident on vertex M{j} and M{m} is the total\n",
      "     |      edge weight in the graph.\n",
      "     |      \n",
      "     |      @attention: method overridden in L{Graph} to allow L{VertexClustering}\n",
      "     |        objects as a parameter. This method is not strictly necessary, since\n",
      "     |        the L{VertexClustering} class provides a variable called C{modularity}.\n",
      "     |      @param membership: the membership vector, e.g. the vertex type index for\n",
      "     |        each vertex.\n",
      "     |      @param weights: optional edge weights or C{None} if all edges are weighed\n",
      "     |        equally.\n",
      "     |      @return: the modularity score. Score larger than 0.3 usually indicates\n",
      "     |        strong community structure.\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: MEJ Newman and M Girvan: Finding and evaluating community structure\n",
      "     |        in networks. Phys Rev E 69 026113, 2004.\n",
      "     |  \n",
      "     |  motifs_randesu(...)\n",
      "     |      motifs_randesu(size=3, cut_prob=None, callback=None)\n",
      "     |      \n",
      "     |      Counts the number of motifs in the graph\n",
      "     |      \n",
      "     |      Motifs are small subgraphs of a given structure in a graph. It is\n",
      "     |      argued that the motif profile (ie. the number of different motifs in\n",
      "     |      the graph) is characteristic for different types of networks and\n",
      "     |      network function is related to the motifs in the graph.\n",
      "     |      \n",
      "     |      This function is able to find the different motifs of size three\n",
      "     |      and four (ie. the number of different subgraphs with three and four\n",
      "     |      vertices) in the network.\n",
      "     |      \n",
      "     |      In a big network the total number of motifs can be very large, so\n",
      "     |      it takes a lot of time to find all of them. In such cases, a sampling\n",
      "     |      method can be used. This function is capable of doing sampling via\n",
      "     |      the I{cut_prob} argument. This argument gives the probability that\n",
      "     |      a branch of the motif search tree will not be explored.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wernicke and F. Rasche: FANMOD: a tool for fast network\n",
      "     |        motif detection, Bioinformatics 22(9), 1152--1153, 2006.\n",
      "     |      \n",
      "     |      @param size: the size of the motifs (3 or 4).\n",
      "     |      @param cut_prob: the cut probabilities for different levels of the search\n",
      "     |        tree. This must be a list of length I{size} or C{None} to find all\n",
      "     |        motifs.\n",
      "     |      @param callback: C{None} or a callable that will be called for every motif\n",
      "     |        found in the graph. The callable must accept three parameters: the graph\n",
      "     |        itself, the list of vertices in the motif and the isomorphy class of the\n",
      "     |        motif (see L{Graph.isoclass()}). The search will stop when the callback\n",
      "     |        returns an object with a non-zero truth value or raises an exception.\n",
      "     |      @return: the list of motifs if I{callback} is C{None}, or C{None} otherwise\n",
      "     |      @see: Graph.motifs_randesu_no()\n",
      "     |  \n",
      "     |  motifs_randesu_estimate(...)\n",
      "     |      motifs_randesu_estimate(size=3, cut_prob=None, sample)\n",
      "     |      \n",
      "     |      Counts the total number of motifs in the graph\n",
      "     |      \n",
      "     |      Motifs are small subgraphs of a given structure in a graph.\n",
      "     |      This function estimates the total number of motifs in a graph without\n",
      "     |      assigning isomorphism classes to them by extrapolating from a random\n",
      "     |      sample of vertices.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wernicke and F. Rasche: FANMOD: a tool for fast network\n",
      "     |        motif detection, Bioinformatics 22(9), 1152--1153, 2006.\n",
      "     |      \n",
      "     |      @param size: the size of the motifs (3 or 4).\n",
      "     |      @param cut_prob: the cut probabilities for different levels of the search\n",
      "     |        tree. This must be a list of length I{size} or C{None} to find all\n",
      "     |        motifs.\n",
      "     |      @param sample: the size of the sample or the vertex IDs of the vertices\n",
      "     |        to be used for sampling.\n",
      "     |      @see: Graph.motifs_randesu()\n",
      "     |  \n",
      "     |  motifs_randesu_no(...)\n",
      "     |      motifs_randesu_no(size=3, cut_prob=None)\n",
      "     |      \n",
      "     |      Counts the total number of motifs in the graph\n",
      "     |      \n",
      "     |      Motifs are small subgraphs of a given structure in a graph.\n",
      "     |      This function counts the total number of motifs in a graph without\n",
      "     |      assigning isomorphism classes to them.\n",
      "     |      \n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wernicke and F. Rasche: FANMOD: a tool for fast network\n",
      "     |        motif detection, Bioinformatics 22(9), 1152--1153, 2006.\n",
      "     |      \n",
      "     |      @param size: the size of the motifs (3 or 4).\n",
      "     |      @param cut_prob: the cut probabilities for different levels of the search\n",
      "     |        tree. This must be a list of length I{size} or C{None} to find all\n",
      "     |        motifs.\n",
      "     |      @see: Graph.motifs_randesu()\n",
      "     |  \n",
      "     |  neighborhood(...)\n",
      "     |      neighborhood(vertices=None, order=1, mode=ALL)\n",
      "     |      \n",
      "     |      For each vertex specified by I{vertices}, returns the\n",
      "     |      vertices reachable from that vertex in at most I{order} steps.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs, or\n",
      "     |        C{None} meaning all the vertices in the graph.\n",
      "     |      @param order: the order of the neighborhood, i.e. the maximum number of\n",
      "     |        steps to take from the seed vertex.\n",
      "     |      @param mode: specifies how to take into account the direction of\n",
      "     |        the edges if a directed graph is analyzed. C{\"out\"} means that\n",
      "     |        only the outgoing edges are followed, so all vertices reachable\n",
      "     |        from the source vertex in at most I{order} steps are counted.\n",
      "     |        C{\"in\"} means that only the incoming edges are followed (in\n",
      "     |        reverse direction of course), so all vertices from which the source\n",
      "     |        vertex is reachable in at most I{order} steps are counted. C{\"all\"}\n",
      "     |        treats directed edges as undirected.\n",
      "     |      @return: a single list specifying the neighborhood if I{vertices}\n",
      "     |        was an integer specifying a single vertex index, or a list of lists\n",
      "     |        if I{vertices} was a list or C{None}.\n",
      "     |  \n",
      "     |  neighborhood_size(...)\n",
      "     |      neighborhood_size(vertices=None, order=1, mode=ALL)\n",
      "     |      \n",
      "     |      For each vertex specified by I{vertices}, returns the number of\n",
      "     |      vertices reachable from that vertex in at most I{order} steps.\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs, or\n",
      "     |        C{None} meaning all the vertices in the graph.\n",
      "     |      @param order: the order of the neighborhood, i.e. the maximum number of\n",
      "     |        steps to take from the seed vertex.\n",
      "     |      @param mode: specifies how to take into account the direction of\n",
      "     |        the edges if a directed graph is analyzed. C{\"out\"} means that\n",
      "     |        only the outgoing edges are followed, so all vertices reachable\n",
      "     |        from the source vertex in at most I{order} steps are counted.\n",
      "     |        C{\"in\"} means that only the incoming edges are followed (in\n",
      "     |        reverse direction of course), so all vertices from which the source\n",
      "     |        vertex is reachable in at most I{order} steps are counted. C{\"all\"}\n",
      "     |        treats directed edges as undirected.\n",
      "     |      @return: a single number specifying the neighborhood size if I{vertices}\n",
      "     |        was an integer specifying a single vertex index, or a list of sizes\n",
      "     |        if I{vertices} was a list or C{None}.\n",
      "     |  \n",
      "     |  neighbors(...)\n",
      "     |      neighbors(vertex, mode=ALL)\n",
      "     |      \n",
      "     |      Returns adjacent vertices to a given vertex.\n",
      "     |      \n",
      "     |      @param vertex: a vertex ID\n",
      "     |      @param mode: whether to return only successors (L{OUT}),\n",
      "     |        predecessors (L{IN}) or both (L{ALL}). Ignored for undirected\n",
      "     |        graphs.\n",
      "     |  \n",
      "     |  path_length_hist(...)\n",
      "     |      path_length_hist(directed=True)\n",
      "     |      \n",
      "     |      Calculates the path length histogram of the graph\n",
      "     |      @attention: this function is wrapped in a more convenient syntax in the\n",
      "     |        derived class L{Graph}. It is advised to use that instead of this version.\n",
      "     |      \n",
      "     |      @param directed: whether to consider directed paths\n",
      "     |      @return: a tuple. The first item of the tuple is a list of path lengths,\n",
      "     |        the M{i}th element of the list contains the number of paths with length\n",
      "     |        M{i+1}. The second item contains the number of unconnected vertex pairs\n",
      "     |        as a float (since it might not fit into an integer)\n",
      "     |  \n",
      "     |  permute_vertices(...)\n",
      "     |      permute_vertices(permutation)\n",
      "     |      \n",
      "     |      Permutes the vertices of the graph according to the given permutation\n",
      "     |      and returns the new graph.\n",
      "     |      \n",
      "     |      Vertex M{k} of the original graph will become vertex M{permutation[k]}\n",
      "     |      in the new graph. No validity checks are performed on the permutation\n",
      "     |      vector.\n",
      "     |      \n",
      "     |      @return: the new graph\n",
      "     |  \n",
      "     |  personalized_pagerank(...)\n",
      "     |      personalized_pagerank(vertices=None, directed=True, damping=0.85,\n",
      "     |              reset=None, reset_vertices=None, weights=None, \n",
      "     |              arpack_options=None, implementation=\"prpack\", niter=1000,\n",
      "     |              eps=0.001)\n",
      "     |      \n",
      "     |      Calculates the personalized PageRank values of a graph.\n",
      "     |      \n",
      "     |      The personalized PageRank calculation is similar to the PageRank\n",
      "     |      calculation, but the random walk is reset to a non-uniform distribution\n",
      "     |      over the vertices in every step with probability M{1-damping} instead of a\n",
      "     |      uniform distribution.\n",
      "     |      \n",
      "     |      @param vertices: the indices of the vertices being queried.\n",
      "     |        C{None} means all of the vertices.\n",
      "     |      @param directed: whether to consider directed paths.\n",
      "     |      @param damping: the damping factor.\n",
      "     |        M{1-damping} is the PageRank value for vertices with no\n",
      "     |        incoming links.\n",
      "     |      @param reset: the distribution over the vertices to be used when resetting\n",
      "     |        the random walk. Can be a sequence, an iterable or a vertex attribute\n",
      "     |        name as long as they return a list of floats whose length is equal to\n",
      "     |        the number of vertices. If C{None}, a uniform distribution is assumed,\n",
      "     |        which makes the method equivalent to the original PageRank algorithm.\n",
      "     |      @param reset_vertices: an alternative way to specify the distribution\n",
      "     |        over the vertices to be used when resetting the random walk. Simply\n",
      "     |        supply a list of vertex IDs here, or a L{VertexSeq} or a L{Vertex}.\n",
      "     |        Resetting will take place using a uniform distribution over the specified\n",
      "     |        vertices.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @param arpack_options: an L{ARPACKOptions} object used to fine-tune\n",
      "     |        the ARPACK eigenvector calculation. If omitted, the module-level\n",
      "     |        variable called C{arpack_options} is used. This argument is\n",
      "     |        ignored if not the ARPACK implementation is used, see the \n",
      "     |        I{implementation} argument.\n",
      "     |      @param implementation: which implementation to use to solve the \n",
      "     |        PageRank eigenproblem. Possible values are:\n",
      "     |      \n",
      "     |          - C{\"prpack\"}: use the PRPACK library. This is a new \n",
      "     |            implementation in igraph 0.7\n",
      "     |      \n",
      "     |          - C{\"arpack\"}: use the ARPACK library. This implementation\n",
      "     |            was used from version 0.5, until version 0.7.\n",
      "     |      \n",
      "     |          - C{\"power\"}: use a simple power method. This is the\n",
      "     |            implementation that was used before igraph version 0.5.\n",
      "     |      \n",
      "     |      @param niter: The number of iterations to use in the power method\n",
      "     |        implementation. It is ignored in the other implementations.\n",
      "     |      @param eps: The power method implementation will consider the\n",
      "     |        calculation as complete if the difference of PageRank values between\n",
      "     |        iterations change less than this value for every node. It is \n",
      "     |        ignored by the other implementations.\n",
      "     |      @return: a list with the personalized PageRank values of the specified\n",
      "     |        vertices.\n",
      "     |  \n",
      "     |  predecessors(...)\n",
      "     |      predecessors(vertex)\n",
      "     |      \n",
      "     |      Returns the predecessors of a given vertex.\n",
      "     |      \n",
      "     |      Equivalent to calling the L{Graph.neighbors} method with type=L{IN}.\n",
      "     |  \n",
      "     |  radius(...)\n",
      "     |      radius(mode=OUT)\n",
      "     |      \n",
      "     |      Calculates the radius of the graph.\n",
      "     |      \n",
      "     |      The radius of a graph is defined as the minimum eccentricity of\n",
      "     |      its vertices (see L{eccentricity()}).\n",
      "     |      @param mode: what kind of paths to consider for the calculation\n",
      "     |        in case of directed graphs. C{OUT} considers paths that follow\n",
      "     |        edge directions, C{IN} considers paths that follow the opposite\n",
      "     |        edge directions, C{ALL} ignores edge directions. The argument is\n",
      "     |        ignored for undirected graphs.\n",
      "     |      @return: the radius\n",
      "     |      @see: L{Graph.eccentricity()}\n",
      "     |  \n",
      "     |  reciprocity(...)\n",
      "     |      reciprocity(ignore_loops=True, mode=\"default\")\n",
      "     |      \n",
      "     |      Reciprocity defines the proportion of mutual connections in a\n",
      "     |      directed graph. It is most commonly defined as the probability\n",
      "     |      that the opposite counterpart of a directed edge is also included\n",
      "     |      in the graph. This measure is calculated if C{mode} is C{\"default\"}.\n",
      "     |      \n",
      "     |      Prior to igraph 0.6, another measure was implemented, defined as\n",
      "     |      the probability of mutual connection between a vertex pair if we\n",
      "     |      know that there is a (possibly non-mutual) connection between them.\n",
      "     |      In other words, (unordered) vertex pairs are classified into three\n",
      "     |      groups: (1) disconnected, (2) non-reciprocally connected and (3)\n",
      "     |      reciprocally connected. The result is the size of group (3), divided\n",
      "     |      by the sum of sizes of groups (2) and (3). This measure is calculated\n",
      "     |      if C{mode} is C{\"ratio\"}.\n",
      "     |      \n",
      "     |      @param ignore_loops: whether loop edges should be ignored.\n",
      "     |      @param mode: the algorithm to use to calculate the reciprocity; see\n",
      "     |        above for more details.\n",
      "     |      @return: the reciprocity of the graph\n",
      "     |  \n",
      "     |  rewire(...)\n",
      "     |      rewire(n=1000, mode=\"simple\")\n",
      "     |      \n",
      "     |      Randomly rewires the graph while preserving the degree distribution.\n",
      "     |      \n",
      "     |      Please note that the rewiring is done \"in-place\", so the original\n",
      "     |      graph will be modified. If you want to preserve the original graph,\n",
      "     |      use the L{copy} method before.\n",
      "     |      \n",
      "     |      @param n: the number of rewiring trials.\n",
      "     |      @param mode: the rewiring algorithm to use. It can either be C{\"simple\"} or\n",
      "     |        C{\"loops\"}; the former does not create or destroy loop edges while the\n",
      "     |        latter does.\n",
      "     |  \n",
      "     |  rewire_edges(...)\n",
      "     |      rewire_edges(prob, loops=False, multiple=False)\n",
      "     |      \n",
      "     |      Rewires the edges of a graph with constant probability.\n",
      "     |      \n",
      "     |      Each endpoint of each edge of the graph will be rewired with a constant\n",
      "     |      probability, given in the first argument.\n",
      "     |      \n",
      "     |      Please note that the rewiring is done \"in-place\", so the original\n",
      "     |      graph will be modified. If you want to preserve the original graph,\n",
      "     |      use the L{copy} method before.\n",
      "     |      \n",
      "     |      @param prob: rewiring probability\n",
      "     |      @param loops: whether the algorithm is allowed to create loop edges\n",
      "     |      @param multiple: whether the algorithm is allowed to create multiple\n",
      "     |        edges.\n",
      "     |  \n",
      "     |  shortest_paths(...)\n",
      "     |      shortest_paths(source=None, target=None, weights=None, mode=OUT)\n",
      "     |      \n",
      "     |      Calculates shortest path lengths for given vertices in a graph.\n",
      "     |      \n",
      "     |      The algorithm used for the calculations is selected automatically:\n",
      "     |      a simple BFS is used for unweighted graphs, Dijkstra's algorithm is\n",
      "     |      used when all the weights are positive. Otherwise, the Bellman-Ford\n",
      "     |      algorithm is used if the number of requested source vertices is larger\n",
      "     |      than 100 and Johnson's algorithm is used otherwise.\n",
      "     |      \n",
      "     |      @param source: a list containing the source vertex IDs which should be\n",
      "     |        included in the result. If C{None}, all vertices will be considered.\n",
      "     |      @param target: a list containing the target vertex IDs which should be\n",
      "     |        included in the result. If C{None}, all vertices will be considered.\n",
      "     |      @param weights: a list containing the edge weights. It can also be\n",
      "     |        an attribute name (edge weights are retrieved from the given\n",
      "     |        attribute) or C{None} (all edges have equal weight).\n",
      "     |      @param mode: the type of shortest paths to be used for the\n",
      "     |        calculation in directed graphs. L{OUT} means only outgoing,\n",
      "     |        L{IN} means only incoming paths. L{ALL} means to consider\n",
      "     |        the directed graph as an undirected one.\n",
      "     |      @return: the shortest path lengths for given vertices in a matrix\n",
      "     |  \n",
      "     |  similarity_dice(...)\n",
      "     |      similarity_dice(vertices=None, pairs=None, mode=IGRAPH_ALL, loops=True)\n",
      "     |      \n",
      "     |      Dice similarity coefficient of vertices.\n",
      "     |      \n",
      "     |      The Dice similarity coefficient of two vertices is twice the number of\n",
      "     |      their common neighbors divided by the sum of their degrees. This\n",
      "     |      coefficient is very similar to the Jaccard coefficient, but usually\n",
      "     |      gives higher similarities than its counterpart.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None} and I{pairs} is also\n",
      "     |        C{None}, all vertices will be considered.\n",
      "     |      @param pairs: the vertex pairs to be analysed. If this is given, I{vertices}\n",
      "     |        must be C{None}, and the similarity values will be calculated only for the\n",
      "     |        given pairs. Vertex pairs must be specified as tuples of vertex IDs.\n",
      "     |      @param mode: which neighbors should be considered for directed graphs.\n",
      "     |        Can be L{ALL}, L{IN} or L{OUT}, ignored for undirected graphs.\n",
      "     |      @param loops: whether vertices should be considered adjacent to\n",
      "     |        themselves. Setting this to C{True} assumes a loop edge for all vertices\n",
      "     |        even if none is present in the graph. Setting this to C{False} may\n",
      "     |        result in strange results: nonadjacent vertices may have larger\n",
      "     |        similarities compared to the case when an edge is added between them --\n",
      "     |        however, this might be exactly the result you want to get.\n",
      "     |      @return: the pairwise similarity coefficients for the vertices specified,\n",
      "     |        in the form of a matrix if C{pairs} is C{None} or in the form of a list\n",
      "     |        if C{pairs} is not C{None}.\n",
      "     |  \n",
      "     |  similarity_inverse_log_weighted(...)\n",
      "     |      similarity_inverse_log_weighted(vertices=None, mode=IGRAPH_ALL)\n",
      "     |      \n",
      "     |      Inverse log-weighted similarity coefficient of vertices.\n",
      "     |      \n",
      "     |      Each vertex is assigned a weight which is 1 / log(degree). The\n",
      "     |      log-weighted similarity of two vertices is the sum of the weights\n",
      "     |      of their common neighbors.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None}, all vertices\n",
      "     |        will be considered.\n",
      "     |      @param mode: which neighbors should be considered for directed graphs.\n",
      "     |        Can be L{ALL}, L{IN} or L{OUT}, ignored for undirected graphs.\n",
      "     |        L{IN} means that the weights are determined by the out-degrees, L{OUT}\n",
      "     |        means that the weights are determined by the in-degrees.\n",
      "     |      @return: the pairwise similarity coefficients for the vertices specified,\n",
      "     |        in the form of a matrix (list of lists).\n",
      "     |  \n",
      "     |  similarity_jaccard(...)\n",
      "     |      similarity_jaccard(vertices=None, pairs=None, mode=IGRAPH_ALL, loops=True)\n",
      "     |      \n",
      "     |      Jaccard similarity coefficient of vertices.\n",
      "     |      \n",
      "     |      The Jaccard similarity coefficient of two vertices is the number of their\n",
      "     |      common neighbors divided by the number of vertices that are adjacent to\n",
      "     |      at least one of them.\n",
      "     |      \n",
      "     |      @param vertices: the vertices to be analysed. If C{None} and I{pairs} is also\n",
      "     |        C{None}, all vertices will be considered.\n",
      "     |      @param pairs: the vertex pairs to be analysed. If this is given, I{vertices}\n",
      "     |        must be C{None}, and the similarity values will be calculated only for the\n",
      "     |        given pairs. Vertex pairs must be specified as tuples of vertex IDs.\n",
      "     |      @param mode: which neighbors should be considered for directed graphs.\n",
      "     |        Can be L{ALL}, L{IN} or L{OUT}, ignored for undirected graphs.\n",
      "     |      @param loops: whether vertices should be considered adjacent to\n",
      "     |        themselves. Setting this to C{True} assumes a loop edge for all vertices\n",
      "     |        even if none is present in the graph. Setting this to C{False} may\n",
      "     |        result in strange results: nonadjacent vertices may have larger\n",
      "     |        similarities compared to the case when an edge is added between them --\n",
      "     |        however, this might be exactly the result you want to get.\n",
      "     |      @return: the pairwise similarity coefficients for the vertices specified,\n",
      "     |        in the form of a matrix if C{pairs} is C{None} or in the form of a list\n",
      "     |        if C{pairs} is not C{None}.\n",
      "     |  \n",
      "     |  simplify(...)\n",
      "     |      simplify(multiple=True, loops=True, combine_edges=None)\n",
      "     |      \n",
      "     |      Simplifies a graph by removing self-loops and/or multiple edges.\n",
      "     |      \n",
      "     |      \n",
      "     |      For example, suppose you have a graph with an edge attribute named\n",
      "     |      C{weight}. C{graph.simplify(combine_edges=max)} will take the\n",
      "     |      maximum of the weights of multiple edges and assign that weight to\n",
      "     |      the collapsed edge. C{graph.simplify(combine_edges=sum)} will\n",
      "     |      take the sum of the weights. You can also write\n",
      "     |      C{graph.simplify(combine_edges=dict(weight=\"sum\"))} or\n",
      "     |      C{graph.simplify(combine_edges=dict(weight=sum))}, since\n",
      "     |      C{sum} is recognised both as a Python built-in function and as\n",
      "     |      a string constant.\n",
      "     |      \n",
      "     |      @param multiple: whether to remove multiple edges.\n",
      "     |      @param loops: whether to remove loops.\n",
      "     |      @param combine_edges: specifies how to combine the attributes of\n",
      "     |        multiple edges between the same pair of vertices into a single\n",
      "     |        attribute. If it is C{None}, only one of the edges will be kept\n",
      "     |        and all the attributes will be lost. If it is a function, the\n",
      "     |        attributes of multiple edges will be collected and passed on to\n",
      "     |        that function which will return the new attribute value that has to\n",
      "     |        be assigned to the single collapsed edge. It can also be one of\n",
      "     |        the following string constants:\n",
      "     |      \n",
      "     |          - C{\"ignore\"}: all the edge attributes will be ignored.\n",
      "     |      \n",
      "     |          - C{\"sum\"}: the sum of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"product\"}: the product of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |          - C{\"mean\"}: the mean of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"median\"}: the median of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"min\"}: the minimum of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"max\"}: the maximum of the edge attribute values will be used for\n",
      "     |            the new edge.\n",
      "     |      \n",
      "     |          - C{\"first\"}: the attribute value of the first edge in the collapsed set\n",
      "     |            will be used for the new edge.\n",
      "     |      \n",
      "     |          - C{\"last\"}: the attribute value of the last edge in the collapsed set\n",
      "     |            will be used for the new edge.\n",
      "     |      \n",
      "     |          - C{\"random\"}: a randomly selected value will be used for the new edge\n",
      "     |      \n",
      "     |          - C{\"concat\"}: the attribute values will be concatenated for the new\n",
      "     |            edge.\n",
      "     |      \n",
      "     |        You can also use a dict mapping edge attribute names to functions or\n",
      "     |        the above string constants if you want to make the behaviour of the\n",
      "     |        simplification process depend on the name of the attribute.\n",
      "     |        C{None} is a special key in this dict, its value will be used for all\n",
      "     |        the attributes not specified explicitly in the dictionary.\n",
      "     |  \n",
      "     |  st_mincut(...)\n",
      "     |      st_mincut(source, target, capacity=None)\n",
      "     |      \n",
      "     |      Calculates the minimum cut between the source and target vertices in a\n",
      "     |      graph.\n",
      "     |      \n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @param capacity: the capacity of the edges. It must be a list or a valid\n",
      "     |        attribute name or C{None}. In the latter case, every edge will have the\n",
      "     |        same capacity.\n",
      "     |      @return: the value of the minimum cut, the IDs of vertices in the\n",
      "     |        first and second partition, and the IDs of edges in the cut,\n",
      "     |        packed in a 4-tuple\n",
      "     |      \n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a list of L{Cut} objects. It is\n",
      "     |        advised to use that.\n",
      "     |  \n",
      "     |  strength(...)\n",
      "     |      strength(vertices, mode=ALL, loops=True, weights=None)\n",
      "     |      \n",
      "     |      Returns the strength (weighted degree) of some vertices from the graph\n",
      "     |      \n",
      "     |      This method accepts a single vertex ID or a list of vertex IDs as a\n",
      "     |      parameter, and returns the strength (that is, the sum of the weights\n",
      "     |      of all incident edges) of the given vertices (in the\n",
      "     |      form of a single integer or a list, depending on the input\n",
      "     |      parameter).\n",
      "     |      \n",
      "     |      @param vertices: a single vertex ID or a list of vertex IDs\n",
      "     |      @param mode: the type of degree to be returned (L{OUT} for\n",
      "     |        out-degrees, L{IN} IN for in-degrees or L{ALL} for the sum of\n",
      "     |        them).\n",
      "     |      @param loops: whether self-loops should be counted.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |  \n",
      "     |  subcomponent(...)\n",
      "     |      subcomponent(v, mode=ALL)\n",
      "     |      \n",
      "     |      Determines the indices of vertices which are in the same component as a given vertex.\n",
      "     |      \n",
      "     |      @param v: the index of the vertex used as the source/destination\n",
      "     |      @param mode: if equals to L{IN}, returns the vertex IDs from\n",
      "     |        where the given vertex can be reached. If equals to L{OUT},\n",
      "     |        returns the vertex IDs which are reachable from the given\n",
      "     |        vertex. If equals to L{ALL}, returns all vertices within the\n",
      "     |        same component as the given vertex, ignoring edge directions.\n",
      "     |        Note that this is not equal to calculating the union of the \n",
      "     |        results of L{IN} and L{OUT}.\n",
      "     |      @return: the indices of vertices which are in the same component as a given vertex.\n",
      "     |  \n",
      "     |  subgraph_edges(...)\n",
      "     |      subgraph_edges(edges, delete_vertices=True)\n",
      "     |      \n",
      "     |      Returns a subgraph spanned by the given edges.\n",
      "     |      \n",
      "     |      @param edges: a list containing the edge IDs which should\n",
      "     |        be included in the result.\n",
      "     |      @param delete_vertices: if C{True}, vertices not incident on\n",
      "     |        any of the specified edges will be deleted from the result.\n",
      "     |        If C{False}, all vertices will be kept.\n",
      "     |      @return: the subgraph\n",
      "     |  \n",
      "     |  subisomorphic_lad(...)\n",
      "     |      subisomorphic_lad(other, domains=None, induced=False, time_limit=0, \n",
      "     |        return_mapping=False)\n",
      "     |      \n",
      "     |      Checks whether a subgraph of the graph is isomorphic to another graph.\n",
      "     |      \n",
      "     |      The optional C{domains} argument may be used to restrict vertices that\n",
      "     |      may match each other. You can also specify whether you are interested\n",
      "     |      in induced subgraphs only or not.\n",
      "     |      \n",
      "     |      @param other: the pattern graph we are looking for in the graph.\n",
      "     |      @param domains: a list of lists, one sublist belonging to each vertex in\n",
      "     |        the template graph. Sublist M{i} contains the indices of the vertices in\n",
      "     |        the original graph that may match vertex M{i} in the template graph.\n",
      "     |        C{None} means that every vertex may match every other vertex.\n",
      "     |      @param induced: whether to consider induced subgraphs only.\n",
      "     |      @param time_limit: an optimal time limit in seconds. Only the integral\n",
      "     |        part of this number is taken into account. If the time limit is\n",
      "     |        exceeded, the method will throw an exception.\n",
      "     |      @param return_mapping: when C{True}, the function will return a tuple,\n",
      "     |        where the first element is a boolean denoting whether a subisomorphism\n",
      "     |        has been found or not, and the second element describes the mapping\n",
      "     |        of the vertices from the template graph to the original graph. When\n",
      "     |        C{False}, only the boolean is returned.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graph\n",
      "     |        contains a subgraph that is isomorphic to the given template, C{False}\n",
      "     |        otherwise. If the mapping is calculated, the result is a tuple, the first\n",
      "     |        element being the above mentioned boolean, and the second element being\n",
      "     |        the mapping from the target to the original graph.\n",
      "     |  \n",
      "     |  subisomorphic_vf2(...)\n",
      "     |      subisomorphic_vf2(other, color1=None, color2=None, edge_color1=None,\n",
      "     |        edge_color2=None, return_mapping_12=False, return_mapping_21=False,\n",
      "     |        callback=None, node_compat_fn=None, edge_compat_fn=None)\n",
      "     |      \n",
      "     |      Checks whether a subgraph of the graph is isomorphic to another graph.\n",
      "     |      \n",
      "     |      Vertex and edge colors may be used to restrict the isomorphisms, as only\n",
      "     |      vertices and edges with the same color will be allowed to match each other.\n",
      "     |      \n",
      "     |      @param other: the other graph with which we want to compare the graph.\n",
      "     |      @param color1: optional vector storing the coloring of the vertices of\n",
      "     |        the first graph. If C{None}, all vertices have the same color.\n",
      "     |      @param color2: optional vector storing the coloring of the vertices of\n",
      "     |        the second graph. If C{None}, all vertices have the same color.\n",
      "     |      @param edge_color1: optional vector storing the coloring of the edges of\n",
      "     |        the first graph. If C{None}, all edges have the same color.\n",
      "     |      @param edge_color2: optional vector storing the coloring of the edges of\n",
      "     |        the second graph. If C{None}, all edges have the same color.\n",
      "     |      @param return_mapping_12: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the first graph to the second. The mapping can contain\n",
      "     |        -1 if a given node is not mapped.\n",
      "     |      @param return_mapping_21: if C{True}, calculates the mapping which maps\n",
      "     |        the vertices of the second graph to the first. The mapping can contain\n",
      "     |        -1 if a given node is not mapped.\n",
      "     |      @param callback: if not C{None}, the subisomorphism search will not stop at\n",
      "     |        the first match; it will call this callback function instead for every\n",
      "     |        subisomorphism found. The callback function must accept four arguments:\n",
      "     |        the first graph, the second graph, a mapping from the nodes of the\n",
      "     |        first graph to the second, and a mapping from the nodes of the second\n",
      "     |        graph to the first. The function must return C{True} if the search\n",
      "     |        should continue or C{False} otherwise.\n",
      "     |      @param node_compat_fn: a function that receives the two graphs and two\n",
      "     |        node indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the nodes given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on node-specific\n",
      "     |        criteria that are too complicated to be represented by node color\n",
      "     |        vectors (i.e. the C{color1} and C{color2} parameters). C{None} means\n",
      "     |        that every node is compatible with every other node.\n",
      "     |      @param edge_compat_fn: a function that receives the two graphs and two\n",
      "     |        edge indices (one from the first graph, one from the second graph) and\n",
      "     |        returns C{True} if the edges given by the two indices are compatible\n",
      "     |        (i.e. they could be matched to each other) or C{False} otherwise. This\n",
      "     |        can be used to restrict the set of isomorphisms based on edge-specific\n",
      "     |        criteria that are too complicated to be represented by edge color\n",
      "     |        vectors (i.e. the C{edge_color1} and C{edge_color2} parameters). C{None}\n",
      "     |        means that every edge is compatible with every other node.\n",
      "     |      @return: if no mapping is calculated, the result is C{True} if the graph\n",
      "     |        contains a subgraph that's isomorphic to the given one, C{False}\n",
      "     |        otherwise. If any or both mappings are calculated, the result is a\n",
      "     |        3-tuple, the first element being the above mentioned boolean, the\n",
      "     |        second element being the 1 -> 2 mapping and the third element being\n",
      "     |        the 2 -> 1 mapping. If the corresponding mapping was not calculated,\n",
      "     |        C{None} is returned in the appropriate element of the 3-tuple.\n",
      "     |  \n",
      "     |  successors(...)\n",
      "     |      successors(vertex)\n",
      "     |      \n",
      "     |      Returns the successors of a given vertex.\n",
      "     |      \n",
      "     |      Equivalent to calling the L{Graph.neighbors} method with type=L{OUT}.\n",
      "     |  \n",
      "     |  to_directed(...)\n",
      "     |      to_directed(mutual=True)\n",
      "     |      \n",
      "     |      Converts an undirected graph to directed.\n",
      "     |      \n",
      "     |      @param mutual: C{True} if mutual directed edges should be\n",
      "     |        created for every undirected edge. If C{False}, a directed\n",
      "     |        edge with arbitrary direction is created.\n",
      "     |  \n",
      "     |  to_undirected(...)\n",
      "     |      to_undirected(mode=\"collapse\", combine_edges=None)\n",
      "     |      \n",
      "     |      Converts a directed graph to undirected.\n",
      "     |      \n",
      "     |      @param mode: specifies what to do with multiple directed edges\n",
      "     |        going between the same vertex pair. C{True} or C{\"collapse\"}\n",
      "     |        means that only a single edge should be created from multiple\n",
      "     |        directed edges. C{False} or C{\"each\"} means that every edge\n",
      "     |        will be kept (with the arrowheads removed). C{\"mutual\"}\n",
      "     |        creates one undirected edge for each mutual directed edge pair.\n",
      "     |      @param combine_edges: specifies how to combine the attributes of\n",
      "     |        multiple edges between the same pair of vertices into a single\n",
      "     |        attribute. See L{Graph.simplify()} for more details.\n",
      "     |  \n",
      "     |  topological_sorting(...)\n",
      "     |      topological_sorting(mode=OUT)\n",
      "     |      \n",
      "     |      Calculates a possible topological sorting of the graph.\n",
      "     |      \n",
      "     |      Returns a partial sorting and issues a warning if the graph is not\n",
      "     |      a directed acyclic graph.\n",
      "     |      \n",
      "     |      @param mode: if L{OUT}, vertices are returned according to the\n",
      "     |        forward topological order -- all vertices come before their\n",
      "     |        successors. If L{IN}, all vertices come before their ancestors.\n",
      "     |      @return: a possible topological ordering as a list\n",
      "     |  \n",
      "     |  transitivity_avglocal_undirected(...)\n",
      "     |      transitivity_avglocal_undirected(mode=\"nan\")\n",
      "     |      \n",
      "     |      Calculates the average of the vertex transitivities of the graph.\n",
      "     |      \n",
      "     |      The transitivity measures the probability that two neighbors of a\n",
      "     |      vertex are connected. In case of the average local transitivity,\n",
      "     |      this probability is calculated for each vertex and then the average\n",
      "     |      is taken. Vertices with less than two neighbors require special\n",
      "     |      treatment, they will either be left out from the calculation or\n",
      "     |      they will be considered as having zero transitivity, depending on\n",
      "     |      the I{mode} parameter.\n",
      "     |      \n",
      "     |      Note that this measure is different from the global transitivity measure\n",
      "     |      (see L{transitivity_undirected()}) as it simply takes the average local\n",
      "     |      transitivity across the whole network.\n",
      "     |      \n",
      "     |      @param mode: defines how to treat vertices with degree less than two.\n",
      "     |        If C{TRANSITIVITT_ZERO} or C{\"zero\"}, these vertices will have\n",
      "     |        zero transitivity. If C{TRANSITIVITY_NAN} or C{\"nan\"}, these\n",
      "     |        vertices will be excluded from the average.\n",
      "     |      @see: L{transitivity_undirected()}, L{transitivity_local_undirected()}\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: D. J. Watts and S. Strogatz: I{Collective dynamics of small-world\n",
      "     |        networks}. Nature 393(6884):440-442, 1998.\n",
      "     |  \n",
      "     |  transitivity_local_undirected(...)\n",
      "     |      transitivity_local_undirected(vertices=None, mode=\"nan\", weights=None)\n",
      "     |      \n",
      "     |      Calculates the local transitivity (clustering coefficient) of the\n",
      "     |      given vertices in the graph.\n",
      "     |      \n",
      "     |      The transitivity measures the probability that two neighbors of a\n",
      "     |      vertex are connected. In case of the local transitivity, this\n",
      "     |      probability is calculated separately for each vertex.\n",
      "     |      \n",
      "     |      Note that this measure is different from the global transitivity\n",
      "     |      measure (see L{transitivity_undirected()}) as it calculates\n",
      "     |      a transitivity value for each vertex individually.\n",
      "     |      \n",
      "     |      The traditional local transitivity measure applies for unweighted graphs\n",
      "     |      only. When the C{weights} argument is given, this function calculates\n",
      "     |      the weighted local transitivity proposed by Barrat et al (see references).\n",
      "     |      \n",
      "     |      @param vertices: a list containing the vertex IDs which should be\n",
      "     |        included in the result. C{None} means all of the vertices.\n",
      "     |      @param mode: defines how to treat vertices with degree less than two.\n",
      "     |        If C{TRANSITIVITT_ZERO} or C{\"zero\"}, these vertices will have\n",
      "     |        zero transitivity. If C{TRANSITIVITY_NAN} or C{\"nan\"}, these\n",
      "     |        vertices will have C{NaN} (not a number) as their transitivity.\n",
      "     |      @param weights: edge weights to be used. Can be a sequence or iterable or\n",
      "     |        even an edge attribute name.\n",
      "     |      @return: the transitivities for the given vertices in a list\n",
      "     |      @see: L{transitivity_undirected()}, L{transitivity_avglocal_undirected()}\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: Watts DJ and Strogatz S: I{Collective dynamics of small-world\n",
      "     |        networks}. Nature 393(6884):440-442, 1998.\n",
      "     |      @ref: Barrat A, Barthelemy M, Pastor-Satorras R and Vespignani A:\n",
      "     |        I{The architecture of complex weighted networks}. PNAS 101, 3747 (2004).\n",
      "     |        U{http://arxiv.org/abs/cond-mat/0311416}.\n",
      "     |  \n",
      "     |  transitivity_undirected(...)\n",
      "     |      transitivity_undirected(mode=\"nan\")\n",
      "     |      \n",
      "     |      Calculates the global transitivity (clustering coefficient) of the\n",
      "     |      graph.\n",
      "     |      \n",
      "     |      The transitivity measures the probability that two neighbors of a\n",
      "     |      vertex are connected. More precisely, this is the ratio of the\n",
      "     |      triangles and connected triplets in the graph. The result is a\n",
      "     |      single real number. Directed graphs are considered as undirected\n",
      "     |      ones.\n",
      "     |      \n",
      "     |      Note that this measure is different from the local transitivity\n",
      "     |      measure (see L{transitivity_local_undirected()}) as it calculates\n",
      "     |      a single value for the whole graph.\n",
      "     |      \n",
      "     |      @param mode: if C{TRANSITIVITY_ZERO} or C{\"zero\"}, the result will\n",
      "     |        be zero if the graph does not have any triplets. If C{\"nan\"} or\n",
      "     |        C{TRANSITIVITY_NAN}, the result will be C{NaN} (not a number).\n",
      "     |      @return: the transitivity\n",
      "     |      @see: L{transitivity_local_undirected()}, L{transitivity_avglocal_undirected()}\n",
      "     |      @newfield ref: Reference\n",
      "     |      @ref: S. Wasserman and K. Faust: I{Social Network Analysis: Methods and\n",
      "     |        Applications}. Cambridge: Cambridge University Press, 1994.\n",
      "     |  \n",
      "     |  triad_census(...)\n",
      "     |      triad_census()\n",
      "     |      \n",
      "     |      Triad census, as defined by Davis and Leinhardt\n",
      "     |      \n",
      "     |      Calculating the triad census means classifying every triplets of\n",
      "     |      vertices in a directed graph. A triplet can be in one of 16 states,\n",
      "     |      these are listed in the documentation of the C interface of igraph.\n",
      "     |      \n",
      "     |      @attention: this function has a more convenient interface in class\n",
      "     |        L{Graph} which wraps the result in a L{TriadCensus} object.\n",
      "     |        It is advised to use that. The name of the triplet classes are\n",
      "     |        also documented there.\n",
      "     |  \n",
      "     |  unfold_tree(...)\n",
      "     |      unfold_tree(sources=None, mode=OUT)\n",
      "     |      \n",
      "     |      Unfolds the graph using a BFS to a tree by duplicating vertices as necessary.\n",
      "     |      \n",
      "     |      @param sources: the source vertices to start the unfolding from. It should be a\n",
      "     |        list of vertex indices, preferably one vertex from each connected component.\n",
      "     |        You can use L{Graph.topological_sorting()} to determine a suitable set. A single\n",
      "     |        vertex index is also accepted.\n",
      "     |      @param mode: which edges to follow during the BFS. C{OUT} follows outgoing edges,\n",
      "     |        C{IN} follows incoming edges, C{ALL} follows both. Ignored for undirected\n",
      "     |        graphs.\n",
      "     |      @return: the unfolded tree graph and a mapping from the new vertex indices to the\n",
      "     |        old ones.\n",
      "     |  \n",
      "     |  union(...)\n",
      "     |      union(graphs)\n",
      "     |      \n",
      "     |      Creates the union of two (or more) graphs.\n",
      "     |      \n",
      "     |      @param graphs: the list of graphs to be united with\n",
      "     |        the current one.\n",
      "     |  \n",
      "     |  vcount(...)\n",
      "     |      vcount()\n",
      "     |      \n",
      "     |      Counts the number of vertices.\n",
      "     |      @return: the number of vertices in the graph.\n",
      "     |      @rtype: integer\n",
      "     |  \n",
      "     |  vertex_attributes(...)\n",
      "     |      vertex_attributes()\n",
      "     |      \n",
      "     |      @return: the attribute name list of the graph's vertices\n",
      "     |  \n",
      "     |  vertex_connectivity(...)\n",
      "     |      vertex_connectivity(source=-1, target=-1, checks=True, neighbors=\"error\")\n",
      "     |      \n",
      "     |      Calculates the vertex connectivity of the graph or between some vertices.\n",
      "     |      \n",
      "     |      The vertex connectivity between two given vertices is the number of vertices\n",
      "     |      that have to be removed in order to disconnect the two vertices into two\n",
      "     |      separate components. This is also the number of vertex disjoint directed\n",
      "     |      paths between the vertices (apart from the source and target vertices of\n",
      "     |      course). The vertex connectivity of the graph is the minimal vertex\n",
      "     |      connectivity over all vertex pairs.\n",
      "     |      \n",
      "     |      This method calculates the vertex connectivity of a given vertex pair if both\n",
      "     |      the source and target vertices are given. If none of them is given (or they\n",
      "     |      are both negative), the overall vertex connectivity is returned.\n",
      "     |      \n",
      "     |      @param source: the source vertex involved in the calculation.\n",
      "     |      @param target: the target vertex involved in the calculation.\n",
      "     |      @param checks: if the whole graph connectivity is calculated and this is\n",
      "     |        C{True}, igraph performs some basic checks before calculation. If the\n",
      "     |        graph is not strongly connected, then the connectivity is obviously\n",
      "     |        zero. If the minimum degree is one, then the connectivity is\n",
      "     |        also one. These simple checks are much faster than checking the entire\n",
      "     |        graph, therefore it is advised to set this to C{True}. The parameter\n",
      "     |        is ignored if the connectivity between two given vertices is computed.\n",
      "     |      @param neighbors: tells igraph what to do when the two vertices are\n",
      "     |        connected. C{\"error\"} raises an exception, C{\"infinity\"} returns\n",
      "     |        infinity, C{\"ignore\"} ignores the edge.\n",
      "     |      @return: the vertex connectivity\n",
      "     |  \n",
      "     |  write_dimacs(...)\n",
      "     |      write_dimacs(f, source, target, capacity=None)\n",
      "     |      \n",
      "     |      Writes the graph in DIMACS format to the given file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param source: the source vertex ID\n",
      "     |      @param target: the target vertex ID\n",
      "     |      @param capacity: the capacities of the edges in a list. If it is not a\n",
      "     |        list, the corresponding edge attribute will be used to retrieve\n",
      "     |        capacities.\n",
      "     |  \n",
      "     |  write_dot(...)\n",
      "     |      write_dot(f)\n",
      "     |      \n",
      "     |      Writes the graph in DOT format to the given file.\n",
      "     |      \n",
      "     |      DOT is the format used by the U{GraphViz <http://www.graphviz.org>}\n",
      "     |      software package.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  write_edgelist(...)\n",
      "     |      write_edgelist(f)\n",
      "     |      \n",
      "     |      Writes the edge list of a graph to a file.\n",
      "     |      \n",
      "     |      Directed edges are written in (from, to) order.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  write_gml(...)\n",
      "     |      write_gml(f, creator=None, ids=None)\n",
      "     |      \n",
      "     |      Writes the graph in GML format to the given file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param creator: optional creator information to be written to the file.\n",
      "     |        If C{None}, the current date and time is added.\n",
      "     |      @param ids: optional numeric vertex IDs to use in the file. This must\n",
      "     |        be a list of integers or C{None}. If C{None}, the C{id} attribute of\n",
      "     |        the vertices are used, or if they don't exist, numeric vertex IDs\n",
      "     |        will be generated automatically.\n",
      "     |  \n",
      "     |  write_graphml(...)\n",
      "     |      write_graphml(f)\n",
      "     |      \n",
      "     |      Writes the graph to a GraphML file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  write_leda(...)\n",
      "     |      write_leda(f, names=\"name\", weights=\"weights\")\n",
      "     |      \n",
      "     |      Writes the graph to a file in LEDA native format.\n",
      "     |      \n",
      "     |      The LEDA format supports at most one attribute per vertex and edge. You can\n",
      "     |      specify which vertex and edge attribute you want to use. Note that the\n",
      "     |      name of the attribute is not saved in the LEDA file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param names: the name of the vertex attribute to be stored along with\n",
      "     |        the vertices. It is usually used to store the vertex names (hence the\n",
      "     |        name of the keyword argument), but you may also use a numeric attribute.\n",
      "     |        If you don't want to store any vertex attributes, supply C{None} here.\n",
      "     |      @param weights: the name of the edge attribute to be stored along with\n",
      "     |        the edges. It is usually used to store the edge weights (hence the\n",
      "     |        name of the keyword argument), but you may also use a string attribute.\n",
      "     |        If you don't want to store any edge attributes, supply C{None} here.\n",
      "     |  \n",
      "     |  write_lgl(...)\n",
      "     |      write_lgl(f, names=\"name\", weights=\"weights\", isolates=True)\n",
      "     |      \n",
      "     |      Writes the edge list of a graph to a file in .lgl format.\n",
      "     |      \n",
      "     |      Note that multiple edges and/or loops break the LGL software,\n",
      "     |      but igraph does not check for this condition. Unless you know\n",
      "     |      that the graph does not have multiple edges and/or loops, it\n",
      "     |      is wise to call L{simplify()} before saving.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param names: the name of the vertex attribute containing the name\n",
      "     |        of the vertices. If you don't want to store vertex names,\n",
      "     |        supply C{None} here.\n",
      "     |      @param weights: the name of the edge attribute containing the weight\n",
      "     |        of the vertices. If you don't want to store weights,\n",
      "     |        supply C{None} here.\n",
      "     |      @param isolates: whether to include isolated vertices in the output.\n",
      "     |  \n",
      "     |  write_ncol(...)\n",
      "     |      write_ncol(f, names=\"name\", weights=\"weights\")\n",
      "     |      \n",
      "     |      Writes the edge list of a graph to a file in .ncol format.\n",
      "     |      \n",
      "     |      Note that multiple edges and/or loops break the LGL software,\n",
      "     |      but igraph does not check for this condition. Unless you know\n",
      "     |      that the graph does not have multiple edges and/or loops, it\n",
      "     |      is wise to call L{simplify()} before saving.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |      @param names: the name of the vertex attribute containing the name\n",
      "     |        of the vertices. If you don't want to store vertex names,\n",
      "     |        supply C{None} here.\n",
      "     |      @param weights: the name of the edge attribute containing the weight\n",
      "     |        of the vertices. If you don't want to store weights,\n",
      "     |        supply C{None} here.\n",
      "     |  \n",
      "     |  write_pajek(...)\n",
      "     |      write_pajek(f)\n",
      "     |      \n",
      "     |      Writes the graph in Pajek format to the given file.\n",
      "     |      \n",
      "     |      @param f: the name of the file to be written or a Python file handle\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes defined here:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "    \n",
      "    class Vertex(__builtin__.object)\n",
      "     |  Class representing a single vertex in a graph.\n",
      "     |  \n",
      "     |  The vertex is referenced by its index, so if the underlying graph\n",
      "     |  changes, the semantics of the vertex object might change as well\n",
      "     |  (if the vertex indices are altered in the original graph).\n",
      "     |  \n",
      "     |  The attributes of the vertex can be accessed by using the vertex\n",
      "     |  as a hash:\n",
      "     |  \n",
      "     |    >>> v[\"color\"] = \"red\"                  #doctest: +SKIP\n",
      "     |    >>> print v[\"color\"]                      #doctest: +SKIP\n",
      "     |    red\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __delitem__(...)\n",
      "     |      x.__delitem__(y) <==> del x[y]\n",
      "     |  \n",
      "     |  __eq__(...)\n",
      "     |      x.__eq__(y) <==> x==y\n",
      "     |  \n",
      "     |  __ge__(...)\n",
      "     |      x.__ge__(y) <==> x>=y\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __gt__(...)\n",
      "     |      x.__gt__(y) <==> x>y\n",
      "     |  \n",
      "     |  __hash__(...)\n",
      "     |      x.__hash__() <==> hash(x)\n",
      "     |  \n",
      "     |  __le__(...)\n",
      "     |      x.__le__(y) <==> x<=y\n",
      "     |  \n",
      "     |  __len__(...)\n",
      "     |      x.__len__() <==> len(x)\n",
      "     |  \n",
      "     |  __lt__(...)\n",
      "     |      x.__lt__(y) <==> x<y\n",
      "     |  \n",
      "     |  __ne__(...)\n",
      "     |      x.__ne__(y) <==> x!=y\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setitem__(...)\n",
      "     |      x.__setitem__(i, y) <==> x[i]=y\n",
      "     |  \n",
      "     |  attribute_names(...)\n",
      "     |      attribute_names() -> list\n",
      "     |      \n",
      "     |      Returns the list of vertex attribute names\n",
      "     |  \n",
      "     |  attributes(...)\n",
      "     |      attributes() -> dict\n",
      "     |      \n",
      "     |      Returns a dict of attribute names and values for the vertex\n",
      "     |  \n",
      "     |  betweenness(...)\n",
      "     |      Proxy method to L{Graph.betweenness()}\n",
      "     |      \n",
      "     |      This method calls the betweenness method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.betweenness() for details.\n",
      "     |  \n",
      "     |  closeness(...)\n",
      "     |      Proxy method to L{Graph.closeness()}\n",
      "     |      \n",
      "     |      This method calls the closeness method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.closeness() for details.\n",
      "     |  \n",
      "     |  constraint(...)\n",
      "     |      Proxy method to L{Graph.constraint()}\n",
      "     |      \n",
      "     |      This method calls the constraint method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.constraint() for details.\n",
      "     |  \n",
      "     |  degree(...)\n",
      "     |      Proxy method to L{Graph.degree()}\n",
      "     |      \n",
      "     |      This method calls the degree method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.degree() for details.\n",
      "     |  \n",
      "     |  delete(...)\n",
      "     |      Proxy method to L{Graph.delete_vertices()}\n",
      "     |      \n",
      "     |      This method calls the delete_vertices method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.delete_vertices() for details.\n",
      "     |  \n",
      "     |  diversity(...)\n",
      "     |      Proxy method to L{Graph.diversity()}\n",
      "     |      \n",
      "     |      This method calls the diversity method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.diversity() for details.\n",
      "     |  \n",
      "     |  eccentricity(...)\n",
      "     |      Proxy method to L{Graph.eccentricity()}\n",
      "     |      \n",
      "     |      This method calls the eccentricity method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.eccentricity() for details.\n",
      "     |  \n",
      "     |  get_shortest_paths(...)\n",
      "     |      Proxy method to L{Graph.get_shortest_paths()}\n",
      "     |      \n",
      "     |      This method calls the get_shortest_paths method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.get_shortest_paths() for details.\n",
      "     |  \n",
      "     |  indegree(...)\n",
      "     |      Proxy method to L{Graph.indegree()}\n",
      "     |      \n",
      "     |      This method calls the indegree method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.indegree() for details.\n",
      "     |  \n",
      "     |  is_minimal_separator(...)\n",
      "     |      Proxy method to L{Graph.is_minimal_separator()}\n",
      "     |      \n",
      "     |      This method calls the is_minimal_separator method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_minimal_separator() for details.\n",
      "     |  \n",
      "     |  is_separator(...)\n",
      "     |      Proxy method to L{Graph.is_separator()}\n",
      "     |      \n",
      "     |      This method calls the is_separator method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_separator() for details.\n",
      "     |  \n",
      "     |  neighbors(...)\n",
      "     |      Proxy method to L{Graph.neighbors()}\n",
      "     |      \n",
      "     |      This method calls the neighbors method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.neighbors() for details.\n",
      "     |  \n",
      "     |  outdegree(...)\n",
      "     |      Proxy method to L{Graph.outdegree()}\n",
      "     |      \n",
      "     |      This method calls the outdegree method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.outdegree() for details.\n",
      "     |  \n",
      "     |  pagerank(...)\n",
      "     |      Proxy method to L{Graph.pagerank()}\n",
      "     |      \n",
      "     |      This method calls the pagerank method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.pagerank() for details.\n",
      "     |  \n",
      "     |  personalized_pagerank(...)\n",
      "     |      Proxy method to L{Graph.personalized_pagerank()}\n",
      "     |      \n",
      "     |      This method calls the personalized_pagerank method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.personalized_pagerank() for details.\n",
      "     |  \n",
      "     |  predecessors(...)\n",
      "     |      Proxy method to L{Graph.predecessors()}\n",
      "     |      \n",
      "     |      This method calls the predecessors method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.predecessors() for details.\n",
      "     |  \n",
      "     |  shortest_paths(...)\n",
      "     |      Proxy method to L{Graph.shortest_paths()}\n",
      "     |      \n",
      "     |      This method calls the shortest_paths method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.shortest_paths() for details.\n",
      "     |  \n",
      "     |  strength(...)\n",
      "     |      Proxy method to L{Graph.strength()}\n",
      "     |      \n",
      "     |      This method calls the strength method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.strength() for details.\n",
      "     |  \n",
      "     |  successors(...)\n",
      "     |      Proxy method to L{Graph.successors()}\n",
      "     |      \n",
      "     |      This method calls the successors method of the L{Graph} class with this vertex as the first argument, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.successors() for details.\n",
      "     |  \n",
      "     |  update_attributes(...)\n",
      "     |      update_attributes(E, **F) -> None\n",
      "     |      \n",
      "     |      Updates the attributes of the vertex from dict/iterable E and F.\n",
      "     |      \n",
      "     |      If E has a C{keys()} method, it does: C{for k in E: self[k] = E[k]}.\n",
      "     |      If E lacks a C{keys()} method, it does: C{for (k, v) in E: self[k] = v}.\n",
      "     |      In either case, this is followed by: C{for k in F: self[k] = F[k]}.\n",
      "     |      \n",
      "     |      This method thus behaves similarly to the C{update()} method of Python\n",
      "     |      dictionaries.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  graph\n",
      "     |      The graph the vertex belongs to\n",
      "     |  \n",
      "     |  index\n",
      "     |      Index of the vertex\n",
      "    \n",
      "    class VertexSeq(igraph.core.VertexSeq)\n",
      "     |  Class representing a sequence of vertices in the graph.\n",
      "     |  \n",
      "     |  This class is most easily accessed by the C{vs} field of the\n",
      "     |  L{Graph} object, which returns an ordered sequence of all vertices in\n",
      "     |  the graph. The vertex sequence can be refined by invoking the\n",
      "     |  L{VertexSeq.select()} method. L{VertexSeq.select()} can also be\n",
      "     |  accessed by simply calling the L{VertexSeq} object.\n",
      "     |  \n",
      "     |  An alternative way to create a vertex sequence referring to a given\n",
      "     |  graph is to use the constructor directly:\n",
      "     |  \n",
      "     |    >>> g = Graph.Full(3)\n",
      "     |    >>> vs = VertexSeq(g)\n",
      "     |    >>> restricted_vs = VertexSeq(g, [0, 1])\n",
      "     |  \n",
      "     |  The individual vertices can be accessed by indexing the vertex sequence\n",
      "     |  object. It can be used as an iterable as well, or even in a list\n",
      "     |  comprehension:\n",
      "     |  \n",
      "     |    >>> g=Graph.Full(3)\n",
      "     |    >>> for v in g.vs:\n",
      "     |    ...   v[\"value\"] = v.index ** 2\n",
      "     |    ...\n",
      "     |    >>> [v[\"value\"] ** 0.5 for v in g.vs]\n",
      "     |    [0.0, 1.0, 2.0]\n",
      "     |  \n",
      "     |  The vertex set can also be used as a dictionary where the keys are the\n",
      "     |  attribute names. The values corresponding to the keys are the values\n",
      "     |  of the given attribute for every vertex selected by the sequence.\n",
      "     |  \n",
      "     |    >>> g=Graph.Full(3)\n",
      "     |    >>> for idx, v in enumerate(g.vs):\n",
      "     |    ...   v[\"weight\"] = idx*(idx+1)\n",
      "     |    ...\n",
      "     |    >>> g.vs[\"weight\"]\n",
      "     |    [0, 2, 6]\n",
      "     |    >>> g.vs.select(1,2)[\"weight\"] = [10, 20]\n",
      "     |    >>> g.vs[\"weight\"]\n",
      "     |    [0, 10, 20]\n",
      "     |  \n",
      "     |  If you specify a sequence that is shorter than the number of vertices in\n",
      "     |  the VertexSeq, the sequence is reused:\n",
      "     |  \n",
      "     |    >>> g = Graph.Tree(7, 2)\n",
      "     |    >>> g.vs[\"color\"] = [\"red\", \"green\"]\n",
      "     |    >>> g.vs[\"color\"]\n",
      "     |    ['red', 'green', 'red', 'green', 'red', 'green', 'red']\n",
      "     |  \n",
      "     |  You can even pass a single string or integer, it will be considered as a\n",
      "     |  sequence of length 1:\n",
      "     |  \n",
      "     |    >>> g.vs[\"color\"] = \"red\"\n",
      "     |    >>> g.vs[\"color\"]\n",
      "     |    ['red', 'red', 'red', 'red', 'red', 'red', 'red']\n",
      "     |  \n",
      "     |  Some methods of the vertex sequences are simply proxy methods to the\n",
      "     |  corresponding methods in the L{Graph} object. One such example is\n",
      "     |  L{VertexSeq.degree()}:\n",
      "     |  \n",
      "     |    >>> g=Graph.Tree(7, 2)\n",
      "     |    >>> g.vs.degree()\n",
      "     |    [2, 3, 3, 1, 1, 1, 1]\n",
      "     |    >>> g.vs.degree() == g.degree()\n",
      "     |    True\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      VertexSeq\n",
      "     |      igraph.core.VertexSeq\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Shorthand notation to select()\n",
      "     |      \n",
      "     |      This method simply passes all its arguments to L{VertexSeq.select()}.\n",
      "     |  \n",
      "     |  attributes(self)\n",
      "     |      Returns the list of all the vertex attributes in the graph\n",
      "     |      associated to this vertex sequence.\n",
      "     |  \n",
      "     |  betweenness(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.betweenness()}\n",
      "     |      \n",
      "     |      This method calls the C{betweenness()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.betweenness() for details.\n",
      "     |  \n",
      "     |  bibcoupling(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.bibcoupling()}\n",
      "     |      \n",
      "     |      This method calls the C{bibcoupling()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.bibcoupling() for details.\n",
      "     |  \n",
      "     |  closeness(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.closeness()}\n",
      "     |      \n",
      "     |      This method calls the C{closeness()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.closeness() for details.\n",
      "     |  \n",
      "     |  cocitation(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.cocitation()}\n",
      "     |      \n",
      "     |      This method calls the C{cocitation()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.cocitation() for details.\n",
      "     |  \n",
      "     |  constraint(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.constraint()}\n",
      "     |      \n",
      "     |      This method calls the C{constraint()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.constraint() for details.\n",
      "     |  \n",
      "     |  degree(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.degree()}\n",
      "     |      \n",
      "     |      This method calls the C{degree()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.degree() for details.\n",
      "     |  \n",
      "     |  delete = delete_vertices(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.delete_vertices()}\n",
      "     |      \n",
      "     |      This method calls the C{delete_vertices()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.delete_vertices() for details.\n",
      "     |  \n",
      "     |  diversity(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.diversity()}\n",
      "     |      \n",
      "     |      This method calls the C{diversity()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.diversity() for details.\n",
      "     |  \n",
      "     |  eccentricity(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.eccentricity()}\n",
      "     |      \n",
      "     |      This method calls the C{eccentricity()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.eccentricity() for details.\n",
      "     |  \n",
      "     |  find(self, *args, **kwds)\n",
      "     |      Returns the first vertex of the vertex sequence that matches some\n",
      "     |      criteria.\n",
      "     |      \n",
      "     |      The selection criteria are equal to the ones allowed by L{VertexSeq.select}.\n",
      "     |      See L{VertexSeq.select} for more details.\n",
      "     |      \n",
      "     |      For instance, to find the first vertex with name C{foo} in graph C{g}:\n",
      "     |      \n",
      "     |          >>> g.vs.find(name=\"foo\")            #doctest:+SKIP\n",
      "     |      \n",
      "     |      To find an arbitrary isolated vertex:\n",
      "     |      \n",
      "     |          >>> g.vs.find(_degree=0)             #doctest:+SKIP\n",
      "     |  \n",
      "     |  get_shortest_paths(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.get_shortest_paths()}\n",
      "     |      \n",
      "     |      This method calls the C{get_shortest_paths()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.get_shortest_paths() for details.\n",
      "     |  \n",
      "     |  indegree(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.indegree()}\n",
      "     |      \n",
      "     |      This method calls the C{indegree()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.indegree() for details.\n",
      "     |  \n",
      "     |  is_minimal_separator(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.is_minimal_separator()}\n",
      "     |      \n",
      "     |      This method calls the C{is_minimal_separator()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_minimal_separator() for details.\n",
      "     |  \n",
      "     |  is_separator(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.is_separator()}\n",
      "     |      \n",
      "     |      This method calls the C{is_separator()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.is_separator() for details.\n",
      "     |  \n",
      "     |  isoclass(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.isoclass()}\n",
      "     |      \n",
      "     |      This method calls the C{isoclass()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.isoclass() for details.\n",
      "     |  \n",
      "     |  maxdegree(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.maxdegree()}\n",
      "     |      \n",
      "     |      This method calls the C{maxdegree()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.maxdegree() for details.\n",
      "     |  \n",
      "     |  outdegree(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.outdegree()}\n",
      "     |      \n",
      "     |      This method calls the C{outdegree()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.outdegree() for details.\n",
      "     |  \n",
      "     |  pagerank(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.pagerank()}\n",
      "     |      \n",
      "     |      This method calls the C{pagerank()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.pagerank() for details.\n",
      "     |  \n",
      "     |  personalized_pagerank(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.personalized_pagerank()}\n",
      "     |      \n",
      "     |      This method calls the C{personalized_pagerank()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.personalized_pagerank() for details.\n",
      "     |  \n",
      "     |  select(self, *args, **kwds)\n",
      "     |      Selects a subset of the vertex sequence based on some criteria\n",
      "     |      \n",
      "     |      The selection criteria can be specified by the positional and the keyword\n",
      "     |      arguments. Positional arguments are always processed before keyword\n",
      "     |      arguments.\n",
      "     |      \n",
      "     |        - If the first positional argument is C{None}, an empty sequence is\n",
      "     |          returned.\n",
      "     |      \n",
      "     |        - If the first positional argument is a callable object, the object\n",
      "     |          will be called for every vertex in the sequence. If it returns\n",
      "     |          C{True}, the vertex will be included, otherwise it will\n",
      "     |          be excluded.\n",
      "     |      \n",
      "     |        - If the first positional argument is an iterable, it must return\n",
      "     |          integers and they will be considered as indices of the current\n",
      "     |          vertex set (NOT the whole vertex set of the graph -- the\n",
      "     |          difference matters when one filters a vertex set that has\n",
      "     |          already been filtered by a previous invocation of\n",
      "     |          L{VertexSeq.select()}. In this case, the indices do not refer\n",
      "     |          directly to the vertices of the graph but to the elements of\n",
      "     |          the filtered vertex sequence.\n",
      "     |      \n",
      "     |        - If the first positional argument is an integer, all remaining\n",
      "     |          arguments are expected to be integers. They are considered as\n",
      "     |          indices of the current vertex set again.\n",
      "     |      \n",
      "     |      Keyword arguments can be used to filter the vertices based on their\n",
      "     |      attributes. The name of the keyword specifies the name of the attribute\n",
      "     |      and the filtering operator, they should be concatenated by an\n",
      "     |      underscore (C{_}) character. Attribute names can also contain\n",
      "     |      underscores, but operator names don't, so the operator is always the\n",
      "     |      largest trailing substring of the keyword name that does not contain\n",
      "     |      an underscore. Possible operators are:\n",
      "     |      \n",
      "     |        - C{eq}: equal to\n",
      "     |      \n",
      "     |        - C{ne}: not equal to\n",
      "     |      \n",
      "     |        - C{lt}: less than\n",
      "     |      \n",
      "     |        - C{gt}: greater than\n",
      "     |      \n",
      "     |        - C{le}: less than or equal to\n",
      "     |      \n",
      "     |        - C{ge}: greater than or equal to\n",
      "     |      \n",
      "     |        - C{in}: checks if the value of an attribute is in a given list\n",
      "     |      \n",
      "     |        - C{notin}: checks if the value of an attribute is not in a given\n",
      "     |          list\n",
      "     |      \n",
      "     |      For instance, if you want to filter vertices with a numeric C{age}\n",
      "     |      property larger than 200, you have to write:\n",
      "     |      \n",
      "     |        >>> g.vs.select(age_gt=200)                   #doctest: +SKIP\n",
      "     |      \n",
      "     |      Similarly, to filter vertices whose C{type} is in a list of predefined\n",
      "     |      types:\n",
      "     |      \n",
      "     |        >>> list_of_types = [\"HR\", \"Finance\", \"Management\"]\n",
      "     |        >>> g.vs.select(type_in=list_of_types)        #doctest: +SKIP\n",
      "     |      \n",
      "     |      If the operator is omitted, it defaults to C{eq}. For instance, the\n",
      "     |      following selector selects vertices whose C{cluster} property equals\n",
      "     |      to 2:\n",
      "     |      \n",
      "     |        >>> g.vs.select(cluster=2)                    #doctest: +SKIP\n",
      "     |      \n",
      "     |      In the case of an unknown operator, it is assumed that the\n",
      "     |      recognized operator is part of the attribute name and the actual\n",
      "     |      operator is C{eq}.\n",
      "     |      \n",
      "     |      Attribute names inferred from keyword arguments are treated specially\n",
      "     |      if they start with an underscore (C{_}). These are not real attributes\n",
      "     |      but refer to specific properties of the vertices, e.g., its degree.\n",
      "     |      The rule is as follows: if an attribute name starts with an underscore,\n",
      "     |      the rest of the name is interpreted as a method of the L{Graph} object.\n",
      "     |      This method is called with the vertex sequence as its first argument\n",
      "     |      (all others left at default values) and vertices are filtered\n",
      "     |      according to the value returned by the method. For instance, if you\n",
      "     |      want to exclude isolated vertices:\n",
      "     |      \n",
      "     |        >>> g = Graph.Famous(\"zachary\")\n",
      "     |        >>> non_isolated = g.vs.select(_degree_gt=0)\n",
      "     |      \n",
      "     |      For properties that take a long time to be computed (e.g., betweenness\n",
      "     |      centrality for large graphs), it is advised to calculate the values\n",
      "     |      in advance and store it in a graph attribute. The same applies when\n",
      "     |      you are selecting based on the same property more than once in the\n",
      "     |      same C{select()} call to avoid calculating it twice unnecessarily.\n",
      "     |      For instance, the following would calculate betweenness centralities\n",
      "     |      twice:\n",
      "     |      \n",
      "     |        >>> edges = g.vs.select(_betweenness_gt=10, _betweenness_lt=30)\n",
      "     |      \n",
      "     |      It is advised to use this instead:\n",
      "     |      \n",
      "     |        >>> g.vs[\"bs\"] = g.betweenness()\n",
      "     |        >>> edges = g.vs.select(bs_gt=10, bs_lt=30)\n",
      "     |      \n",
      "     |      @return: the new, filtered vertex sequence\n",
      "     |  \n",
      "     |  shortest_paths(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.shortest_paths()}\n",
      "     |      \n",
      "     |      This method calls the C{shortest_paths()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.shortest_paths() for details.\n",
      "     |  \n",
      "     |  similarity_dice(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.similarity_dice()}\n",
      "     |      \n",
      "     |      This method calls the C{similarity_dice()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.similarity_dice() for details.\n",
      "     |  \n",
      "     |  similarity_jaccard(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.similarity_jaccard()}\n",
      "     |      \n",
      "     |      This method calls the C{similarity_jaccard()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.similarity_jaccard() for details.\n",
      "     |  \n",
      "     |  subgraph(*args, **kwds)\n",
      "     |      Proxy method to L{Graph.subgraph()}\n",
      "     |      \n",
      "     |      This method calls the C{subgraph()} method of the L{Graph} class\n",
      "     |      restricted to this sequence, and returns the result.\n",
      "     |      \n",
      "     |      @see: Graph.subgraph() for details.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from igraph.core.VertexSeq:\n",
      "     |  \n",
      "     |  __delitem__(...)\n",
      "     |      x.__delitem__(y) <==> del x[y]\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __init__(...)\n",
      "     |      x.__init__(...) initializes x; see help(type(x)) for signature\n",
      "     |  \n",
      "     |  __len__(...)\n",
      "     |      x.__len__() <==> len(x)\n",
      "     |  \n",
      "     |  __setitem__(...)\n",
      "     |      x.__setitem__(i, y) <==> x[i]=y\n",
      "     |  \n",
      "     |  attribute_names(...)\n",
      "     |      attribute_names() -> list\n",
      "     |      \n",
      "     |      Returns the attribute name list of the graph's vertices\n",
      "     |  \n",
      "     |  get_attribute_values(...)\n",
      "     |      get_attribute_values(attrname) -> list\n",
      "     |      Returns the value of a given vertex attribute for all vertices in a list.\n",
      "     |      \n",
      "     |      The values stored in the list are exactly the same objects that are stored\n",
      "     |      in the vertex attribute, meaning that in the case of mutable objects,\n",
      "     |      the modification of the list element does affect the attribute stored in\n",
      "     |      the vertex. In the case of immutable objects, modification of the list\n",
      "     |      does not affect the attribute values.\n",
      "     |      \n",
      "     |      @param attrname: the name of the attribute\n",
      "     |  \n",
      "     |  set_attribute_values(...)\n",
      "     |      set_attribute_values(attrname, values) -> list\n",
      "     |      Sets the value of a given vertex attribute for all vertices\n",
      "     |      \n",
      "     |      @param attrname: the name of the attribute\n",
      "     |      @param values: the new attribute values in a list\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from igraph.core.VertexSeq:\n",
      "     |  \n",
      "     |  graph\n",
      "     |      The graph the vertex sequence belongs to\n",
      "     |  \n",
      "     |  indices\n",
      "     |      The vertex indices in this vertex sequence\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from igraph.core.VertexSeq:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "\n",
      "FUNCTIONS\n",
      "    StringIO(...)\n",
      "        StringIO([s]) -- Return a StringIO-like stream for reading or writing\n",
      "    \n",
      "    autocurve(graph, attribute='curved', default=0)\n",
      "        Calculates curvature values for each of the edges in the graph to make\n",
      "        sure that multiple edges are shown properly on a graph plot.\n",
      "        \n",
      "        This function checks the multiplicity of each edge in the graph and\n",
      "        assigns curvature values (numbers between -1 and 1, corresponding to\n",
      "        CCW (-1), straight (0) and CW (1) curved edges) to them. The assigned\n",
      "        values are either stored in an edge attribute or returned as a list,\n",
      "        depending on the value of the I{attribute} argument.\n",
      "        \n",
      "        @param graph: the graph on which the calculation will be run\n",
      "        @param attribute: the name of the edge attribute to save the curvature\n",
      "          values to. The default value is C{curved}, which is the name of the\n",
      "          edge attribute the default graph plotter checks to decide whether an\n",
      "          edge should be curved on the plot or not. If I{attribute} is C{None},\n",
      "          the result will not be stored.\n",
      "        @param default: the default curvature for single edges. Zero means that\n",
      "          single edges will be straight. If you want single edges to be curved\n",
      "          as well, try passing 0.5 or -0.5 here.\n",
      "        @return: the list of curvature values if I{attribute} is C{None},\n",
      "          otherwise C{None}.\n",
      "    \n",
      "    community_to_membership(...)\n",
      "        community_to_membership(merges, nodes, steps, return_csize=False)\n",
      "    \n",
      "    convex_hull(...)\n",
      "        convex_hull(vs, coords=False)\n",
      "        \n",
      "        Calculates the convex hull of a given point set.\n",
      "        \n",
      "        @param vs: the point set as a list of lists\n",
      "        @param coords: if C{True}, the function returns the\n",
      "          coordinates of the corners of the convex hull polygon,\n",
      "          otherwise returns the corner indices.\n",
      "        @return: either the hull's corner coordinates or the point\n",
      "          indices corresponding to them, depending on the C{coords}\n",
      "          parameter.\n",
      "    \n",
      "    cos(...)\n",
      "        cos(x)\n",
      "        \n",
      "        Return the cosine of x (measured in radians).\n",
      "    \n",
      "    deprecated(message)\n",
      "        Prints a warning message related to the deprecation of some igraph\n",
      "        feature.\n",
      "    \n",
      "    is_degree_sequence(...)\n",
      "        is_degree_sequence(out_deg, in_deg=None)\n",
      "        \n",
      "        Returns whether a list of degrees can be a degree sequence of some graph.\n",
      "        \n",
      "        Note that it is not required for the graph to be simple; in other words,\n",
      "        this function may return C{True} for degree sequences that can be realized\n",
      "        using one or more multiple or loop edges only.\n",
      "        \n",
      "        In particular, this function checks whether\n",
      "        \n",
      "          - all the degrees are non-negative\n",
      "          - for undirected graphs, the sum of degrees are even\n",
      "          - for directed graphs, the two degree sequences are of the same length and\n",
      "            equal sums\n",
      "        \n",
      "        @param out_deg: the list of degrees. For directed graphs, this list must\n",
      "          contain the out-degrees of the vertices.\n",
      "        @param in_deg: the list of in-degrees for directed graphs. This parameter\n",
      "          must be C{None} for undirected graphs.\n",
      "        @return: C{True} if there exists some graph that can realize the given degree\n",
      "          sequence, C{False} otherwise.@see: L{is_graphical_degree_sequence()} if you do not want to allow multiple\n",
      "          or loop edges.\n",
      "    \n",
      "    is_graphical_degree_sequence(...)\n",
      "        is_graphical_degree_sequence(out_deg, in_deg=None)\n",
      "        \n",
      "        Returns whether a list of degrees can be a degree sequence of some simple graph.\n",
      "        \n",
      "        Note that it is required for the graph to be simple; in other words,\n",
      "        this function will return C{False} for degree sequences that cannot be realized\n",
      "        without using one or more multiple or loop edges.\n",
      "        \n",
      "        @param out_deg: the list of degrees. For directed graphs, this list must\n",
      "          contain the out-degrees of the vertices.\n",
      "        @param in_deg: the list of in-degrees for directed graphs. This parameter\n",
      "          must be C{None} for undirected graphs.\n",
      "        @return: C{True} if there exists some simple graph that can realize the given\n",
      "          degree sequence, C{False} otherwise.\n",
      "        @see: L{is_degree_sequence()} if you want to allow multiple or loop edges.\n",
      "    \n",
      "    load = read(filename, *args, **kwds)\n",
      "        Loads a graph from the given filename.\n",
      "        \n",
      "        This is just a convenience function, calls L{Graph.Read} directly.\n",
      "        All arguments are passed unchanged to L{Graph.Read}\n",
      "        \n",
      "        @param filename: the name of the file to be loaded\n",
      "    \n",
      "    read(filename, *args, **kwds)\n",
      "        Loads a graph from the given filename.\n",
      "        \n",
      "        This is just a convenience function, calls L{Graph.Read} directly.\n",
      "        All arguments are passed unchanged to L{Graph.Read}\n",
      "        \n",
      "        @param filename: the name of the file to be loaded\n",
      "    \n",
      "    save = write(graph, filename, *args, **kwds)\n",
      "        Saves a graph to the given file.\n",
      "        \n",
      "        This is just a convenience function, calls L{Graph.write} directly.\n",
      "        All arguments are passed unchanged to L{Graph.write}\n",
      "        \n",
      "        @param graph: the graph to be saved\n",
      "        @param filename: the name of the file to be written\n",
      "    \n",
      "    set_progress_handler(...)\n",
      "        set_progress_handler(handler)\n",
      "        \n",
      "        Sets the handler to be called when igraph is performing a long operation.\n",
      "        @param handler: the progress handler function. It must accept two\n",
      "          arguments, the first is the message informing the user about\n",
      "          what igraph is doing right now, the second is the actual\n",
      "          progress information (a percentage).\n",
      "    \n",
      "    set_random_number_generator(...)\n",
      "        set_random_number_generator(generator)\n",
      "        \n",
      "        Sets the random number generator used by igraph.\n",
      "        @param generator: the generator to be used. It must be a Python object\n",
      "          with at least three attributes: C{random}, C{randint} and C{gauss}.\n",
      "          Each of them must be callable and their signature and behaviour\n",
      "          must be identical to C{random.random}, C{random.randint} and\n",
      "          C{random.gauss}. By default, igraph uses the C{random} module for\n",
      "          random number generation, but you can supply your alternative\n",
      "          implementation here. If the given generator is C{None}, igraph\n",
      "          reverts to the default Mersenne twister generator implemented in the\n",
      "          C layer, which might be slightly faster than calling back to Python\n",
      "          for random numbers, but you cannot set its seed or save its state.\n",
      "    \n",
      "    set_status_handler(...)\n",
      "        set_status_handler(handler)\n",
      "        \n",
      "        Sets the handler to be called when igraph tries to display a status\n",
      "        message.\n",
      "        \n",
      "        This is used to communicate the progress of some calculations where\n",
      "        no reasonable progress percentage can be given (so it is not possible\n",
      "        to use the progress handler).\n",
      "        \n",
      "        @param handler: the status handler function. It must accept a single\n",
      "          argument, the message that informs the user about what igraph is\n",
      "          doing right now.\n",
      "    \n",
      "    sin(...)\n",
      "        sin(x)\n",
      "        \n",
      "        Return the sine of x (measured in radians).\n",
      "    \n",
      "    summary(obj, stream=None, *args, **kwds)\n",
      "        Prints a summary of object o to a given stream\n",
      "        \n",
      "        Positional and keyword arguments not explicitly mentioned here are passed\n",
      "        on to the underlying C{summary()} method of the object if it has any.\n",
      "        \n",
      "        @param obj: the object about which a human-readable summary is requested.\n",
      "        @param stream: the stream to be used. If C{None}, the standard output\n",
      "          will be used.\n",
      "    \n",
      "    warn(...)\n",
      "        Issue a warning, or maybe ignore it or raise an exception.\n",
      "    \n",
      "    write(graph, filename, *args, **kwds)\n",
      "        Saves a graph to the given file.\n",
      "        \n",
      "        This is just a convenience function, calls L{Graph.write} directly.\n",
      "        All arguments are passed unchanged to L{Graph.write}\n",
      "        \n",
      "        @param graph: the graph to be saved\n",
      "        @param filename: the name of the file to be written\n",
      "\n",
      "DATA\n",
      "    ADJ_DIRECTED = 0\n",
      "    ADJ_LOWER = 3\n",
      "    ADJ_MAX = 1\n",
      "    ADJ_MIN = 4\n",
      "    ADJ_PLUS = 5\n",
      "    ADJ_UNDIRECTED = 1\n",
      "    ADJ_UPPER = 2\n",
      "    ALL = 3\n",
      "    BLISS_F = 0\n",
      "    BLISS_FL = 1\n",
      "    BLISS_FLM = 4\n",
      "    BLISS_FM = 3\n",
      "    BLISS_FS = 2\n",
      "    BLISS_FSM = 5\n",
      "    GET_ADJACENCY_BOTH = 2\n",
      "    GET_ADJACENCY_LOWER = 1\n",
      "    GET_ADJACENCY_UPPER = 0\n",
      "    IN = 2\n",
      "    Nexus = <igraph.remote.nexus.NexusConnection object>\n",
      "    OUT = 1\n",
      "    REWIRING_SIMPLE = 0\n",
      "    REWIRING_SIMPLE_LOOPS = 1\n",
      "    STAR_IN = 1\n",
      "    STAR_MUTUAL = 3\n",
      "    STAR_OUT = 0\n",
      "    STAR_UNDIRECTED = 2\n",
      "    STRONG = 2\n",
      "    TRANSITIVITY_NAN = 0\n",
      "    TRANSITIVITY_ZERO = 1\n",
      "    TREE_IN = 1\n",
      "    TREE_OUT = 0\n",
      "    TREE_UNDIRECTED = 2\n",
      "    WEAK = 1\n",
      "    __build_date__ = 'Mar  6 2017'\n",
      "    __license__ = u'\\nCopyright (C) 2006-2012  Tam\\xe1s Nepusz <nt...treet...\n",
      "    __version__ = '0.7.1'\n",
      "    arpack_options = <igraph.ARPACKOptions object>\n",
      "    config = <igraph.configuration.Configuration object>\n",
      "    dbl_epsilon = 2.220446049250313e-16\n",
      "    known_colors = {'alice blue': (0.9411764705882353, 0.9725490196078431,...\n",
      "    name = 'write_svg'\n",
      "    palettes = {'gray': <GradientPalette with 256 colors>, 'heat': <Advanc...\n",
      "    pi = 3.141592653589793\n",
      "    with_statement = _Feature((2, 5, 0, 'alpha', 1), (2, 6, 0, 'alpha', 0)...\n",
      "\n",
      "VERSION\n",
      "    0.7.1\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!pip install python-igraph\n",
    "import igraph\n",
    "# help(igraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def convert_to_igraph(tree):\n",
    "    #Convert a Biopython Tree object to an igraph Graph.\n",
    "    def add_edge(graph, node1, node2):\n",
    "        graph.add_edge(node1.name, node2.name)\n",
    "           \n",
    "\n",
    "    def build_subgraph(graph, top):\n",
    "        \"\"\"Traverse  the Tree, and retrieve  graph edges and nodes.\"\"\"\n",
    "        for clade in top:\n",
    "            graph.add_vertex(name=clade.root.name)\n",
    "            add_edge(graph, top.root, clade.root)\n",
    "            build_subgraph(graph, clade)\n",
    "\n",
    "    if tree.rooted:\n",
    "        G = igraph.Graph(directed=True)\n",
    "    else:\n",
    "        G = igraph.Graph()\n",
    "    G.add_vertex(name=str(tree.root))\n",
    "    build_subgraph(G, tree.root)\n",
    "    return G            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "G=convert_to_igraph(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get Edges and Vertices:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Edges=[e.tuple   for e in G.es]\n",
    "\n",
    "V=[v for v in G.vs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "node_colors=[inner_node_color if 'Inner'  in v['name'] else set_node_color(v['name']) for v in V]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract tree node labels and get the list of labels to be displayed on hover:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "labels=[v['name']   for v in V]\n",
    "display_labels=['' if 'Inner' in label else label for label in labels]\n",
    "node_size=[1 if d_label == '' else 12 for d_label in display_labels]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set   Kamada-Kawai graph layout:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "124"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "layt=G.layout('kk')\n",
    "N=len(layt)# N is equal to len(G.vs)\n",
    "N\n",
    "# help(G.layout())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "###  Define the Plotly objects to plot the corresponding graph ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import plotly.plotly as py\n",
    "from plotly.graph_objs import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Xn=[layt[k][0] for k in range(N)]\n",
    "Yn=[layt[k][1] for k in range(N)]\n",
    "\n",
    "Xe=[]\n",
    "Ye=[]\n",
    "for e in Edges:\n",
    "    Xe+=[layt[e[0]][0],layt[e[1]][0], None]\n",
    "    Ye+=[layt[e[0]][1],layt[e[1]][1], None]  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trace1=Scatter(x=Xe,\n",
    "               y=Ye,\n",
    "               mode='lines',\n",
    "               line=Line(color=inner_node_color, width=1),\n",
    "               hoverinfo='none'\n",
    "               )\n",
    "trace2=Scatter(x=Xn,\n",
    "               y=Yn,\n",
    "               mode='markers',\n",
    "               name='',\n",
    "               marker=Marker(symbol='dot',\n",
    "                             size=node_size, \n",
    "                             color=node_colors,\n",
    "                             line=Line(color='rgb(50,50,50)', width=0.5)\n",
    "                             ),\n",
    "               text=display_labels,\n",
    "               hoverinfo='text'\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "axis=dict(showbackground=False,# \n",
    "          showline=False,  \n",
    "          zeroline=False,\n",
    "          showgrid=False,\n",
    "          showticklabels=False,\n",
    "          title='' \n",
    "          )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set Plotly plot layout:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "layout = Layout(\n",
    "         title=\"Unrooted tree from sequence alignments,<br> with Biopython, Igraph and  Plotly\", \n",
    "         width=800,\n",
    "         height=800,\n",
    "         showlegend=False,\n",
    "         xaxis=XAxis(axis),\n",
    "         yaxis=YAxis(axis), \n",
    "         \n",
    "     margin=Margin(\n",
    "        t=100\n",
    "    ),\n",
    "    hovermode='closest',\n",
    "     annotations=Annotations([\n",
    "           Annotation(\n",
    "           showarrow=False, \n",
    "            text=\"Data source: <a href='http://kinase.com/human/kinome/groups/agc.aln'> [1]</a>\",  \n",
    "            xref='paper',     \n",
    "            yref='paper',     \n",
    "            x=0.2,  \n",
    "            y=-0.1,  \n",
    "            xanchor='left',   \n",
    "            yanchor='bottom',  \n",
    "            font=Font(\n",
    "            size=14 \n",
    "            )     \n",
    "            )\n",
    "        ]),\n",
    "     \n",
    "      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "data=Data([trace1, trace2])\n",
    "fig=Figure(data=data, layout=layout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "PlotlyError",
     "evalue": "Sign in failed.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPlotlyError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-32-c00418de2265>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msign_in\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'empet'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'api-key'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilename\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'unrooted-tree'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/pgreisen/anaconda2/lib/python2.7/site-packages/plotly/plotly/plotly.pyc\u001b[0m in \u001b[0;36msign_in\u001b[0;34m(username, api_key, **kwargs)\u001b[0m\n\u001b[1;32m     56\u001b[0m         \u001b[0mv2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcurrent\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     57\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mexceptions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPlotlyRequestError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 58\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mexceptions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPlotlyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Sign in failed.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     59\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[0mupdate_plot_options\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate_session_plot_options\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mPlotlyError\u001b[0m: Sign in failed."
     ]
    }
   ],
   "source": [
    "#py.sign_in('empet', 'api-key')\n",
    "py.offline.iplot(fig, filename='unrooted-tree')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "    /*body {\n",
       "        background-color: #F5F5F5;\n",
       "    }*/\n",
       "    div.cell{\n",
       "        width: 900px;\n",
       "        margin-left: 13% !important;\n",
       "        margin-right: auto;\n",
       "    }\n",
       "    #notebook li { /* More space between bullet points */\n",
       "    margin-top:0.8em;\n",
       "    }\n",
       "\n",
       "    h1 {\n",
       "        font-family: 'Alegreya Sans', sans-serif;\n",
       "    }\n",
       "    .text_cell_render h1 {\n",
       "        font-weight: 200;\n",
       "        font-size: 40pt;\n",
       "        line-height: 100%;\n",
       "        color: rgb(8, 66, 133);\n",
       "        margin-bottom: 0em;\n",
       "        margin-top: 0em;\n",
       "        display: block;\n",
       "    }\n",
       "    h2 {\n",
       "        font-family: 'Fenix', serif;\n",
       "        text-indent:1em;\n",
       "        text-align:center;\n",
       "    }\n",
       "    .text_cell_render h2 {\n",
       "        font-weight: 200;\n",
       "        font-size: 28pt;\n",
       "        line-height: 100%;\n",
       "        color: rgb(8, 66, 133);\n",
       "        margin-bottom: 1.5em;\n",
       "        margin-top: 0.5em;\n",
       "        display: block;\n",
       "    }\n",
       "    h3 {\n",
       "        font-family: 'Fenix', serif;\n",
       "        %margin-top:12px;\n",
       "        %margin-bottom: 3px;\n",
       "    }\n",
       "    .text_cell_render h3 {\n",
       "        font-weight: 300;\n",
       "        font-size: 18pt;\n",
       "        line-height: 100%;\n",
       "        color: rgb(8, 66, 133);\n",
       "        margin-bottom: 0.5em;\n",
       "        margin-top: 2em;\n",
       "        display: block;\n",
       "    }\n",
       "    h4 {\n",
       "        font-family: 'Fenix', serif;\n",
       "    }\n",
       "    .text_cell_render h4 {\n",
       "        font-weight: 300;\n",
       "        font-size: 16pt;\n",
       "        color: rgb(8, 66, 133);\n",
       "        margin-bottom: 0.5em;\n",
       "        margin-top: 0.5em;\n",
       "        display: block;\n",
       "    }\n",
       "    h5 {\n",
       "        font-family: 'Alegreya Sans', sans-serif;\n",
       "    }\n",
       "    .text_cell_render h5 {\n",
       "        font-weight: 300;\n",
       "        font-style: normal;\n",
       "        font-size: 16pt;\n",
       "        margin-bottom: 0em;\n",
       "        margin-top: 1.5em;\n",
       "        display: block;\n",
       "        }\n",
       "    div.text_cell_render{\n",
       "        font-family: 'Alegreya Sans',Computer Modern, \"Helvetica Neue\", Arial, Helvetica, Geneva, sans-serif;\n",
       "        line-height: 145%;\n",
       "        font-size: 130%;\n",
       "        width:900px;\n",
       "        margin-left:auto;\n",
       "        margin-right:auto;\n",
       "        %text-align:justify;\n",
       "        %text-justify:inter-word;\n",
       "    }\n",
       "    \n",
       "    \n",
       "    code{\n",
       "      font-size: 78%;\n",
       "    }\n",
       "    .rendered_html code{\n",
       "        background-color: transparent;\n",
       "        white-space: inherit;   \n",
       "    }\n",
       "    .prompt{\n",
       "        display: None;\n",
       "     }\n",
       "    .rendered_html code{\n",
       "    background-color: transparent;\n",
       "    }\n",
       "\n",
       "    blockquote{\n",
       "      display:block;\n",
       "      background: #f3f3f3;\n",
       "      font-family: \"Open sans\",verdana,arial,sans-serif;\n",
       "      width:610px;\n",
       "      padding: 15px 15px 15px 15px;\n",
       "      text-align:justify;\n",
       "      text-justify:inter-word;\n",
       "      }\n",
       "      blockquote p {\n",
       "        margin-bottom: 0;\n",
       "        line-height: 125%;\n",
       "        font-size: 100%;\n",
       "      }\n",
       "   /* element.style {\n",
       "    } */\n",
       "</style>\n",
       "<script>\n",
       "    MathJax.Hub.Config({\n",
       "                        TeX: {\n",
       "                           extensions: [\"AMSmath.js\"]\n",
       "                           },\n",
       "                tex2jax: {\n",
       "                    inlineMath: [ [\"$\",\"$\"], [\"\\\\(\",\"\\\\)\"] ],\n",
       "                    displayMath: [ [\"$$\",\"$$\"], [\"\\\\[\",\"\\\\]\"] ]\n",
       "                },\n",
       "                displayAlign: \"center\", // Change this to \"center\" to center equations.\n",
       "                \"HTML-CSS\": {\n",
       "                    styles: {\".MathJax_Display\": {\"margin\": 4}}\n",
       "                }\n",
       "        });\n",
       "</script>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.core.display import HTML\n",
    "def  css_styling():\n",
    "    styles = open(\"./custom.css\", \"r\").read()\n",
    "    return HTML(styles)\n",
    "css_styling()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
